<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="The yt Project" />
  <meta name="author" content="Matthew Turk" />
  <meta name="author" content="Nathan J Goldbaum" />
  <meta name="author" content="John A. ZuHone" />
  <meta name="author" content="Cameron Hummels" />
  <meta name="author" content="Suoqing Ji" />
  <meta name="author" content="Meagan Lang" />
  <meta name="author" content="Madicken Munk" />
  <meta name="author" content="Britton Smith" />
  <meta name="author" content="Kacper Kowalik" />
  <meta name="author" content="Miguel de Val-Borro" />
  <meta name="author" content="Jared W. Coughlin" />
  <meta name="author" content="Corentin Cadiou" />
  <meta name="author" content="Michael Zingale" />
  <meta name="author" content="Leigh Orf" />
  <meta name="author" content="Kelton Halbert" />
  <meta name="author" content="Clément Robert" />
  <meta name="author" content="Christopher Havlin" />
  <meta name="author" content="Stephanie Tonnesen" />
  <meta name="author" content="Andrew Myers" />
  <meta name="author" content="Alex Gurvich" />
  <meta name="author" content="Desika Narayanan" />
  <meta name="author" content="Samuel W. Skillman" />
  <meta name="author" content="Axel Huebl" />
  <meta name="author" content="Elliott Biondo" />
  <meta name="author" content="Andrew Wetzel" />
  <meta name="author" content="Clayton Strawn" />
  <meta name="author" content="Alexander Lindsay" />
  <meta name="author" content="Gabriel Altay" />
  <meta name="author" content="Erwin T. Lau" />
  <meta name="author" content="Aaron Smith" />
  <meta name="author" content="Ji-hoon Kim" />
  <meta name="author" content="Hsi-Yu Schive" />
  <meta name="author" content="Navaneeth S" />
  <meta name="author" content="Brian W. O’Shea" />
  <meta name="author" content="Tom Abel" />
  <meta name="author" content="Yash Gondhalekar" />
  <meta name="author" content="William J Gray" />
  <meta name="author" content="Nickolay Y. Gnedin" />
  <meta name="author" content="Cristian Joana" />
  <meta name="author" content="Yuan Li" />
  <meta name="author" content="Ryan Jeffrey Farber" />
  <meta name="author" content="Jonah M Miller" />
  <meta name="author" content="Michael Ryan" />
  <meta name="author" content="Devin W. Silvia" />
  <meta name="author" content="Robert Jackson" />
  <meta name="author" content="Kenz Arraki" />
  <meta name="author" content="Alankar Dutta" />
  <meta name="author" content="Ritali Ghosh" />
  <meta name="author" content="Shaokun Xie" />
  <meta name="author" content="Jill P. Naiman" />
  <meta name="author" content="Ronan Hix" />
  <meta name="author" content="Josh Borrow" />
  <meta name="author" content="Bili Dong" />
  <meta name="author" content="Ole Streicher" />
  <meta name="author" content="Stuart Mumford" />
  <meta name="author" content="Benjamin Keller" />
  <meta name="author" content="Benjamin Thompson" />
  <meta name="author" content="Philipp Grete" />
  <meta name="author" content="John H. Wise" />
  <meta name="author" content="Shin-Rong Tsai" />
  <meta name="author" content="Nastasha Anna Wijers" />
  <meta name="author" content="Add Yourself" />
  <meta name="dcterms.date" content="2025-01-13" />
  <meta name="keywords" content="markdown, publishing, manubot" />
  <title>Introducing yt 4.0: Analysis and Visualization of Volumetric Data</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      {   }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span.al { color: #ff0000; font-weight: bold; } /* Alert */
    code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
    code span.at { color: #7d9029; } /* Attribute */
    code span.bn { color: #40a070; } /* BaseN */
    code span.bu { color: #008000; } /* BuiltIn */
    code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #4070a0; } /* Char */
    code span.cn { color: #880000; } /* Constant */
    code span.co { color: #60a0b0; font-style: italic; } /* Comment */
    code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
    code span.do { color: #ba2121; font-style: italic; } /* Documentation */
    code span.dt { color: #902000; } /* DataType */
    code span.dv { color: #40a070; } /* DecVal */
    code span.er { color: #ff0000; font-weight: bold; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #40a070; } /* Float */
    code span.fu { color: #06287e; } /* Function */
    code span.im { color: #008000; font-weight: bold; } /* Import */
    code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
    code span.kw { color: #007020; font-weight: bold; } /* Keyword */
    code span.op { color: #666666; } /* Operator */
    code span.ot { color: #007020; } /* Other */
    code span.pp { color: #bc7a00; } /* Preprocessor */
    code span.sc { color: #4070a0; } /* SpecialChar */
    code span.ss { color: #bb6688; } /* SpecialString */
    code span.st { color: #4070a0; } /* String */
    code span.va { color: #19177c; } /* Variable */
    code span.vs { color: #4070a0; } /* VerbatimString */
    code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta property="og:type" content="article" />
  <meta name="dc.title" content="Introducing yt 4.0: Analysis and Visualization of Volumetric Data" />
  <meta name="citation_title" content="Introducing yt 4.0: Analysis and Visualization of Volumetric Data" />
  <meta property="og:title" content="Introducing yt 4.0: Analysis and Visualization of Volumetric Data" />
  <meta property="twitter:title" content="Introducing yt 4.0: Analysis and Visualization of Volumetric Data" />
  <meta name="dc.date" content="2025-01-13" />
  <meta name="citation_publication_date" content="2025-01-13" />
  <meta property="article:published_time" content="2025-01-13" />
  <meta name="dc.modified" content="2025-01-13T01:32:40+00:00" />
  <meta property="article:modified_time" content="2025-01-13T01:32:40+00:00" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="The yt Project" />
  <meta name="citation_author_institution" content="NumFOCUS" />
  <meta name="twitter:creator" content="@yt_astro" />
  <meta name="citation_author" content="Matthew Turk" />
  <meta name="citation_author_institution" content="School of Information Sciences, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_institution" content="Department of Astronomy, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_institution" content="National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0002-5294-0198" />
  <meta name="citation_author" content="Nathan J Goldbaum" />
  <meta name="citation_author_institution" content="National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0001-5557-267X" />
  <meta name="twitter:creator" content="@njgoldbaum" />
  <meta name="citation_author" content="John A. ZuHone" />
  <meta name="citation_author_institution" content="Harvard-Smithsonian Center for Astrophysics" />
  <meta name="citation_author_orcid" content="0000-0003-3175-2347" />
  <meta name="twitter:creator" content="@astrojaz" />
  <meta name="citation_author" content="Cameron Hummels" />
  <meta name="citation_author_institution" content="Department of Theoretical Astrophysics, California Institute of Technology" />
  <meta name="citation_author_orcid" content="0000-0002-3817-8133" />
  <meta name="citation_author" content="Suoqing Ji" />
  <meta name="citation_author_institution" content="Physics Department, University of California Santa Barbara" />
  <meta name="citation_author_orcid" content="0000-0001-9658-0588" />
  <meta name="citation_author" content="Meagan Lang" />
  <meta name="citation_author_institution" content="National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0002-2058-2816" />
  <meta name="citation_author" content="Madicken Munk" />
  <meta name="citation_author_institution" content="National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0003-0117-5366" />
  <meta name="citation_author" content="Britton Smith" />
  <meta name="citation_author_institution" content="University of Edinburgh" />
  <meta name="citation_author_orcid" content="0000-0002-6804-630X" />
  <meta name="citation_author" content="Kacper Kowalik" />
  <meta name="citation_author_institution" content="National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0003-1709-3744" />
  <meta name="citation_author" content="Miguel de Val-Borro" />
  <meta name="citation_author_institution" content="Planetary Science Institute" />
  <meta name="citation_author_orcid" content="000-0002-0455-9384" />
  <meta name="citation_author" content="Jared W. Coughlin" />
  <meta name="citation_author_institution" content="National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0002-4373-4114" />
  <meta name="citation_author" content="Corentin Cadiou" />
  <meta name="citation_author_institution" content="Department of Physics, division of Astrophysics, Lund University" />
  <meta name="citation_author_institution" content="Institut d&#39;Astrophysique de Paris" />
  <meta name="citation_author_orcid" content="0000-0003-2285-0332" />
  <meta name="twitter:creator" content="@cphyc" />
  <meta name="citation_author" content="Michael Zingale" />
  <meta name="citation_author_institution" content="Stony Brook University" />
  <meta name="citation_author_orcid" content="0000-0001-8401-030X" />
  <meta name="citation_author" content="Leigh Orf" />
  <meta name="citation_author_institution" content="Space Science and Engineering Center, University of Wisconsin - Madison" />
  <meta name="citation_author_orcid" content="0000-0002-2677-6427" />
  <meta name="citation_author" content="Kelton Halbert" />
  <meta name="citation_author_institution" content="Cooperative Institute for Meteorological Satellite Studies, The University of Wisconsin, Madison" />
  <meta name="citation_author_orcid" content="0000-0001-6898-2731" />
  <meta name="citation_author" content="Clément Robert" />
  <meta name="citation_author_institution" content="Institut de Planétologie et d&#39;Astrophysique de Grenoble (IPAG), Université Grenoble Alpes, France" />
  <meta name="citation_author_orcid" content="0000-0001-8629-7068" />
  <meta name="citation_author" content="Christopher Havlin" />
  <meta name="citation_author_institution" content="School of Information Sciences, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0003-0585-8236" />
  <meta name="citation_author" content="Stephanie Tonnesen" />
  <meta name="citation_author_institution" content="Center for Computational Astrophysics, Flatiron Institute" />
  <meta name="citation_author_orcid" content="0000-0002-8710-9206" />
  <meta name="citation_author" content="Andrew Myers" />
  <meta name="citation_author_institution" content="Lawrence Berkeley National Laboratory" />
  <meta name="citation_author_orcid" content="0000-0001-8427-8330" />
  <meta name="citation_author" content="Alex Gurvich" />
  <meta name="citation_author_orcid" content="0000-0002-6145-3674" />
  <meta name="citation_author" content="Desika Narayanan" />
  <meta name="citation_author_institution" content="Department of Astronomy, University of Florida, 211 Bryant Space Sciences Center, Gainesville, FL 32611 USA" />
  <meta name="citation_author_institution" content="University of Florida Informatics Institute, 432 Newell Drive, CISE Bldg E251, Gainesville, FL 32611" />
  <meta name="citation_author_institution" content="Cosmic Dawn Center at the Niels Bohr Institute, University of Copenhagen and DTU-Space, Technical University of Denmark" />
  <meta name="citation_author_orcid" content="0000-0002-7064-4309" />
  <meta name="citation_author" content="Samuel W. Skillman" />
  <meta name="citation_author_institution" content="Outer Loop LLC" />
  <meta name="citation_author_orcid" content="0000-0002-7626-522X" />
  <meta name="citation_author" content="Axel Huebl" />
  <meta name="citation_author_institution" content="Lawrence Berkeley National Laboratory" />
  <meta name="citation_author_orcid" content="0000-0003-1943-7141" />
  <meta name="citation_author" content="Elliott Biondo" />
  <meta name="citation_author_institution" content="Oak Ridge National Laboratory" />
  <meta name="citation_author_orcid" content="0000-0002-9088-1360" />
  <meta name="citation_author" content="Andrew Wetzel" />
  <meta name="citation_author_institution" content="University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0003-0603-8942" />
  <meta name="citation_author" content="Clayton Strawn" />
  <meta name="citation_author_institution" content="UC Santa Cruz" />
  <meta name="citation_author_orcid" content="0000-0001-9695-4017" />
  <meta name="citation_author" content="Alexander Lindsay" />
  <meta name="citation_author_institution" content="Idaho National Laboratory" />
  <meta name="citation_author_orcid" content="0000-0002-6988-2123" />
  <meta name="citation_author" content="Gabriel Altay" />
  <meta name="citation_author_institution" content="Free Agent" />
  <meta name="citation_author_orcid" content="0000-0002-4120-2907" />
  <meta name="citation_author" content="Erwin T. Lau" />
  <meta name="citation_author_institution" content="Center for Astrophysics - Harvard &amp; Smithsonian" />
  <meta name="citation_author_institution" content="University of Miami" />
  <meta name="citation_author_orcid" content="0000-0001-8914-8885" />
  <meta name="citation_author" content="Aaron Smith" />
  <meta name="citation_author_institution" content="Center for Astrophysics - Harvard &amp; Smithsonian" />
  <meta name="citation_author_orcid" content="0000-0002-2838-9033" />
  <meta name="citation_author" content="Ji-hoon Kim" />
  <meta name="citation_author_institution" content="Center for Theoretical Physics, Seoul National University" />
  <meta name="citation_author_orcid" content="0000-0003-4464-1160" />
  <meta name="citation_author" content="Hsi-Yu Schive" />
  <meta name="citation_author_institution" content="Institute of Astrophysics, National Taiwan University, Taipei 10617, Taiwan" />
  <meta name="citation_author_institution" content="Physics Division, National Center for Theoretical Sciences, Taipei 10617, Taiwan" />
  <meta name="citation_author_orcid" content="0000-0002-1249-279X" />
  <meta name="citation_author" content="Navaneeth S" />
  <meta name="citation_author_institution" content="Indian Institute of Technology Kharagpur" />
  <meta name="citation_author_orcid" content="0009-0007-6922-0369" />
  <meta name="citation_author" content="Brian W. O&#39;Shea" />
  <meta name="citation_author_institution" content="Michigan State University" />
  <meta name="citation_author_orcid" content="0000-0002-2786-0348" />
  <meta name="citation_author" content="Tom Abel" />
  <meta name="citation_author_institution" content="Kavli Institute for Particle Astrophysics and Cosmology, Stanford University" />
  <meta name="citation_author_orcid" content="0000-0002-5969-1251" />
  <meta name="citation_author" content="Yash Gondhalekar" />
  <meta name="citation_author_institution" content="Birla Institute of Technology and Science, Pilani, Sancoale, Goa 403726, India" />
  <meta name="citation_author_orcid" content="0000-0002-6646-4225" />
  <meta name="citation_author" content="William J Gray" />
  <meta name="citation_author_orcid" content="0000-0001-9014-3125" />
  <meta name="citation_author" content="Nickolay Y. Gnedin" />
  <meta name="citation_author_institution" content="Fermilab" />
  <meta name="citation_author_orcid" content="0000-0001-5925-4580" />
  <meta name="citation_author" content="Cristian Joana" />
  <meta name="citation_author_institution" content="Institute of theoretical physics, Chinese Academy of Science" />
  <meta name="citation_author_orcid" content="0000-0003-4642-3028" />
  <meta name="citation_author" content="Yuan Li" />
  <meta name="citation_author_institution" content="University of North Texas" />
  <meta name="citation_author_orcid" content="0000-0001-5262-6150" />
  <meta name="citation_author" content="Ryan Jeffrey Farber" />
  <meta name="citation_author_institution" content="Max Planck Institute for Astrophysics" />
  <meta name="citation_author_orcid" content="0000-0002-0649-9055" />
  <meta name="citation_author" content="Jonah M Miller" />
  <meta name="citation_author_institution" content="Los Alamos National Laboratory" />
  <meta name="citation_author_orcid" content="0000-0001-6432-7860" />
  <meta name="citation_author" content="Michael Ryan" />
  <meta name="citation_author_institution" content="Penn State University" />
  <meta name="citation_author_orcid" content="0000-0002-0378-5195" />
  <meta name="citation_author" content="Devin W. Silvia" />
  <meta name="citation_author_institution" content="Michigan State University" />
  <meta name="citation_author_orcid" content="0000-0002-4109-9313" />
  <meta name="citation_author" content="Robert Jackson" />
  <meta name="citation_author_institution" content="Argonne National Laboratory" />
  <meta name="citation_author_orcid" content="0000-0003-2518-1234" />
  <meta name="citation_author" content="Kenz Arraki" />
  <meta name="citation_author_institution" content="none" />
  <meta name="citation_author_orcid" content="0000-0002-3012-1167" />
  <meta name="citation_author" content="Alankar Dutta" />
  <meta name="citation_author_institution" content="Indian Institute of Science, Bangalore, India" />
  <meta name="citation_author_orcid" content="0000-0002-9287-4033" />
  <meta name="citation_author" content="Ritali Ghosh" />
  <meta name="citation_author_institution" content="Indian Institute of Science" />
  <meta name="citation_author_orcid" content="0000-0001-8643-7104" />
  <meta name="citation_author" content="Shaokun Xie" />
  <meta name="citation_author_institution" content="Shanghai Astronomical Observatory, Chinese Academy of Sciences" />
  <meta name="citation_author_institution" content="School of Astronomy and Space Sciences, University of Chinese Academy of Sciences" />
  <meta name="citation_author_orcid" content="0000-0001-5624-6008" />
  <meta name="citation_author" content="Jill P. Naiman" />
  <meta name="citation_author_institution" content="School of Information Sciences, University of Illinois, Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0002-9397-6189" />
  <meta name="citation_author" content="Ronan Hix" />
  <meta name="citation_author_institution" content="University of Maryland, Oak Ridge National Lab" />
  <meta name="citation_author_orcid" content="0000-0002-7047-3730" />
  <meta name="citation_author" content="Josh Borrow" />
  <meta name="citation_author_institution" content="Department of Physics, Kavli Institute for Astrophysics and Space Research, Massachusetts Institute of Technology, Cambridge, MA 02139, USA" />
  <meta name="citation_author_orcid" content="0000-0002-1327-1921" />
  <meta name="citation_author" content="Bili Dong" />
  <meta name="citation_author_institution" content="Department of Physics, Center for Astrophysics and Space Sciences, University of California at San Diego" />
  <meta name="citation_author_orcid" content="0000-0001-5081-9039" />
  <meta name="citation_author" content="Ole Streicher" />
  <meta name="citation_author_institution" content="Leibniz Institute for Astrophysics Potsdam (AIP)" />
  <meta name="citation_author_orcid" content="0000-0001-7751-1843" />
  <meta name="citation_author" content="Stuart Mumford" />
  <meta name="citation_author_institution" content="Aperio Software Ltd, UK" />
  <meta name="citation_author_orcid" content="0000-0003-4217-4642" />
  <meta name="citation_author" content="Benjamin Keller" />
  <meta name="citation_author_institution" content="Department of Physics and Materials Science, University of Memphis" />
  <meta name="citation_author_orcid" content="0000-0002-9642-7193" />
  <meta name="citation_author" content="Benjamin Thompson" />
  <meta name="citation_author_institution" content="Jeremiah Horrocks Institute, University of Central Lancashire, Preston, Lancashire, PR1 2HE, UK" />
  <meta name="citation_author_institution" content="Institute for Computational Astrophysics, Dept of Astronomy &amp; Physics, Saint Mary’s University, Halifax, BH3 3C3, Canada" />
  <meta name="citation_author_orcid" content="0000-0003-4383-9183" />
  <meta name="citation_author" content="Philipp Grete" />
  <meta name="citation_author_institution" content="University of Hamburg" />
  <meta name="citation_author_orcid" content="0000-0003-3555-9886" />
  <meta name="citation_author" content="John H. Wise" />
  <meta name="citation_author_institution" content="Center for Relativistic Astrophysics, School of Physics, Georgia Institute of Technology, Atlanta, GA 30332, USA" />
  <meta name="citation_author_orcid" content="0000-0003-1173-8847" />
  <meta name="citation_author" content="Shin-Rong Tsai" />
  <meta name="citation_author_institution" content="Department of Physics, National Taiwan University" />
  <meta name="citation_author_institution" content="School of Information Sciences, University of Illinois at Urbana-Champaign" />
  <meta name="citation_author_orcid" content="0000-0003-4635-6259" />
  <meta name="citation_author" content="Nastasha Anna Wijers" />
  <meta name="citation_author_institution" content="CIERA and Department of Physics and Astronomy, Northwestern University, 1800 Sherman Ave, Evanston, IL 60201, USA" />
  <meta name="citation_author_orcid" content="0000-0001-6374-7185" />
  <meta name="citation_author" content="Add Yourself" />
  <meta name="citation_author_institution" content="Your University" />
  <meta name="citation_author_orcid" content="XXXX-XXXX-XXXX-XXXX" />
  <link rel="canonical" href="https://yt-project.github.io/yt-4.0-paper/" />
  <meta property="og:url" content="https://yt-project.github.io/yt-4.0-paper/" />
  <meta property="twitter:url" content="https://yt-project.github.io/yt-4.0-paper/" />
  <meta name="citation_fulltext_html_url" content="https://yt-project.github.io/yt-4.0-paper/" />
  <meta name="citation_pdf_url" content="https://yt-project.github.io/yt-4.0-paper/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://yt-project.github.io/yt-4.0-paper/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://yt-project.github.io/yt-4.0-paper/v/d4a957c9370c8780f1d5c6b691a2a316b2bfcdb5/" />
  <meta name="manubot_html_url_versioned" content="https://yt-project.github.io/yt-4.0-paper/v/d4a957c9370c8780f1d5c6b691a2a316b2bfcdb5/" />
  <meta name="manubot_pdf_url_versioned" content="https://yt-project.github.io/yt-4.0-paper/v/d4a957c9370c8780f1d5c6b691a2a316b2bfcdb5/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<!--
  D3 Plugin (third-party) 
  
  Allows you to create complex, dynamic,interactive, data-driven visualizations
  in SVG. See https://d3js.org/.
-->

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/d3/6.2.0/d3.min.js"
  integrity="sha512-C2RveGuPIWqkaLAluvoxyiaN1XYNe5ss11urhZWZYBUA9Ydgj+hfGKPcxCzTwut1/fmjEZR7Ac35f2aycT8Ogw=="
  crossorigin="anonymous"
></script>
<!-- PaperJS plugin -->

<script 
    src="https://cdnjs.cloudflare.com/ajax/libs/acorn/8.2.1/acorn.min.js"
    integrity="sha512-SaR3AskCC6XMUHYTuq+O7Pat3lTYepJeW/yaMA1nXl8y6i2B5F1NpkJ/qM5RcXcL9MQ8VOdGvvww7Dp7rDR3aw=="
    crossorigin="anonymous"
>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // Acorn is a parser for javascript used by PaperJS.  Loading a recent
    // version of it enables PaperJS scripts to include modern ECMAScript 
    // features such as arrow functions.

    // https://github.com/acornjs/acorn/
</script>

<script
    src="https://cdnjs.cloudflare.com/ajax/libs/paper.js/0.12.15/paper-full.min.js"
    integrity="sha512-ovjLI1ZcZe6bw+ImQ21r+sv8q/Vwob2kq7tFidK6E1LWfi0T4uobbmpfEU1//a9h9o5Kkt+MnMWf6rWlg0EiMw=="
    crossorigin="anonymous"
>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // PaperJS is an implementation of PaperScript that enables interactive
    // geometric diagrams.

    // https://paperjs.org/
</script>
<!--
  vega-lite Plugin (third-party) 
  
  Allows you to use a declarative grammar to describe interactive visualizations.
  See https://vega.github.io/
-->

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/vega/5.25.0/vega.min.js"
  integrity="sha512-XX9YBZNtGu/Bz3d3zstE6vcEf4coHTGyxTwzU+0DNR7kHoD37krFcyWnqKTdzV6yWS9a5g3H3a71bM4ctzE1kA=="
  crossorigin="anonymous"
  referrerpolicy="no-referrer"
></script>
<script
  src="https://cdnjs.cloudflare.com/ajax/libs/vega-lite/5.14.1/vega-lite.min.js"
  integrity="sha512-tOoHntHr+Zq7t/jBpDkIQJG6q+fWZVzoSNOpoU4g+TEh4I4cTZuCPNSLGXP9KC9mWSyQ8QMEem8PMHSX9zPq9w=="
  crossorigin="anonymous"
  referrerpolicy="no-referrer"
></script>
<script
  src="https://cdnjs.cloudflare.com/ajax/libs/vega-embed/6.22.2/vega-embed.min.js"
  integrity="sha512-XGkG4fTLMPA3Lp9FUGv7MFNvzpAzC/QgC//cOSxjfhmj8fZjF8iyqn9rgXjwHiY1xqWFzn+BnH8R0oGClEIPvw=="
  crossorigin="anonymous"
  referrerpolicy="no-referrer"
></script>
<!-- ThreeJS plugin -->

<script type="importmap">
    {
        "imports": {
            "three": "https://unpkg.com/three@0.152.2/build/three.module.js",
            "three/addons/": "https://unpkg.com/three@0.152.2/examples/jsm/"
        }
    }
</script>
<script type="module">
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // ThreeJS is a library for displaying three-dimensional scenes in the browser.
    // We use an importmap to allow for relative intra-package imports, but
    // unfortunately this does not (currently) allow an SRI.

    // https://threejs.org/
    import * as THREE from 'three';
    import { OrbitControls } from 'three/addons/controls/OrbitControls.js';
    async function start() {
        window.THREE = THREE;
        window.OrbitControls = OrbitControls;
        const LoadedEvent = new Event("THREELoaded");
        LoadedEvent.THREE = THREE;
        LoadedEvent.OrbitControls = OrbitControls;
        document.dispatchEvent(LoadedEvent);
    }
    window.addEventListener("load", start);
</script>
<header id="title-block-header">
<h1 class="title">Introducing yt 4.0: Analysis and Visualization of Volumetric Data</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://yt-project.github.io/yt-4.0-paper/v/d4a957c9370c8780f1d5c6b691a2a316b2bfcdb5/">permalink</a>)
was automatically generated
from <a href="https://github.com/yt-project/yt-4.0-paper/tree/d4a957c9370c8780f1d5c6b691a2a316b2bfcdb5">yt-project/yt-4.0-paper@d4a957c</a>
on January 13, 2025.
</em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>The yt Project</strong>
<br>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/yt-project">yt-project</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/yt_astro">yt_astro</a>
<br>
<small>
NumFOCUS
</small></p></li>
<li><p><strong>Matthew Turk</strong>
<sup><a href="#correspondence">✉</a></sup><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-5294-0198">0000-0002-5294-0198</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/MatthewTurk">MatthewTurk</a>
· <img src="images/mastodon.svg" class="inline_icon" width="16" height="16" alt="Mastodon icon" />
<a href="https://mastodon.social/@powersoffour">@powersoffour@mastodon.social</a>
<br>
<small>
School of Information Sciences, University of Illinois at Urbana-Champaign; Department of Astronomy, University of Illinois at Urbana-Champaign; National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign
</small></p></li>
<li><p><strong>Nathan J Goldbaum</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5557-267X">0000-0001-5557-267X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/ngoldbaum">ngoldbaum</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/njgoldbaum">njgoldbaum</a>
<br>
<small>
National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign
</small></p></li>
<li><p><strong>John A. ZuHone</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-3175-2347">0000-0003-3175-2347</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/jzuhone">jzuhone</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/astrojaz">astrojaz</a>
<br>
<small>
Harvard-Smithsonian Center for Astrophysics
</small></p></li>
<li><p><strong>Cameron Hummels</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-3817-8133">0000-0002-3817-8133</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/chummels">chummels</a>
<br>
<small>
Department of Theoretical Astrophysics, California Institute of Technology
</small></p></li>
<li><p><strong>Suoqing Ji</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-9658-0588">0000-0001-9658-0588</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/jisuoqing">jisuoqing</a>
<br>
<small>
Physics Department, University of California Santa Barbara
</small></p></li>
<li><p><strong>Meagan Lang</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2058-2816">0000-0002-2058-2816</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/langmm">langmm</a>
<br>
<small>
National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign
</small></p></li>
<li><p><strong>Madicken Munk</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-0117-5366">0000-0003-0117-5366</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/munkm">munkm</a>
<br>
<small>
National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign
</small></p></li>
<li><p><strong>Britton Smith</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-6804-630X">0000-0002-6804-630X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/brittonsmith">brittonsmith</a>
<br>
<small>
University of Edinburgh
</small></p></li>
<li><p><strong>Kacper Kowalik</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-1709-3744">0000-0003-1709-3744</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Xarthisius">Xarthisius</a>
<br>
<small>
National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign
</small></p></li>
<li><p><strong>Miguel de Val-Borro</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/000-0002-0455-9384">000-0002-0455-9384</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/migueldvb">migueldvb</a>
<br>
<small>
Planetary Science Institute
</small></p></li>
<li><p><strong>Jared W. Coughlin</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-4373-4114">0000-0002-4373-4114</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/jcoughlin11">jcoughlin11</a>
<br>
<small>
National Center for Supercomputing Applications, University of Illinois at Urbana-Champaign
</small></p></li>
<li><p><strong>Corentin Cadiou</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-2285-0332">0000-0003-2285-0332</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/cphyc">cphyc</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/cphyc">cphyc</a>
<br>
<small>
Department of Physics, division of Astrophysics, Lund University; Institut d’Astrophysique de Paris
</small></p></li>
<li><p><strong>Michael Zingale</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8401-030X">0000-0001-8401-030X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/zingale">zingale</a>
<br>
<small>
Stony Brook University
</small></p></li>
<li><p><strong>Leigh Orf</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2677-6427">0000-0002-2677-6427</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/leighorf">leighorf</a>
<br>
<small>
Space Science and Engineering Center, University of Wisconsin - Madison
</small></p></li>
<li><p><strong>Kelton Halbert</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-6898-2731">0000-0001-6898-2731</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/keltonhalbert">keltonhalbert</a>
<br>
<small>
Cooperative Institute for Meteorological Satellite Studies, The University of Wisconsin, Madison
</small></p></li>
<li><p><strong>Clément Robert</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8629-7068">0000-0001-8629-7068</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/neutrinoceros">neutrinoceros</a>
<br>
<small>
Institut de Planétologie et d’Astrophysique de Grenoble (IPAG), Université Grenoble Alpes, France
</small></p></li>
<li><p><strong>Christopher Havlin</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-0585-8236">0000-0003-0585-8236</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/chrishavlin">chrishavlin</a>
<br>
<small>
School of Information Sciences, University of Illinois at Urbana-Champaign
</small></p></li>
<li><p><strong>Stephanie Tonnesen</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8710-9206">0000-0002-8710-9206</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/stonnes">stonnes</a>
<br>
<small>
Center for Computational Astrophysics, Flatiron Institute
</small></p></li>
<li><p><strong>Andrew Myers</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8427-8330">0000-0001-8427-8330</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/atmyers">atmyers</a>
<br>
<small>
Lawrence Berkeley National Laboratory
</small></p></li>
<li><p><strong>Alex Gurvich</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-6145-3674">0000-0002-6145-3674</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/agurvich">agurvich</a>
<br>
<small>
</small></p></li>
<li><p><strong>Desika Narayanan</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-7064-4309">0000-0002-7064-4309</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/dnarayanan">dnarayanan</a>
<br>
<small>
Department of Astronomy, University of Florida, 211 Bryant Space Sciences Center, Gainesville, FL 32611 USA; University of Florida Informatics Institute, 432 Newell Drive, CISE Bldg E251, Gainesville, FL 32611; Cosmic Dawn Center at the Niels Bohr Institute, University of Copenhagen and DTU-Space, Technical University of Denmark
</small></p></li>
<li><p><strong>Samuel W. Skillman</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-7626-522X">0000-0002-7626-522X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/samskillman">samskillman</a>
<br>
<small>
Outer Loop LLC
</small></p></li>
<li><p><strong>Axel Huebl</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-1943-7141">0000-0003-1943-7141</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/ax3l">ax3l</a>
<br>
<small>
Lawrence Berkeley National Laboratory
</small></p></li>
<li><p><strong>Elliott Biondo</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-9088-1360">0000-0002-9088-1360</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/elliottbiondo">elliottbiondo</a>
<br>
<small>
Oak Ridge National Laboratory
</small></p></li>
<li><p><strong>Andrew Wetzel</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-0603-8942">0000-0003-0603-8942</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/arwetzel">arwetzel</a>
<br>
<small>
University of California, Davis
</small></p></li>
<li><p><strong>Clayton Strawn</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-9695-4017">0000-0001-9695-4017</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/claytonstrawn">claytonstrawn</a>
<br>
<small>
UC Santa Cruz
</small></p></li>
<li><p><strong>Alexander Lindsay</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-6988-2123">0000-0002-6988-2123</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/lindsayad">lindsayad</a>
<br>
<small>
Idaho National Laboratory
</small></p></li>
<li><p><strong>Gabriel Altay</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-4120-2907">0000-0002-4120-2907</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/galtay">galtay</a>
<br>
<small>
Free Agent
</small></p></li>
<li><p><strong>Erwin T. Lau</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8914-8885">0000-0001-8914-8885</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/ethlau">ethlau</a>
<br>
<small>
Center for Astrophysics - Harvard &amp; Smithsonian; University of Miami
</small></p></li>
<li><p><strong>Aaron Smith</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2838-9033">0000-0002-2838-9033</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/astrosmith">astrosmith</a>
<br>
<small>
Center for Astrophysics - Harvard &amp; Smithsonian
</small></p></li>
<li><p><strong>Ji-hoon Kim</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4464-1160">0000-0003-4464-1160</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/mornkr">mornkr</a>
<br>
<small>
Center for Theoretical Physics, Seoul National University
</small></p></li>
<li><p><strong>Hsi-Yu Schive</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-1249-279X">0000-0002-1249-279X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/hyschive">hyschive</a>
<br>
<small>
Institute of Astrophysics, National Taiwan University, Taipei 10617, Taiwan; Physics Division, National Center for Theoretical Sciences, Taipei 10617, Taiwan
</small></p></li>
<li><p><strong>Navaneeth S</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0009-0007-6922-0369">0009-0007-6922-0369</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/themousepotato">themousepotato</a>
<br>
<small>
Indian Institute of Technology Kharagpur
</small></p></li>
<li><p><strong>Brian W. O’Shea</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2786-0348">0000-0002-2786-0348</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/bwoshea">bwoshea</a>
<br>
<small>
Michigan State University
</small></p></li>
<li><p><strong>Tom Abel</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-5969-1251">0000-0002-5969-1251</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/yipihey">yipihey</a>
<br>
<small>
Kavli Institute for Particle Astrophysics and Cosmology, Stanford University
</small></p></li>
<li><p><strong>Yash Gondhalekar</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-6646-4225">0000-0002-6646-4225</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Yash-10">Yash-10</a>
<br>
<small>
Birla Institute of Technology and Science, Pilani, Sancoale, Goa 403726, India
</small></p></li>
<li><p><strong>William J Gray</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-9014-3125">0000-0001-9014-3125</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/"></a>
<br>
<small>
</small></p></li>
<li><p><strong>Nickolay Y. Gnedin</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5925-4580">0000-0001-5925-4580</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/ngnedin">ngnedin</a>
<br>
<small>
Fermilab
</small></p></li>
<li><p><strong>Cristian Joana</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4642-3028">0000-0003-4642-3028</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/cjoana">cjoana</a>
<br>
<small>
Institute of theoretical physics, Chinese Academy of Science
</small></p></li>
<li><p><strong>Yuan Li</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5262-6150">0000-0001-5262-6150</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/"></a>
<br>
<small>
University of North Texas
</small></p></li>
<li><p><strong>Ryan Jeffrey Farber</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-0649-9055">0000-0002-0649-9055</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/rjfarber">rjfarber</a>
<br>
<small>
Max Planck Institute for Astrophysics
</small></p></li>
<li><p><strong>Jonah M Miller</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-6432-7860">0000-0001-6432-7860</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Yurlungur">Yurlungur</a>
<br>
<small>
Los Alamos National Laboratory
</small></p></li>
<li><p><strong>Michael Ryan</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-0378-5195">0000-0002-0378-5195</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/mtryan83">mtryan83</a>
<br>
<small>
Penn State University
</small></p></li>
<li><p><strong>Devin W. Silvia</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-4109-9313">0000-0002-4109-9313</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/devinsilvia">devinsilvia</a>
<br>
<small>
Michigan State University
</small></p></li>
<li><p><strong>Robert Jackson</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-2518-1234">0000-0003-2518-1234</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/rcjackson">rcjackson</a>
<br>
<small>
Argonne National Laboratory
</small></p></li>
<li><p><strong>Kenz Arraki</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-3012-1167">0000-0002-3012-1167</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/karraki">karraki</a>
<br>
<small>
none
</small></p></li>
<li><p><strong>Alankar Dutta</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-9287-4033">0000-0002-9287-4033</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/dutta-alankar">dutta-alankar</a>
<br>
<small>
Indian Institute of Science, Bangalore, India
</small></p></li>
<li><p><strong>Ritali Ghosh</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8643-7104">0000-0001-8643-7104</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/RitaliG">RitaliG</a>
<br>
<small>
Indian Institute of Science
</small></p></li>
<li><p><strong>Shaokun Xie</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5624-6008">0000-0001-5624-6008</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/xshaokun">xshaokun</a>
<br>
<small>
Shanghai Astronomical Observatory, Chinese Academy of Sciences; School of Astronomy and Space Sciences, University of Chinese Academy of Sciences
</small></p></li>
<li><p><strong>Jill P. Naiman</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-9397-6189">0000-0002-9397-6189</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/jnaiman">jnaiman</a>
<br>
<small>
School of Information Sciences, University of Illinois, Urbana-Champaign
</small></p></li>
<li><p><strong>Ronan Hix</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-7047-3730">0000-0002-7047-3730</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Ronan-Hix">Ronan-Hix</a>
<br>
<small>
University of Maryland, Oak Ridge National Lab
</small></p></li>
<li><p><strong>Josh Borrow</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-1327-1921">0000-0002-1327-1921</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/jborrow">jborrow</a>
<br>
<small>
Department of Physics, Kavli Institute for Astrophysics and Space Research, Massachusetts Institute of Technology, Cambridge, MA 02139, USA
</small></p></li>
<li><p><strong>Bili Dong</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5081-9039">0000-0001-5081-9039</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/qobilidop">qobilidop</a>
<br>
<small>
Department of Physics, Center for Astrophysics and Space Sciences, University of California at San Diego
</small></p></li>
<li><p><strong>Ole Streicher</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-7751-1843">0000-0001-7751-1843</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/olebole">olebole</a>
<br>
<small>
Leibniz Institute for Astrophysics Potsdam (AIP)
</small></p></li>
<li><p><strong>Stuart Mumford</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4217-4642">0000-0003-4217-4642</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Cadair">Cadair</a>
<br>
<small>
Aperio Software Ltd, UK
</small></p></li>
<li><p><strong>Benjamin Keller</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-9642-7193">0000-0002-9642-7193</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/bwkeller">bwkeller</a>
<br>
<small>
Department of Physics and Materials Science, University of Memphis
</small></p></li>
<li><p><strong>Benjamin Thompson</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4383-9183">0000-0003-4383-9183</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/cosmosquark">cosmosquark</a>
<br>
<small>
Jeremiah Horrocks Institute, University of Central Lancashire, Preston, Lancashire, PR1 2HE, UK; Institute for Computational Astrophysics, Dept of Astronomy &amp; Physics, Saint Mary’s University, Halifax, BH3 3C3, Canada
· Funded by STFC PhD Studentship programme (ST/F007701/1)
</small></p></li>
<li><p><strong>Philipp Grete</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-3555-9886">0000-0003-3555-9886</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/pgrete">pgrete</a>
<br>
<small>
University of Hamburg
· Funded by European Union’s Horizon 2020 (Marie Skłodowska-Curie grant agreement No 101030214)
</small></p></li>
<li><p><strong>John H. Wise</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-1173-8847">0000-0003-1173-8847</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/jwise77">jwise77</a>
<br>
<small>
Center for Relativistic Astrophysics, School of Physics, Georgia Institute of Technology, Atlanta, GA 30332, USA
· Funded by NASA Grants 80NSSC20K0520, 80NSSC21K1053; NSF Grants OAC-1835213, AST-2108020
</small></p></li>
<li><p><strong>Shin-Rong Tsai</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4635-6259">0000-0003-4635-6259</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/cindytsai">cindytsai</a>
<br>
<small>
Department of Physics, National Taiwan University; School of Information Sciences, University of Illinois at Urbana-Champaign
</small></p></li>
<li><p><strong>Nastasha Anna Wijers</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-6374-7185">0000-0001-6374-7185</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/nastasha-w">nastasha-w</a>
<br>
<small>
CIERA and Department of Physics and Astronomy, Northwestern University, 1800 Sherman Ave, Evanston, IL 60201, USA
· Funded by CIERA Postdoctoral Fellowship
</small></p></li>
<li><p><strong>Add Yourself</strong>
<br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/XXXX-XXXX-XXXX-XXXX">XXXX-XXXX-XXXX-XXXX</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/yournamehere">yournamehere</a>
<br>
<small>
Your University
· Funded by Grant XXXXXXXX
</small></p></li>
</ul>
<div id="correspondence">
<p>✉ — Correspondence possible via <a href="https://github.com/yt-project/yt-4.0-paper/issues">GitHub Issues</a>
or email to
Matthew Turk &lt;mjturk@illinois.edu&gt;.</p>
</div>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<p>We present the current version of the <code>yt</code> software package.
<code>yt</code> is an open-source, community-developed platform for analysis of volumetric data, with readers for several dozen data formats, indexing systems for gridded data, adaptive mesh refinement data, unstructured mesh data, discrete and particle formats, and octree-based data, as well as the combination of these.
We describe the systems implemented in <code>yt</code> to facilitate a “science-first” approach to data analysis, wherein the emphasis is on the meaning and interpretation of the data as opposed to its discretization or layout.</p>
<h1 id="authorship-policy">Authorship Policy</h1>
<p>We note that the author list for this paper is, by design, extensive.
We have separated the authors into those that contributed to the text (whose names are ordered <strong>somehow TBD</strong>) and those that are members of the <code>yt</code> community.
The authors from each group have been indicated in the respective author affiliations.</p>
<p>This paper was developed collaboratively, using the Manubot <span class="citation" data-cites="cTN2TQIL">[<a href="#ref-cTN2TQIL" role="doc-biblioref">1</a>]</span> system for collaborating on and reviewing contributed text.</p>
<blockquote>
<p>To add yourself to the author list, please follow the instructions in our
<a href="https://github.com/yt-project/yt-4.0-paper/blob/master/README.md#authorship-policy">README</a>.</p>
</blockquote>
<h2 id="sec:introduction">Introduction</h2>
<p>The process of transforming data into understanding constitutes the vast majority of time, energy, and intellectual effort spent during scientific inquiry.
This is true across domains, whether data is the product of a computational simulation, a telescope observation, the synthesis of sensors distributed across the Earth, or a collection of images of the human brain.
Data, by themselves, do not reflect an understanding of the Universe or its underlying physical properties; rather, they are recordings, or measurements, of the state of systems as observed.
Even for computational simulations, such as simulations of star formation in the galaxy, this is true: these simulations encode information about a discretization of a model, rather than the model itself.</p>
<p>Bridging the gap between this discretization and the physical understanding requires accessing data, manipulating and interrogating this data, and then applying to this data a sense of understanding.
Somehow, bits stored on a disk must become, in our minds, a galaxy undergoing a starburst.</p>
<p>This process is both mediated and impeded by computational tools.
When those tools align with our mental model of how data exists, they can allow us to work more efficiently, asking questions of data and building sophisticated scientific inquiry.
However, when they do not, they can cause frustration, delays, and most worryingly, incorrect or misinterpreted results.
When viewing this from the perspective of the landscape of inquiry, the most startling realization is that the questions a computational tool enables individuals to ask shapes the questions they think to ask.</p>
<p>In <span class="citation" data-cites="1GOm01ggN">[<a href="#ref-1GOm01ggN" role="doc-biblioref">2</a>]</span>, the analysis platform <code>yt</code> was described.
At the time, <code>yt</code> was focused on analyzing and visualizing the output of grid-based adaptive mesh refinement hydrodynamic simulations; while these were used to study many different physical phenomena, they all were laid out in roughly the same way, in rectilinear meshes of data.
In this paper, we present the current version of <code>yt</code>, which enables identical scripts to analyze and visualize data stored as rectilinear grids as before, but additionally particle or discrete data, octree-based data, and data stored as unstructured meshes.
This has been the result of a large-scale effort to rewrite the underlying machinery within <code>yt</code> for accessing data, indexing that data, and providing it in efficient ways to higher-level routines, as discussed in Section Something.
While this was underway, <code>yt</code> has also been considerably reinstrumented with <a href="#sec:units">metadata-aware array infrastructure</a>, the <a href="#sec:vr">volume rendering infrastructure</a> has been rewritten to be more user-friendly and capable, and support for <a href="#sec:noncartesian">non-Cartesian geometries</a> has been added.</p>
<p>The single biggest update/addition to <code>yt</code> since that paper was published has not been technical in nature.
In the intervening years, a directed and intense community-building effort has resulted in the contributions from over a hundred different individuals, many of them early-stage researchers, and a <a href="#sec:community">thriving community of both users and developers</a>.
This is the crowning achievement of development, as we have attempted to build <code>yt</code> into a tool that enables inquiry from a technical level as well as fosters a supportive, friendly community of individuals engaged in self-directed inquiry.</p>
<h2 id="sec:community">Community Building</h2>
<p>Choosing a software package for a particular purpose involves evaluating several differentiating factors; these factors include the functionality of a package, the performance of a package, the user-friendliness, and even the ability of an individual to find help, engage with others and feel a sense of participation <span class="citation" data-cites="XY0j7dm0">[<a href="#ref-XY0j7dm0" role="doc-biblioref">3</a>]</span>.
The development, fostering and design of the community around <code>yt</code> is deemed to be both crucial to the success or failure of <code>yt</code>, and in many ways inseparable from its functionality.</p>
<h3 id="composition">Composition</h3>
<p>There are several rough categories of individuals engaged in development and utilization of <code>yt</code>.
As a result of its API-first design, there are few if any individuals who use <code>yt</code> that do not do so through the scripting interface; this means that the vast (if not exclusive) majority of individuals who interact with the functionality in <code>yt</code> are doing so by writing their own scripts, modules, and code, and arguably engaging in a value-added development process of their own.
The majority of individuals using <code>yt</code> at present are in astronomy and astrophysics, typically fields of simulation, although representatives from other domains are increasingly participating in development and using yt for their own domain-specific problems</p>
<p>Making the distinction somewhat more clearly, there are individuals who have built their own scripts and utilized them as well as individuals who have contributed changes or modules to the primary <code>yt</code> codebase.
In addition, there is an emerging set of projects that build on <code>yt</code> as infrastructure to conduct scientific analysis.
These developers are largely driven by their own pragmatic scientific needs, and they constitute the majority of developers (by number) that contribute to the code base.
The majority of these individuals are early- to mid-career researchers, typically graduate students, postdocs, and assistant professors.</p>
<p>In recent years, there has emerged a more coherent contingency of individuals who participate in both pragmatically-focused development of modules and functionality for their own benefit as well as modules or overall improvement that is supplemental or even external to their own research agenda.
Sections of the code base receiving such improvements include unit handling, plotting code, infrastructure for loading disparate datasets, and so on.
At this time we do not know of any individuals funded to work on <code>yt</code> completely independent of a scientific or scholarly goal.</p>
<p>The composition of the community, particularly with a mixture of timelines for goal-setting and completion, can at times cause frustrations and difficulties.
For instance, the response to “Can this feature be implemented?” often includes an invitation for the questioner to collaborate on developing that feature and submitting it to the codebase.
Developing a schedule of releases is an act of consensus building, both deciding what bugs are critical to fix in the timeline of a release as well as building consensus on what features should be considered blockers for a new release.
The intersection of this with academic deadlines (for instance job application season) requires balance and care.</p>
<h3 id="types-of-tasks">Types of Tasks</h3>
<p>When evaluating the level of engagement, we consider a few different classifications of tasks that are performed by individuals in the community, and evaluate these based on how they flow into greater engagement.</p>
<ul>
<li>Filing issues</li>
<li>Participating in mailing list discussions</li>
<li>Issuing a pull request</li>
<li>Writing documentation</li>
<li>Participating in code review</li>
<li>Closing bug reports</li>
<li>Drafting an enhancement proposal</li>
</ul>
<p>While there are other activities that individuals can participate in, these are the typical activities we see among participants in the community.
The order, flowing from the first to the last, is the typical flow we see for an individual coming to participate in the community.
The first step is typically to file an issue or bug report (occasionally these are requests for new features), followed by participating in development-focused discussion on mailing lists.
The next level of engagement typically involves the development of a new piece of functionality, refinement of existing code, or issuing a fix for a bug or issue.
These take the form of pull requests (described in greater detail <a href="#sec:development">here</a>) that can be reviewed and added to the code base.</p>
<p>The next level of engagement centers around tasks that are not fully-aligned with pragmatic, code-driven scientific inquiry.
The development of documentation is often viewed as orthogonal to the scientific process, and typically requires an iterative writing process.
Participation in code review, providing comments, feedback and suggestions to other authors, is another somewhat orthogonal task; it doesn’t necessarily directly benefit the developer doing the reviewing (although it might) and it does not necessarily result in academic rewards (citations, authorship, etc).
But, it does arise from a pragmatic (ensuring code reliability) or altruistic (the public good of the software) motivation, and is thus a deeper level of engagement.</p>
<p>The final two activities, drafting enhancement proposals and closing bug reports, are the most engaged, and often the most removed from the academic motivation structure.
Developing an <a href="#sec:ytep">enhancement proposal</a> for <code>yt</code> means iterating with other developers on the motivation behind and implementation of a large piece of functionality; it requires both motivation to engage with the community and the patience to build consensus among stakeholders.
Closing bug reports – and the development work associated with identifying, tracking and fixing bugs – requires patience and often repeated engagement with stakeholders.</p>
<h3 id="engagement-metrics">Engagement Metrics</h3>
<p>We include here plots of the level of engagement on mailing list discussions and the citation count of the original method paper.</p>
<h2 id="sec:governance">Project Governance</h2>
<p>Between the publication of the first paper and this paper, the <code>yt</code> project instituted a form of governance involving a steering committee, a set of “members” of the project, and a defined process for developing improvements and enhancements (the YTEP, or <code>yt</code>-enhancement-proposal process).
YTEPs are discussed in <a href="#sec:ytep">1.4.6</a>.
The systems developed account for a number of important procedures, mostly related to decision-making, but do not address pressing community needs such as community standards for conduct, changes in committee composition, sub-project coordination, or the transition of members and developers to “emeritus” status.</p>
<p>This governance structure, however, serves as a process for the <code>yt</code> <em>project</em>, rather than just the software itself.
Much like other software projects with community-building as core goals, the <code>yt</code> project encompasses the core package itself (largely the subject of this paper), affiliated software projects (such as <span class="citation" data-cites="v58DzoTC">[<a href="#ref-v58DzoTC" role="doc-biblioref">4</a>]</span> ) but also, crucially, the broader community and the spaces in which discussions, tutorials, inquiries and development take place.
As such, the governance structure was designed to be more holistic than addressing so-called “Commit Privileges” and access control.</p>
<p>There are several components of the yt governance structure that we highlight here, but direct the interested reader to our <a href="https://github.com/yt-project/governance">governance source repository</a> for a more detailed understanding.</p>
<h3 id="conflicts-of-interest">Conflicts of Interest</h3>
<p><code>yt</code> project development is distributed across universities and organizations, where in some cases different members of the development team may find themselves collaborating, or even applying for similar funding or academic positions.
To minimize real or perceived conflicts of interest among the community, a policy regarding conflicts of interest is applied to members of the <code>yt</code> steering committee.</p>
<blockquote>
<p>It is expected that the Steering Committee Members will be
employed at a wide range of companies, universities and non-profit
organizations.
Because of this, it is possible that Members will have conflict
of interests.
Such conflict of interests include, but are not limited to:</p>
<ul>
<li>Financial interests, such as investments, employment or contracting work, outside of The Project that may influence their work on The Project.</li>
<li>Access to proprietary information of their employer that could potentially leak into their work with the Project.</li>
</ul>
<p>All members of the Steering Committee shall disclose to the rest of the Council any conflict of interest they may have.
Members with a conflict of interest in a particular issue may participate in Committee discussions on that issue, but must recuse themselves from voting on the issue.</p>
</blockquote>
<p>This also addresses the possibility of funded work conflicting with the development of <code>yt</code> as a whole, or project wellness.
As of the time of writing, this policy has not been exercised, but providing it in writing will ensure that the conduct of the steering committee is appropriate.</p>
<h3 id="project-members-and-groups">Project Members and Groups</h3>
<p>Individuals interact with projects in different ways at different times; as careers progress and personal lives shift, the degree of engagement that one can expect will change as well.
To reflect this, as well as to codify the different means by which people can participate in the community, the <code>yt</code> governance structure allows for “members” to transition between different levels of activity.</p>
<p>The base level of “membership” in the <code>yt</code> project is that of “project members.”
This is, by design, a broad category of individuals who have demonstrated their willingness to participate in the community, either through participating in community discussions, helping other members, contributing code or documentation, or reviewing the contributions of others.
The initial seed of the project members was chosen by a flat cutoff of 50 changesets included in the repository; since that time, members are nominated by other members and must receive at least three positive votes.
This membership, unless either explicitly revoked (via five supporting member votes and a steering council vote) lasts for life, although may be transitioned into “emeritus” status.
Within the “members” group, a set of several sub-groups are identified that follow a rough, orchard-like metaphor.</p>
<p>The first of these groups is the <strong>fertilizer</strong> group.
(We whimsically note that the choice of name is not meant as a value judgment!)
This is the high-level organizer group, including the steering committee and any other leadership roles.
This group tends to the growth of the community, ensuring that there are opportunities for new members, and that the project continues as expected.
Explicitly identified in the governance structure is that this is a group meant to train new members and to facilitate its membership to grow and change over time.</p>
<p>The <strong>arboretum</strong> is the largest group, and is the starting point for new members of the project.
The arboretum includes all project members and steering committee members.
This group facilitates and develops contributions to the code base, encouraging diversity of participation.</p>
<p>The <strong>greenhouse</strong> group includes those new members of the project, who are still developing their participation but have contributed to the community (for instance with a bug fix or documentation enhancement.)
Members of the greenhouse group do not necessarily have high-level project responsibilities, but are encouraged to seek involvement and participate in triage meetings and other discussions.</p>
<p>In recognition that temporary obligations, changes in activity level and other interest migrations happen, there is a group entitled the <strong>fallow fields</strong> for members who wish to temporarily step back from their responsibilities.
Many different reasons exist for stepping into the fallow fields – burnout, new or enhanced work responsibilities, family or care duties, or even simply a desire to do other things.
Transitioning to the fallow fields is not permanent, and participation in the project can either resume or the individual can choose to move on.</p>
<p>Finally, we have designated a class of project members as <strong>emeritus</strong> (the only non-orchard name utilized!) for individuals who have moved on to other interests or activities.
The yt project has had a number of individuals and contributors who have later gone on to work in industry or non-profits, or who simply have opted to leave the community.
The emeritus designation is a manner by which their previous contributions can still be recognized and honored, while also recognizing that they are no longer active participants.</p>
<h3 id="decision-making">Decision Making</h3>
<p>The majority of decisions in the yt project are small, day-to-day decisions about code quality, merging pull requests, how to respond to questions and other items for which it is typically quite easy to identify a consensus.
For those situations where consensus is not possible, deciding upon and implementing a course of action can be one of the more difficult issues in a project; members of the community each bring their own preferences, requirements, needs and desires to a project.
These decisions might be regarding the future of specific areas of the project’s code base – for instance, “should the software volume rendering be split into a separate repository?” or “how should we decide the line to draw for excising old frontends?”
Typically, the YTEP process is sufficient, even in cases where a consensus cannot be reached, as it allows a large amount of discussion.
In cases where it cannot, the decision making process utilizes a multi-tier system of votes, including the YTEP and mailing list as venues.
If a 2/3 majority is unable to be reached, the steering committee is allowed approximately a month to identify a consensus among themselves; failing this, a majority vote will be reached.</p>
<p>While the full mechanism of the yt project decision making process has not been exercised in its entirety, its presence has provided a “safety net” for discussion.
For further details, the <a href="https://yt-project.github.io/governance/voting.html">voting process</a> section of the <a href="https://yt-project.github.io/governance/">governance documents</a> lay out the steps, involved groups and timeline.</p>
<h2 id="sec:development">Development Procedure</h2>
<p><code>yt</code> is developed openly.
During the Spring of 2017, development transitioned from occurring on <a href="https://bitbucket.org/yt_analysis/">Bitbucket</a> to <a href="https://github.com/yt-project/">GitHub</a>, and the source code management system was changed from <a href="https://www.mercurial-scm.org/">Mercurial</a> to <a href="https://git-scm.org/">git</a>.
Development occurs through the “pull request” model, wherein changes to the codebase are made and then requested to be included in the primary repository.
Typically, there are two branches of development, and occasionally three. The first of these is the “stable” branch, which is much slower-paced, and typically only modified during the release periods.
The second is that of “main” (formerly “master”, which is the conventional term in git terminology, and renamed in early 2021; the corresponding mercurial term would be “default”) which is where current development takes place.
The “main” branch is meant to be for development proceeding that does not drastically disrupt usage patterns.
Occasionally, such as during the development of <code>yt</code> 4.0, a third branch is included in the primary repository.
This development branch is open for large and potentially disruptive changes, but in order to centralize code review and developer attention it takes place there.
For instance, during the development of <code>yt</code> 4.0, the branch <code>yt-4.0</code> was where the global mesh was removed and where the units subsystem was removed and replaced with <code>unyt</code>.</p>
<p>This three-pronged approach generally has suited the community; the process of backporting changes from the “main” branch to the “stable” branch can be time-consuming.
However, balancing the needs of a community requiring stable methods for analyzing data against the ease of development suggests that this is a toll worth paying.</p>
<p>In general, the development of <code>yt</code> is reasonably top-heavy, with the majority of contributions coming from a core group of individuals.
We discuss the implications of this on sustainability in Section <a href="#sec:sustainability">1.19</a>, and provide here a graph of the contributions over time.
Of particular note is that the development history of yt is also highly bifurcated between version control systems and developer practice.
In the past, yt developers tended to commit frequently and include all of the individual development history of individual features or bug fixes.
Recent practice, however, is more inclined toward commit “squashing,” where multiple commits are combined into a single commit with the same net effect, or commit rebasing, where changes are included linearly rather than through a branched history.
One result of this is in figures such as the top row of Figure <a href="#fig:commit-graph">1</a>, some contributors appear to have made a smaller quantity of contributions than an informed observer would recognize.
Specifically, this applies to Clément Robert, who has contributed a considerable amount of change to the code base but has done so in a way that does not maximize the “statistics” presented below.
This particular bias, toward contributions measured in count, is one that affects other members of the community as well, especially those whose participation is through community engagement, documentation, tutorials, and mentoring, rather than through direct modifications of the code base.
To mitigate this shortcoming, we present the number of pull requests merged into the code base, as a function of time, as well as the time between their creation and their merge, in the lower row of Figure <a href="#fig:commit-graph">1</a>.
This demonstrates that in many cases, the number of discrete contributions to the codebase varies greatly depending on the developer, and we believe gives a more informed perception of the activity in the code base.</p>
<div id="figure-commit-graph" class="center">

</div>
<div id="fig:commit-graph" class="fignos">
<figure>
<img src="images/blank.svg" width="1" alt="Figure 1: Commits and pull requests to the source code as a function of time." />
<figcaption aria-hidden="true"><span>Figure 1:</span> Commits and pull requests to the source code as a function of time.</figcaption>
</figure>
</div>
<script>
vegaEmbed('#figure-commit-graph', "images/yt_repo.vl", {"actions": false});
</script>
<p>In Figure <a href="#fig:pr-closing-time">2</a> we have plotted distribution of pull requests based on the time between their creation and their merge.
The longest time between opening a pull request and merging it was nearly four years; this was the addition of the <code>cf_radial</code> frontend, which occurred in fits and starts over a very long period of time.
The next longest pull request durations are for splitting the code used for bitmap indexing (see <a href="#sec:point_indexing">1.9</a>) and a per-field configuration system.
This includes only those pull requests that occurred on GitHub.</p>
<div id="figure-pr-closing-time" class="center">

</div>
<div id="fig:pr-closing-time" class="fignos">
<figure>
<img src="images/blank.svg" width="1" alt="Figure 2: The distribution of pull requests as a function of how long it took to close them." />
<figcaption aria-hidden="true"><span>Figure 2:</span> The distribution of pull requests as a function of how long it took to close them.</figcaption>
</figure>
</div>
<script>
vegaEmbed('#figure-pr-closing-time', "images/pr_times.vl", {"actions": false});
</script>
<h3 id="sec:unit_testing">Unit Testing</h3>
<p>The <code>yt</code> codebase includes a number of unit tests; although extensive, their existence post-dates the initial development of the code, and they largely work around the extant APIs at the time of their creation.
Most modern recommendations for developing scientific software emphasize isolated components, well-structured interfaces, and few side effects.
While the development process attempts to emphasize development of isolated APIs and well-constrained unit tests, the balance struck between enabling contribution from junior developers and ensuring the (subjective) standards of the code base does not always fall on the side of rigid design.</p>
<p>Many of the <code>yt</code> APIs that are tested require the existence of a “dataset.”
For instance, the testing of whether objects are correctly selected by a sphere selector (which absolutely <em>could</em> be tested in isolation, were the APIs more separable) is done via creating several different sets of mock datasets of different organizations and shapes and testing whether or not they correctly choose the data points to be included.
To support these operations, the <code>yt</code> testing utilities provide helper functions for creating mock datasets that have different geometric configurations and different collections of “fields” included in their set of primitive values.
Many of the tests are parameterized against the types and organizations of the datasets, the decomposition across mock processors, and the underlying values of the fields.
This ensures that we check against errors and bugs that may depend on behavior that varies as the number of processors or the organization of the data changes.
One example of this would be in the selection of grid values for a single grid of size <span class="math inline">\(128^3\)</span>.
The values selected in this should match the values selected in the same grid decomposed into eight sets of <span class="math inline">\(64^3\)</span> cells, or 64 sets of <span class="math inline">\(32^3\)</span> cells.</p>
<p>The mechanism by which fields are tested is somewhat more extensive, touching on two different needs.
The first need is that of accuracy – fields with known answers, or fields that can be written to be decomposed into primitive, non-optimized operations, are tested for correctness.
The second need is that of dependency calculation; all fields should have their dependencies correctly detected.
For example, if a dataset has primitive fields for “mass” and “velocity,” the calculation of momentum should require both.
If the dataset includes a “momentum” field, then that should be detected as well.
This dependency calculation enables <code>yt</code> to consolidate IO tasks and read as much data as possible in each pass over the full dataset.
In addition to this, fields are tested to ensure that the values generated for them are independent of the organization of the dataset.
Like in the example above, the “momentum” field for a fixed set of values should be identical regardless of the decomposition of the individual cell elements.</p>
<p>Wherever possible, analytical solutions are preferred.
For processes like surface extraction, this might include ensuring that fixed radii extraction produce the correct spherical region.
For streamlines, it might include computing the analytical solution to an integration along a known vector field.
And for projections, it would mean that integrating the path with a weight of “one” should result in a uniform set of values equal to the path length across the domain.</p>
<p>At present, the unit tests in <code>yt</code> take a considerable amount of time to run, and are using the nosetests framework. Modern Python practice is to use the newer pytest framework, and efforts are underway to port <code>yt</code> to utilize pytest, and in the process, attempt to reduce overall runtime.</p>
<h3 id="sec:answer_testing">Answer Testing</h3>
<p>The most time-consuming part of the testing process is what we refer to as “answer testing.”
Because so much of <code>yt</code> is focused on computing analysis results, and because some of these analysis results simultaneously depend on specific IO routines, selection routines, and many “frontend-specific” pieces of code, we have built a system for ensuring that for a given set of analysis operations, the result of a set of operations does not change beyond a fixed (typically quite small) tolerance.</p>
<p>In general, we allow three different classes of answers, against which we compare results from the current version of the code:</p>
<ol type="1">
<li>Data values which should not ever change unless an explicit decision is made (i.e., raw data values accessed from on-disk fields)</li>
<li>Lightly-processed data values which we do not anticipate any deviation from exact, ordered values (i.e., averages, extrema, etc.)</li>
<li>Higher-level processed values which may differently accumulate error across platforms and architectures but are still within fine-grained (<span class="math inline">\(\equiv 10^{-7}\)</span>) tolerance (i.e., images, pixelized projections, etc.)</li>
</ol>
<p>In the first case, we can utilize hashing functions (such as MD5 and SHA) to guarantee consistency across executions.
Typically, however, we store the full set of values to facilitate easy comparison.
In the latter two cases, we apply unit-aware relative tolerances.
This allows for changes in unit assignment to be isolated from changes in value, and furthermore allows minor migration of values.
A recent incident in which these tests needed to be changed resulted from minor differences as a result of consolidating operations within a loop to conserve memory; the code in question was converted to Cython and the drift was on the scale of <span class="math inline">\(10^{-10}\)</span>.</p>
<p>For small-scale answer tests, results are stored in a separate repository that is cloned as a subrepository of the principle <code>yt</code> repository.
When a new set of answers are needed, they are submitted via pull request, and the changeset hash used for answer validation is updated in the main repository.
This allows a different cadence, and also enables individuals <em>not</em> interested in updating answer values to avoid cloning the subrepository and its full history.
Larger dataset answers are stored in on our continuous integration servers; a YAML file in the main <code>yt</code> repository stores the current version number for those answers, which is incremented (via a pull request) when needed.
Requiring this clear decision-making process allows for both collaborative discussion and community governance over the degree of answer drift allowed.</p>
<h3 id="code-review">Code Review</h3>
<p>Code review in <code>yt</code> is conducted on a line-by-line basis, as well as on a higher-level regarding pull requests.
The workflow for code review roughly follows this outline:</p>
<ol type="1">
<li>A pull request is issued. When a new pull request is issued, a template is provided that includes a description of the change, requesting information about its compliance with coding standards, etc.</li>
<li>The pull request is automatically marked as unmergeable until a team member applies the correct component label.</li>
<li>Code is reviewed, line-by-line, and suggestions are made by humans. Code linting, where specific behaviors are identified (such as inconsistent naming, unused variables, unreachable code sections, etc) is automated.</li>
<li>This process is iterated, ensuring that tests, results accuracy and coding standards are maintained.</li>
</ol>
<p>One increasing issue with the code review process is ensuring that changes are reviewed with appropriate urgency; larger pull requests tend to languish without review, as the requirements for review necessarily add burden to the maintainers.
“Bugfix” changes formally require only one reviewer, whereas the <code>yt</code> guidelines suggest that larger changes require review from two different team members.</p>
<p>One of the most pressing bottlenecks with code review is that the time it takes for tests to pass is much longer than the typical time span during which code review takes place.
Because tests are often required to be run on the <em>current</em> version of the code, not the version of the code against which the pull request has been issued, they are often re-initiated following a merge.
This results in a pull request being merged, and then whatever pull request is next to be reviewed has to wait until the tests (now updated with the newly accepted pull request) pass.
To alleviate this, we have recently begun utilizing the <a href="https://github.blog/changelog/2021-02-04-pull-request-auto-merge-is-now-generally-available/">“auto-merge” feature</a> provided by GitHub.
This allows a maintainer to mark a pull request as “queued” to be merged once a set of requirements – such as tests passing, approval messages, comment resolution and so forth – are met.
By queuing up pull requests for merging, it allows maintainers to mark a set of pull requests as ready to be merged, and then when they meet the appropriate (automated and asynchronous) criteria, they will be merged.</p>
<h3 id="code-styling-and-linting">Code Styling and Linting</h3>
<p>For code included in <code>yt</code>, a set of styles are enforced.
The term “linting” is used to describe applying automated checks to enforce sytlistic consistency, as well as to flag potential errors that can be detected through static analysis of the code.</p>
<p>We rely on the <a href="https://pre-commit.com"><code>pre-commit</code></a> framework, which enables automated checks as well as automatic fixes at commit time.
This tool is an opt-in so not every “drive-by” contributor has to learn and install it, but continuous linting is provided by <a href="https://pre-commit.ci">pre-commit.ci</a> so styling errors cannot slip in.</p>
<p>We configure <code>pre-commit</code> to run a mixture of formatters and static checkers.
The former modify the code in place, while the latter only report errors and so-called “code smells” (such as undefined variables, unused imports, bare <code>except</code> statements…) but require human developers fix them.</p>
<p>Our suite of formatter most prominently includes
<a href="https://black.readthedocs.io/en/stable/"><code>black</code></a>,
<a href="https://pycqa.github.io/isort/"><code>isort</code></a> and
<a href="https://github.com/asottile/pyupgrade"><code>pyupgrade</code></a>.
<code>black</code> has been designed to maximize readability with as few free parameters as possible
(In many ways, the fact that most of the <code>yt</code> code developers did not utilize this style before it was enforced likely enabled its uptake, as it was seen as a choice that “made everyone compromise.”),
while <code>isort</code> ensures that all <code>import</code> statements are sorted (according to alphabetical order within a first/second/third-party categorization), and <code>pyupgrade</code> modernizes some Python idioms according to our minimal support version of the language.
In particular, <code>pyupgrade</code> enforces the usage of modern Python <a href="https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals">“f-strings”</a> since we do not support Python version older than 3.6.</p>
<p>For static code analysis we rely on the <a href="https://flake8.pycqa.org/en/latest/"><code>flake8</code></a> framework.
Effort is underway to enable using <code>mypy</code> for the specialized task of type checking.</p>
<p>All changes that can be applied via automation (specifically, code formatting) are accessible from within the GitHub pull request interface, and are again provided by <a href="https://pre-commit.ci">pre-commit.ci</a>.
This allows drive-by contributions to have their pull requests updated inline by an automated process, reducing the need to manually install packages to apply the changes.</p>
<p>We note that in <span class="citation" data-cites="9uqQt00R">[<a href="#ref-9uqQt00R" role="doc-biblioref">5</a>]</span> evidence is presented that code review bots can lead to a reduction in rejected pull requests, as well as decreased communication among developers.
While yt is not necessarily the perfect laboratory for this, as the project as a whole does not have an extensive history of declining pull requests, we have anecdotally noted that discussion around ‘code nits’ and stylistic issues has been considerably reduced, leading to what developers describe as expedited pull requests.</p>
<h3 id="type-hinting">Type Hinting</h3>
<p>Code included in <code>yt</code> is not required to utilize <a href="https://docs.python.org/3/library/typing.html">type hinting</a>.
However, new contributions are allowed to include type hinting, and work is underway to develop an ontology of types as well as a base set of utilities and types that would be used both internally, and possibly be made available to external package developers.</p>
<h3 id="sec:ytep">YTEP Process</h3>
<p>YTEPs, or “<code>yt</code>-enhancement proposal” are vehicles for collaborative decision-making in the project.
During periods of rapid development, the needs of the community for stability have to be balanced against desires for change; the YTEP process was implemented to facilitate stakeholder feedback, allow for discussion of design decisions, and to prompt detailed thinking about how and why things should be implemented.
We have modeled this process against that used in the AstroPy community (“APE”).
To create a new proposal for a large change to <code>yt</code>, or to document a decision-making process, individuals prepare a description of the background, motivation for the change, the steps to implementation, and potential alternative approaches.
The proposal is discussed through the pull-request process, and once discussion has concluded it is added to the <a href="https://github.com/yt-project/ytep">repository</a> of YTEPs that is auto-built and <a href="https://ytep.readthedocs.org/">deployed</a>.</p>
<p>Implemented shortly after the first paper on <code>yt</code> was released, the YTEP process experienced a fairly pronounced period of usage during the transition between versions 2.0 and 3.0 of <code>yt</code>, and has since been utilized considerably less.
The accepted YTEPs have included implementing the chunking system, developing a units system, removing legacy components, and implementing a code of conduct.
Below, we include a table of current YTEPs as of this writing.</p>
<table>
<colgroup>
<col style="width: 6%" />
<col style="width: 35%" />
<col style="width: 12%" />
<col style="width: 44%" />
</colgroup>
<thead>
<tr class="header">
<th>Number</th>
<th>YTEP Title</th>
<th>Created</th>
<th>Authors</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0001</td>
<td>IO Chunking</td>
<td>November 26, 2012</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0002</td>
<td>Profile Plotter</td>
<td>December 5, 2012</td>
<td>Matthew Turk</td>
</tr>
<tr class="odd">
<td>0003</td>
<td>Standardizing field names</td>
<td>December 11, 2012</td>
<td>Casey Stark, Nathan Goldbaum, Matthew Turk</td>
</tr>
<tr class="even">
<td>0005</td>
<td>Octrees for Fluids and Particles</td>
<td>December 24, 2012</td>
<td>Matthew Turk</td>
</tr>
<tr class="odd">
<td>0006</td>
<td>Periodicity</td>
<td>January 10, 2013</td>
<td>Matthew Turk, Nathan Goldbaum</td>
</tr>
<tr class="even">
<td>0007</td>
<td>Automatic Pull Requests’ validation</td>
<td>February 21, 2013</td>
<td>Kacper Kowalik</td>
</tr>
<tr class="odd">
<td>0008</td>
<td>Release Schedule</td>
<td>February 21, 2013</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0009</td>
<td>AMRKDTree for Data Sources</td>
<td>February 28, 2012</td>
<td>Sam Skillman</td>
</tr>
<tr class="odd">
<td>0010</td>
<td>Refactoring for Volume Rendering and Movie Generation</td>
<td>March 3, 2013</td>
<td>Cameron Hummels</td>
</tr>
<tr class="even">
<td>0011</td>
<td>Symbol units in yt</td>
<td>March 7, 2013</td>
<td>Nathan Goldbaum, Casey Stark, Anna Rosen, Matthew Turk</td>
</tr>
<tr class="odd">
<td>0012</td>
<td>Halo Redesign</td>
<td>March 7, 2013</td>
<td>Britton Smith, Cameron Hummels, Chris Moody, Mark Richardson, Yu Lu</td>
</tr>
<tr class="even">
<td>0013</td>
<td>Deposited Particle Fields</td>
<td>April 25, 2013</td>
<td>Chris Moody, Matthew Turk, Britton Smith, Doug Rudd, Sam Leitner</td>
</tr>
<tr class="odd">
<td>0014</td>
<td>Field Filters</td>
<td>July 2nd, 2013</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0015</td>
<td>Transfer Function Refactor</td>
<td>August 13, 2013</td>
<td>Sam Skillman</td>
</tr>
<tr class="odd">
<td>0016</td>
<td>Volume Traversal</td>
<td>September 10, 2013</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0017</td>
<td>Domain-Specific Output Types</td>
<td>September 18, 2013</td>
<td>Matthew Turk and Anthony Scopatz</td>
</tr>
<tr class="odd">
<td>0018</td>
<td>Changing dict-like access to Static Output</td>
<td>September 18, 2013</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0019</td>
<td>Reduce items in main import</td>
<td>October 2, 2013</td>
<td>Matthew Turk</td>
</tr>
<tr class="odd">
<td>0020</td>
<td>Removing PlotCollection</td>
<td>March 18, 2014</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0021</td>
<td>Particle-Only Plots</td>
<td>August 29, 2014</td>
<td>Andrew Myers</td>
</tr>
<tr class="odd">
<td>0022</td>
<td>Benchmarks</td>
<td>January 19, 2015</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0023</td>
<td>yt Community Code of Conduct</td>
<td>July 11, 2015</td>
<td>Britton Smith</td>
</tr>
<tr class="odd">
<td>0024</td>
<td>Alternative Smoothing Kernels</td>
<td>August 1, 2015</td>
<td>Bili Dong</td>
</tr>
<tr class="even">
<td>0025</td>
<td>The ytdata Frontend</td>
<td>August 31, 2015</td>
<td>Britton Smith</td>
</tr>
<tr class="odd">
<td>0026</td>
<td>NumPy-like Operations</td>
<td>September 21, 2015</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0027</td>
<td>Non-Spatial Data</td>
<td>December 1, 2015</td>
<td>Matthew Turk, Nathan Goldbaum, John ZuHone</td>
</tr>
<tr class="odd">
<td>0028</td>
<td>Alternative Unit Systems</td>
<td>December 8, 2015</td>
<td>John ZuHone, Nathan Goldbaum, Matthew Turk</td>
</tr>
<tr class="even">
<td>0029</td>
<td>Extension Packages</td>
<td>January 25, 2016</td>
<td>Matthew Turk</td>
</tr>
<tr class="odd">
<td>0031</td>
<td>Unstructured Mesh</td>
<td>December 18, 2014</td>
<td>Matthew Turk</td>
</tr>
<tr class="even">
<td>0032</td>
<td>Removing the global octree mesh for particle data</td>
<td>February 9 2017</td>
<td>Nathan Goldbaum, Meagan Lang, Matthew Turk</td>
</tr>
<tr class="odd">
<td>0033</td>
<td>Dropping Python2 Support</td>
<td>November 28, 2017</td>
<td>Nathan Goldbaum</td>
</tr>
<tr class="even">
<td>0034</td>
<td>yt FITS Image Standard</td>
<td>September 9, 2018</td>
<td>John ZuHone</td>
</tr>
<tr class="odd">
<td>0036</td>
<td>Migrating from nose to pytest</td>
<td>September 30, 2019</td>
<td>Jared Coughlin</td>
</tr>
<tr class="even">
<td>0037</td>
<td>Code Styling</td>
<td>May 18, 2020</td>
<td>Clément Robert</td>
</tr>
<tr class="odd">
<td>1000</td>
<td>GitHub Migration</td>
<td>March 25, 2017</td>
<td>Lots of folks</td>
</tr>
<tr class="even">
<td>1776</td>
<td>Team Infrastructure</td>
<td>August 24, 2014</td>
<td>Britton Smith</td>
</tr>
<tr class="odd">
<td>3000</td>
<td>Let’s all start using yt 3.0!</td>
<td>October 30, 2013</td>
<td>Matthew Turk</td>
</tr>
</tbody>
</table>
<h2 id="indexing-and-geometry">Indexing and Geometry</h2>
<p>yt is designed for analysis and visualization of datasets that describe “natural” or “physical” phenomena; more generally, yt is designed to analyze data that can be characterized by a metric of some type.
The most common use case, by far, is that of data that is described in a Cartesian space, by the orthogonal axes of x, y and z.
However, for reasons related to naturalness of coordinate systems and relevance to physical phenomena, datasets are also frequently organized in other coordinate systems, such as cylindrical polar (<span class="math inline">\(r\)</span>, <span class="math inline">\(z\)</span> and <span class="math inline">\(\theta\)</span>), spherical (<span class="math inline">\(r\)</span>, <span class="math inline">\(\theta\)</span> and <span class="math inline">\(\phi\)</span>) and variants such as geographic (latitude, longitude and altitude).</p>
<p>Importantly, however, yt distinguishes between the <em>coordinate</em> space a dataset describes and the natural or <em>index</em> space by which its organization is described.
This distinction is the most relevant among datasets and data formats where the organization is <em>implicit</em>, rather than <em>explicit</em>; for instance, in a grid patch dataset, data variable locations are often only specified implicitly.
For a grid volume that covers a given region, the relationship between the “index” value of a cell (for instance, <span class="math inline">\(i,j,k\)</span>) and its position in space (for instance, <span class="math inline">\(x, y, z\)</span> or <span class="math inline">\(r, \theta, \phi\)</span>) requires transformation between a logically-Cartesian decomposition of the space and the potentially-non Cartesian space that it represents.</p>
<p>In Figure <a href="#fig:coordinate_space">3</a> we demonstrate one possible mapping.
We note that the specific data layout is not optimized for IO throughput, and is unlikely to be exactly replicated in real world formats.
In this case, the data points may be laid out sequentially on disk (or in memory) and a mapping function translates these into position and extent in the coordinate system, here cylindrical coordinates.
For instance, there may be a cell that spans <span class="math inline">\(r\)</span> from <span id="coordinate_info_r" style="font-family: monospace;">0.375 to 0.5</span> and
<span class="math inline">\(\theta\)</span> from <span id="coordinate_info_theta" style="font-family: monospace;">45.0 to 52.5</span>, which is defined by the array values defined in cell <span id="coordinate_info_ij" style="font-family: monospace;">1, 4</span>.</p>
<div id="fig:coordinate_space" class="fignos">
<figure>
<img src="images/indexing/coordinate_space.svg" alt="Figure 3: Index space to coordinate space mapping. On the left is an example of how data points may be laid out on disk and on the right is how these points might be translated into a (cylindrical) coordinate space." />
<figcaption aria-hidden="true"><span>Figure 3:</span> Index space to coordinate space mapping. On the left is an example of how data points may be laid out on disk and on the right is how these points might be translated into a (cylindrical) coordinate space.</figcaption>
</figure>
</div>
<script>
document.addEventListener("SVGLoaded", function(event) { 
  d3.selectAll("svg#index_coord_figure .cell").classed("active");
  d3.selectAll("svg#index_coord_figure .cell")
    .on("mouseover", (event) => {
      dataset = event.target.dataset;
      d3.select("#coordinate_info_r").text(dataset["r-0"] + " to " + dataset["r-1"]);
      d3.select("#coordinate_info_theta").text(dataset["theta-0"] + " to " + dataset["theta-1"]);
      d3.select("#coordinate_info_ij").text(dataset["cellI"] + ", " + dataset["cellJ"]);
      d3.selectAll("svg#index_coord_figure .cell").classed("highlighted", (d, i, n) => (n[i].dataset["cell"] == dataset["cell"]));
    });
});
</script>
<h3 id="sec:abstraction-coordinates">Abstraction of Coordinate Systems</h3>
<p>yt provides a system for defining relationships between index-space and coordinate-space.
During instantiation of a <code>Dataset</code> object, a helper object (<code>coordinates</code>, a subclass of <code>CoordinateHandler</code>) is created.
This helper object tracks the correspondence between numerical axes and spatial axes (for instance, even in some Cartesian datasets, axis 0 corresponds to <span class="math inline">\(z\)</span> rather than <span class="math inline">\(x\)</span>), the names of axes, and the transformation and pixelization methods for visualization.
In addition to these helper functions, the coordinate handler provides definitions for derived fields that describe local cell width (and orthogonal path length), positions in coordinate space as computed by index space coordinates, volumes, and surface areas.
These coordinate handlers also provide transformations between different spaces, albeit using the somewhat undesirable method of conversion to reference cartesian frames and subsequent conversion to local coordinate frames.</p>
<p>At present, coordinate spaces are defined in the spaces enumerated in Table <a href="#tbl:coord-systems">1</a>.
While these are representative of the most common spatial representations, additional representations (such as those that include a non-trivial mapping between coordinates and index values) are possible to implement.</p>
<div id="tbl:coord-systems" class="tablenos">
<table id="tbl:coord-systems">
<caption><span>Table 1:</span> Extant coordinate systems; in all cases, value ranges should be taken to describe extent rather than specific boundary points. </caption>
<colgroup>
<col style="width: 35%" />
<col style="width: 64%" />
</colgroup>
<thead>
<tr class="header">
<th>Coordinate system</th>
<th>Axes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Cartesian coordinates</td>
<td><span class="math inline">\(x, y, z\)</span></td>
</tr>
<tr class="even">
<td>Cylindrical polar coordinates</td>
<td><span class="math inline">\(r, \theta \in [0, 2 \pi], z\)</span></td>
</tr>
<tr class="odd">
<td>Spherical coordinates</td>
<td><span class="math inline">\(r, \theta, \phi\)</span></td>
</tr>
<tr class="even">
<td>Geographic coordinates</td>
<td>latitude <span class="math inline">\(\in [0, 180]\)</span>, longitude <span class="math inline">\(\in [0, 360]\)</span>, altitude</td>
</tr>
<tr class="odd">
<td>Internal geographic coordinates</td>
<td>latitude, longitude, depth</td>
</tr>
<tr class="even">
<td>Spectral cube</td>
<td>Image <span class="math inline">\(x\)</span>, Image <span class="math inline">\(y\)</span> and <span class="math inline">\(\nu\)</span></td>
</tr>
</tbody>
</table>
</div>
<p>Future developments may involve code generation for arbitrary coordinate systems, using SymPy or other libraries.
Independent of the visualization methods (which can often be reused), the development of coordinate systems is largely rote, applying straightforward mathematics to construct derived field definitions.
As such, using mechanisms in SymPy for construction of relationships between coordinate systems may be a feasible method of developing code-generation for coordinate system handlers in <code>yt</code>.</p>
<h2 id="sec:data_objects">Data Objects</h2>
<p>The basic principles by which <code>yt</code> operates are built on the notion of selecting data (through coarse and subsequent fine-grained indexing of data sources such as files), accessing that data in a memory-efficient fashion, and then processing that data into either a resultant set of quantitative data or a visualization.</p>
<p>Selections in <code>yt</code> are usually spatial in nature, although several non-spatial mechanisms focused on queries can be utilized as well.
These objects which conduct selection are selectors, and are designed to provide as small of an API as possible, to enable ease of development and deployment of new selectors.</p>
<p>Selectors require defining several functions, with the option of defining additional functions for optimization, that return true or false whether a given point is or is not included in the selected region.
These functions include selection of a rectilinear grid (or any point within that grid), selection of a point with zero extent and selection of a point with a non-zero spherical radius.</p>
<p>The base selector object utilizes these routines during a selection operation to maximize the amount of code reused between particle, patch, and octree selection of data.
These three types of data are selected through specific routines designed to minimize the number of times that the selection function must be called, as they can be quite expensive.</p>
<p>Selecting data from a grid is a two-step process.
The first step is identifying which grids intersect a given data selector; this is done through a sequence of bounding box intersection checks.
Within a given grid, the cells which are intersected are identified.
This results in the selection routine being called once for each grid object in the simulation and once for each cell located within an intersecting grid.
This can be conducted hierarchically, but due to implementation details around how the grid index is stored this is not yet cost effective.</p>
<p>Selecting data from an octree-organized dataset utilizes a recursive scheme that selects individual oct nodes, then for each cell within that oct, determining which cells must be selected or child nodes recursed into.
This system is designed to allow for having leaf nodes of varying cells-per-side, for instance 1, 2, 4, 8, etc.
However, the number of nodes is fixed at 8, with subdivision always occurring at the midplane.</p>
<p>The final mechanism by which data is selected is for discrete data points, typically particles in astrophysical simulations.
At present, this is done by first identifying which data files intersect with a given selector, then selecting individual points.
There is no hierarchical data selection conducted in this system, as we do not yet allow for re-ordering of data on disk or in-memory which would facilitate hierarchical selection through the use of operations such as Morton indices.</p>
<h3 id="selection-routines">Selection Routines</h3>
<p>Given these set of hierarchical selection methods, all of which are designed to provide opportunities for early-termination, each <em>geometric</em> selector object is required to implement a small set of methods to expose its functionality to the hierarchical selection process.
Duplicative functions often result from attempts to avoid expensive calculations that take into account boundary conditions such as periodicity and reflectivity unless necessary.
Additionally, by providing some routines as options, we can in some instances specialize them for the specific geometric operation.</p>
<ul>
<li><code>select_cell(cell_center, cell_width)</code>: this function, which is somewhat degenerate with <code>select_bbox</code>, returns whether a given “cell,” defined by its center and its width along each dimension, is included within the selection. In situations where the cells are spaced logarithmically, rather than linearly, this may produce slightly reduced accuracy for near-misses and glancing-selections.</li>
<li><code>select_point(position)</code>: this function returns whether or not a point of zero-extent is included within the selection. This has some degeneracy with <code>select_sphere</code>.</li>
<li><code>select_sphere(position, radius)</code>: This is equivalent to the <code>select_point</code> function, except that any point within the specified radius is included within the selector object.</li>
<li><code>select_bbox(lower_left, upper_right)</code>: Determine overlap with an axis-aligned bounding box. Particularly for hierarchical selection methods, determining whether or not a bounding box overlaps with a geometric selector can lead to early-termination of some selection operations.<br />
</li>
<li><code>select_bbox_edge(lower_left, upper_right)</code>: This is a special-case of the bounding box routine that provides information as to whether or not the <em>entire</em> bounding box is included or just a <em>partial</em> portion of the bounding box.</li>
</ul>
<p>We demonstrate a handful of selection operations on a low-resolution dataset below.
In each of these figures, the rectangular regions outlined in gray and black indicate individual grid patches, where data may vary in resolution and cell size.
In Figure <a href="#fig:reg2">4</a> we illustrate the selection of a rectangular prism (i.e., a <code>region</code>, like in Section <a href="#sec:dobj-region">1.6.2.17</a>.
In Figure <a href="#fig:sp2">5</a>, we illustrate the selection of a sphere (i.e., a <code>sphere</code>, like in Section <a href="#sec:dobj-sphere">1.6.2.20</a>.
And, to demonstrate yt’s ability to construct boolean selectors from these objects (i.e., Section <a href="#sec:dobj-bool">1.6.2.2</a> we show what the logical <code>NOT</code> of these two objects would produce in <a href="#fig:reg2_not_sp2">6</a>.
We note in particular that while these regions are constructed from geometric selections, the data points are selected by the inclusion of their center points, leading to slightly irregular edges; this is by design.</p>
<div id="fig:reg2" class="fignos">
<figure>
<img src="images/selectors/reg2.svg" alt="Figure 4: A selection of data in a low-resolution simulation from a rectangular prism." />
<figcaption aria-hidden="true"><span>Figure 4:</span> A selection of data in a low-resolution simulation from a rectangular prism.</figcaption>
</figure>
</div>
<div id="fig:sp2" class="fignos">
<figure>
<img src="images/selectors/sp2.svg" alt="Figure 5: A selection of data in a low-resolution simulation from a sphere." />
<figcaption aria-hidden="true"><span>Figure 5:</span> A selection of data in a low-resolution simulation from a sphere.</figcaption>
</figure>
</div>
<div id="fig:reg2_not_sp2" class="fignos">
<figure>
<img src="images/selectors/reg2_not_sp2.svg" alt="Figure 6: The logical A AND NOT B for regions A and B from Figures 4 and 5 respectively." />
<figcaption aria-hidden="true"><span>Figure 6:</span> The logical <code>A AND NOT B</code> for regions <code>A</code> and <code>B</code> from Figures <a href="#fig:reg2">4</a> and <a href="#fig:sp2">5</a> respectively.</figcaption>
</figure>
</div>
<h3 id="fast-and-slow-paths">Fast and Slow Paths</h3>
<p>Given an ensemble of objects, the simplest way of testing for inclusion in a selector is to call the operation <code>select_cell</code> on each individual object.
Where the objects are organized in a regular fashion, for instance a “grid” that contains many “cells,” we can apply both “first pass” and “second pass” fast-path operations.
The “first pass” checks whether or not the given ensemble of objects is included, and only iterates inward if there is partial or total inclusion.
The “second pass” fast pass is specialized to both the organization of the objects <em>and</em> the selector itself, and is used to determine whether either only a specific (and well-defined) subset of the objects is included or the entirety of them.</p>
<p>For instance, we can examine the specific case of selecting grid cells within a rectangular prism.
When we select a “grid” of cells within a rectangular prism, we can have either total inclusion, partial inclusion, or full exclusion.
In the case of full inclusion, where the entire grid is included within the selector, we simply sidestep the specific inclusion checks completely and return a full mask of cells to utilize.
In the case of partial inclusion, we can often determine the “start” and “end” indices of inclusion in the rectangular prism by examining the intersection volume.
This allows us to avoid many costly individual <code>select_cell</code> calls.</p>
<p>With discrete point selection (and for our purposes, often unstructured mesh falls into this category) we often do not have the same organizing principle on which we can rely.
However, utilizing hierarchical bitmap indexing we can often organize subsets of particles into collections of cells which may or may not be contiguous.
In this situation, we can check for full inclusion within data objects, although we are not able to identify start and stop indices as the data are not assumed to be organized spatially independent of how we have indexed them.</p>
<p>At present, the objects listed in Table <a href="#tbl:selection-objects">2</a> are provided as selectors in yt.
We do make a distinction between “selection” operations and “reduction” or “construction” operations (such as projections and smoothing/resampling), but have included both here for consistency.
Additionally, some have been marked as not “user-facing,” in the sense that they are not expected to be constructed directly by users, but instead are utilized internally for indexing purposes.
In columns to the right, we provide information as to whether there is an available “fast” path for grid objects.</p>
<div id="tbl:selection-objects" class="tablenos">
<table id="tbl:selection-objects">
<caption><span>Table 2:</span> Selection objects and their types. </caption>
<thead>
<tr class="header">
<th>Object Name</th>
<th>Object Type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Arbitrary grid</td>
<td>Resampling</td>
</tr>
<tr class="even">
<td>Boolean object</td>
<td>Selection (Base Class)</td>
</tr>
<tr class="odd">
<td>Covering grid</td>
<td>Resampling</td>
</tr>
<tr class="even">
<td>Cut region</td>
<td>Selection</td>
</tr>
<tr class="odd">
<td>Cutting plane</td>
<td>Selection</td>
</tr>
<tr class="even">
<td>Data collection</td>
<td>Selection</td>
</tr>
<tr class="odd">
<td>Disk</td>
<td>Selection</td>
</tr>
<tr class="even">
<td>Ellipsoid</td>
<td>Selection</td>
</tr>
<tr class="odd">
<td>Intersection</td>
<td>Selection (Bool)</td>
</tr>
<tr class="even">
<td>Octree</td>
<td>Internal index</td>
</tr>
<tr class="odd">
<td>Orthogonal ray</td>
<td>Selection</td>
</tr>
<tr class="even">
<td>Particle projection</td>
<td>Reduction</td>
</tr>
<tr class="odd">
<td>Point</td>
<td>Selection</td>
</tr>
<tr class="even">
<td>Quadtree projection</td>
<td>Reduction</td>
</tr>
<tr class="odd">
<td>Ray</td>
<td>Selection</td>
</tr>
<tr class="even">
<td>Rectangular Prism</td>
<td>Selection</td>
</tr>
<tr class="odd">
<td>Slice</td>
<td>Selection</td>
</tr>
<tr class="even">
<td>Smoothed covering grid</td>
<td>Resampling</td>
</tr>
<tr class="odd">
<td>Sphere</td>
<td>Selection</td>
</tr>
<tr class="even">
<td>Streamline</td>
<td>Selection</td>
</tr>
<tr class="odd">
<td>Surface</td>
<td>Selection</td>
</tr>
<tr class="even">
<td>Union</td>
<td>Selection (Bool)</td>
</tr>
</tbody>
</table>
</div>
<h4 id="sec:dobj-arbitrary_grid">Arbitrary grid</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Left edge</li>
<li>Right edge</li>
<li>Active Dimensions</li>
</ul>
<p>A 3D region with arbitrary bounds and dimensions. In contrast to the
Covering Grid, this object accepts a left edge, a right edge, and
dimensions. This allows it to be used for creating 3D particle
deposition fields that are independent of the underlying mesh, whether
that is yt-generated or from the simulation data. For example,
arbitrary boxes around particles can be drawn and particle deposition
fields can be created. This object will refuse to generate any fluid
fields.</p>
<h4 id="sec:dobj-bool">Bool</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Operation</li>
<li>Data object 1</li>
<li>Data object 2</li>
</ul>
<p>This is a boolean operation, accepting AND, OR, XOR, and NOT for
combining multiple data objects. This object is not designed to be
created directly; it is designed to be created implicitly by using one
of the bitwise operations (&amp;, |, ^, ~) on one or two other data
objects. These correspond to the appropriate boolean operations, and
the resultant object can be nested.</p>
<h4 id="sec:dobj-covering_grid">Covering grid</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Level</li>
<li>Left edge</li>
<li>Active Dimensions</li>
</ul>
<p>A 3D region with all data extracted to a single, specified resolution.
Left edge should align with a cell boundary, but defaults to the
closest cell boundary.</p>
<h4 id="sec:dobj-cut_region">Cut region</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Base object</li>
<li>Conditionals</li>
</ul>
<p>This is a data object designed to allow individuals to apply logical
operations to fields and filter as a result of those cuts.</p>
<h4 id="sec:dobj-cutting">Cutting</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Normal</li>
<li>Center</li>
</ul>
<p>This is a data object corresponding to an oblique slice through the
simulation domain. This object is typically accessed through the
<code>cutting</code> object that hangs off of index objects. A cutting plane is
an oblique plane through the data, defined by a normal vector and a
coordinate. It attempts to guess an ‘north’ vector, which can be
overridden, and then it pixelizes the appropriate data onto the plane
without interpolation.</p>
<h4 id="sec:dobj-data_collection">Data collection</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Object List</li>
</ul>
<p>By selecting an arbitrary <em>object_list</em>, we can act on those grids.
Child cells are not returned.</p>
<h4 id="sec:dobj-disk">Disk</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Center</li>
<li>Normal vector</li>
<li>Radius</li>
<li>Height</li>
</ul>
<p>By providing a <em>center</em>, a <em>normal</em>, a <em>radius</em> and a <em>height</em> we can
define a cylinder of any proportion. Only cells whose centers are
within the cylinder will be selected.</p>
<h4 id="sec:dobj-ellipsoid">Ellipsoid</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Center</li>
<li>a</li>
<li>b</li>
<li>c</li>
<li>e0</li>
<li>tilt</li>
</ul>
<p>By providing a <em>center</em>,<em>A</em>,<em>B</em>,<em>C</em>,<em>e0</em>,<em>tilt</em> we can define a
ellipsoid of any proportion. Only cells whose centers are within the
ellipsoid will be selected.</p>
<h4 id="sec:dobj-intersection">Intersection</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Data objects</li>
</ul>
<p>This is a more efficient method of selecting the intersection of
multiple data selection objects. Creating one of these objects
returns the intersection of all of the sub-objects; it is designed to
be a faster method than chaining &amp; (“and”) operations to create a
single, large intersection.</p>
<h4 id="sec:dobj-minimal_sphere">Minimal sphere</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Center</li>
<li>Radius</li>
</ul>
<p>Build the smallest sphere that encompasses a set of points.</p>
<h4 id="sec:dobj-octree">Octree</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Left edge</li>
<li>Right edge</li>
<li>Particle count refinement criteria</li>
</ul>
<p>A 3D region with all the data filled into an octree. This container
will mean deposit particle fields onto octs using a kernel and SPH
smoothing.</p>
<h4 id="sec:dobj-ortho_ray">Ortho ray</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Axis</li>
<li>Coords</li>
</ul>
<p>This is an orthogonal ray cast through the entire domain, at a
specific coordinate. This object is typically accessed through the
<code>ortho_ray</code> object that hangs off of index objects. The resulting
arrays have their dimensionality reduced to one, and an ordered list
of points at an (x,y) tuple along <code>axis</code> are available.</p>
<h4 id="sec:dobj-particle_proj">Particle proj</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Axis</li>
<li>Field</li>
<li>Weight field</li>
</ul>
<p>A projection operation optimized for SPH particle data.</p>
<h4 id="sec:dobj-point">Point</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>P</li>
</ul>
<p>A 0-dimensional object defined by a single point</p>
<h4 id="sec:dobj-quad_proj">Quad proj</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Axis</li>
<li>Field</li>
<li>Weight field</li>
</ul>
<p>This is a data object corresponding to a line integral through the
simulation domain. This object is typically accessed through the
<code>proj</code> object that hangs off of index objects. YTQuadTreeProj is a
projection of a <code>field</code> along an <code>axis</code>. The field can have an
associated <code>weight_field</code>, in which case the values are multiplied by
a weight before being summed, and then divided by the sum of that
weight; the two fundamental modes of operating are direct line
integral (no weighting) and average along a line of sight (weighting.)
What makes <code>proj</code> different from the standard projection mechanism is
that it utilizes a quadtree data structure, rather than the old
mechanism for projections. It will not run in parallel, but serial
runs should be substantially faster. Note also that lines of sight
are integrated at every projected finest-level cell.</p>
<h4 id="sec:dobj-ray">Ray</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Start point</li>
<li>End point</li>
</ul>
<p>This is an arbitrarily-aligned ray cast through the entire domain, at
a specific coordinate. This object is typically accessed through the
<code>ray</code> object that hangs off of index objects. The resulting arrays
have their dimensionality reduced to one, and an ordered list of
points at an (x,y) tuple along <code>axis</code> are available, as is the <code>t</code>
field, which corresponds to a unitless measurement along the ray from
start to end.</p>
<h4 id="sec:dobj-region">Region</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Center</li>
<li>Left edge</li>
<li>Right edge</li>
</ul>
<p>A 3D region of data with an arbitrary center. Takes an array of three
<em>left_edge</em> coordinates, three <em>right_edge</em> coordinates, and a
<em>center</em> that can be anywhere in the domain. If the selected region
extends past the edges of the domain, no data will be found there,
though the object’s <code>left_edge</code> or <code>right_edge</code> are not modified.</p>
<h4 id="sec:dobj-slice">Slice</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Axis</li>
<li>Coord</li>
</ul>
<p>This is a data object corresponding to a slice through the simulation
domain. This object is typically accessed through the <code>slice</code> object
that hangs off of index objects. Slice is an orthogonal slice through
the data, taking all the points at the finest resolution available and
then indexing them. It is more appropriately thought of as a slice
‘operator’ than an object, however, as its field and coordinate can
both change.</p>
<h4 id="sec:dobj-smoothed_covering_grid">Smoothed covering grid</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Level</li>
<li>Left edge</li>
<li>Active Dimensions</li>
</ul>
<p>A 3D region with all data extracted and interpolated to a single,
specified resolution. (Identical to covering_grid, except that it
interpolates.) Smoothed covering grids start at level 0,
interpolating to fill the region to level 1, replacing any cells
actually covered by level 1 data, and then recursively repeating this
process until it reaches the specified <code>level</code>.</p>
<h4 id="sec:dobj-sphere">Sphere</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Center</li>
<li>Radius</li>
</ul>
<p>A sphere of points defined by a <em>center</em> and a <em>radius</em>.</p>
<h4 id="sec:dobj-streamline">Streamline</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Positions</li>
</ul>
<p>This is a streamline, which is a set of points defined as being
parallel to some vector field. This object is typically accessed
through the Streamlines.path function. The resulting arrays have
their dimensionality reduced to one, and an ordered list of points at
an (x,y) tuple along <code>axis</code> are available, as is the <code>t</code> field, which
corresponds to a unitless measurement along the ray from start to end.</p>
<h4 id="sec:dobj-surface">Surface</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Data source</li>
<li>Surface field</li>
<li>Field value</li>
</ul>
<p>This surface object identifies isocontours on a cell-by-cell basis,
with no consideration of global connectedness, and returns the
vertices of the Triangles in that isocontour. This object simply
returns the vertices of all the triangles calculated by the <a href="https://en.wikipedia.org/wiki/Marching_cubes">marching
cubes</a> algorithm; for
more complex operations, such as identifying connected sets of cells
above a given threshold, see the extract_connected_sets function.
This is more useful for calculating, for instance, total isocontour
area, or visualizing in an external program (such as
<a href="http://www.meshlab.net">MeshLab</a>.) The object has the properties .vertices
and will sample values if a field is requested. The values are
interpolated to the center of a given face.</p>
<h4 id="sec:dobj-union">Union</h4>
<p><em>Arguments</em>:</p>
<ul>
<li>Data objects</li>
</ul>
<p>This is a more efficient method of selecting the union of multiple
data selection objects. Creating one of these objects returns the
union of all of the sub-objects; it is designed to be a faster method
than chaining | (or) operations to create a single, large union.</p>
<h2 id="processing-and-analysis-of-data">Processing and Analysis of Data</h2>
<p><code>yt</code> provides several interfaces for accessing the data available in a given dataset.
As described in <a href="#sec:data_objects">1.6</a>, the primary means of accessing data is through “data objects” that apply selections to the dataset.
These objects present dictionary-like interfaces that return data; below, we describe what options are available for the data that is returned (<a href="#sec:field_system">1.7.1</a>), as well as high-level interfaces for applying aggregations and reductions (<a href="#sec:arraylike_operations">1.7.2</a>).</p>
<h3 id="sec:field_system">Field System</h3>
<p>In <code>yt</code>, there are three types of “fields” that define values at a given spatial location.
The first of these is an “on-disk” field, representing the raw, unmodified (except potentially up-cast to 64 bit precision) values read from the data storage that defines the dataset, such as files or bucket storage; while <code>yt</code> does provide routines for reading these fields, they are passed largely unmodified and so we do not discuss them in depth.
The second type of field is a “derived field,” which is a functional definition of how to process or combine one or more fields that exist in the dataset.
Finally, providing the closure necessary for these derived fields to be accessed independently of their naming convention are “alias fields” that provide mappings between platform- or format-specific names for fields and those used internally in <code>yt</code>.</p>
<p>Fields are also defined by their “sampling type” to distinguish between those fields defined in a volume-filling fashion (i.e., cell-based fields) and those that are defined by discrete samples that may or may not require closure or convolution functions to be applied.
Fields that are defined as a collection of discrete samples can be combined or filtered differently than those that are defined in a volume-filling manner, as described in <a href="#sec:particle_filters">1.7.1.3</a> and <a href="#sec:particle_unions">1.7.1.4</a>.</p>
<h4 id="sec:field_aliases">Field Aliases</h4>
<p>Small differences in naming fields can prove disproportionately challenging for writing platform-neutral analysis code.
For instance, if one platform names the “density” field <code>dens</code> and another refers to it as <code>Density</code> (or, as we have seen in one platform, even the unicode character for <span class="math inline">\(\rho\)</span>) then any platform-independent derived field that utilizes density must be defined multiple times to refer to this fundamentally identical quantity.
(An important note here is that in many cases, the reverse problem is true – some codes may refer to things with the same name but with different underlying definitions, which provides an additional challenge to the analysis process by requiring <a href="https://www.youtube.com/watch?v=WZLkcFns8Ks">disambiguation</a>.)</p>
<p>To address this issue, <code>yt</code> defines a set of fundamental fields, along with a naming convention for extensibility, that are provided as “aliases” for the dataset-specific field names.
This enables a consistent ontology to be defined for fields in <code>yt</code>, upon which the remainder of derived fields can rely.
Typically these are defined by the authors of a given dataset format frontend, wherein a translation or lookup table is provided to match the on-disk fields to those expected by <code>yt</code>.</p>
<p>In some cases, it is through a combination of derived and aliased fields that the full set of data is made available to the researcher; for instance, some datasets do not store velocity as a quantity on disk, but instead store momentum.
In this case, momentum is aliased from the on-disk field to the <code>yt</code> field, and then a derived field is generated to seamlessly provide access to the velocity field wherever it is needed.</p>
<h4 id="sec:derived_fields">Derived Fields</h4>
<p>In addition to the fields that are defined in the dataset, <code>yt</code> recognizes that there exist essentially infinite fields <em>in potentia</em> that can be defined.
For instance, commonly in astrophysics datasets the “density” of different elemental abundances are stored (which provides a natural conservation scheme with the density) in the dataset.
A simple derived field might be defined to provide the “fraction” of a given field:
<span class="math display">\[ f_{X} \equiv \frac{\rho_{X}}{\rho} \]</span>
<code>yt</code> provides the ability to define this as a derived field in a functional form.
For instance, if the density of helium is stored as the field-tuple <code>("gas", "helium_density")</code> we can define the function as:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> _helium_fraction(field, data):</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> data[<span class="st">&quot;gas&quot;</span>, <span class="st">&quot;helium_density&quot;</span>] <span class="op">/</span> data[<span class="st">&quot;gas&quot;</span>, <span class="st">&quot;density&quot;</span>]</span></code></pre></div>
<p>Note that here, the argument <code>field</code> is a field definition object and <code>data</code> is a data object which we are using for our selection.
This is the form that derived fields in <code>yt</code> take; these can be supplied to the function <code>add_field</code> (or they can use <code>derived_field</code> as a decorator) and they will become available for all data objects.</p>
<p>These fields can accept parameters (associated with the base data object used for selection) and can require that spatial information is made available to the derived field; this can enable the calculation of finite-difference stencils for operations such as averaging and operators such as the gradient.</p>
<p>Derived fields are an extremely integral component of <code>yt</code> and are the gateway to enabling low-memory overhead calculations and sharing of analysis code.
In addition, <code>yt</code> includes a large number of fields available, many of which are dynamically constructed according to metadata available in the dataset, to jump-start analysis.
Researchers using <code>yt</code> can load a dataset and immediately compute, for instance, the velocity divergence and <code>yt</code> will construct the appropriate finite different stencil, fill in any missing zones at the edge of individual boundaries, and return an array that can be accessed, visualized or processed.</p>
<h4 id="sec:particle_filters">Particle Filters</h4>
<p>Many of the data formats that <code>yt</code> accepts define particles as mixtures of a single set of attributes (such as position, velocity, etc) and then a “type” – for instance, intermingling dark matter particles with “star” particles.
Where simulations are concerned, this can produce much more efficient code; since particles are typically evolved in the same fashion, storing them adjacent in memory can speedup operations such as time evolution steps.
However, when reading the data in, they often need to be handled in fundamentally different ways.
The analysis of dark matter particles in a galaxy, for instance, needs to be conducted differently than the analysis of collisional particles, or particles that arise from other phenomena (such as gas).
<code>yt</code> provides a method for creating new “particle types” on the fly and applying existing derived fields to them.
By adding a new “filter” method, particles that meet this criteria (“high-mass Black Holes,” for instance, or “star clusters more than 1 billion years old”) are accessible in a new field tuple.
This enables all existing memory-conservative operations to act on them.</p>
<p>This filter, for example, checks and returns only those particles whose field <code>particle_type</code> is set to a value of 2.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="at">@yt.particle_filter</span>(requires<span class="op">=</span>[<span class="st">&quot;particle_type&quot;</span>], filtered_type<span class="op">=</span><span class="st">&quot;all&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> stars(pfilter, data):</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">filter</span> <span class="op">=</span> data[(pfilter.filtered_type, <span class="st">&quot;particle_type&quot;</span>)] <span class="op">==</span> <span class="dv">2</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">filter</span></span></code></pre></div>
<p>In this case, <code>yt</code> also infers the name of the newly filtered type from the name of the function, and they become stars.
Now all existing operations will work on field-tuples beginning with <code>"stars"</code> as their field type.</p>
<h4 id="sec:particle_unions">Particle Unions</h4>
<p>The opposite operation to that in <a href="#sec:particle_filters">1.7.1.3</a> is also accessible, by which multiple particle types can be combined and viewed as a single logical type.
For instance, if “star particles” and “black hole” particles are distinct in a simulation of galaxy formation, they can be combined into a logical union:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>u <span class="op">=</span> ParticleUnion(<span class="st">&quot;massive_objects&quot;</span>, [<span class="st">&quot;bh&quot;</span>, <span class="st">&quot;stars&quot;</span>])</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>ds.add_particle_union(u)</span></code></pre></div>
<p>Since unions are restricted to combinations <em>in full</em> of different types, their creation requires only specification of the particle types to combine.
The set of available fields is the intersection of the fields available for <em>all</em> the combined types.
If both particle types share fields A and B but only one shares C, the union will only have fields A and B accessible to it.</p>
<h4 id="sec:field_detection">Field Detection</h4>
<p><code>yt</code> determines at dataset instantiation time the fields that are available to be computed.
This provides the ability for researchers to <em>query</em> what fields are available, and additionally as a side-effect it provides information to the <code>yt</code> IO routines which fields need to be computed for a given derived field.
By utilizing this information, <code>yt</code> can “resolve” all required fields when a derived field is requested.
As such, it is able to identify that <code>("gas", "velocity_divergence")</code> relies on the velocity fields along each axis.
If these are the fields that exist in the dataset, the resolution process concludes here.
If, however, they need to be computed from the momentum and density fields, those become the fields that are read from the dataset.</p>
<p>This resolution of field dependencies enables <code>yt</code> to read only the fields that are necessary <em>and</em> to do so in a single pass over a file, reducing the initialization and seeking time within a file.
Particularly in environments where metadata operations (required for an <code>open</code> system call) or seek operations (where dataset chunks may need to be looked up within a file as indexed by a header) are expensive, this can have significant impact on the overall performance, and by operating on a chunk-by-chunk basis, it further reduces the need to store multiple fields in memory simultaneously.</p>
<p>This computation does, however, come with an overhead.
Detecting the fields that are required (and thus determining which fields are available) can be expensive, as many small sympy objects are created in the unit handling subsystem and many redundant calculations performed in the <code>yt</code>-specific field resolution code.
This is an area of great interest for future optimizations, as the current situation benefits the access of large derived fields over iteration over many small datasets.
In particular, an enormous amount of time in the unit testing framework is spent detecting fields for datasets that are only used once and then discarded.</p>
<h3 id="sec:arraylike_operations">Array-like Operations</h3>
<p>In <code>yt</code>, a newly-constructed data selector contains no data – this enables data selectors for large regions, in extremely large datasets, to be lightweight and cheap to construct.
By ensuring that these objects don’t immediately consume resources, they can be manipulated and operated on in a high-level fashion, without taxing the computational power.
While these data objects <em>can</em> return the full set of data they include, <code>yt</code> also provides array-like operations that do not require immediate access to the full set of numerical values, and which align with the mental-model for data processing that <code>yt</code> exposes.
As an example, consider the following two operations:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>dd <span class="op">=</span> ds.all_data()</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>dd[<span class="st">&quot;gas&quot;</span>, <span class="st">&quot;density&quot;</span>].<span class="bu">max</span>()</span></code></pre></div>
<p>and</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>dd <span class="op">=</span> ds.all_data()</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>dd.<span class="bu">max</span>((<span class="st">&quot;gas&quot;</span>, <span class="st">&quot;density&quot;</span>))</span></code></pre></div>
<p>Both are available in <code>yt</code>.
As a side-effect of Python’s object model, the first will access the <code>("gas", "density")</code> item in the object <code>dd</code>, itself a concatenated numpy array, and then execute the <code>max</code> method on it.
The second will call the <code>max</code> method on the data object, supplying to it the name of the field.
This allows <code>yt</code> to decide how to decompose, parallelize and process the data in a memory-efficient way, and spread across multiple processors.
Additionally, by emphasizing that the “maximum” is being taken on the data object, rather than the numerical data, other operations can be exposed that build on the underlying data organization.
For instance, taking the maximum along a given (spatial) axis:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sp <span class="op">=</span> ds.sphere(<span class="st">&quot;center&quot;</span>, (<span class="fl">10.0</span>, <span class="st">&quot;m&quot;</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>sp.<span class="bu">max</span>((<span class="st">&quot;gas&quot;</span>, <span class="st">&quot;temperature&quot;</span>), axis<span class="op">=</span><span class="st">&quot;z&quot;</span>)</span></code></pre></div>
<p>This translates our meaning – find the maximum value along the z-axis – into a dimensionality reduction operation that uses <code>yt</code>’s built-in “projection” method.
These operations, on data objects (rather than the underlying arrays of values that are accessible through them) provide dataframe-like methods for querying very large, spatially registered data.</p>
<p>The array-like operations utilized in <code>yt</code> attempt to map to conceptually similar operations in numpy.
Unlike numpy, however, these utilize <code>yt</code>’s dataset-aware “chunking” operations, in a manner philosophically similar to the chunking operations used in the parallel computation library dask.
Below, we outline the three classes of operations that are available, based on the type of their return value.</p>
<h4 id="sec:arrayops-scalar">Reduction to Scalars</h4>
<p>Traditional array operations that map from an array to a scalar are accessible utilizing familiar syntax. These include:</p>
<ul>
<li><code>min(field_specification)</code>, <code>max(field_specification)</code>, and <code>ptp(field_specification)</code></li>
<li><code>argmin(field_specification, axis)</code>, and <code>argmax(field_specification, axis)</code></li>
<li><code>mean(field_specification, weight)</code>, <code>std(field_specification, weight)</code>, and <code>sum(field_specification)</code></li>
</ul>
<p>In addition to the advantages of allowing the parallelism and memory management be handled by <code>yt</code>, these operations are also able to accept multiple fields.
This allows multiple fields to be queried in a single pass over the data, rather than multiple passes.
Additionally, the <code>min</code> and <code>max</code> operations will automatically cache the results during a single pass, which means that calling <code>max</code> immediately after <code>min</code> (and vice versa) on the same data object and field will not require a recomputation.</p>
<p>In the case of <code>argmin</code> and <code>argmax</code>, the default returned “axis” will be the spatial coordinates of the minimum or maximum field value (respectively).<br />
However, by specifying an axis or set of axes that correspond to fields, the field values will be queried at these minimum or maximum points.
This allows, for instance, to query the value of “density” at the minimum “temperature.”
The operations <code>mean</code> and <code>sum</code> are available here in a non-spatial form, where they simply compute the scalar reduction independent of the spatial registration of the dataset.</p>
<h4 id="sec:arrayops-vector">Reduction to Vectors</h4>
<ul>
<li><code>profile(axes, fields, profile_specification)</code></li>
</ul>
<p>The <code>profile</code> operation provides weighted or unweighted histogramming in one or two dimensions.
This function accepts the axes along which to compute the histogram as well as the fields to compute, and information about whether the binning should be an accumulation, an average, or a weighted average.
These operations are described in more detail in <strong>reference profile section</strong>.</p>
<h4 id="sec:arrayops-remap">Remapping Operations</h4>
<ul>
<li><code>mean(field_specification, weight, axis)</code></li>
<li><code>sum(field_specification, axis)</code></li>
<li><code>integrate(field_specification, weight, axis)</code></li>
</ul>
<p>These functions map directly to different methods used by the projection data object.
Both <code>mean</code> and <code>sum</code>, when supplied a spatial axis, will compute a dimensionally-reduced projection, remapped into a pixel coordinate plane.
Importantly, if the dataset is a finite-volume dataset (grid, octree, etc), the results of these operations will be a variable-resolution mesh, rather than a fixed resolution image buffer.</p>
<h2 id="abstracting-simulation-types">Abstracting Simulation Types</h2>
<h3 id="sec:chunking">Chunking and Decomposition Strategies</h3>
<p>Reading data, particularly data that will not be utilized in a computation, can incur substantial overhead, particularly if the data is spread over multiple files on a networked filesystem, where metadata queries can dominate the cost of IO.
<code>yt</code> takes the approach of building a coarse-grained index based on the discretization method of the data (particle, grid, octree, unstructured mesh), combining this with datapoint-level indexing for selection processes.</p>
<p>To supplement this, methods in <code>yt</code> that process data utilize a system of data “chunking,” whereby segments of data identified during coarse-grained indexing are subdivided by one of a few different schemes and yielded to the iterating function; these schemes can include a limited number of tuning parameters or arguments.
These three chunking methods are <code>all</code>, <code>spatial</code> and <code>io</code>.
The <code>all</code> method simply returns a single, one-dimensional array, and the number of chunks is always exactly one; this enables both non-parallel algorithms and simple access to small datasets.
<code>spatial</code> chunking yields three-dimensional arrays.
For grid-based datasets, these are the grids, while for particle and octree datasets they are leaf-by-leaf collections of particles or mesh values.
Optionally, the <code>spatial</code> chunking method can return “ghost zones” around regions, for computation of stencils.
The final type of chunking, <code>io</code>, is designed to iterate over sets of data in a manner that is most conducive to pipelined IO.
These will not always be load-balanced in size of the returned chunks, however.
In some cases, <code>io</code> chunking may return one file at a time (in the case of spreading items across many different files), while in others it may be returning sub-components of a single file.
This chunking type is the most common strategy for parallel-decomposition.</p>
<p>Necessarily, both indexing and selection methods must be implemented to expose these different chunking interfaces; <code>yt</code> utilizes specific methods for each of the primary data types that it can access.
We detail these below, specifically describing how they are implemented and how they can be improved in future iterations.</p>
<h3 id="grid-analysis">Grid Analysis</h3>
<div id="fig:grid_organization" class="fignos">
<figure>
<img src="" alt="Figure 7: The grid structure of the simulation IsolatedGalaxy" />
<figcaption aria-hidden="true"><span>Figure 7:</span> The grid structure of the simulation <code>IsolatedGalaxy</code></figcaption>
</figure>
</div>
<script>
document.addEventListener("THREELoaded", function(event) {
    const THREE = event.THREE;
    //const OrbitControls = event.OrbitControls;
    document.querySelector("div#fig\\:grid_organization img").remove();
    const container = document.createElement("div");
    const width = 400;
    const height = 400;
    container.style.border = 'solid 1px black';
    container.style.width = `${width}px`;
    container.style.height = `${height}px`;
    container.id = 'grid_scene';
    document.querySelector("div#fig\\:grid_organization").prepend(container);
    fetch('./images/grid_scene.json')
        .then((response) => response.json())
        .then((json) => {
            const scene = new THREE.ObjectLoader().parse( json );
            const renderer = new THREE.WebGLRenderer();
            const camera = scene.children[0];
            // We set this to 400 so that it doesn't need to be rendered
            // onscreen before it can be set up correctly.
            renderer.setSize(width, height);
            renderer.setPixelRatio(window.devicePixelRatio);
            const controls = new OrbitControls( camera, renderer.domElement );
            container.append(renderer.domElement);
            camera.position.set(2.5, 2.5, 2.5);
            camera.lookAt(0.5, 0.5, 0.5);
            controls.target.set(0.5, 0.5, 0.5);
            controls.update();
            renderer.render(scene, camera);
            animate = () => {
                requestAnimationFrame( animate );
                controls.update();
                renderer.render( scene, camera );
            }
            animate();
    });
});
</script>
<p><code>yt</code> was originally written to support the Enzo code, which is a patch-based Adaptive Mesh Refinement (AMR) simulation platform.
In Figure <a href="#fig:grid_organization">7</a> the grid structure of one of the standard <code>yt</code> example datasets, <code>IsolatedGalaxy</code>, can be seen.
Analysis of grid-based data is the most frequent application of <code>yt</code>.
While we discuss much of the techniques implemented for datasets consisting of multiple, potentially overlapping grids, <code>yt</code> also supports single-grid datasets (such as FITS cubes) and is able to decompose them for parallel analysis.</p>
<p><code>yt</code> also supports other grid patch codes, listed in the section on <a href="#sec:data_formats_and_frontends">frontends</a>.</p>
<p><code>yt</code> supports several different “features” of patch-based codes.
These include grids that span multiple parent objects, grids that overlap with coarser data (i.e., AMR), grids that overlap with other grids that provide the same level of resolution of data (i.e., grids at the same AMR level), refinement factors that vary based on level, and edge, and vertex-centered data.
For the cases of overlapping grids (either on the same or higher refinement levels) masks are generated that indicate which data is considered authoritative.</p>
<p>As noted in <a href="#sec:data_objects">Data Objects</a>, the process of selecting points is multi-step, starting at coarse selection that may be at the file level, and proceeding to selection of specific data points that are included in a selector.
For grid-based data, the coarse selection stage proceeds in an extremely simple fashion, by iterating over flat arrays of left and right grid edges and creating a bitmap of the selected grids.
Because this method – while not taking advantage of any data structures of even mild sophistication – is able to take advantage of pipelining and cache-optimization, we have found that it is sufficiently performant in most geometries up to approximately <span class="math inline">\(10^6\)</span> grid objects.
In those cases, the distinction between “wide and shallow” grid structures (where refinement occurs essentially everywhere, but not to a great degree) and “thin and deep” grid structures (where refinement occurs in essentially one location but to very high levels), as well as the specific selection process, impact the overall performance.
The second-stage selection occurs within individual grids, where points are selected based on the data point center.
In the case of cell-centered data, this returns an array of size <span class="math inline">\(N\)</span> where <span class="math inline">\(N\)</span> is the number of points selected; in the case of 3D vertex-centered data, this would be <span class="math inline">\((N,8)\)</span>.</p>
<p>Indexing grid data in <code>yt</code> is optimized for systems of grids that tend to have larger grid patches, rather than smaller; specifically, in <code>yt</code> each grid patch consists of a Python object, which adds a bit of overhead.
In the limit of many more cells than grid objects, this overhead is small, but in cases where the number of grids is <span class="math inline">\(\sim 10^7\)</span> this can become prohibitive.
These cases are becoming more common even for medium-scale simulations.</p>
<p>To address both the memory overhead and the Python overhead, as well as more generally address potential scalability issues with grid selection, several tentative explorations have been made into an implementation of a more sophisticated “grid visitors” indexing and selection method, drawing on the approach used by the oct-visitors (described below in Section #sec:octree_analysis).
These were an attempt to unify the selection methods between octrees and grids, to reduce the overall code duplication and implementation overhead.
Each process – selection, copying of data, generation of coordinates – is represented by an instance of a <code>GridVisitor</code> object.
A spatial tree is constructed, wherein parent/child relationships are established between grids.
The tree is recursively traversed, and for all selected points the object is called.
This allows grids, their relationships, and the data masks to be stored in structures and forms that are both optimized and compressed.
This method is essential for scaling to a large number of grid patches; the storage requirements of a single grid patch Python object are around 1K per object (about one gigabyte per million grids), whereas the optimized storage reduces this to approximately 140 bytes (about one gigabyte per eight million grids), with further reductions possible; for selection operations, we are also able to reduce the number of temporary arrays and utilize compressed mask representation, bringing peak memory usage down further.
The spatial-tree optimization substantially increases performance for “wide and shallow” dataset selection.
However, while such an implementation may be possible, the previous attempts were stymied by performance and maintenance considerations for the grid code, in particular related to the masking of “child” zones in an efficient and straightforward manner.</p>
<h3 id="sec:octree_analysis">Octree Analysis</h3>
<p><code>yt</code> supports octree-based AMR datasets (primarily RAMSES and ART, but also the output from the octree-based radiative transfer code Hyperion).
<code>yt</code> stores a copy of the octree using a pointer-based approach, where each oct points to its eight children (if refined).
The octs living at the coarsest level of the simulation are stored as a uniform grid. For domain-decomposed datasets, each domain is represented as a sparse octree, where the root octs are stored as a list and efficiently accessed using a binary search, ensuring each root oct is found in <span class="math inline">\(O(\log(N))\)</span> time, where <span class="math inline">\(N\)</span> is the number of root octs in the domain.
Each oct is represented as structure that contains the on-file location of the oct (<code>file_ind</code>) and its in-memory location (<code>domain_ind</code>), the index of the domain it belongs to (<code>domain</code>) and a list of pointers to its children (up to eight in 3D). This requires at most 88 bytes per oct.</p>
<p>In order to load data within a given region, a two-step approach is followed.
First, the cells within the region of interest, as described in <a href="#sec:data_objects">Data Objects</a> are selected. <code>yt</code> relies on an oct-visitor machinery combined with selection routines.
The tree is recursively traversed depth-first starting from the root grid, following only those branches that may intersect with the selected region.
At the tip of each branch, the up-to-eight leaf cells are visited.
In a first pass, the number of selected cells is computed and in a second pass, the on-file location of their parent oct is stored.
Second, <code>yt</code> relies on the on-file location obtained from the octree traversal to lazily read data from disk.
This ensures that only the minimal amount of data is being read and is particularly efficient when accessing a region spanning a small number of domains and/or a small number of refinement levels.</p>
<p>Recently, <code>yt</code> has been extended to fully support accessing neighboring cells.
This is achieved by computing one-cell thick quantities around each oct, which emulates the “ghost zones” found in patch-based codes. This approach has the advantage of abstracting the octree structure and provides a common interface to create derived fields, as described in <a href="#sec:derived_fields">1.7.1.2</a>.
The 56 neighbors (<span class="math inline">\(4^3 - 2^3\)</span>) surrounding each oct are found by performing a search in the octree, which finds any neighbor in <span class="math inline">\(O(\mathrm{level})\)</span>, where <span class="math inline">\(\mathrm{level}\)</span> is the level of the central oct.
The search is illustrated on Figure <a href="#fig:binary-search">8</a>.
Other optimizations are possible, but not yet implemented in <code>yt</code>, that trade computational time with memory, for example by storing the tree as a fully-threaded structure (i.e. store pointers to the 6 neighbors sharing a face with each oct), or by starting at a central oct and searching “upwards and outwards.”</p>
<div id="fig:binary-search" class="fignos">
<figure>
<img src="images/octree/binary_tree_research_2D.svg" alt="Figure 8: Illustration of a binary search through a quadtree. The search starts at the root level (level = 1 here) and recursively selects the quad that contains the point until reaching a leaf. The procedure is easily generalized in 3D." />
<figcaption aria-hidden="true"><span>Figure 8:</span> Illustration of a binary search through a quadtree. The search starts at the root level (level = 1 here) and recursively selects the quad that contains the point until reaching a leaf.
The procedure is easily generalized in 3D.</figcaption>
</figure>
</div>
<div id="fig:octree-gradient" class="fignos">
<figure>
<img src="images/octree/gradient_computation.svg" alt="Figure 9: Scheme of the AMR structure used to estimate the gradient of a quantity in the central oct (red). Octs are represented in thick lines, cells in thin lines and virtual cells in dashed lines. Left panel: The virtual cell values on a 4^3 grid are interpolated from the nearest cell in the AMR grid. If the nearest cell is at the same (or coarser) level, its value is used directly. Note that virtual cells f_{31} and f_{32} have the value of the the actual coarser cell (green). If the cell is refined, the mean of its children is used (for example f_{20} is the mean of all the blue cells). Right panel: Gradients are estimated using a first-order finite difference centered scheme on the 4^3 virtual cells, here illustrated for a gradient along the x direction." />
<figcaption aria-hidden="true"><span>Figure 9:</span> Scheme of the AMR structure used to estimate the gradient of a quantity in the central oct (red). Octs are represented in thick lines, cells in thin lines and virtual cells in dashed lines. <em>Left panel:</em> The virtual cell values on a <span class="math inline">\(4^3\)</span> grid are interpolated from the nearest cell in the AMR grid. If the nearest cell is at the same (or coarser) level, its value is used directly. Note that virtual cells <span class="math inline">\(f_{31}\)</span> and <span class="math inline">\(f_{32}\)</span> have the value of the the actual coarser cell (green). If the cell is refined, the mean of its children is used (for example <span class="math inline">\(f_{20}\)</span> is the mean of all the blue cells). <em>Right panel:</em> Gradients are estimated using a first-order finite difference centered scheme on the <span class="math inline">\(4^3\)</span> virtual cells, here illustrated for a gradient along the <span class="math inline">\(x\)</span> direction.</figcaption>
</figure>
</div>
<h3 id="sec:sph-analysis">SPH Analysis</h3>
<p>Smoothed Particle Hydrodynamics (SPH) is a commonly-used method for solving equations of hydrodynamics in astrophysics (as well as many other fields!) from a lagrangian perspective.
This provides many advantages over grid-based discretizations, but also poses somewhat different challenges for analysis and visualization.
While a full description of SPH is outside the scope of this paper, there are a handful of crucial and important pieces of information that we will review.
For more information, we refer interested readers to this comprehensive review of Smoothed Particle Hydrodynamics by Daniel Price <span class="citation" data-cites="fZ2Zii88">[<a href="#ref-fZ2Zii88" role="doc-biblioref">6</a>]</span>, the one by Volker Springel <span class="citation" data-cites="NSGlchZc">[<a href="#ref-NSGlchZc" role="doc-biblioref">7</a>]</span>, or to the SPLASH method paper by Daniel Price <span class="citation" data-cites="1A6gLV5Q0">[<a href="#ref-1A6gLV5Q0" role="doc-biblioref">8</a>]</span>.
SPH defines field quantities <em>at</em> a set of moving points, allowing generic field values (i.e., between the points) to be computed by integrating over all the elements in the domain using a special-purpose kernel; this method is an exact interpolation between the discretized points by applying the smoothing kernel.
Formally, this is represented as:
<span class="math display">\[
A(\mathbf{r}) = \int A(\mathbf{r}&#39;)W(|\mathbf{r} - \mathbf{r}&#39;|,h)\mathrm{d}V(\mathbf{r}&#39;)
\]</span>
This is then reduced to a sum over the particles (the discretization points):
<span class="math display">\[
A(\mathbf{r}, h) = \sum_j V_j A_j W(|\mathbf{r} - \mathbf{r}_j|, h)
\]</span>
In these equations, <span class="math inline">\(A\)</span> is the field, W is the weighting function (the ‘kernel’) and <span class="math inline">\(h\)</span> is the smoothing length.
This weighting function typically takes the shape of a Gaussian, approximated through various spline functions (truncated at some radius), but that is not strictly necessary <span class="citation" data-cites="Yng3Bhre">[<a href="#ref-Yng3Bhre" role="doc-biblioref">9</a>]</span>.
The quantity <span class="math inline">\(h\)</span>, the smoothing length, has previously been referred to as the ‘half-smoothing length’ (hence the variable name <code>hsml</code> used in many contexts), but formally represents the full-width-half-maximum of the Gaussian approximated by the spline kernels, with the ratio <span class="math inline">\(\ell_{\rm FWHM} / h = \sqrt{2\ln2}\)</span> <span class="citation" data-cites="GC8jwjG1">[<a href="#ref-GC8jwjG1" role="doc-biblioref">10</a>]</span>.
The value of <span class="math inline">\(h\)</span> in adaptive simulations is typically allowed to vary (which is extremely common in astronomy, though this is not true when SPH is applied to other domains where the density of the fluid remains roughly fixed) such that a constraint equation, e.g.
<span class="math display">\[
  n(\mathbf{r}, h_i) = \sum_j W(|\mathbf{r} - \mathbf{r}_j|, h_i) = \left(\frac{\eta}{h_i}\right)^{n_{\rm D}},
\]</span>
is satisfied, where <span class="math inline">\(\eta\)</span> is a fixed constant that sets the spatial resolution of the simulation and <span class="math inline">\(n_{\rm D}\)</span> is the number of dimensions.
Various codes may change this constraint equation for differing purposes, such as confining the integral to a fixed number of particles, or those that evaluate the constraint based on pressure criteria, and so forth.
Many codes store the maximal radial extent of the kernel (known as the kernel extent, and often represented by <span class="math inline">\(H\)</span>), as this is what is used in neighbour finding operations.
The drawback to this choice is that the specific value of <span class="math inline">\(H\)</span> is kernel-dependent, as some may cut off at much shorter distances than others when representing the same Gaussian.
As different codes and methods make use of different kernels for various reasons (ones with larger cut-off radii can produce less noisy results, but are computationally more costly), some codes now employ <span class="math inline">\(h=H/\gamma_{\rm K}\)</span>, where <span class="math inline">\(\gamma_{\rm K}\)</span> is a kernel-dependent quantity.</p>
<p>In general, there are two approaches to defining the finite set of particles that contribute to a field at a given point.
In “scatter” methods, computing a field at a given point is conducted by iterating over particles and identifying those whose smoothing length overlaps with a given point.
In the alternate method “gather,” the outer and inner loops of the algorithm are essentially swapped; for every point at which a field value is sought, the local smoothing length is computed and then all particles within that smoothing length are used in the computation.</p>
<p>For the purposes of <em>post-processing</em> analysis and visualization, the most important criteria for applying SPH to a set of particles are to ensure that the function that computes field values at a given location is <em>identical</em> to that used inside the simulation code (or as close as possible) and that the calculations are conducted in as short a time as possible.
This set of dual requirements has led to yt implementing a flexible system for defining the smoothing kernel used, whether or not a normalization step is applied to SPH quantities, and the option to use either “scatter” or “gather” methods for computing field values at fixed locations.</p>
<p>Previous versions of <code>yt</code> provided analysis of SPH data through a hybrid approach that mixed pure-SPH analysis with octree-based gridding and indexing that used particle density as a guide for the necessary resolution.
Although the present, yt 4.0 series does not utilize octrees for particles, a description of the previous implementation is useful to provide both historical information and modern motivation for the “demeshening” initiative that led to the current code base.</p>
<p>In practice, what this meant was that when a dataset was loaded, the particles positions were converted into one dimensional Morton index values, sorted, and by using a process of identifying the longest prefix in their bitwise representations, an octree (that mapped directly onto the Morton index values) was constructed.
This octree was controlled by two parameters: <code>n_ref</code>, which specified the number of particles in an octree leaf node necessary to refine that node into eight sub-octants, and <code>over_refine_factor</code>, which specified the number of cells that each leaf node represented.
For instance, if <code>n_ref</code> was set to 64 (the default), any octree node containing 64 particles would be refined into eight child nodes; if <code>over_refine_factor</code> was set to N, each leaf node would consist of a set of zones that were <span class="math inline">\(2^N\)</span> zones on a side (i.e., the default <code>over_refine_factor</code> produced eight mesh elements total).</p>
<p>Constructing these octrees using morton indices, if the entire set of particles could be stored in memory simultaneously, was <em>extremely</em> efficient.
To do so, the particles merely needed to be converted into a morton index via fast, bit-level operations, those index values sorted, and then processed in order to identify the greatest common bit-prefix.
Because two successive particles with identical index values would share an octree location, looking for sequences of identical prefix values (i.e., lower-level octree colocation) naturally produces an octree.
When fluid quantities such as density were requested in the yt 3.0 series, the values were computed on the mesh defined by the octree; increasing the <code>over_refine_factor</code> and decreasing the <code>n_ref</code> would serve to increase the resolution.
While this produced mostly-acceptable visualizations, and in particular produced dynamically-resolved visualizations, it posed several problems for both visualization and analysis.
The first, and arguably the most important, is that the strict locality requirements for refinement produced artifacts at leaf node boundaries.
This resulted in incorrect and unphysical visualizations of hydrodynamic quantities, affecting most obviously those regions at the edges of clusters of gas particles.
These were mitigated in regions of highly-clustered gas particles, but visual artifacts were still clear, as <code>yt</code> was applying a visualization suited for finite volume elements to Lagrangian particles.</p>
<p>With the 4.0 series, <code>yt</code> no longer utilizes octrees for analyzing, meshing or visualizing SPH data.
Indexing, for the purposes of fast, memory-efficient access to subsets of the data, is provided by a bitmap index using the Morton indices of the particles, as described in <a href="#sec:point_indexing">Bitmap Indexing</a>.
For the purposes of visualization, any quantity requiring smoothing over nearest neighbors is computed on-demand at each pixel in the output image; this provides much higher resolution than the previous method, which was both subject to free parameters and required the construction of a 3D fluid field that was then collapsed to 2D for visualization.
In many cases, this is also considerably more performant, as constructing a full-domain 3D fluid field is avoided, thus reducing both memory requirements and the number of floating point calculations.</p>
<p>Development of this new method was referred to internally as “the demeshening,” as it served to eliminate the global (octree) mesh.
In order to facilitate the massive, type and dimensionality-specific spatial queries necessary for performing millions of queries as efficiently as possible, and with as little overhead as possible, <code>yt</code> packages a kD-tree written in Cython that can be called from either Cython or Python, and which provides low-level APIs for querying from within tight loops.
Whereas previously, constructing a projection or a slice would slice through an octree mesh and provide the results from that variable resolution mesh, the current version of <code>yt</code>’s SPH machinery will instead construct a pixel plane and smooth the appropriately identified particles onto that pixel plane.
This produces much higher-fidelity results (see Figure <a href="#fig:demesh-comparison">10</a>), but a current limitation is that whenever the pixel plane is changed, the particles must be re-deposited; this puts it at odds with the similar machinery for octree and patch-based datasets, which provide a “read-once-pixelize-many” approach.</p>
<p>The octree method – while not incapable of utilizing different normalization and particle search methods – was less flexible than the current, de-meshened approach.
For instance, the method of SPH particle identification (i.e., so-called “scatter” or “gather” methods for correlating particles with positions) is now flexible and able to be set at runtime.
The normalization (if used) can take into account global quantities, local quantities, and is flexible based on the field being smoothed.</p>
<div id="fig:demesh-comparison" class="fignos">
<figure>
<img src="images/yt34comparison.png" alt="Figure 10: Comparisons between the older, octree-based method used in version 3.0 of yt (left) and the newer, “demeshened” algorithm used in yt 4.0 and beyond (right). The left image clearly shows artifacts from the octree structure imposed on the underlying dataset by yt, and the right hand side is much smoother, with more definition at individual pixels. The difference in color bars is notable as well, accounted for by the different normalization methods." />
<figcaption aria-hidden="true"><span>Figure 10:</span> Comparisons between the older, octree-based method used in version 3.0 of <code>yt</code> (left) and the newer, “demeshened” algorithm used in yt 4.0 and beyond (right). The left image clearly shows artifacts from the octree structure imposed on the underlying dataset by <code>yt</code>, and the right hand side is much smoother, with more definition at individual pixels. The difference in color bars is notable as well, accounted for by the different normalization methods.</figcaption>
</figure>
</div>
<p>Some additional differences between SPH analysis and the analysis of finite volume data are present when utilizing data selectors.
For instance, 3D data selectors as applied to finite volume codes only select those cells whose centers fall within the data selector.
2D and 1D data selectors (such as slices and rays) also include those finite volume cells that the selector passes through.
However, with SPH data, the selection methods in 2D and 3D will always include those particles whose spheres of influence, defined by the appropriate smoothing lengths, are within or overlapping with the data selector.
This is somewhat counter to the expectations set by the grid codes, but aligns with the need to have a fully self-contained data-container for computing field values.
For instance, this means that a “ray” object (often used to compute, for instance, the column density in a cosmological simulation) will in fact include a set of particles within a (potentially) varying impact parameter.
This can be seen in diagram form in Figure <a href="#fig:sph_ray_tracing">11</a>.
We note that, as described in the SPLASH method paper <span class="citation" data-cites="1A6gLV5Q0">[<a href="#ref-1A6gLV5Q0" role="doc-biblioref">8</a>]</span>, the kernel interpolation can be computed using the (dimensionless) ratio between the impact parameter of the ray and the smoothing length of the particle.</p>
<canvas id="sph-ray-figure" style="width:100%;height:100%;min-height: 500px;" hidpi="true">
</canvas>
<script type="text/paperscript" canvas="sph-ray-figure" src="images/sph_ray_tracing.js"></script>
<div id="fig:sph_ray_tracing" class="fignos">
<figure>
<img src="images/blank.svg" width="1" alt="Figure 11: A cartoon diagram of a ray passing through a collection of particles. The radius of the particle is indicative of its smoothing length. As can be seen, the individual particles each contribute different amounts as a result of their smoothing length, the chord-length as the ray passes through the circle, and the values within each particle." />
<figcaption aria-hidden="true"><span>Figure 11:</span> A cartoon diagram of a ray passing through a collection of particles. The radius of the particle is indicative of its smoothing length. As can be seen, the individual particles each contribute different amounts as a result of their smoothing length, the chord-length as the ray passes through the circle, and the values within each particle.</figcaption>
</figure>
</div>
<p>Other than these differences, which have been intentionally made to align the results with the expected results from the underlying discretization method, the APIs for access to particle data and finite volume data are identical, and they provide broadly identical functionality, where the disparities are typically in functionality such as volume rendering.
This allows a single analysis script, or package (such as Trident), to utilize a high-level API to address both gridded and Lagrangian data while still ensuring that the results are high-fidelity and representative of the underlying methods.</p>
<h3 id="sec:unstructured_mesh">Unstructured Mesh Analysis</h3>
<p><code>yt</code> has support for several different types of unstructured mesh elements.
Typically, these are supplied as a set of coordinate points (vertices) and connectivity between those vertices.
<code>yt</code> is able to interpret three types of elements (and their 2D counterparts): tetrahedral elements (4 faces, 4 vertices), wedge elements (5 faces, 6 vertices) and hexahedral elements (6 faces, 8 vertices).
These vertices can serve as control points, where values are defined at those locations; in finite element simulations, there can be additional control points for higher-order solutions.
(For a deeper investigation of the way finite elements are defined and how this corresponds to real-space coordinates, we suggest starting with the <a href="https://www-users.cse.umn.edu/~arnold/femtable/index.html">periodic table of the finite elements</a> which provides both visual reference and a set of citations for further exploration; further explanation can be found in the <a href="https://sinews.siam.org/Details-Page/periodic-table-of-the-finite-elements">SIAM News Article</a> describing the table.)</p>
<h4 id="data-access-for-unstructured-mesh">Data Access for Unstructured Mesh</h4>
<p>Similar to how <code>yt</code> manages data access for particle and finite volume datasets, for unstructured mesh datasets <code>yt</code> identifies each element <em>collection</em> as a chunk.
This means that for situations where you have multiple meshes, composed of individual elements, each will represent its own chunk as well as its own mesh object.
For example, in MOOSE-based simulations with multiple connectivity arrays, each will be a different “field type” – typically named <code>connect1</code>, <code>connect2</code>, etc.
These are then joined (similar to how <a href="#sec:particle_unions">1.7.1.4</a> are defined) into collections that include all of the elements of different types.</p>
<p>A few items are of particular note in the implementation of finite element mesh analysis in <code>yt</code>.
The first is that <code>yt</code> supports direct, native higher-order finite element visualization.
Visualization of unstructured meshes, and finite element frameworks, utilizes its own set of custom pixelization routines that are dependent not only on the element type but the order of the calculation.</p>
<p>The second item that is of relevance is that <code>yt</code> is able to apply “displacement” vectors to the elements; these displacement vectors can vary with time, and thus element position and shape can vary over the course of a simulation.
By providing appropriate arguments, <code>yt</code> can scale these displacement vectors (either with scalars or vector values) to exaggerate or distort their application, and in addition a vector offset can be applied to the vertices in the dataset.
Scaling and offsetting are both applied on a per-mesh basis, enabling individual collections of elements to be scaled individually.</p>
<p>One of the most important optimizations that has yet to be applied to the unstructured mesh support in <code>yt</code> is in the “coarse” indexing process of selection.
While fine-grained indexing and selection is applied, the process of checking which meshes (i.e., coarse chunks) may intersect a given selector currently passes everything through to the next stage; this is highly-inefficient, and an important target for future optimization.</p>
<h4 id="sec:unstructured_mesh_sampling">Sampling Mesh Elements</h4>
<p>The pixelization routines in <code>yt</code> for unstructured mesh elements rely on computing <span class="math inline">\(f(x,y,z)\)</span> for all locations within an element that appear in the image plane.
To properly conduct this pixelization, as well as to utilize software or hardware volume rendering, we have to construct a high-fidelity sampling system that can accept data of different orders, connectivity, and shape.
This utilizes a multi-step process that is mediated by subclasses of the Cython-based class, <code>ElementMapper</code>.
All <code>ElementMapper</code> subclasses need to provide two functions, one to transform a “physical” position <span class="math inline">\((x, y, z)\)</span> to the position within the reference “unit” element (<span class="math inline">\(x&#39;, y&#39;, z&#39;)\)</span>, and one to sample the value at a position in the “unit” element (<span class="math inline">\(f(x&#39;, y&#39;, z&#39;)\)</span>) given a set of vertex or control point values.
Where hand-written optimizations for these functions are not available, classes are autogenerated from high-level shape function definitions; functions for both the sampling method and a Jacobian are generated using SymPy and output to Cython, where they are compiled ahead of time.
In <a href="#tbl:finite_element_types">3</a> we enumerate the types of finite elements supported at present.</p>
<div id="tbl:finite_element_types" class="tablenos">
<table id="tbl:finite_element_types">
<caption><span>Table 3:</span> Finite element types supported in <code>yt</code>. </caption>
<thead>
<tr class="header">
<th>Type</th>
<th># Dims</th>
<th># Vertices</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>P1</code></td>
<td>1</td>
<td>2</td>
<td>Linear</td>
</tr>
<tr class="even">
<td><code>P1</code></td>
<td>2</td>
<td>4</td>
<td>Linear Triangular</td>
</tr>
<tr class="odd">
<td><code>Q1</code></td>
<td>2</td>
<td>4</td>
<td>Linear Quadrilateral</td>
</tr>
<tr class="even">
<td><code>T2</code></td>
<td>2</td>
<td>6</td>
<td>Quadratic Triangular</td>
</tr>
<tr class="odd">
<td><code>Q2</code></td>
<td>2</td>
<td>9</td>
<td>Quadratic Quadrilateral</td>
</tr>
<tr class="even">
<td><code>P1</code></td>
<td>3</td>
<td>8</td>
<td>Linear Tetrahedral</td>
</tr>
<tr class="odd">
<td><code>Q1</code></td>
<td>3</td>
<td>8</td>
<td>Linear Hexahedral</td>
</tr>
<tr class="even">
<td><code>W1</code></td>
<td>3</td>
<td>6</td>
<td>Linear Wedge</td>
</tr>
<tr class="odd">
<td><code>Tet2</code></td>
<td>3</td>
<td>10</td>
<td>Quadratic Tetrahedral</td>
</tr>
<tr class="even">
<td><code>S2</code></td>
<td>3</td>
<td>20</td>
<td>Quadratic Hexahedral</td>
</tr>
</tbody>
</table>
</div>
<p>To conduct pixelization of a slice or to compute values for volume rendering, <code>yt</code> first computes bounding boxes for the individual values.
Once a pixel has been identified as being “within” a particular element (which also takes into account the shape of higher-order elements, rather than assuming a flat set of planes) the pixelizer has to compute the value at that location.
In order to compute intra-element values at a position <span class="math inline">\((x, y, z)\)</span> the position within a <em>reference</em> element <span class="math inline">\((x&#39;, y&#39;, z&#39;)\)</span> must first be computed, and then the value solved for given the values at the vertices.
This is conducted within the function <code>sample_at_real_point</code>, which is defined for each <code>ElementMapper</code>.</p>
<div id="fig:finite_element_higher_order" class="fignos">
<figure>
<img src="images/fem_example.png" alt="Figure 12: Example of a finite element mesh with higher-order tetrahedral elements, including a zoom-in on one of the elements" />
<figcaption aria-hidden="true"><span>Figure 12:</span> Example of a finite element mesh with higher-order tetrahedral elements, including a zoom-in on one of the elements</figcaption>
</figure>
</div>
<p>Of particular note is that, as listed in Table <a href="#tbl:finite_element_types">3</a> , <code>yt</code> has support for higher-order element types.
In Figure <a href="#fig:finite_element_higher_order">12</a>, an example of this is displayed.
On the left of the figure is a slice plot through a 2nd-order tetrahedral mesh.
On the right, we have zoomed in on the edge of the boundary of the element mesh.
In both, the mesh elements have been outlined in black.
As is clearly visible in the second plot, <code>yt</code> is applying higher-order methods for computing pixel values; not only through non-linear interpolation of field values, but also in the shape of the elements, which extend outside the <em>linear</em> boundaries of the tetrahedral elements.</p>
<h3 id="sec:noncartesian">Non-Cartesian Coordinates</h3>
<p>In Section <a href="#sec:abstraction-coordinates">1.5.1</a>, we describe the relationship between the internal ‘index’ space that <code>yt</code> uses for referencing values and the natural ‘data’ space that the values represent.
The abstraction of the coordinate systems and the relationship between index-space and data-space provides the ability to convert between the two; however, constructing visualizations and annotations requires an additional level of complexity.</p>
<p>The single most important shortcoming in the analysis of non-cartesian datasets in <code>yt</code> is that the data selection operators almost exclusively function on the coordinates in <em>index</em> space, rather than in <em>data</em> space.
As such, subselecting datasets by utilizing traditional geometric selectors in <code>yt</code> is much less useful than it should be; for example, selecting a sphere (see <a href="#sec:dobj-sphere">1.6.2.20</a>) applies spherical selections in index space, which result in a decidedly non-spherical object.
Selections of objects such as <a href="#sec:dobj-region">1.6.2.17</a> do make considerably more sense, however, as they are often thought of as sweeping data along coordinate axes; the region object itself will naturally select wedges in a spherical domain, for instance.
Future versions of <code>yt</code> will likely introduce means of more clearly selecting objects in coordinate space, for more natural subsetting of data.
It is still possible to apply data selection based on field values, which can include the coordinate-space field coordinates (such as <span class="math inline">\(r, \theta, \phi\)</span>).</p>
<p>Despite these weak spots, however, <code>yt</code> does provide a number of routines that are specific to non-cartesian datasets, including pixelizers for cylindrical and spherical coordinate systems.
(See <a href="#sec:pixelization">1.11.39</a> for more detail on this process.)
Pixelizers that take variable-resolution data along the <span class="math inline">\(r\)</span> and <span class="math inline">\(\theta\)</span> axes have been made available (for slicing along a conical section of a sphere or along the <span class="math inline">\(z\)</span> axis of a cylinder) as well as very simple projections from the surface of a sphere to a flat image (specifically utilizing the Aitoff projection).
<code>yt</code> also provides access to Cartopy <span class="citation" data-cites="LwVDbUsQ">[<a href="#ref-LwVDbUsQ" role="doc-biblioref">11</a>]</span> for more advanced or featureful map projections, as well as overlaying continents and other geographic shapes.
In Figure <a href="#fig:spherical_data">13</a> we demonstrate some of the native, built-in functionality <code>yt</code> provides for non-cartesian data.
This dataset, a simulation of magnetically-driven winds in a protoplanetary disk, was conducted in spherical coordinates.
Here, we have used <code>yt</code>’s functionality for overplotting streamlines as well as line integral convolution on irregular meshes to display the data in its native resolution and as a slice along the azimuthal axis of the simulation domain.</p>
<div id="fig:spherical_data" class="fignos">
<figure>
<img src="images/spherical_data.png" alt="Figure 13: Spherical data from a protoplanetary disk, overlaid with annotations supplied by yt to demonstrate both the magnetic field and velocity structure of the data. Data are used, with permission, from a simulation based on the ones described in [12]." />
<figcaption aria-hidden="true"><span>Figure 13:</span> Spherical data from a protoplanetary disk, overlaid with annotations supplied by <code>yt</code> to demonstrate both the magnetic field and velocity structure of the data.
Data are used, with permission, from a simulation based on the ones described in <span class="citation" data-cites="190ab0NKz">[<a href="#ref-190ab0NKz" role="doc-biblioref">12</a>]</span>.</figcaption>
</figure>
</div>
<h2 id="sec:point_indexing">Indexing Discrete-Point Datasets</h2>
<p>Advances in both hardware and software facilitate astrophysical datasets of growing complexity and size.
The datasets produced by numerical simulations can currently reach sizes of $$100 Tbytes split across hundreds of files <span class="citation" data-cites="xbCL1tS1">[e.g. <a href="#ref-xbCL1tS1" role="doc-biblioref">13</a>]</span>.
For even simple analysis tasks, the cost of incrementally reading datasets this large into memory is quite high.
This problem is not limited to theoretical work.
During operations the Large Synoptic Survey Telescope (LSST) will produce 15 Tbytes of data each night <span class="citation" data-cites="gw5rMEhN">[<a href="#ref-gw5rMEhN" role="doc-biblioref">14</a>]</span>.
In order to analyze such large datasets, we need innovative techniques for quickly indexing and selecting data without loading the entire dataset into memory.
We present a technique for using Morton bitmap indexes to map files and accelerate data analysis.</p>
<h3 id="sec:bitmap_theory">Theory and Background</h3>
<h4 id="sec:decomp">Domain Partitioning Between Files</h4>
<p>A common analysis task is the selection of data within a subset of the full domain; we use the term “selector" to refer to the selection operator.
If the dataset is split across multiple files, either due to size constraints or to allow for parallel I/O, such selections require every file to be loaded and parsed in order to assemble all of the data within the selection criteria.
This process can be very costly in terms of both the memory required to store the data and the time required to read each file.
However, if the contents of the files are mapped in advanced, only the files touched by the selection will need to be loaded.
This is particularly effective for partitioning schemes that are localized within the domain.
If each file contains data that are localized to one part of the domain, selections of contiguous sub-sections within the domain will require fewer files to be loaded.
Figure <a href="#fig:files">14</a> shows four examples of possible partitions of a two-dimensional spatial domain split equally between 8 files.</p>
<div id="fig:files" class="fignos">
<figure>
<img src="images/bitmap/files.png" alt="Figure 14: Examples of four different schemes for partitioning a 2D domain between 8 files. Each color represents a different file." />
<figcaption aria-hidden="true"><span>Figure 14:</span> Examples of four different schemes for partitioning a 2D domain between 8 files.
Each color represents a different file.</figcaption>
</figure>
</div>
<p>Panel (a) is an example where random parts of the domain are contained within each file.
In such a case, many files will need to be loaded for contiguous selections within the domain.
In panel (b), the domain was split between the files along the <span class="math inline">\(x\)</span> dimension.
Fewer files will need to be loaded for queries along the <span class="math inline">\(y\)</span>-dimension, but contiguous selection in <span class="math inline">\(x\)</span> will still require a greater number of files since the partition is not well localized in that dimension.
Panels (c) and (d) are both examples of partitioning the domain between the files along a space filling curve [Morton and Hilbert curves respectively; <span class="citation" data-cites="PSH5f9xg">[<a href="#ref-PSH5f9xg" role="doc-biblioref">15</a>]</span>; <span class="citation" data-cites="oIYFC4kD">[<a href="#ref-oIYFC4kD" role="doc-biblioref">16</a>]</span>].
These partitions have the greatest chance of limiting the number of files that must be loaded for a contiguous selection with slightly improved localization for the Hilbert curve.
Consequently, Hilbert curves have also been used for load-balancing in parallel simulation codes like Gadget-2 <span class="citation" data-cites="I01oLUIj">[<a href="#ref-I01oLUIj" role="doc-biblioref">17</a>]</span> and RAMSES <span class="citation" data-cites="1D5nekPGx">[<a href="#ref-1D5nekPGx" role="doc-biblioref">18</a>]</span>.</p>
<p>Figure <a href="#fig:selector1">15</a> shows examples of three selections within the above domain partitions.</p>
<div id="fig:selector1" class="fignos">
<figure>
<img src="images/bitmap/selector1.png" alt="Figure 15: Examples of file selection for four different domain partitions and three different shaded selectors. The number of files above each images is the number of files that must be loaded in order to get all of the data within the selected region." />
<figcaption aria-hidden="true"><span>Figure 15:</span> Examples of file selection for four different domain partitions and three different shaded selectors.
The number of files above each images is the number of files that must be loaded in order to get all of the data within the selected region.</figcaption>
</figure>
</div>
<div id="fig:selector5" class="fignos">
<figure>
<img src="images/bitmap/selector5.png" alt="Figure 16: Examples of file selection for four different domain partitions and three different shaded selectors. The number of files above each images is the number of files that must be loaded in order to get all of the data within the selected region." />
<figcaption aria-hidden="true"><span>Figure 16:</span> Examples of file selection for four different domain partitions and three different shaded selectors.
The number of files above each images is the number of files that must be loaded in order to get all of the data within the selected region.</figcaption>
</figure>
</div>
<div id="fig:selectors4" class="fignos">
<figure>
<img src="images/bitmap/selector4.png" alt="Figure 17: Examples of file selection for four different domain partitions and three different shaded selectors. The number of files above each images is the number of files that must be loaded in order to get all of the data within the selected region." />
<figcaption aria-hidden="true"><span>Figure 17:</span> Examples of file selection for four different domain partitions and three different shaded selectors.
The number of files above each images is the number of files that must be loaded in order to get all of the data within the selected region.</figcaption>
</figure>
</div>
<p>For the smallest selector (first row), the random domain decomposition (a) already requires half of the files to be loaded while more localized schemes require much fewer.
Similarly, while the sliced domain partition (b), requires the fewest files to be loaded when the selector is oriented in the same direction as the slicing (second row), it requires <em>all</em> of the files when the selector is perpendicular to the slicing (third row).
While some datasets may have information on the domain range covered by each file, the partitioning scheme used for simulation output is often decided at runtime, can be system dependent, and may be imperfect.</p>
<p>Files are often partitioned for parallel I/O such that each processor outputs data on the portion of the domain it is responsible for processing.
To limit the cost of communication between processors, the domain will be split across processors such that neighboring processors are responsible for neighboring parts of the domain.
This means that, although the overall partitioning scheme may be known for a given dataset, the exact order of the files will be dependent on the configuration of the processors at runtime.</p>
<p>The partitioning can also be imperfect if the domain decomposition is not perfect at the time of output.
For instance, in astrophysical N-body simulations, it is possible for particles to travel from one processor’s domain to another.
In this case, the partition will only be perfect directly following an update to the domain decomposition.</p>
<p>In cases where the exact file organization is not known or imperfect, it is advantageous to map the files post-process in order to speed up selections for analysis.
Although the same result can be achieved by re-sorting the data itself, creating the map can be less computationally less expensive than re-sorting the data, can be saved for use with multiple selections, and does not required write access; this is typically not feasible, especially in the case of datasets shared by large, distributed communities.</p>
<h2 id="morton-indices">Morton Indices</h2>
<p>Morton ordering maps multidimensional data onto a one-dimensional space filling curve <span class="citation" data-cites="PSH5f9xg">[<a href="#ref-PSH5f9xg" role="doc-biblioref">15</a>]</span>.
This is done by breaking up the domain into cells where each cell’s position within the <span class="math inline">\(N\)</span>-dimensional domain can be described by <span class="math inline">\(N\)</span> integers.
The Morton index of the cell is then created by interleaving the bits of the <span class="math inline">\(N\)</span> integers to create a single integer that fully describes the cell’s position (see panel (b) Figure <a href="#fig:zorder">18</a>).
As seen in panel (a) of Figure <a href="#fig:zorder">18</a>, ordering of the cells by their Morton indices forms a space filling Z-curve.</p>
<div id="fig:zorder" class="fignos">
<figure>
<img src="images/bitmap/zorder.png" alt="Figure 18: Example of 3rd order Morton curve in two dimensions. The bits of the x and y indices are interleaved to generate a single integer that fully describes the cell’s location within the two-dimensional domain to within 1/2^{3}th of the domain in each dimension." />
<figcaption aria-hidden="true"><span>Figure 18:</span> Example of 3rd order Morton curve in two dimensions.
The bits of the <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> indices are interleaved to generate a single integer that fully describes the cell’s location within the two-dimensional domain to within <span class="math inline">\(1/2^{3}\)</span>th of the domain in each dimension.</figcaption>
</figure>
</div>
<p>The precision of a single Morton index is only limited by the size of the integer used to store it.
For instance, 64-bit Morton indices in 3 dimensions can be localized to <span class="math inline">\(1/2^{21}\)</span>th of the domain in each dimension (<span class="math inline">\(3\times21\)</span> bits = 63 bits).
If the domain is binarily divided into subcells to some order <span class="math inline">\(k\)</span> in each dimension (i.e.
<span class="math inline">\(2^{Nk}\)</span> cells), coarser Morton indices can be obtained by simply masking lower bits.
Morton ordering has been used to speed up quadtree construction <span class="citation" data-cites="LpXfagAT">[<a href="#ref-LpXfagAT" role="doc-biblioref">19</a>]</span>, nearest neighbor searches <span class="citation" data-cites="12vJ72aHG">[<a href="#ref-12vJ72aHG" role="doc-biblioref">20</a>]</span>, and range queries <span class="citation" data-cites="4mimKxQe">[<a href="#ref-4mimKxQe" role="doc-biblioref">21</a>]</span>.
By recording the indices of the cells containing data from each file within a dataset, Morton indices can also be used to construct one-dimensional maps of an <span class="math inline">\(N\)</span>-dimensional dataset that can be represented as bitmaps.</p>
<h4 id="bitmaps-ewah-compression">Bitmaps &amp; EWAH Compression</h4>
<p>Bitmap indexes use the values of single bits within an array of bits to describe dataset properties.
This form requires minimal memory and can be filtered using computationally inexpensive boolean operations.
Bitmap indexes have long been popular for use with large data warehouses <span class="citation" data-cites="f1rS5tZk YV4eT8wZ DMkimxcD">[<a href="#ref-f1rS5tZk" role="doc-biblioref">22</a>,<a href="#ref-YV4eT8wZ" role="doc-biblioref">23</a>,<a href="#ref-DMkimxcD" role="doc-biblioref">24</a>]</span>.
However, as scientific datasets have become larger and more complex, they have also begun to gain traction in a diverse array of scientific fields including geosciences <span class="citation" data-cites="itaUSTKC">[<a href="#ref-itaUSTKC" role="doc-biblioref">25</a>]</span>, earth sciences, rocket science <span class="citation" data-cites="QMaCupmb mMJ3UeWp">[<a href="#ref-QMaCupmb" role="doc-biblioref">26</a>,<a href="#ref-mMJ3UeWp" role="doc-biblioref">27</a>]</span>, high-energy physics <span class="citation" data-cites="VzYAeBg3">[<a href="#ref-VzYAeBg3" role="doc-biblioref">28</a>]</span>, and combustion <span class="citation" data-cites="AFyCFou3">[<a href="#ref-AFyCFou3" role="doc-biblioref">29</a>]</span>.</p>
<p>In cases where data attributes can take on a finite set of values, one bitmap is constructed for each possible attribute value.
Within the bitmap each bit specifies whether or not the corresponding data point has that value.
In this way, queries for data with a single attribute value require consulting only one bitmap and queries of multiple attributes/values can be done using boolean AND operations on the corresponding bitmaps.
In the case of scientific data, which often contains floating point value attributes, the attributes must be binned prior to constructing the bitmaps <span class="citation" data-cites="Sju7gfgZ uibMnWfE pe5S32KG">[<a href="#ref-Sju7gfgZ" role="doc-biblioref">30</a>,<a href="#ref-uibMnWfE" role="doc-biblioref">31</a>,<a href="#ref-pe5S32KG" role="doc-biblioref">32</a>]</span>.
Here, Morton indices are used to bin N-dimensional floating point data onto one-dimension.
As a result, each file can be described by one bitmap.</p>
<p>For each file within a dataset, the Morton indices touched by the data within that file can then be stored in a bitmap index for future searches where the value of bit <span class="math inline">\(j\)</span> indicates whether or not Morton index <span class="math inline">\(j\)</span> is touched by the file in question.
For Morton indexing of order <span class="math inline">\(k\)</span>, this would result in a bitmap of length <span class="math inline">\(2^{Nk}\)</span> bits per file.
For large bitmaps, this can become costly in terms of memory and the time required to perform bitmap operations.
However, Enhanced Word-Aligned Hybrid (EWAH) compression can be used to limit these costs, particularly when the domain is densely or sparsely populated in localized regions <span class="citation" data-cites="yuSdhiDQ zXsFmNjY 2qfATif3">[<a href="#ref-yuSdhiDQ" role="doc-biblioref">33</a>,<a href="#ref-zXsFmNjY" role="doc-biblioref">34</a>,<a href="#ref-2qfATif3" role="doc-biblioref">35</a>]</span>.</p>
<p>An EWAH compressed bitmap will be smaller when there are long sequences, or “runs,” of identical values.
This means that an EWAH compressed bitmap will be smallest if either all or none of its bits are set.
An uncompressed bitmap would require the same, maximum, amount of memory in both of these cases.
The locality of Morton indices takes advantage of the EWAH compression.
If there are regions of the domain that are densely/sparsely populated, there Z-order space filling curve ensures that the bits denoting those regions will be adjacent, increasing the likelihood that there will be runs of identical (set/unset) bits and limiting the size of the compressed bitmaps.</p>
<h4 id="collisions">Collisions</h4>
<p>It is possible that two files will contain data within the same Morton cell.
This would mean that any time that cell is touched by a selection, both files would need to be loaded even if the selection only touches data from one of the files.
Figure <a href="#fig:collision">19</a> provides an example of collisions between two files.
In panel (a) of Figure <a href="#fig:collision">19</a>, purple cells are those that contain data from both files, a collision, for a 3rd order Morton index.
Any selector that contained one of those cells would need to load all of the data from both files, even if it only selected part of the cell.
Where the data is highly-concentrated in a central region (for instance, in a galaxy formation simulation with particles centrally-concentrated) this can mean that some regions suffer from worst-case scenario collision.</p>
<div id="fig:collision" class="fignos">
<figure>
<img src="images/bitmap/collisions.png" alt="Figure 19: Examples of a collision between two files. The red points and blue points are contained by two different files. The larger grid in both panels denotes the boundaries of 3rd order Morton cells. The cells containing points from either file are shaded accordingly such that cells containing points from both files are purple. The smaller grids within these cells on the right are the boundaries of 2nd order Morton cells refining the collisions." />
<figcaption aria-hidden="true"><span>Figure 19:</span> Examples of a collision between two files.
The red points and blue points are contained by two different files.
The larger grid in both panels denotes the boundaries of 3rd order Morton cells.
The cells containing points from either file are shaded accordingly such that cells containing points from both files are purple.
The smaller grids within these cells on the right are the boundaries of 2nd order Morton cells refining the collisions.</figcaption>
</figure>
</div>
<p>Collisions can be limited by either increasing the order of the index or allowing for multi-resolution indexes <span class="citation" data-cites="QMaCupmb mMJ3UeWp">[<a href="#ref-QMaCupmb" role="doc-biblioref">26</a>,<a href="#ref-mMJ3UeWp" role="doc-biblioref">27</a>]</span>.
Panel (b) of Figure <a href="#fig:collision">19</a> demonstrates an example of nesting a second index within cells that contain collisions.
In those cells which contained collisions, a 2nd order Morton index was added.
Those cells with collisions at the level of the refined index (purple cells in panel (b)) cover a much smaller portion of the domain than the cells with collisions at the level of the coarse index (purple cells in panel (a)).
This means that any given selection is less likely to contain a collision and it will be less likely for a selector to require both files to be loaded unless it actually touches data from both files.</p>
<p>Increasing the order of the coarse index has the same effect as nesting a second refined index within cells with collisions, but can also increases the size of the resulting map and the time it takes to identify files touched by a selection.
However, if the order of the coarse index is too small or the order of the refined index too large, this too can increase the cost of a selection in terms of memory and time.
Section <a href="#sec:test_order">1.10.2.1</a> discusses this tradeoff and how to choose index orders.</p>
<p>Collisions are more common for file partitioning schemes that are not localized.
Figure <a href="#fig:collision_files">20</a> shows an example of collisions for the different partitioning schemes discussed in Section <a href="#sec:decomp">1.9.1.1</a>.</p>
<div id="fig:collision_files" class="fignos">
<figure>
<img src="images/bitmap/index.png" alt="Figure 20: Examples of collisions for four different domain partitioning schemes. The heavy black lines denote 1st order Morton cells. The presence of more that one file (color) within a Morton cell indicates a collision. NOTE: from an accessibility standpoint, I must note that using a smooth color gradient may be counter productive for colorblind readers as it makes it a lot harder to perveice that neighbouring regions belong to different files." />
<figcaption aria-hidden="true"><span>Figure 20:</span> Examples of collisions for four different domain partitioning schemes.
The heavy black lines denote 1st order Morton cells.
The presence of more that one file (color) within a Morton cell indicates a collision.
<strong>NOTE: from an accessibility standpoint, I must note that using a smooth color gradient may be counter productive for colorblind readers as it makes it a lot harder to perveice that neighbouring regions belong to different files.</strong></figcaption>
</figure>
</div>
<p>For the random domain partition in panel (a), every cell within a 1st order Morton index will contain data from all 8 files.
This means that any selection using a 1st order bitmap index will require every file to be loaded.
For the more localized partitions in panels (c) and (d), only two files touch each Morton cell.</p>
<h4 id="ghost-zones">Ghost Zones</h4>
<p>It is often the case that, in selecting a region, additional padding around the region should be included in the selection.
This is particularly useful for algorithms that need information about neighboring points in the domain [e.g.
gas properties in simulations using Smoothed Particle Hydrodynamics; SPH; <span class="citation" data-cites="1C6sv2xA4">[<a href="#ref-1C6sv2xA4" role="doc-biblioref">36</a>]</span>; <span class="citation" data-cites="DQa79Dfo">[<a href="#ref-DQa79Dfo" role="doc-biblioref">37</a>]</span>; <span class="citation" data-cites="I01oLUIj">[<a href="#ref-I01oLUIj" role="doc-biblioref">17</a>]</span>].
For Morton indices, this is straightforward as the indices neighboring Morton cells can be found by incrementing the bits corresponding to each dimension.
We have included the ability to pad selectors with some number of Morton cells referred to as ‘ghost zones’.
Those files that touch ghost zones, but not the selector itself are referred to below as ‘ghost files’.</p>
<p>Depending on how the domain is split between files, the inclusion of ghost zones may or may not increase the number of files that need to be loaded.
Figure <a href="#fig:ghosts">21</a> shows an example of a ghost zone around the first selector from Figure <a href="#fig:selector1">15</a>.</p>
<div id="fig:ghosts" class="fignos">
<figure>
<img src="images/bitmap/ghosts.png" alt="Figure 21: Examples of a selector ghost zone with a width of one Morton cell at an index order of 3 for four different domain partitioning schemes. The shaded circular region is the selector and the shaded box is the ghost zone. Different partitioning schemes will lead to different numbers of ghost files. NOTE: same as figure 13" />
<figcaption aria-hidden="true"><span>Figure 21:</span> Examples of a selector ghost zone with a width of one Morton cell at an index order of 3 for four different domain partitioning schemes.
The shaded circular region is the selector and the shaded box is the ghost zone.
Different partitioning schemes will lead to different numbers of ghost files.
<strong>NOTE: same as figure 13</strong></figcaption>
</figure>
</div>
<p>The ghost zone has a width of one Morton cell at an index order of 3 and contains the same part of the domain in each case.
However, due to differences in how the domain was partitioned between the files in the four cases, the number of additional ghost files touched by the ghost zone in each case is different.
This will also depend on the order of the index to which ghost zones are added.
Ghost zones added at the order of the coarse index will be larger than those added at the order of the refined index and will have a higher probability of touching additional files.
While including ghost zones is advantageous when neighbor info is needed, it also increases the computational cost of identifying files (see Section <a href="#sec:tests">1.10.2</a>).</p>
<h3 id="sec:methods">Methods</h3>
<p>The basic procedure for constructing the bitmap index is as follows:</p>
<ol type="1">
<li><p><strong>Compute coarse indices.</strong> For each file in the data set, read in
the data and compute the indices of Morton cells at a given coarse
order that are touched by data contained within that file. These
coarse indices are then stored by setting the corresponding bits in
an EWAH compressed bitmap.</p></li>
<li><p><strong>Find collisions.</strong> The indices of coarse cells that are touched by
data in more than one file (collisions) are located using bitwise
operations on the file bitmaps. These indices are also stored in an
EWAH compressed bitmap.</p></li>
<li><p><strong>Compute refined indices.</strong> For each file in the data set, read in
the data and compute the indices of Morton cells at a given refined
order within coarse cells with collisions that are touched by that
file. These refined indices are stored in a map from coarse Morton
index to an EWAH compressed bitmap of refined Morton indices within
that cell.</p></li>
<li><p><strong>Output bitmaps.</strong> The EWAH compressed bitmaps for the coarse
indices, refined indices, and collisions are saved to an external
index file.</p></li>
</ol>
<p>For large datasets and/or high levels of refinement, this can be a time consuming process; however, it must only be done once.
For future selections, the bitmap can be quickly loaded and used to identify files in less time than would be required to load and query each file within the dataset individually.
Selection using a loaded bitmap goes as follows:</p>
<ol type="1">
<li><p><strong>Construct selector bitmap.</strong> In the same way each file was mapped,
the indices of Morton cells touched by the selector are stored in a
bitmap. This is done by checking for intersection of the selector
with Morton cells at the order of the coarse bitmaps. For contiguous
selectors, this is done at lower order (parent) cells first and
continued recursively until the order of the coarse bitmap is
reached.</p>
<ul>
<li><p>If a cell is completely within the selector, all of its child
cells at the coarse order are added to the bitmap.</p></li>
<li><p>If a cell intersects the edge of the selector, child cells at
increasing orders are checked until the order of the coarse
bitmaps are reached. If the cell is at the coarse order and
there is a collision between two files, a refined bitmap is the
constructed for the selector in the same manner.</p></li>
</ul></li>
<li><p><strong>Find files intersecting the selector.</strong> Bitwise operations with
the coarse file bitmaps are then used to efficiently identify files
that intersect the selector within coarse cells. If the coarse cells
within the intersection with a file all have collisions with other
files, bitwise operations with the refined file bitmaps are then
used to determine if the file is selected at the order of the
refined index.</p></li>
</ol>
<p>If ghost zones are desired, the neighbors of cells that intersect the edge of the selector are added to a separate bitmap.
For cells without collisions, the neighbors are added at the coarse bitmap order.
If there are collisions, the neighbors are added at the refined bitmap order.</p>
<h3 id="sec:tests">Tests</h3>
<p>The utility of using Morton index bitmaps for mapping files to decrease query times was tested on artificial N-body simulation datasets containing <span class="math inline">\(1024^3\)</span> points in three dimensions, distributed between 512 files.
For each test a Morton index bitmap was constructed for the dataset and used to identify files touched by cube shaped three-dimensional selectors.
The performance is assessed in terms of the number of files identified and the average time required to identify them across 10 runs.
If fewer files are touched, fewer files will need to be loaded during analysis of a selected region and the overall fraction of time spent on I/O will be lower.
If less time is required to identify the files touched by a given selector, more selections can be made using the same computational resources.
This was done for varying index orders (Section <a href="#sec:test_order">1.10.2.1</a>), selector sizes (Section <a href="#sec:test_size">1.10.2.4</a>), and partitions of the domain between files (Section <a href="#sec:test_decomp">1.10.2.5</a>).</p>
<h4 id="sec:test_order">Index Order</h4>
<h4 id="sec:test_order1">Overall Refinement</h4>
<p>The order of the Morton indices used to map the files determines the time required to identify files and the number of collisions that will occur between files.
Higher order indices will result in fewer collisions, but will take longer to query, as seen in Figure <a href="#fig:test_order1">22</a> Six selectors of varying sizes and positions within the domain where used to identify files based on Morton index bitmaps of varying order.
The test dataset was split across the files using a Hilbert curve of order 6 with 10% scatter between Hilbert cells to simulate an imperfect domain decomposition as can occur if particle positions are updated and output prior to updating the domain decomposition.</p>
<div id="fig:test_order1" class="fignos">
<figure>
<img src="images/bitmap/vary_order1_or0.png" alt="Figure 22: Dependence of query time (top), fraction of files selected/cells with collisions (middle), and index size (bottom) on the total refinement of the bitmap index. The solid black lines correspond to the query times and files identified by just the selectors. The dashed blue lines correspond to the query times and additional files selected when a ghost zone with the width of one Morton cell is added around the selectors. The dash-dotted line in the middle panel shows the fraction of cells with collisions between files. NOTE: maybe this would make more sense as scatter plots ? (connecting dots to fill a continuous yet unreachable space of non-integer-limited x values seems unappropriate." />
<figcaption aria-hidden="true"><span>Figure 22:</span> Dependence of query time (top), fraction of files selected/cells with collisions (middle), and index size (bottom) on the total refinement of the bitmap index.
The solid black lines correspond to the query times and files identified by just the selectors.
The dashed blue lines correspond to the query times and additional files selected when a ghost zone with the width of one Morton cell is added around the selectors.
The dash-dotted line in the middle panel shows the fraction of cells with collisions between files.
<strong>NOTE: maybe this would make more sense as scatter plots ? (connecting dots to fill a continuous yet unreachable space of non-integer-limited x values seems unappropriate.</strong></figcaption>
</figure>
</div>
<p>Below a bitmap index order of 4, there are collisions between multiple files within every cell, resulting in a larger number of files being identified.
However, as the order increases, the number of collisions drops and the file count plateaus at <span class="math inline">\(\sim25\)</span>%.
This translates to a <span class="math inline">\(\sim75\)</span>% reduction in the memory and time required for processing files, a significant increase in performance.
For a 7th order bitmap index, selection requires <span class="math inline">\(&gt;100\times\)</span> the time that the same selection took using a 6th order index, but there is no change in the number of files indicated.
A 6th order index is sufficient to identify the minimal set of files touched by the selectors in this case because the dataset was partitioned between the files along a 6th order Hilbert space filling curve.
While it is generally true that the time required to identify files using a bitmap index will increase exponentially with the size of the index, the order of the index that results in the minimal number of files for any dataset will depend upon how the domain is partitioned between files (see Section <a href="#sec:test_decomp">1.10.2.5</a>).
The memory required to store the index for the test dataset scales according to <span class="math inline">\(\propto2^{2k}\)</span>, for a <span class="math inline">\(k^{\mbox{th}}\)</span> order index.
If uncompressed bitmaps had been used instead of EWAH compressed bitmap, the memory would have scaled with the total number of cells contained within the 3-dimensional test domain (<span class="math inline">\(2^{3k}\)</span>).</p>
<h4 id="sec:test_order2">Collision Refinement</h4>
<p>Increasing the refinement of the primary index does so for the entire domain and, as seen in Section <a href="#sec:test_order1">1.10.2.2</a>, can become costly in terms of the memory required to store the bitmap and the time required to perform operations.
However, it is also possible to increase refinement by nesting a second Morton bitmap index within those cells of the primary index that contain collisions.
As the nested indexes will contain a smaller portion of the domain and data, they will be less complex and can be compressed more efficiently than the primary index covering the entire domain.
This enhanced compression means that, although a greater overall number of EWAH compressed bitmaps will need to be utilized (one for the coarse index and one for each collision within the coarse index), less space will be needed to store the bitmap and bitwise operations will be faster.
Figure <a href="#fig:test_order2">23</a> shows the results for adding a secondary index of varying order with the overall refinement order of the index (primary index order + secondary index order) held constant at 6.
The test dataset and selectors applied were the same as in Section <a href="#sec:test_order1">1.10.2.2</a>.</p>
<div id="fig:test_order2" class="fignos">
<figure>
<img src="images/bitmap/vary_order2_to6.png" alt="Figure 23: Dependence of query time (top), fraction of files selected/cells with collisions (middle), and memory required to store the index (bottom) on the order of the secondary index used to refine collisions. In the middle panel, the solid black line corresponds with the fraction of files identified, the dash-dotted blue line is the fraction of cells at the first index level that have collisions, and the dotted red line is the fraction of cells at the second index level that have collisions." />
<figcaption aria-hidden="true"><span>Figure 23:</span> Dependence of query time (top), fraction of files selected/cells with collisions (middle), and memory required to store the index (bottom) on the order of the secondary index used to refine collisions.
In the middle panel, the solid black line corresponds with the fraction of files identified, the dash-dotted blue line is the fraction of cells at the first index level that have collisions, and the dotted red line is the fraction of cells at the second index level that have collisions.</figcaption>
</figure>
</div>
<p>When the order of the second refined index is low, the first index is larger resulting in fewer cells with collisions at the first index and more at the second.
The reverse is true when the order of second index is higher.
As the overall order is held constant, the same number of files are identified regardless of the orders of the first and second indexes.
The time required to identify the files is minimized when cells within the first index become saturated with collisions.
For secondary indexes of order 2 or lower, the large increase in performance offered for increases in the index order results from the reduction in the total complexity of the index which translates to shorter times for bitwise operations and less memory required for storage.
Above 2nd order, the overhead from storing and accessing more complex EWAH compressed bitmaps for each collision begins to flatten the memory scaling and increase the time required for queries.
However, selections using higher order secondary indices still require less time than in the case where only a single index is used.</p>
<p>The optimal value for the orders of the first and second indexes will depend upon the dataset in question.
The density of data points within the test dataset used here is relatively uniform throughout the domain and does not need a high level of refinement at collisions.
However, if a datasetwere less uniform with concentrations of points, the optimal order of the second index for performance may be higher.</p>
<h4 id="sec:test_size">Selector Size</h4>
<p>The time required to identify files touched by a selection will also depend upon the size of the region being selected.
Larger selectors will intersect more indices and more files, resulting in more bitmap operations.
Figure <a href="#fig:test_size">24</a> shows the result from varying the selector size.
The same test datasetfrom Section <a href="#sec:test_order">1.10.2.1</a> was used.
A bitmap index with a 4th order primary Morton index and 2nd order secondary Morton index was used in all cases.
Each cube selector was placed at the center of the domain and scaled along each dimension to some fraction of the total domain.</p>
<div id="fig:test_size" class="fignos">
<figure>
<img src="images/bitmap/vary_selector.png" alt="Figure 24: Dependence of query time (top) and number of files selected (bottom) on selector width in terms of the total domain width. The solid black lines correspond to the query times and files identified by the selectors alone. The dashed blue lines correspond the query times and additional files identified when a ghost zone with a width of one cell is added to the selector." />
<figcaption aria-hidden="true"><span>Figure 24:</span> Dependence of query time (top) and number of files selected (bottom) on selector width in terms of the total domain width.
The solid black lines correspond to the query times and files identified by the selectors alone.
The dashed blue lines correspond the query times and additional files identified when a ghost zone with a width of one cell is added to the selector.</figcaption>
</figure>
</div>
<p>As the selector increases in size, it touches a greater number of files, resulting in longer query times.
The number of files touched increases in steps due to the way the test dataset was partitioned between files.
Using the Hilbert curve, the domain covered by any one file is localized and will have a rectangular shape.
This results in an ordered structure that is similar along all dimensions.
An increase in the number of files touched indicates that the selector has grown past a file boundary in all directions.
It is just prior to these jumps that ghost files are present.
If the selector edge is near a file boundary, ghost zones have the potential to overlap the domains contained by neighboring files that are not already touched by the selector.
For such a highly ordered dataset, the ghost zones will only identify additional files for selectors that are nearing the edges of file boundaries.
However, queries including ghost zones require slightly more time even when this is not the case.</p>
<h4 id="sec:test_decomp">Domain Partitioning</h4>
<p>As discussed in Section <a href="#sec:decomp">1.9.1.1</a>, a bitmap index is more effective in cases where the domain is partitioned between files in a localized way.
If files contain non-contiguous parts of the domain, contiguous selections will require more files to be loaded.
Figure <a href="#fig:test_decomp">25</a> shows results for four different partitioning schemes.
All four data sets cover the same three-dimensional domain with <span class="math inline">\(1024^3\)</span> points split across 512 files.
The Hilbert dataset is the same one used in previous tests (see Section <a href="#sec:test_order">1.10.2.1</a> for a description).
The Morton dataset is constructed in a similar way to the Hilbert dataset with file partitions occurring along a 6th order Morton curve and including a 10% scatter of points between Morton cells.
The sliced dataset is partitioned in slices along one dimension with 10% scatter of points between adjacent slices.
Files in the random dataset contain a random sample of points, uniformly distributed across the domain.</p>
<div id="fig:test_decomp" class="fignos">
<figure>
<img src="images/bitmap/vary_decomp_to0.png" alt="Figure 25: Dependence of query time (top), the number of files selected (middle), and the size of the index (bottom) on index order for different domain partitioning between files. The dotted magenta lines are for a randomly partitioned dataset, the cyan dashed-dotted lines are a dataset partitioned by equal slices alone one dimension, the dashed red lines are a dataset partitioned along an 6th order Morton curve, and the solid blue lines are a dataset partitioned along a 6th order Hilbert curve. NOTE: it looks like the Morton curve is almost always behind the Hilbert one. Can you use zorder to put it at the frontrow instead so that we can see both ?" />
<figcaption aria-hidden="true"><span>Figure 25:</span> Dependence of query time (top), the number of files selected (middle), and the size of the index (bottom) on index order for different domain partitioning between files.
The dotted magenta lines are for a randomly partitioned dataset, the cyan dashed-dotted lines are a dataset partitioned by equal slices alone one dimension, the dashed red lines are a dataset partitioned along an 6th order Morton curve, and the solid blue lines are a dataset partitioned along a 6th order Hilbert curve.
<strong>NOTE: it looks like the Morton curve is almost always behind the Hilbert one. Can you use zorder to put it at the frontrow instead so that we can see both ?</strong></figcaption>
</figure>
</div>
<p>Many more files are identified for the random dataset than those datasets with localized partitioning of the domain.
Above an order of 3, very few files could be excluded for the random dataset.
This was not true for the localized partitioning schemes.
At the highest order, only <span class="math inline">\(\sim20-30\)</span>% of the files within these datasets would need to be loaded in order to get all of the data within the selected regions, while <span class="math inline">\(\sim80\)</span>% of the files in the random dataset would be required.
The smallest fraction of files were identified for the Hilbert and Morton datasets, with a slightly greater fraction being identified for the sliced dataset.
For a 6th order index and below, queries on the Morton and Hilbert partitioned datasets are the fastest.
An index order of 7 provides refinement beyond the 6th order curves used to partition the dataset between the files and the required for queries on these datasets increases dramatically.
The sliced dataset performed particularly well in this case because the selectors used were cubes and did not preferentially select along any one dimension.</p>
<p>Overall, this technique offers a considerable improvement in performance over other methods that require reading, evaluating and discarding all of the particles.</p>
<h3 id="sec:discuss">Summary &amp; Discussion</h3>
<p>Mapping files using Morton bitmap indexes speeds up analysis of large datasets split across multiple files by reducing the number of files that need to be loaded in order to perform operations on a subset of the full domain.
The time required for making selections using the bitmap index is minimal for even large datasets and can be optimized by partitioning the domain between files in a localized way and using an index or indexes of appropriate order for the dataset.</p>
<p>Without an index, queries require loading the data contained in every file into memory and then searching the data for those points that are selected by the query.
With a bitmap index, queries require loading the index, using it to identify the files touched by the query, reading in the data contained within the identified files, and searching the data for points selected by the query.
In this way, the bitmap index can decrease the computational cost of reading in the data and selecting data points if it identifies a subset of the total number of files.
While using an existing bitmap index decreases the time required for queries in this case, constructing a bitmap index can be more computationally expensive than directly querying the data without a bitmap index.
Therefore, in the case where only a small number of selections need to be made, it will be more efficient to perform direct queries of the data than to construct and utilize the bitmap index.</p>
<p>Bitmap indexing is particularly useful in astronomy and astrophysics.
Output from N-body simulations is often split between multiple files to take advantage of parallel I/O and the domain decomposition generally leads to localized partitioning between files <span class="citation" data-cites="DQa79Dfo I01oLUIj 18bL4U0kM">[<a href="#ref-I01oLUIj" role="doc-biblioref">17</a>,<a href="#ref-DQa79Dfo" role="doc-biblioref">37</a>,<a href="#ref-18bL4U0kM" role="doc-biblioref">38</a>]</span>.</p>
<p>Currently, this technique is most useful for datasets split across multiple files.
However, it can also be applied to single files by dividing the file’s contents into chunks.
As in the multi-file case, the single file would need to be organized such that chunks were localized within the domain to take full advantage of the bitmaps.
In addition, while the current implementation of this method is designed for three-dimensional spatial datasets like those produced by astrophysical simulations, the same methods can be applied to non-spatial datasets with arbitrary dimensionality.</p>
<h3 id="sec:code">Code</h3>
<p>These procedures have been incorporated into yt as of version 4.0, but then as of version 4.2 have been extracted into a standalone package entitled <a href="https://pypi.org/project/ewah-bool-utils/"><code>ewah-bool-utils</code></a>.
It is our hope that this will expand its application to non-yt uses.
The open source EWAHBoolArray C++ package is used for implementing EWAH bitmaps <span class="citation" data-cites="zXsFmNjY 2qfATif3">[<a href="#ref-zXsFmNjY" role="doc-biblioref">34</a>,<a href="#ref-2qfATif3" role="doc-biblioref">35</a>]</span> and exposed to Python using Cython <span class="citation" data-cites="RNeSxk82">[<a href="#ref-RNeSxk82" role="doc-biblioref">39</a>]</span>.
The authors would like to thank Daniel Lemire for his open source EWAH implementation.</p>
<h2 id="sec:data_formats_and_frontends">Data Formats and Frontends</h2>
<h3 id="adaptahop">adaptahop</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> Halo finding format</p>
<h3 id="ahf">ahf</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> Halo finding format</p>
<h3 id="amrvac">amrvac</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> MPI-AMRVAC is a parallel adaptive mesh refinement framework aimed at solving (primarily hyperbolic) partial differential equations by a number of different numerical schemes.
The emphasis is on (near) conservation laws and on shock-dominated problems in particular.</p>
<p><strong>Primary Citations:</strong> <span class="citation" data-cites="zZcl1XfZ aP30JL4t ooR7tjmM">[<a href="#ref-zZcl1XfZ" role="doc-biblioref">40</a>,<a href="#ref-aP30JL4t" role="doc-biblioref">41</a>,<a href="#ref-ooR7tjmM" role="doc-biblioref">42</a>]</span></p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="qhbp5GQg 13jLWXZs6 rbiXHtk4">[<a href="#ref-qhbp5GQg" role="doc-biblioref">43</a>,<a href="#ref-13jLWXZs6" role="doc-biblioref">44</a>,<a href="#ref-rbiXHtk4" role="doc-biblioref">45</a>]</span></p>
<h3 id="arepo">arepo</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> While listed as a “particle” dataset, AREPO data is properly described as a moving mesh.
For visualization purposes, yt treats it similarly to an SPH dataset; some efforts have been made to properly apply nearest-neighbor interpolation, but these have not been mainlined in the yt visualization routines as of yet.</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="11kNsPLrg K2xKlXYH">[<a href="#ref-11kNsPLrg" role="doc-biblioref">46</a>,<a href="#ref-K2xKlXYH" role="doc-biblioref">47</a>]</span></p>
<h3 id="art">art</h3>
<p><strong>Index Type:</strong> octree, particle</p>
<p><strong>Comments:</strong> ART, or ‘NMSU-ART’ as it is occasionally referred to internally, is an octree-based code that is often used for cosmology simulations and galaxy formation simulations.
yt supports two sets of data formats in the NMSU-ART family.
The first is a particle-only format, where the particles are stored independently of any gas structure.
The second is the hydro data format, wherein gas values are stored organized into an octree.
Each of these constitutes a different indexing system, sharing similar datatypes, units, and so forth.</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="11kNsPLrg 19CnTqnJi">[<a href="#ref-11kNsPLrg" role="doc-biblioref">46</a>,<a href="#ref-19CnTqnJi" role="doc-biblioref">48</a>]</span></p>
<h3 id="artio">artio</h3>
<p><strong>Index Type:</strong> octree</p>
<p><strong>Comments:</strong> The ARTIO frontend, which shares some ancestry with the <code>art</code> or ‘NMSU-ART’ frontend, utilizes a special purpose library for input/output operations called <code>libartio</code>, which is bundled with yt.</p>
<p>Utilizing <code>libartio</code> is interesting from a historical perspective, as it provided an opportunity to reorganize the way that yt understood indexing of octree datasets.
Whereas previously yt required a full understanding of data distribution, the <code>artio</code> frontend outsourced much of that to the internal library, which it regarded as essentially an opaque API.
Reducing the amount of internal bookkeepping that yt had to do provided a lighter-weight interface to the data, which in turn has enabled other frontends to be created using opaque APIs.</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="8yudbbM3">[<a href="#ref-8yudbbM3" role="doc-biblioref">49</a>]</span></p>
<h3 id="athena">athena</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="12IBqVsGq">[<a href="#ref-12IBqVsGq" role="doc-biblioref">50</a>]</span></p>
<h3 id="athena_pp">athena_pp</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="159ZGwZYu">[<a href="#ref-159ZGwZYu" role="doc-biblioref">51</a>]</span></p>
<h3 id="boxlib">boxlib</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> BoxLib is the predecessor to the AMReX adaptive mesh refinement library <span class="citation" data-cites="juPUBGqN">[<a href="#ref-juPUBGqN" role="doc-biblioref">52</a>]</span> and supports all BoxLib and AMReX codes that use the native BoxLib binary output. AMReX implements patch-based AMR following Berger and Colella <span class="citation" data-cites="ETcy2ZP4">[<a href="#ref-ETcy2ZP4" role="doc-biblioref">53</a>]</span> with typical refinement jumps of 2 or 4 between levels (this factor can vary in the level hierarchy).</p>
<p><strong>Primary Citations:</strong> <span class="citation" data-cites="juPUBGqN 9y1W7z3a Ul6Zyy8M 15s0VliuP">[<a href="#ref-juPUBGqN" role="doc-biblioref">52</a>,<a href="#ref-9y1W7z3a" role="doc-biblioref">54</a>,<a href="#ref-Ul6Zyy8M" role="doc-biblioref">55</a>,<a href="#ref-15s0VliuP" role="doc-biblioref">56</a>]</span></p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="1HoHM1Gnp Igp8qi2S 1HiWhAc3l rfkfEHN0 aETt4JEy whh3mxKo">[<a href="#ref-1HoHM1Gnp" role="doc-biblioref">57</a>,<a href="#ref-Igp8qi2S" role="doc-biblioref">58</a>,<a href="#ref-1HiWhAc3l" role="doc-biblioref">59</a>,<a href="#ref-rfkfEHN0" role="doc-biblioref">60</a>,<a href="#ref-aETt4JEy" role="doc-biblioref">61</a>,<a href="#ref-whh3mxKo" role="doc-biblioref">62</a>]</span></p>
<h3 id="cholla">cholla</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> CHOLLA is a uniform-resolution grid code operating almost entirely on GPUs.</p>
<p><strong>Primary Citations:</strong> <span class="citation" data-cites="rUJIl3i2">[<a href="#ref-rUJIl3i2" role="doc-biblioref">63</a>]</span></p>
<h3 id="chombo">chombo</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="MDEDwL4n Mz86NzmT pHJo1lO1">[<a href="#ref-MDEDwL4n" role="doc-biblioref">64</a>,<a href="#ref-Mz86NzmT" role="doc-biblioref">65</a>,<a href="#ref-pHJo1lO1" role="doc-biblioref">66</a>]</span></p>
<h3 id="eagle">eagle</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="158VPhStY tZ9VJJVi qsP6SRrG 158VPhStY">[<a href="#ref-158VPhStY" role="doc-biblioref">67</a>,<a href="#ref-158VPhStY" role="doc-biblioref">67</a>,<a href="#ref-tZ9VJJVi" role="doc-biblioref">68</a>,<a href="#ref-qsP6SRrG" role="doc-biblioref">69</a>]</span></p>
<h3 id="enzo">enzo</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> Enzo was the very first code supported by yt, and while many remnants of that have been removed, many still remain.
During the development of yt 3.0 in particular, many “enzo-isms” were removed; however, as Enzo uses a patch-based grid data format, much of the naming conventions for grid objects still mirror those in use in Enzo.
These include names such as <code>GridLeftEdge</code> and <code>ActiveDimensions</code> (which refers to the dimensions not including ghost zones).</p>
<p>There are several ‘sub-frontends’ within the Enzo frontend; these include dedicated 1D and 2D readers, as well as historical frontends (of varying functionality) for reading HDF4, ‘unpacked’ HDF5 and other sub-dialects of Enzo.</p>
<p><strong>Primary Citations:</strong> <span class="citation" data-cites="1AOjjAzwL">[<a href="#ref-1AOjjAzwL" role="doc-biblioref">70</a>]</span></p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="11kNsPLrg 12pekKHkD XEFlUFJ9 EKVFk61S QmIwzRzQ nKZ32a6">[<a href="#ref-11kNsPLrg" role="doc-biblioref">46</a>,<a href="#ref-12pekKHkD" role="doc-biblioref">71</a>,<a href="#ref-XEFlUFJ9" role="doc-biblioref">72</a>,<a href="#ref-EKVFk61S" role="doc-biblioref">73</a>,<a href="#ref-QmIwzRzQ" role="doc-biblioref">74</a>,<a href="#ref-nKZ32a6" role="doc-biblioref">75</a>]</span></p>
<h3 id="enzo_e">enzo_e</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> Enzo-E is a nearly full-rewrite of the Enzo infrastructure designed for emerging and future architectures.</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="AQXulJGy">[<a href="#ref-AQXulJGy" role="doc-biblioref">76</a>]</span></p>
<h3 id="exodus_ii">exodus_ii</h3>
<p><strong>Index Type:</strong> unstructured mesh</p>
<p><strong>Comments:</strong> Exodus II, and its output in NetCDF4 format, was the initial driver for higher-order unstructured mesh support in yt.
The original driver for developing this format was for simulations of nuclear reactors.</p>
<h3 id="fits">fits</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> None</p>
<h3 id="flash">flash</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> FLASH is a block-refinement adaptive mesh refinement code used in many different fields, including astrophysics and high-energy density physics.</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="1CqhBqTLw 1BmyNSium 3RHUErzU m3lUExr1 dqw1JuL wzYhh0PP">[<a href="#ref-1CqhBqTLw" role="doc-biblioref">77</a>,<a href="#ref-1BmyNSium" role="doc-biblioref">78</a>,<a href="#ref-3RHUErzU" role="doc-biblioref">79</a>,<a href="#ref-m3lUExr1" role="doc-biblioref">80</a>,<a href="#ref-dqw1JuL" role="doc-biblioref">81</a>,<a href="#ref-wzYhh0PP" role="doc-biblioref">82</a>]</span></p>
<h3 id="gadget">gadget</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="UiYJOB8x">[<a href="#ref-UiYJOB8x" role="doc-biblioref">83</a>]</span></p>
<h3 id="gadget_fof">gadget_fof</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> Halo finding</p>
<h3 id="gamer">gamer</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Primary Citations:</strong> <span class="citation" data-cites="12qB7HhOD">[<a href="#ref-12qB7HhOD" role="doc-biblioref">84</a>]</span></p>
<h3 id="gdf">gdf</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> Custom built for yt</p>
<h3 id="gizmo">gizmo</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="19Yvj8KDs 11kNsPLrg">[<a href="#ref-11kNsPLrg" role="doc-biblioref">46</a>,<a href="#ref-19Yvj8KDs" role="doc-biblioref">85</a>]</span></p>
<h3 id="halo_catalog">halo_catalog</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<h3 id="http_stream">http_stream</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<h3 id="moab">moab</h3>
<p><strong>Index Type:</strong> unstructured mesh</p>
<p><strong>Comments:</strong> None</p>
<h3 id="nc4_cm1">nc4_cm1</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> None</p>
<h3 id="open_pmd">open_pmd</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<h3 id="owls">owls</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<h3 id="owls_subfind">owls_subfind</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<h3 id="ramses">ramses</h3>
<p><strong>Index Type:</strong> octree</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="P48PyrZ1 19Fc0UYAA WeeFmczr zrKJQN1c KeH9yqn6 14vlGPzZS Gj37qtQ OgdghMdw 1f52Kyt lBI1SKpf jwp8Rkvm 39ZLqJPZ aLNHrylH 11r9ZRs87 11kNsPLrg RMe7IYaY 8ElS8qaE NLdlIida WeeFmczr JV9vVQtk Z9ORyk4e JV9vVQtk">[<a href="#ref-11kNsPLrg" role="doc-biblioref">46</a>,<a href="#ref-P48PyrZ1" role="doc-biblioref">86</a>,<a href="#ref-19Fc0UYAA" role="doc-biblioref">87</a>,<a href="#ref-WeeFmczr" role="doc-biblioref">88</a>,<a href="#ref-WeeFmczr" role="doc-biblioref">88</a>,<a href="#ref-zrKJQN1c" role="doc-biblioref">89</a>,<a href="#ref-KeH9yqn6" role="doc-biblioref">90</a>,<a href="#ref-14vlGPzZS" role="doc-biblioref">91</a>,<a href="#ref-Gj37qtQ" role="doc-biblioref">92</a>,<a href="#ref-OgdghMdw" role="doc-biblioref">93</a>,<a href="#ref-1f52Kyt" role="doc-biblioref">94</a>,<a href="#ref-lBI1SKpf" role="doc-biblioref">95</a>,<a href="#ref-jwp8Rkvm" role="doc-biblioref">96</a>,<a href="#ref-39ZLqJPZ" role="doc-biblioref">97</a>,<a href="#ref-aLNHrylH" role="doc-biblioref">98</a>,<a href="#ref-11r9ZRs87" role="doc-biblioref">99</a>,<a href="#ref-RMe7IYaY" role="doc-biblioref">100</a>,<a href="#ref-8ElS8qaE" role="doc-biblioref">101</a>,<a href="#ref-NLdlIida" role="doc-biblioref">102</a>,<a href="#ref-JV9vVQtk" role="doc-biblioref">103</a>,<a href="#ref-JV9vVQtk" role="doc-biblioref">103</a>,<a href="#ref-Z9ORyk4e" role="doc-biblioref">104</a>]</span></p>
<h3 id="rockstar">rockstar</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<h3 id="sdf">sdf</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="1Ce48o4Xl">[<a href="#ref-1Ce48o4Xl" role="doc-biblioref">105</a>]</span></p>
<h3 id="sph">sph</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> A base frontend from which other SPH-type frontends can be defined.</p>
<h3 id="stream">stream</h3>
<p><strong>Index Type:</strong> multi</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="VVJcMOza 16ZVBcAT2">[<a href="#ref-VVJcMOza" role="doc-biblioref">106</a>,<a href="#ref-16ZVBcAT2" role="doc-biblioref">107</a>]</span></p>
<h3 id="swift">swift</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Primary Citations:</strong> <span class="citation" data-cites="10pVv5dtC">[<a href="#ref-10pVv5dtC" role="doc-biblioref">108</a>]</span></p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="17ydwlTdn 100I0ZLek">[<a href="#ref-17ydwlTdn" role="doc-biblioref">109</a>,<a href="#ref-100I0ZLek" role="doc-biblioref">110</a>]</span></p>
<h3 id="tipsy">tipsy</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> None</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="YCT49Lcv 100MO1Ryn">[<a href="#ref-YCT49Lcv" role="doc-biblioref">111</a>,<a href="#ref-100MO1Ryn" role="doc-biblioref">112</a>]</span></p>
<h3 id="ytdata">ytdata</h3>
<p><strong>Index Type:</strong> particle</p>
<p><strong>Comments:</strong> An internal frontend for re-loading saved data and derived data</p>
<p><strong>Usage Citations:</strong> <span class="citation" data-cites="kWd16rLo">[<a href="#ref-kWd16rLo" role="doc-biblioref">113</a>]</span></p>
<h3 id="parthenon">parthenon</h3>
<p><strong>Index Type:</strong> grid</p>
<p><strong>Comments:</strong> Parthenon is a performance portable block-structured adaptive mesh refinement framework and the basis for various downstream codes such as AthenaPK, Phoebus, or KHARMA.</p>
<p><strong>Primary Citations:</strong> <span class="citation" data-cites="kEEsXbtq">[<a href="#ref-kEEsXbtq" role="doc-biblioref">114</a>]</span></p>
<p>The primary method by which researchers interact with their data in <code>yt</code> is via visualization; from the standpoint of the library, however, this is a side-effect of the various analysis, regularization and data-processing algorithms that are implemented within <code>yt</code>.
Nearly all of the visualization that is done using <code>yt</code> utilizes the matplotlib library for actual deposition of pixels into an image format, although all of the <em>input</em> to that deposition is conducted by <code>yt</code>.
Making this distinction is important, because it underscores the relationship between the different libraries and how they exist in the ecosystem of scientific software; <code>yt</code> does not replace matplotlib, but rather, augments it by providing a grammar of analysis of volumetric data and defining how that grammar is translated into visual representations as presented by matplotlib.</p>
<p><strong>CC: discuss ray traversal for patch-based datasets + oct-based datasets.</strong></p>
<h3 id="sec:pixelization">Pixelizing Variable-Mesh Objects</h3>
<p>The results of either projecting or slicing through a logically-cartesian finite volume dataset is represented in <code>yt</code> as a collection of pixel positions and widths.
These objects, hereafter referred to as exposing the “variable mesh” interface (as originated in HippoDraw), are not typically suitable for direct visualization.
Many visualization libraries, including matplotlib, would necessarily regard these as collections of patches of fixed size, supplying them to the underlying engine.
To optimize for repeated rendering, <code>yt</code> provides its own “pixelization” routines that take advantage of the input data structures.
These “pixelizers” (or “rasterizers”) can account for periodic data, variable resolution, overlapping and disjoint datasets, and non-Cartesian coordinate systems.</p>
<p>The pixelizers in <code>yt</code> are implemented in Cython, and they accept an input “image plane” buffer (with extent) as well as the variable mesh to be deposited.
Pixelizers exist for cartesian coordinates, cylindrical and spherical coordinates, off-axis cartesian planes, and for the Mollweide orthographic projection.
Each of these pixelizers follows a roughly identical process for depositing source pixels into the image plane.
The outer loop is over the input pixels, <span class="math inline">\(p_i\)</span>, composed of <span class="math inline">\(x_i, y_i, dx_i, dy_i, v_i\)</span>, where <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> refer to the coordinate system; in practice this means they may actually represent the <span class="math inline">\(r\)</span>, <span class="math inline">\(\theta\)</span>, <span class="math inline">\(\phi\)</span> or other coordinates.</p>
<ol type="1">
<li>Compute left and right edges of the bounding box for this pixel in the resolution of the image plane</li>
<li>Iterate over the first image plane coordinate from the left edge to the right edge of the bounding box</li>
<li>Iterate over the second image plane coordinate from the left edge to the right edge of the bounding box</li>
<li>Map from the coordinate system to the image plane and deposit <span class="math inline">\(v\)</span></li>
</ol>
<p>In practice, this is a fast operation, as long as the inner loops are sufficiently well determined; for instance, when depositing an input pixel with a width of <span class="math inline">\(w\)</span> into an image plane where the pixel width corresponds to a width of <span class="math inline">\(w/16\)</span>, only <span class="math inline">\(16^2\)</span> pixels (with a high-degree of sequential ordering) have to be iterated over.
The spherical and cylindrical pixelization routines operate similarly, but are somewhat degraded by a lower degree of locality in the final mapping from coordinate system to image plane.</p>
<p>Recent work has been done to port the pixelization routines to Rust and compiling these to WebAssembly, resulting in the development of the Widgyts project.
Widgyts provides a browser-side Jupyterlab interface to the pixelization routines, enabling extremely low-latency exploration of datasets.</p>
<h3 id="higher-order-unstructured-mesh-elements">Higher-Order Unstructured Mesh Elements</h3>
<h3 id="sec:software-volume-rendering">Software Volume Rendering</h3>
<p>The volume rendering is based on classical concepts for rendering 3D objects, and relies on the notion of a scene, a camera and an object to render.
The object to render can be any data container of supported AMR datasets (either patch-based and octree-based datasets).
The implementation of the volume rendering is based on integrating a transfer function along individual “rays.”
By defining a traversal function, cells can be ordered and a sampling function and accumulation can be applied to them.
Below, we describe the traversal operations, and then the different “sampling” operations that can be applied.</p>
<p>Conceptually, it is easies to think about this as constructing a global “ordering” of the cells, and then an in-order application of the sampler function with affiliated ray-specific data storage.
In reality, however, these operations are interleaved and decomposed to enable task- and data-based parallelism.
Data is distributed to different processors wherever possible based on different decomposition strategies, and within the traversal algorithms OpenMP parallelism is utilized to allow for individual rays to be integrated simultaneously.
Typically, multiple “sampling points” are identified within each element to be sampled; the default number is 5.
We have found that the image quality improves somewhat as this number is increased, as the interpolation allows for finer-grained variation, but 5 is typically sufficient for all but the highest resolution data and highest-gradient data.</p>
<h4 id="lenses-for-volume-rendering">Lenses for Volume Rendering</h4>
<p>In addition to specifying traversal algorithms, the software volume rendering in yt relies on the concept of “lenses.”
A lens, in yt terminology, is a descriptive object from which emanate the rays that are integrated.
The traditional lens is that of the perspective projection, wherein a plane of rays are sent outward.
The simplest lens, however, is that of a plane-parallel set of rays, each of which is emitted by an individual pixel with constant direction.
Additional lenses yt provides include
a stereo projection lens (for stereoscopic 3D volume rendering),
a fisheye lens (for hemispheric rendering suitable for planetarium domes),
a spherical lens (for spherical rendering suitable for full-sphere projections (such as for “<span class="math inline">\(360^{\circ}\)</span> YouTube videos),
and the stereoscopic spherical lens.</p>
<p>Each of these lens objects provides a functional form of ray direction and origin.
In some cases, an inverse function can be provided, which in some cases enables the traversal function to more intelligently check intersection and reduce computation time.</p>
<h4 id="patch-based-ray-traversal">Patch-based ray traversal</h4>
<p>Unlike taking standard “projections” of datasets, the process of volume rendering requires that elements be traversed in order, accumulating values that may be reliant on <em>previous</em> values.
While this is straightforward to implement for grids <span class="citation" data-cites="8BOQ3JKi">[<a href="#ref-8BOQ3JKi" role="doc-biblioref">115</a>]</span>, in situations where there are overlapping sets of values the process must be handled more carefully.</p>
<canvas id="vrFigure" style="width:100%;height:100%;min-height: 500px;" hidpi="true">
</canvas>
<script type="text/paperscript" canvas="vrFigure" src="images/volume_rendering/vr-figure.js"></script>
<p>To ensure that cells are traversed in-order, and to ensure that only the highest-resolution cells are used in cases where grids overlap, we decompose the grid hierarchy into a kD-tree.
To construct the kD-tree, each grid at the root level is added to the tree.
Wherever these grids include “child” grids (i.e., grids that include higher-resolution data that overlaps with the grids) a cutting plane is inserted in the kD-tree such that the child grid is isolated.
This process is then applied recursively until all of the grids in the dataset (or its data subselection) have been added.
By inserting cutting planes at child-grid boundaries, the tree can be specified to include only “leaf” node data, thus ensuring that any traversal will <em>only</em> include the highest-resolution data available, and thus can be traversed in order.</p>
<p>kD-trees also provide so-called “viewpoint traversals,” which allows the grids to be ordered in either front-to-back or back-to-front order.
These traversals can also be ordered to allow for radially-emanating traversals, for such cases as hemispheric renderings.</p>
<h4 id="octree-ray-traversal">Octree ray traversal</h4>
<p>Casting rays through an octree can be achieved efficiently by relying on the octree structure.
In order to abstract away the underlying layout of the data, we first construct an octree that contains all leaf cells in the data container.
We store all cells as octs with no children, and mark them with their position within the data container, going from <span class="math inline">\(0\)</span> to <span class="math inline">\(N_\mathrm{cell}-1\)</span>.
Octs that are inserted in the process of building the tree are not marked nor indexed.
We also compute the vertex-centered data for all cells in the container.
Note that, contrary to the octree utilized to index the data from octree datasets, this octree may span multiple domains and contains all levels from the root level (that contains a single oct the size of the simulation domain) all the way down to the leaf cells.
We then cast rays off the camera, and for each ray, we compute the ordered list of the <span class="math inline">\(N\)</span> cells it intersects with together with the intersection points along the ray <span class="citation" data-cites="FQhLVjwO">[<a href="#ref-FQhLVjwO" role="doc-biblioref">116</a>]</span>.
In the following, we will write the coordinate along the ray as <span class="math inline">\(t\)</span>, with the camera located at <span class="math inline">\(t=0\)</span>.
In general the tree may contain holes (this may happen if the data container is a region selector), so that the exit coordinate out of a cell may not coincide with the entry coordinate through the next cell. In practice, we solve this by storing for each cell both the entry and exit coordinates of the ray.</p>
<p>The algorithm relies on the fact that if a ray passes through an oct and intersects with its six faces at coordinate along the ray <span class="math inline">\(t_{xi}, t_{yi}, t_{zi}\)</span> (on entry) and <span class="math inline">\(t_{xo}, t_{yo}, t_{zo}\)</span> (on exit), then its intersection with the inner cells’ faces can be computed explicitly from these six values and their half point <span class="math inline">\((t_{xi}+t_{xo})/2, (t_{yi}+t_{yo})/2, (t_{zi}+t_{zo})/2\)</span>. This implies that each call to the algorithm only need computing one division and a few simple arithmetic comparisons.
It also uses the fact that for a given oct, we can compute which cell the ray will intersect with first, and from any given cell, which cell the the ray will intersect next.</p>
<p>The algorithm then works as follows. If the ray does not intersect with the root oct, then the algorithm returns an empty list of cell crossed and <span class="math inline">\(t\)</span> values.
Otherwise, initialize an empty list of cells traversed and <span class="math inline">\(t\)</span>-values. <em>a)</em> Find the intersection of the ray with all six faces of the oct. <em>b)</em> If the current oct is marked, store the entry and exit <span class="math inline">\(t\)</span>-values and the index of the oct in their respective list and return. <em>c)</em> If the current oct is not marked, find the first cell the ray intersects with and call the algorithm recursively (starting at step <em>a)</em> with the oct contained in the cell, if any). <em>d)</em> Find the next cell within the oct. If there is no next cell, return. Otherwise, call the algorithm recursively (starting at step <em>a)</em> with the oct contained in the cell, if any) then go back to <em>d)</em>.
On exit of the algorithm, we then have a list of cells and <span class="math inline">\(t\)</span>-values.
For each cell in the list, we then call the sampler with the vertex-centered values and the entry and exit coordinates.</p>
<p>An example of the volume rendering of a galaxy in a zoom-in cosmological simulation made with RAMSES is shown on Figure <a href="#fig:ramses-volume-rendering">26</a>.</p>
<div id="fig:ramses-volume-rendering" class="fignos">
<figure>
<img src="images/volume_rendering/ramses-volume-rendering-galaxy.png" alt="Figure 26: Volume rendering of gas density isocontours around a galaxy in a cosmological zoom-in simulation performed with RAMSES. Adapted from [117]." />
<figcaption aria-hidden="true"><span>Figure 26:</span> Volume rendering of gas density isocontours around a galaxy in a cosmological zoom-in simulation performed with RAMSES. Adapted from <span class="citation" data-cites="zRE8zhVC">[<a href="#ref-zRE8zhVC" role="doc-biblioref">117</a>]</span>.</figcaption>
</figure>
</div>
<h4 id="sampling-functions">Sampling functions</h4>
<p>As a ray passes through each sampling point, we apply a sampling function.
In the yt software volume renderer, these cython subclasses of the <code>ImageSampler</code> base class.
Each defines a setup process and a sample process.
Once a ray arrives at a sampling point (for instance, as it walks across a grid and encounters values) the sampler is called with the data it is accessing, information about the ray, the “enter” and “exit” parameters of the sample region, and the accumulated data so far.</p>
<p>Each <code>ImageSampler</code> object implements a different method of integration through the volume.
Below, we discuss each in turn.
To do so, we apply the convention that our volume rendering ray-casting starts at a position defined by the parameter <span class="math inline">\(t\)</span>, starting at the position <span class="math inline">\(\mathbf{\vec{x}}(t_0) \equiv \mathbf{\vec{x}}_0\)</span> and terminates at <span class="math inline">\(\mathbf{\vec{x}}(t_1) \equiv \mathbf{\vec{x}}_1\)</span>.
Field values are taken as <span class="math inline">\(f(\mathbf{\vec{x}})\)</span>, and the distance through with a ray has passed is defined as the path length.</p>
<h5 id="projection-sampler">Projection Sampler</h5>
<p>This sampler object conducts a simple integration through the domain.
<span class="math display">\[
v(\mathbf{\vec{x}}(t_1)) = \int_{t_0}^{t_1} f(x, y, z) dl
\]</span></p>
<p>Because the sampling function includes no dependence on previous values, it can be sampled completely independently and out of order.
This allows for much simpler data-based parallelism.
There are two formulations of this sampler in yt; the first is a piecewise-linear interpolation scheme, wherein the field is assumed to be constant within each sampling volume.
This method requires neither vertex-centered data <em>nor</em> multiple sampling points and is considerably faster to conduct.
The second method available, which is the default, is that of an interpolated projection.
This sampler requires vertex-centered data and utilizes multiple samples.
Because the arithmetic is reasonably simple and the data can be supplied in any order, it is still reasonably fast to conduct.</p>
<p>This sampler function operates independently of color channels, exclusively in floating-point space.
This allows colormapping to be conducted at the end of the process and allows for high dynamic range beyond that typically enabled by RGBA channels.</p>
<h5 id="transfer-function-sampler">Transfer Function Sampler</h5>
<p>The transfer function sampler is the most complex of the sampling functions available in yt, as it affords the most flexibility.
Unfortunately, this also results in a cumbersome and difficult-to-describe setup process, wherein the simple operations are straightforward to accomplish but the more complex operations are subject to trial-and-error.</p>
<p>The transfer function sampler is designed for constructing images; as such, it provides integration for up to six channels of data.
Typically, this is used in the four-channel mode, corresponding to red, green, blue and alpha (RGBA) data.
(It is important to note, however, that this is not by any means the only mechanism that these could be assigned; the values are accumulated in floating point and could correspond to any output data, not just these three colors and one alpha.)
In six-channel mode, typically the channels are for red, green, blue, and a corresponding alpha channel for each of those colors, accumulated independently.</p>
<p>The transfer function is set up around the concept of “field interpolation tables.”
These are 1D arrays, typically 256 elements, that include floating point values as well as concrete bounds in data-space.
These tables provide a mapping <span class="math inline">\(f(v) \rightarrow u\)</span>, where <span class="math inline">\(v\)</span> is a local field value, as well as an optional “weight” value drawn from a second table.
Careful construction of these tables can produce volume renderings via isocontours, where specific values in field space are highlighted with distinct colors.
Indeed, this is the most common method of applying this sampler, as it allows a straightforward way of seeing “through” the outer layers and viewing nesting structures simultaneously.</p>
<p>We use notation <span class="math inline">\(C_i^0\)</span> for the input values for a given channel (where <span class="math inline">\(i\)</span> is one of <span class="math inline">\(R, G, B, A\)</span>) and <span class="math inline">\(C_i^1\)</span> for the output.
With the sampled field value (<span class="math inline">\(f(t)\)</span>) and optional weight field value (<span class="math inline">\(w(t)\)</span>) we can then define a sample, <span class="math inline">\(S_i\)</span> for each channel including alpha.
we compute the updated values for each sampling point via:</p>
<p><span class="math display">\[
C_i^0
a_t \leftarrow \mathrm{max}(1 - dt * S_A, 0) \\
C_i^1 \leftarrow S_i \delta t + a_t C_i^0
\]</span></p>
<p>This is the case for “grey” opacity.
In the use case of differing opacity for each channel, we modify this such that we utilize additional alpha channels for each primary channel.
<span class="math inline">\(a_t\)</span> is then defined individually for each output channel.</p>
<p>This particular set of transfer functions, with what amounts to multi-channel, multi-field sampling in both output and opacity, can be specified to perform complex functions such as variable scattering, different weights for different channels, etc.</p>
<h3 id="hardware-accelerated-volume-rendering">Hardware-accelerated Volume Rendering</h3>
<p>Software volume rendering, as described above, provides a number of affordances for careful visualization.
Specifically, as the code and kernels are written in languages that are similar to more traditional languages such as C and C++, the barrier to entry for describing a new sampling system can be lower.
That being said, when examining responsiveness, software volume rendering is rarely – if ever – competitive with hardware-based volume rendering, such as that accelerated through graphics processing units (GPUs) and using OpenGL, Metal, Vulkan, or one of the higher-level platforms for graphics.
Fluid interactivity is essentially inaccessible for software volume rendering except on the smallest of datasets.
And yet, fluid interactivity enables much deeper exploration of the parameter space of visualization, as well as the ability to immerse oneself in data.</p>
<p>To support more interactive visualization (in addition to that described in <a href="#sec:jupyter_integration">1.13.3</a>) a basic system for conducting hardware-accelerate, OpenGL-based visualization of <code>yt</code>-supported data has been developed in an external package, <a href="https://github.com/yt-project/yt_idv/"><code>yt_idv</code></a>.
<code>yt_idv</code> is built on <code>PyOpenGL</code> and provides support for grid-based, particle-based and finite element datasets, with a heavy emphasis on the grid-based datasets.
While the process of volume rendering is interesting, with many different fascinating areas of inquiry and opportunities for optimization, <code>yt_idv</code> is notable for its architecture more than its algorithms and optimization.</p>
<p><code>yt_idv</code> is built using the Traitlets library, and utilizes the immediate-mode graphical user interface system “Dear ImGUI” for presenting a user interface.
This allows for the system to be largely data-driven; frame redraws are only executed when a parameter changes (such as the camera path, transfer function, etc) and new parameters for the visualization can be easily exposed to the user interface.
Furthermore, this allows shaders to be dynamically recompiled only as necessary.</p>
<p>In addition to this, the shaders themselves are specified in a configuration file that links GLSL source files into multi-component shaders, allowing declarative construction of shader pipelines and improving interoperability.
For ray-casting routines, only a small kernel needs to be modified to change the functionality.
This allows the relatively complex process of casting rays through multiple volumes to be hidden, much as described in <a href="#sec:software-volume-rendering">1.11.41</a>.
Annotations can be added simply, and an event loop has been enabled so that users can interact with the running visualization interface through IPython.</p>
<p>While this project includes a number of additional features designed for accessibility of data and in-depth coupling of visualization with quantitative analysis, they extend beyond the scope of this paper.</p>
<h2 id="sec:units">Unitful Arrays and Quantities</h2>
<p>At a basic level, <code>yt</code> is an engine for converting data created by a simulation code into a physically meaningful result.
Attaching physical units to simulation data makes it possible to perform dimensional analysis on the simulation data, adding new opportunities for catching errors in a data processing pipeline.
In addition, it becomes straightforward to convert data from one unit system to another.</p>
<p>In <code>yt</code> 4.0 we handle units in an automatic fashion, with a symbolic units system.
Originally part of <code>yt</code> itself since version 3.0, the unit system has now been split off into a standalone package, <code>unyt</code> <span class="citation" data-cites="QcHRZTY3">[<a href="#ref-QcHRZTY3" role="doc-biblioref">118</a>]</span>.
<code>unyt</code> leverages the symbolic math library <code>sympy</code> in combination with NumPy.
In what follows, we will refer to <code>unyt</code> when describing the unit implementation, unless we are describing a <code>yt</code>-specific extension.</p>
<p>Instead of returning a NumPy <code>ndarray</code> when users query <code>yt</code> data objects for fields, we return a <code>unyt_array</code>, a subclass of <code>ndarray</code>.
<code>unyt_array</code> preserves <code>ndarray</code>’s array interface, including deep and shallow copies, broadcasting, views, and mathematical operations.
Augmenting <code>ndarray</code>, <code>unyt_array</code> attaches unit metadata to the array data, enabling runtime checking of unit consistency in arithmetic operations between <code>unyt_array</code> instances, and making it trivial to compose new units using algebraic operations.</p>
<p>As a trivial example, when one queries a data object (here given the generic name <code>dd</code>) for the density field, we get back a <code>unyt_array</code>, including both the simulation data for the density field, and the units of the density field, in this case <span class="math inline">\(\rm{g}/\rm{cm}^3\)</span>:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> dd[<span class="st">&#39;density&#39;</span>] </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>unyt_array([<span class="fl">4.92e-31</span>, <span class="fl">4.94e-31</span>, <span class="fl">4.93e-31</span>, ...,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>            <span class="fl">1.12e-25</span>, <span class="fl">1.59e-25</span>, <span class="fl">1.09e-24</span>], <span class="st">&#39;g/cm**3&#39;</span>)</span></code></pre></div>
<p>One of the nicest aspects of the unit system is that the symbolic algebra for unitful operations is performed automatically by <code>sympy</code>:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(dd[<span class="st">&#39;mass&#39;</span>]<span class="op">/</span>dd[<span class="st">&#39;cell_volume&#39;</span>])</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  [<span class="fl">4.92e-31</span> <span class="fl">4.94e-31</span> <span class="fl">4.93e-31</span> ... </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>   <span class="fl">1.12e-25</span> <span class="fl">1.59e-25</span> <span class="fl">1.09e-24</span>] g<span class="op">/</span>cm<span class="op">**</span><span class="dv">3</span></span></code></pre></div>
<p><code>unyt_array</code> is primarily useful for attaching units to NumPy <code>ndarray</code> instances.
For scalar data, we have created the <code>unyt_quantity</code> class.
<code>unyt_quantity</code> is a subclass of <code>unyt_array</code> with the requirement that the “array data” associated with the instance is a single scalar value.
<code>unyt_quantity</code> is primarily useful for physical constants and ensures that the units are propagated correctly when composing quantities from arrays, physical constants, and unitless scalars:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> boltzmann_constant</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(dd[<span class="st">&#39;temperature&#39;</span>]<span class="op">*</span>boltzmann_constant)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1.28e-12</span> <span class="fl">1.29e-12</span> <span class="fl">1.29e-12</span> ... </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a> <span class="fl">1.63e-12</span> <span class="fl">1.59e-12</span> <span class="fl">1.40e-12</span>] erg</span></code></pre></div>
<p>In what follows, we will refer to both array and scalar quantities as arrays generically for the purpose of brevity.</p>
<h3 id="sec:units-implementation">Basic implementation</h3>
<p>The unit implementation in <code>unyt</code> is based upon the notion of dimensionality.
<code>unyt</code> has 8 base dimensions: <code>mass</code>, <code>length</code>, <code>time</code>, <code>temperature</code>, <code>current_mks</code>, <code>luminous_intensity</code>, <code>angle</code>, and <code>logarithmic</code>.
The unitless <code>dimensionless</code> dimension is also technically a base dimension, although a trivial one.
This facilitates the creation of dimensionless unit symbols to represent quantities like metallicity that are formally dimensionless, but convenient to represent in a unit system.
We provide sympy <code>Symbol</code> objects for the base dimensions.
The dimensionality of all other units are <code>sympy</code> <code>Expr</code> objects made up of the base dimension objects and the <code>sympy</code> operation objects <code>Mul</code> and <code>Pow</code>.</p>
<p>This collection of dimensions is admittedly somewhat idiosyncratic.
In the SI system of units, there is no base dimension of angle (the radian is dimensionless formally defined as meter/meter), and there is a dimension of “amount of substance”, for which the base unit is mole, but in <code>yt</code> the <code>mol</code> unit is treated as dimensionless.
The <code>logarithmic</code> dimension is a special case which will be described in more detail below.</p>
<p>For each dimension, we choose a base unit.
The default base units for the first six dimensions mentioned above in <code>unyt</code> are from the SI system: kilograms, meters, seconds, kelvin, ampere, and candela.
The default base unit for the <code>angle</code> dimension is radian, and the default base unit for the <code>logarithmic</code> dimension is the “neper”, which is a logarithmic unit for ratios of quantities such as field strenth or power.
All other units can be described as combinations of these base units along with a conversion factor to equivalent base units.</p>
<p>For historical and practical reasons, the default base units in <code>yt</code> itself are the “Gaussian” centimeters-grams-seconds (CGS) system, where grams and centimeters serve as the different base units for mass and length, the other base units remaining the same as SI, though formally there are no independent base units for luminous intensity or current in CGS systems.
CGS was chosen for <code>yt</code> to stay consistent with the rest of the <code>yt</code> codebase prior to <code>yt</code> 3.0 and to reflect the standard practice in astrophysics.</p>
<p>In any case, using a <strong>physical</strong> unit system makes it possible to compare quantities and arrays produced by different datasets, possibly with different conversion factors to CGS and to code units.
We go into more detail on this point below, where we describe the different unit systems in yt.</p>
<p>Let us first take some common units as examples: gram (<code>g</code>), erg (<code>erg</code>), and solar mass per cubic megaparsec (<code>Msun / Mpc**3</code>).
<code>g</code> is an “atomic” CGS base unit, <code>erg</code> is an atomic unit in CGS, but is not a base unit, and <code>Msun / Mpc**3</code> is a combination of atomic units, which are not in CGS, and one of them even has an SI prefix.
The dimensions of <code>g</code> are <code>mass</code> and the CGS conversion factor is 1.
The dimensions of <code>erg</code> are <code>mass * length**2 / time**2</code> and the CGS conversion factor is 1.
The dimensions of <code>Msun / Mpc**3</code> are <code>mass / length**3</code> and the CGS conversion factor is about <span class="math inline">\(6.8 \times 10^{-41}\)</span>.</p>
<p>We use the <code>UnitRegistry</code> class to define all valid atomic units.
All unit registries contain a unit symbol lookup table containing for each unit in a tuple the conversion factor to SI units, the dimensionality, the additive offset for the unit, a string representing how the unit should be displayed in plots and other media, and a boolean dictating whether the unit is SI-prefixable.
Here is what it would look like with the above units:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>{ </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;g&quot;</span>:    (<span class="fl">1.0e-3</span>, dimensions.mass, <span class="fl">0.0</span>, <span class="vs">r&quot;\rm</span><span class="sc">{g}</span><span class="vs">&quot;</span>, <span class="va">True</span>),</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;erg&quot;</span>:  (<span class="fl">1.0e-7</span>, dimensions.energy, <span class="fl">0.0</span>, <span class="vs">r&quot;\rm</span><span class="sc">{erg}</span><span class="vs">&quot;</span>, <span class="va">True</span>),</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;Msun&quot;</span>: (mass_sun_kg, dimensions.mass, <span class="fl">0.0</span>, <span class="vs">r&quot;\rm</span><span class="sc">{M}</span><span class="vs">_\odot&quot;</span>, <span class="va">False</span>),</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;pc&quot;</span>:   (m_per_pc, dimensions.length, <span class="fl">0.0</span>, <span class="vs">r&quot;\rm</span><span class="sc">{pc}</span><span class="vs">&quot;</span>, <span class="va">True</span>), </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;degC&quot;</span>: (<span class="fl">1.0</span>, dimensions.temperature, <span class="op">-</span><span class="fl">273.15</span>, <span class="vs">r&quot;^\circ\rm</span><span class="sc">{C}</span><span class="vs">&quot;</span>, <span class="va">True</span>),</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Note that we only define <strong>atomic</strong> units here.
There should be no operations or SI prefixes in the registry symbol strings.
When we parse non-atomic units like <code>Msun/Mpc**3</code>, we use the registry to look up the symbols.
The unit system in <code>yt</code> knows how to handle units like <code>Mpc</code> by looking up unit symbols with and without prefixes and modify the conversion factor appropriately.
Note also that in most cases the “additive offset” will be 0–it applies mainly to temperature scales like Fahrenheit and Celsius where the zero-point of the scale are not equivalent between the two scales, and are also not equivalent to “zero thermal energy,” for which the relevant temperature units are Rankine and Kelvin, respectively.</p>
<p>We construct a <code>Unit</code> object by providing a string containing atomic unit symbols, combined with operations in Python syntax, and the registry those atomic unit symbols are defined in.
We use <code>sympy</code>’s string parsing features to create the unit expression from the user-provided string.
Each <code>Unit</code> object also has its own unit <code>sympy</code> <code>Expr</code> object, a dimensionality <code>Expr</code> object, a <code>UnitRegistry</code> instance, a scalar conversion factor to SI units, additive offset value, and display representation.
Among other methods, <code>Unit</code> defines the <code>mul</code>, <code>div</code>, <code>pow</code>, and <code>eq</code> operations with other unit objects, making it easy to compose compound units algebraically.</p>
<p>The <code>UnitRegistry</code> class provides methods which allows users to <code>add</code>, <code>remove</code>, and <code>modify</code> atomic unit definitions present in <code>UnitRegistry</code> objects.
In general, unit registries should only be adjusted inside of a code frontend, since otherwise quantities and arrays might be created with inconsistent unit metadata.
Once a <code>Unit</code> object is created, it will not receive updates if the original unit registry is modified.</p>
<h4 id="sec:code-units">Code units</h4>
<p>When a <code>Dataset</code> object is created, it will itself instantiate and set up a <code>UnitRegistry</code> class (stored in its <code>unit_registry</code> attribute) that contains a full set of units that are defined for the simulation.
This registry includes not only the aforementioned concrete physical units like <code>cm</code> or <code>K</code>, but also unit symbols that correspond to the unit system used internally in the simulation.
These units are instantiated via the concept of “code units”, which are tightly coupled to on-disk parameters.
Examples of such units are <code>code_length</code>, <code>code_mass</code>, <code>code_time</code>, <code>code_velocity</code>, <code>code_density</code>, <code>code_magnetic</code>, <code>code_pressure</code>, and <code>code_metallicity</code>.
When creating a <code>Dataset</code>’s <code>UnitRegistry</code>, <code>yt</code> is able to also determine conversions between these units and physical units in both directions.
Code units therefore preserve dimensionality: an array or quantity that has units of <code>cm</code> will be convertible to <code>code_length</code>, for example, and vice-versa.</p>
<p>On-disk fields are presented in these unconverted code units.
To obtain on-disk data, a user need only query a data object using an on-disk field name:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> yt</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds <span class="op">=</span> yt.load(<span class="st">&quot;Enzo_64/DD0043/data0043&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> dd <span class="op">=</span> ds.all_data()</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(dd[<span class="st">&quot;enzo&quot;</span>, <span class="st">&quot;Density&quot;</span>])</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>[<span class="fl">6.74e-02</span> <span class="fl">6.12e-02</span> <span class="fl">8.92e-02</span> ... </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a> <span class="fl">9.09e+01</span> <span class="fl">5.66e+01</span> <span class="fl">4.27e+01</span>] code_mass<span class="op">/</span>code_length<span class="op">**</span><span class="dv">3</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># (&quot;gas&quot;, &quot;density&quot;) is the same field, but presented in physical units</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(dd[<span class="st">&quot;gas&quot;</span>, <span class="st">&quot;density&quot;</span>])</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1.92e-31</span> <span class="fl">1.74e-31</span> <span class="fl">2.54e-31</span> ... </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a> <span class="fl">2.59e-28</span> <span class="fl">1.61e-28</span> <span class="fl">1.22e-28</span>] g<span class="op">/</span>cm<span class="op">**</span><span class="dv">3</span></span></code></pre></div>
<p>Here, the first data object query is returned in code units, while the second is returned in CGS units.
This is because <code>("enzo", "Density")</code> is an on-disk field, while <code>("gas", "density")</code> is an internal <code>yt</code> field, aliased to the former.</p>
<p>In the <code>set_code_units</code> method of the <code>Dataset</code> base class, the predefined <code>code_mass</code>, <code>code_length</code>, <code>code_time</code>, <code>code_velocity</code>, and <code>code_magnetic</code> symbols are adjusted to the appropriate values, and <code>length_unit</code>, <code>time_unit</code>, <code>mass_unit</code>, <code>velocity_unit</code>, and <code>magnetic_unit</code> attributes are attached to the <code>Dataset</code> instance.
If there are frontend-specific code units, they should also be defined in subclasses by extending this function.</p>
<p>When code units are set up for a <code>Dataset</code>, <code>unyt_quantity</code> instances corresponding to these code units are also set up and added as attributes to the <code>Dataset</code>.
Examples are <code>length_unit</code>, <code>time_unit</code>, <code>mass_unit</code>, and <code>temperature_unit</code>:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> yt</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds <span class="op">=</span> yt.load(<span class="st">&quot;Enzo_64/DD0043/data0043&quot;</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(ds.length_unit)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="dv">128</span> Mpccm<span class="op">/</span>h</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(ds.length_unit.in_cgs())</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fl">5.555172850264421e+26</span> cm</span></code></pre></div>
<p>Optionally, <code>velocity_unit</code>, <code>density_unit</code>, <code>pressure_unit</code>, and <code>magnetic_unit</code> may be defined as well, if the units for these fields cannot be inferred from the mass, length, and time units.</p>
<h4 id="sec:handling-cosmological-units">Handling cosmological units</h4>
<p>A special case of astrophysical hydrodynamical simulations are cosmological simulations, which attempt to simulate the evolution of structure on cosmological length and time scales.
Such simulations use “comoving coordinates”, which is a notion of length that does not change with the expansion of the universe, in contrast to normal physical distances which do.</p>
<p>If we detect that we are loading a cosmological simulation performed in comoving coordinates, extra comoving units are added to the dataset’s unit registry.
Comoving length unit symbols are still named following the pattern <code>&lt;length symbol&gt;cm</code>, i.e. <code>Mpccm</code>.
The <span class="math inline">\(h\)</span> symbol is treated as a base unit, <code>h</code>, which defaults to unity.
The <code>Dataset</code> updates the <code>h</code> symbol to the correct value when loading a cosmological simulation.</p>
<p>The change of physical length is illustrated by the following example, of two datasets from the same simulation but at different times/scale factors/redshifts:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> yt</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds1 <span class="op">=</span> yt.load(<span class="st">&#39;Enzo_64/DD0002/data0002&#39;</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds2 <span class="op">=</span> yt.load(<span class="st">&#39;Enzo_64/DD0043/data0043&#39;</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(ds1.length_unit, ds2.length_unit)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="dv">128</span> Mpccm<span class="op">/</span>h, <span class="dv">128</span> Mpccm<span class="op">/</span>h</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(ds1.length_unit.in_cgs())</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fl">6.26145538088e+25</span> cm</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(ds2.length_unit.in_cgs())</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="fl">5.55517285026e+26</span> cm </span></code></pre></div>
<h3 id="sec:unit-systems">Unit systems</h3>
<p>As noted above, the default unit system in <code>yt</code> for almost all <code>Dataset</code>s is the Gaussian CGS system.
However, <code>yt</code> also includes other unit systems which can be specified, in which case all fields loaded from a <code>Dataset</code> will be converted from their code units to this set of units.
The available unit systems in <code>yt</code> are:</p>
<ul>
<li><code>"cgs"</code>: “Gaussian” centimeter-grams-seconds</li>
<li><code>"mks"</code>: The SI unit system, or meters-kilograms-seconds-amperes</li>
<li><code>"imperial"</code>: The “imperial” unit system, or the “US customary” system, incorporating miles, pounds, seconds, etc.</li>
<li><code>"galactic"</code>: A unit system appropriate for studies of systems on galactic or extragalactic scales, with length units of kiloparsecs, mass units of solar masses, and time units of megayears, etc.</li>
<li><code>"solar"</code>: A unit system appropriate for solar system objects, with length units of astronomical units, mass units of Earth mass, and time units of years.</li>
<li><code>"geometrized"</code>: A unit system appropriate for general relativistic calculations, in which <span class="math inline">\(G\)</span> = <span class="math inline">\(c\)</span> = 1.</li>
<li><code>"planck"</code>: A unit system in Planck units where <span class="math inline">\(G\)</span> = <span class="math inline">\(c\)</span> = <span class="math inline">\(\hbar\)</span> = <span class="math inline">\(k_B\)</span> = 1.</li>
</ul>
<p>Additionally, each <code>Dataset</code> instance has its own <code>"code"</code> unit system.</p>
<p>To load a dataset with a different unit system, simply pass a valid value to the <code>unit_system</code> keywoard argument when calling <code>yt.load</code>:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> yt </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds <span class="op">=</span> yt.load(<span class="st">&quot;Enzo_64/DD0043/data0043&quot;</span>, unit_system<span class="op">=</span><span class="st">&quot;mks&quot;</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> dd <span class="op">=</span> ds.all_data()</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(dd[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;density&quot;</span>])</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1.92555234e-28</span> <span class="fl">1.74617139e-28</span> <span class="fl">2.54742946e-28</span> ... <span class="fl">2.59560386e-25</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a> <span class="fl">1.61728874e-25</span> <span class="fl">1.22032913e-25</span>] kg<span class="op">/</span>m<span class="op">**</span><span class="dv">3</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds2 <span class="op">=</span> yt.load(<span class="st">&quot;Enzo_64/DD0043/data0043&quot;</span>, unit_system<span class="op">=</span><span class="st">&quot;galactic&quot;</span>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> dd2 <span class="op">=</span> ds2.all_data()</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(dd2[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;density&quot;</span>])</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>[<span class="fl">2.84511447e+00</span> <span class="fl">2.58006879e+00</span> <span class="fl">3.76397373e+00</span> ... <span class="fl">3.83515418e+03</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a> <span class="fl">2.38963725e+03</span> <span class="fl">1.80310657e+03</span>] Msun<span class="op">/</span>kpc<span class="op">**</span><span class="dv">3</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds3 <span class="op">=</span> yt.load(<span class="st">&quot;Enzo_64/DD0043/data0043&quot;</span>, unit_system<span class="op">=</span><span class="st">&quot;code&quot;</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> dd3 <span class="op">=</span> ds3.all_data()</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(dd3[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;density&quot;</span>])</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>[<span class="fl">6.74992726e-02</span> <span class="fl">6.12111635e-02</span> <span class="fl">8.92988636e-02</span> ... <span class="fl">9.09875931e+01</span></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a> <span class="fl">5.66932465e+01</span> <span class="fl">4.27780263e+01</span>] code_mass<span class="op">/</span>code_length<span class="op">**</span><span class="dv">3</span></span></code></pre></div>
<p>At the time of writing, two frontends in <code>yt</code> use the <code>"mks"</code> unit system by default, and one leaves the units in “code” units by default.</p>
<p>It is also possible to define a new unit system entirely, with required arguments of a name for the system, and length, mass, time, temperature, and angular units:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> UnitSystem</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> atomic_unit_system <span class="op">=</span> UnitSystem(<span class="st">&#39;atomic&#39;</span>, <span class="st">&#39;nm&#39;</span>, <span class="st">&#39;mp&#39;</span>, <span class="st">&#39;fs&#39;</span>, <span class="st">&#39;nK&#39;</span>, <span class="st">&#39;rad&#39;</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> atomic_unit_system[<span class="st">&#39;energy&#39;</span>] <span class="op">=</span> <span class="st">&#39;eV&#39;</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> atomic_unit_system</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>atomic Unit System</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a> Base Units:</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  length: nm</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  mass: mp</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  time: fs</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  temperature: nK</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  angle: rad</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  current_mks: A</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  luminous_intensity: cd</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  logarithmic: Np</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a> Other Units:</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  energy: eV</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(atomic_unit_system)</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>atomic</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> atomic_unit_system[<span class="st">&#39;number_density&#39;</span>]</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>nm<span class="op">**</span>(<span class="op">-</span><span class="dv">3</span>)</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> atomic_unit_system[<span class="st">&#39;angular_momentum&#39;</span>]</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>mp<span class="op">*</span>nm<span class="op">**</span><span class="dv">2</span><span class="op">/</span>fs</span></code></pre></div>
<p>Note that for dimensions not specified explicitly in the call to <code>UnitSystem</code> (in this example number density and angular momentum), their units were derived automatically.
Now, this user-defined unit system can be used in <code>yt</code>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds4 <span class="op">=</span> yt.load(<span class="st">&quot;Enzo_64/DD0043/data0043&quot;</span>, unit_system<span class="op">=</span><span class="st">&quot;atomic&quot;</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> dd4 <span class="op">=</span> ds4.all_data()</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(dd4[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;density&quot;</span>])</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1.15045218e-28</span> <span class="fl">1.04327815e-28</span> <span class="fl">1.52200266e-28</span> ... <span class="fl">1.55078523e-25</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a> <span class="fl">9.66275139e-26</span> <span class="fl">7.29105244e-26</span>] mp<span class="op">/</span>nm<span class="op">**</span><span class="dv">3</span></span></code></pre></div>
<h3 id="sec:unit-conversions">Unit conversions</h3>
<p>Aside from the ability to carry units through calculations, the other basic need for having a units implementation in <code>yt</code> is to be able to perform conversions of arrays from one unit to another of the same dimension.
Creating a new array or quantity from an existing one can be done using the <code>in_units</code> method (or the <code>to</code> method, which is a convenient alias):</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> yt </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds <span class="op">=</span> yt.load(<span class="st">&quot;GasSloshing/sloshing_nomag2_hdf5_plt_cnt_0100&quot;</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> sp <span class="op">=</span> ds.sphere(<span class="st">&quot;c&quot;</span>, (<span class="fl">100.0</span>, <span class="st">&quot;kpc&quot;</span>))</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(sp[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;velocity_x&quot;</span>])</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>[<span class="op">-</span><span class="fl">4793397.</span>  <span class="op">-</span><span class="fl">5297494.5</span> <span class="op">-</span><span class="fl">4997635.</span>  ... <span class="fl">10608344.</span>  <span class="fl">10382381.</span>  <span class="fl">10529207.</span> ] cm<span class="op">/</span>s</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(sp[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;velocity_x&quot;</span>].in_units(<span class="st">&quot;mile/hr&quot;</span>))</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>[<span class="op">-</span><span class="fl">107225.23711525</span> <span class="op">-</span><span class="fl">118501.57704009</span> <span class="op">-</span><span class="fl">111793.91105941</span> ...  <span class="fl">237301.89692198</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="fl">232247.24856836</span>  <span class="fl">235531.65264853</span>] mile<span class="op">/</span>hr</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(sp[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;pressure&quot;</span>])</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1.62223415e-10</span> <span class="fl">1.60880725e-10</span> <span class="fl">1.62334618e-10</span> ... <span class="fl">1.54101079e-10</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a> <span class="fl">1.52756530e-10</span> <span class="fl">1.53220436e-10</span>] dyn<span class="op">/</span>cm<span class="op">**</span><span class="dv">2</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(sp[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;pressure&quot;</span>].to(<span class="st">&quot;J/m**3&quot;</span>))</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1.62223415e-11</span> <span class="fl">1.60880725e-11</span> <span class="fl">1.62334618e-11</span> ... <span class="fl">1.54101079e-11</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a> <span class="fl">1.52756530e-11</span> <span class="fl">1.53220436e-11</span>] J<span class="op">/</span>m<span class="op">**</span><span class="dv">3</span></span></code></pre></div>
<p>Conversions to other unit systems can also be achieved with the methods <code>in_cgs</code> for Gaussian CGS units, <code>in_mks</code> for SI units, and the more general <code>in_base</code> for conversion to any valid unit system (see <a href="#sec:unit-systems">Unit systems</a>).</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> unyt <span class="im">as</span> u </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> l <span class="op">=</span> np.array([<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">100</span>])<span class="op">*</span>u.Mpc</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(l.in_cgs())</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>[<span class="fl">3.08567758e+25</span> <span class="fl">3.08567758e+26</span> <span class="fl">3.08567758e+26</span>] cm</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(l.in_mks())</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>[<span class="fl">3.08567758e+23</span> <span class="fl">3.08567758e+24</span> <span class="fl">3.08567758e+24</span>] m</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(l.in_base(<span class="st">&quot;imperial&quot;</span>))</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1.01236141e+24</span> <span class="fl">1.01236141e+25</span> <span class="fl">1.01236141e+25</span>] ft</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(l.in_base(<span class="st">&quot;geometrized&quot;</span>))</span></code></pre></div>
<p>To convert an array in-place instead of making a copy, use <code>convert_to_units</code>:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> momentum_y <span class="op">=</span> sp[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;mass&quot;</span>]<span class="op">*</span>sp[<span class="st">&quot;gas&quot;</span>,<span class="st">&quot;velocity_y&quot;</span>]</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(momentum_y)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>[<span class="fl">1.45031068e+47</span> <span class="fl">1.40467746e+47</span> <span class="fl">1.42059875e+47</span> ... <span class="fl">1.01206589e+47</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a> <span class="fl">9.94539437e+46</span> <span class="fl">1.06969711e+47</span>] cm<span class="op">*</span>g<span class="op">/</span>s</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> momentum_y.convert_to_units(<span class="st">&quot;Msun*kpc/Myr&quot;</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(momentum_y)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>[<span class="fl">745945.76032301</span> <span class="fl">722474.98923595</span> <span class="fl">730663.86660289</span> ... <span class="fl">520541.05697294</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a> <span class="fl">511526.58818102</span> <span class="fl">550182.82006007</span>] Msun<span class="op">*</span>kpc<span class="op">/</span>Myr</span></code></pre></div>
<p>Naturally, attempting to convert an array to units with a different dimension raises an error (in most cases, see below for exceptions):</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(momentum_y.to(<span class="st">&quot;degC&quot;</span>))</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>Traceback (most recent call last):</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;&lt;stdin&gt;&quot;</span>, line <span class="dv">1</span>, <span class="kw">in</span> <span class="op">&lt;</span>module<span class="op">&gt;</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/array.py&quot;</span>, line <span class="dv">947</span>, <span class="kw">in</span> to</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="va">self</span>.in_units(units, equivalence<span class="op">=</span>equivalence, <span class="op">**</span>kwargs)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>           <span class="op">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/array.py&quot;</span>, line <span class="dv">874</span>, <span class="kw">in</span> in_units</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    (conversion_factor, offset) <span class="op">=</span> <span class="va">self</span>.units.get_conversion_factor(</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>                                  <span class="op">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py&quot;</span>, line <span class="dv">694</span>, <span class="kw">in</span> get_conversion_factor</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> _get_conversion_factor(<span class="va">self</span>, other_units, dtype)</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>           <span class="op">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py&quot;</span>, line <span class="dv">939</span>, <span class="kw">in</span> _get_conversion_factor</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">raise</span> UnitConversionError(</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>unyt.exceptions.UnitConversionError: Cannot convert between <span class="st">&#39;Msun*kpc/Myr&#39;</span> (dim <span class="st">&#39;(length)*(mass)/(time)&#39;</span>) <span class="kw">and</span> <span class="st">&#39;°C&#39;</span> (dim <span class="st">&#39;(temperature)&#39;</span>).</span></code></pre></div>
<h4 id="sec:equivalencies">Equivalencies</h4>
<p>In certain circumstances, conversion from one quantity to another with different dimensions is desired.
This is the case for a number of frequently encountered, physically-motivated transformations which involve physical constants, some examples of which are:</p>
<ul>
<li>Conversions between temperature and energy via <span class="math inline">\(E = k_BT\)</span></li>
<li>Conversions between wavelength, frequency, and energy for light via <span class="math inline">\(E = h\nu = hc/\lambda\)</span></li>
<li>Conversions between mass and energy via <span class="math inline">\(E = mc^2\)</span></li>
<li>Conversions between density and number density via <span class="math inline">\(n = \rho/({\mu}m_p)\)</span></li>
</ul>
<p>As described above, conversions between quantities in different units is generally not allowed, but these special cases can be handled using unit equivalencies in <code>yt</code>.
This is done using the <code>to_equivalent</code> method, which takes the unit in a different dimension that you want to convert to, and the equivalence you would like to use:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> unyt <span class="im">as</span> u</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> kT <span class="op">=</span> <span class="fl">5.0</span><span class="op">*</span>u.keV</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># converting kT (energy) to T (temperature)</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(kT.to_equivalent(<span class="st">&quot;K&quot;</span>, <span class="st">&quot;thermal&quot;</span>))</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="fl">34813557.843240075</span> K</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># converting kT (energy) to Angstrom (wavelength)</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(kT.to_equivalent(<span class="st">&quot;angstrom&quot;</span>, <span class="st">&quot;spectral&quot;</span>))</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="fl">4.132806438406553</span> Å</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># print the proton&#39;s compton wavelength in picometers </span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(u.mp.to_equivalent(<span class="st">&quot;pm&quot;</span>, <span class="st">&quot;compton&quot;</span>))</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="fl">0.00132141</span> pm</span></code></pre></div>
<p>These equivalencies are straightforward–more information may be required in other cases.
For example, to convert temperature to sound speed via <span class="math inline">\(c_s = \sqrt{\gamma{k_BT}/(\mu{m_p})}\)</span>, there are two parameters which may be optionally changed, <span class="math inline">\(\gamma\)</span> and <span class="math inline">\(\mu\)</span>, which are the ratio of specific heats and the mean molecular weight, respectively:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> unyt <span class="im">as</span> u</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> T <span class="op">=</span> <span class="fl">1.0e6</span><span class="op">*</span>u.K</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># default gamma = 5/3 and mu = 0.6</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(T.to_equivalent(<span class="st">&quot;km/s&quot;</span>, <span class="st">&quot;sound_speed&quot;</span>))</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="fl">151.372499274441</span> km<span class="op">/</span>s</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(T.to_equivalent(<span class="st">&quot;km/s&quot;</span>, <span class="st">&quot;sound_speed&quot;</span>, mu<span class="op">=</span><span class="fl">1.0</span>))</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="fl">117.25263375274085</span> km<span class="op">/</span>s</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(T.to_equivalent(<span class="st">&quot;km/s&quot;</span>, <span class="st">&quot;sound_speed&quot;</span>, gamma<span class="op">=</span><span class="fl">4.</span><span class="op">/</span><span class="fl">3.</span>))</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="fl">135.39167932067505</span> km<span class="op">/</span>s</span></code></pre></div>
<p>For convenience, the same operations can actually be carried out simply using the <code>to</code> method described above:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> unyt <span class="im">as</span> u</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> kT <span class="op">=</span> <span class="fl">5.0</span><span class="op">*</span>u.keV</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(kT.to(<span class="st">&quot;K&quot;</span>, <span class="st">&quot;thermal&quot;</span>))</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fl">34813557.843240075</span> K</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> T <span class="op">=</span> <span class="fl">1.0e6</span><span class="op">*</span>u.K</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(T.to(<span class="st">&quot;km/s&quot;</span>, <span class="st">&quot;sound_speed&quot;</span>, gamma<span class="op">=</span><span class="fl">4.</span><span class="op">/</span><span class="fl">3.</span>))</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="fl">135.39167932067505</span> km<span class="op">/</span>s</span></code></pre></div>
<h4 id="conversions-between-electromagnetic-units-in-different-systems">Conversions between electromagnetic units in different systems</h4>
<p>Conversions between different electromagnetic units require special handling in many cases.
The reason for this is that the units for many electromagnetic quantities do not have the same dimensions between the Gaussian CGS and the SI unit systems.
For illustration, let us consider the units of magnetic field strength.
Superficially, it would appear that there is a simple conversion between the cgs units of gauss (<span class="math inline">\(\rm{G}\)</span>) and the SI units of tesla (<span class="math inline">\(\rm{T}\)</span>), since numerically <span class="math inline">\(1~\rm{G} = 10^{-4}~\rm{T}\)</span>.
However, if we examine the base units, we find that they have different dimensions:</p>
<p><span class="math display">\[
\mathrm{CGS:} 1~\mathrm{G} = 1~\frac{\sqrt{\mathrm{g}}}{\sqrt{\mathrm{cm}}\cdot\mathrm{s}} \\
\mathrm{SI:} 1~\mathrm{T} = 1~\frac{\mathrm{kg}}{\mathrm{A}\cdot{\mathrm{s}}}
\]</span></p>
<p>The reason for this has already been noted above: the SI system has a base unit of current (the ampere), and all other electromagnetic units in SI are derived from it and combinations of other units.
<code>unyt</code> handles conversions between electromagnetic quantities with <em>atomic</em> units between CGS and SI systems, under the hood despite this difference:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> unyt <span class="im">as</span> u</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> I <span class="op">=</span> <span class="fl">10.0</span><span class="op">*</span>u.A</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(I.units.dimensions)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>(current_mks)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> Icgs <span class="op">=</span> I.to(<span class="st">&quot;statA&quot;</span>)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(Icgs)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="fl">2997924580.0</span> statA</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(Icgs.units.dimensions)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>(length)<span class="op">**</span>(<span class="dv">3</span><span class="op">/</span><span class="dv">2</span>)<span class="op">*</span>sqrt((mass))<span class="op">/</span>(time)<span class="op">**</span><span class="dv">2</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> B <span class="op">=</span> <span class="fl">1.0e-4</span><span class="op">*</span>u.T</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(B.units.dimensions)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>(mass)<span class="op">/</span>((current_mks)<span class="op">*</span>(time)<span class="op">**</span><span class="dv">2</span>)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> Bcgs <span class="op">=</span> B.to(<span class="st">&quot;G&quot;</span>)</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(Bcgs)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a><span class="fl">1.0</span> G</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(Bcgs.units.dimensions)</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>sqrt((mass))<span class="op">/</span>(sqrt((length))<span class="op">*</span>(time))</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> Vcgs <span class="op">=</span> <span class="fl">1.0</span><span class="op">*</span>u.statV</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(Vcgs.units.dimensions)</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>sqrt((length))<span class="op">*</span>sqrt((mass))<span class="op">/</span>(time)</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> Vsi <span class="op">=</span> Vcgs.to(<span class="st">&quot;V&quot;</span>)</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(Vsi)</span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="fl">0.0033356409519815205</span> V</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(Vsi.units.dimensions)</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>(length)<span class="op">**</span><span class="dv">2</span><span class="op">*</span>(mass)<span class="op">/</span>((current_mks)<span class="op">*</span>(time)<span class="op">**</span><span class="dv">3</span>)</span></code></pre></div>
<p>Because it is more difficult to handle non-atomic conversions easily, these will still raise errors:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> unyt <span class="im">as</span> u</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># 3D current density in SI units</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> J <span class="op">=</span> <span class="fl">1.0</span><span class="op">*</span>u.A<span class="op">/</span>u.m<span class="op">**</span><span class="dv">3</span> </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># attempt to convert to the same quantity in Gaussian CGS units</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(J.to(<span class="st">&quot;statA/cm**3&quot;</span>))</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>Traceback (most recent call last):</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;&lt;stdin&gt;&quot;</span>, line <span class="dv">1</span>, <span class="kw">in</span> <span class="op">&lt;</span>module<span class="op">&gt;</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/array.py&quot;</span>, line <span class="dv">947</span>, <span class="kw">in</span> to</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="va">self</span>.in_units(units, equivalence<span class="op">=</span>equivalence, <span class="op">**</span>kwargs)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>           <span class="op">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/array.py&quot;</span>, line <span class="dv">874</span>, <span class="kw">in</span> in_units</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>    (conversion_factor, offset) <span class="op">=</span> <span class="va">self</span>.units.get_conversion_factor(</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>                                  <span class="op">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py&quot;</span>, line <span class="dv">694</span>, <span class="kw">in</span> get_conversion_factor</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> _get_conversion_factor(<span class="va">self</span>, other_units, dtype)</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>           <span class="op">^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>  File <span class="st">&quot;/Users/jzuhone/mambaforge/envs/py311/lib/python3.11/site-packages/unyt/unit_object.py&quot;</span>, line <span class="dv">939</span>, <span class="kw">in</span> _get_conversion_factor</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">raise</span> UnitConversionError(</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>unyt.exceptions.UnitConversionError: Cannot convert between <span class="st">&#39;A/m**3&#39;</span> (dim <span class="st">&#39;(current_mks)/(length)**3&#39;</span>) <span class="kw">and</span> <span class="st">&#39;statA/cm**3&#39;</span> (dim <span class="st">&#39;sqrt((mass))/((length)**(3/2)*(time)**2)&#39;</span>).</span></code></pre></div>
<p>For these cases, it is recommended to convert atomic electromagnetic units separately first in equations, if necessary.</p>
<h3 id="physical-constants">Physical Constants</h3>
<p><code>unyt</code> contains a large number of built-in physical constants and
properties of astronomical objects as <code>unyt_quantity</code> instances, which
can be used in computations. The simplest way to use them is to import
them directly from the <code>unyt</code> namespace:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> clight, kboltz, h, Mearth, G</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="cf">for</span> const <span class="kw">in</span> [clight, kboltz, h, Mearth, G]:</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>...    <span class="bu">print</span>(const)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fl">299792458.0</span> m<span class="op">/</span>s</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="fl">1.3806488e-23</span> kg<span class="op">*</span>m<span class="op">**</span><span class="dv">2</span><span class="op">/</span>(K<span class="op">*</span>s<span class="op">**</span><span class="dv">2</span>)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="fl">6.62606957e-34</span> kg<span class="op">*</span>m<span class="op">**</span><span class="dv">2</span><span class="op">/</span>s</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="fl">6.045644495102106e+24</span> kg</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="fl">6.67408e-11</span> m<span class="op">**</span><span class="dv">3</span><span class="op">/</span>(kg<span class="op">*</span>s<span class="op">**</span><span class="dv">2</span>)</span></code></pre></div>
<p>Here’s an example of using constants in a computation:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> unyt_quantity</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> Msun, G</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> d <span class="op">=</span> unyt_quantity(<span class="fl">1.0</span>, <span class="st">&quot;AU&quot;</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> T <span class="op">=</span> np.sqrt(<span class="fl">4.0</span><span class="op">*</span>np.pi<span class="op">**</span><span class="dv">2</span><span class="op">*</span>d<span class="op">**</span><span class="dv">3</span><span class="op">/</span>(G<span class="op">*</span>Msun)) <span class="co"># Kepler&#39;s third law</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(T.to(<span class="st">&quot;yr&quot;</span>))</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="fl">1.000033863000043</span> yr</span></code></pre></div>
<p>Some physical constants are represented by Greek letters or other characters–these can be imported either using ASCII characters or their non-ASCII representations:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> σ_T, ε_0</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> eps_0, sigma_thomson <span class="co"># ASCII versions</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(ε_0) <span class="co"># This is the vacuum permittivity constant</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="fl">8.854187817620389e-12</span> A<span class="op">**</span><span class="dv">2</span><span class="op">*</span>s<span class="op">**</span><span class="dv">4</span><span class="op">/</span>(kg<span class="op">*</span>m<span class="op">**</span><span class="dv">3</span>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(σ_T) <span class="co"># This is the Thomson cross section</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="fl">6.65245854533e-29</span> m<span class="op">**</span><span class="dv">2</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(σ_T <span class="op">==</span> sigma_thomson)</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="va">True</span></span></code></pre></div>
<h3 id="sec:creating-ytarray-and-ytquantity-instances">Creating unyt_array and unyt_quantity instances</h3>
<p>Aside from converting an array to new units as described above, there are two further ways to create new array and quantity objects: via a constructor, and by multiplying scalar data by a unit quantity.</p>
<h4 id="sec:class-constructor">Class constructor</h4>
<p>The primary internal interface for creating new arrays and quantities is through the class constructor for <code>unyt_array</code>.
The constructor takes three arguments.
The first argument is the input data, which can be an integer, float, list, or array.
The second argument, <code>input_units</code>, is a unit specification which must be a string or <code>Unit</code> instance.
Last, users may optionally supply a <code>UnitRegistry</code> instance, which will be attached to the array.
If no <code>UnitRegistry</code> is supplied, a default unit registry is used instead.
Unit specification strings must be algebraic combinations of unit symbol names, using standard Python mathematical syntax (i.e. <code>**</code> for the power function, not <code>^</code>).</p>
<p>Here is a simple example of <code>unyt_array</code> creation:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> unyt_array, unyt_quantity </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> unyt_array([<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>], <span class="st">&#39;cm&#39;</span>) </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>unyt_array([<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>]) cm</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> unyt_quantity(<span class="dv">3</span>, <span class="st">&#39;J&#39;</span>) </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>unyt_quantity(<span class="dv">3</span>, <span class="st">&#39;J&#39;</span>) </span></code></pre></div>
<p>In addition to the class constructor, we have also defined two convenience functions, <code>quan</code>, and <code>arr</code>, for quantity and array creation that are attached to the <code>Dataset</code> class.
These were added to syntactically simplify the creation of arrays with the <code>UnitRegistry</code> instance associated with a dataset.
These functions work exactly like the <code>unyt_array</code> and <code>unyt_quantity</code> constructors, but pass the <code>UnitRegistry</code> instance attached to the dataset to the underlying constructor call.
For example:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> yt</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds <span class="op">=</span> yt.load(<span class="st">&quot;Enzo_64/DD0043/data0043&quot;</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> ds.arr([<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>], <span class="st">&#39;code_length&#39;</span>).in_cgs() </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>unyt_array([<span class="fl">5.55e+26</span>, <span class="fl">1.11e+27</span>, <span class="fl">1.66e+27</span>], <span class="st">&#39;cm&#39;</span>)</span></code></pre></div>
<p>The above example illustrates that the array is being created using <code>ds.unit_registry</code>, rather than the <code>default_unit_registry</code> in <code>yt</code>.</p>
<h4 id="sec:multiplication">Multiplication</h4>
<p>New <code>unyt_array</code> and <code>unyt_quantity</code> instances can also be created by multiplying <code>unyt_array</code>, <code>unyt_quantity</code>, or <code>Unit</code> instances by <code>float</code> or <code>ndarray</code> instances.
To make it easier to create arrays using this mechanism, we have populated the <code>unyt</code> namespace with predefined <code>Unit</code> instances that correspond to common unit symbol names.
For example:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> meter, gram, kilogram, second, joule </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> kilogram <span class="op">*</span> meter<span class="op">**</span><span class="dv">2</span> <span class="op">==</span> joule </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="va">True</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> m, kg, s, W </span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> kg<span class="op">*</span>m<span class="op">**</span><span class="dv">2</span><span class="op">/</span>s<span class="op">**</span><span class="dv">3</span> <span class="op">==</span> W</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="va">True</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> kilometer </span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> three_kilometers <span class="op">=</span> <span class="dv">3</span><span class="op">*</span>kilometer </span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(three_kilometers)</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="fl">3.0</span> km</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> gram, kilogram </span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> my_g <span class="op">=</span> <span class="fl">1.0</span><span class="op">*</span>gram</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> my_kg <span class="op">=</span> <span class="fl">1.0</span><span class="op">*</span>kilogram</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(my_g<span class="op">+</span>my_kg)</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a><span class="fl">1001.0</span> g </span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(my_kg<span class="op">+</span>my_g)</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a><span class="fl">1.001</span> kg </span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(my_kg<span class="op">/</span>my_g)</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a><span class="fl">1000.0</span> dimensionless</span></code></pre></div>
<h3 id="sec:importing-exporting-units">Importing and exporting units</h3>
<p>Arrays from <code>unyt</code> can be exported and imported in a few different ways.</p>
<h4 id="writing-unyt_arrays-to-disk">Writing <code>unyt_array</code>s to disk</h4>
<p><code>unyt_array</code>s can be serialized to disk in either ASCII or HDF5 format.
For ASCII, use <code>unyt.savetxt</code> (with similar syntax as <code>numpy.savetxt</code>) to write one or more arrays:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> unyt <span class="im">as</span> u</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> a <span class="op">=</span> [<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>]<span class="op">*</span>u.cm</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> b <span class="op">=</span> [<span class="dv">8</span>, <span class="dv">10</span>, <span class="dv">12</span>]<span class="op">*</span>u.cm<span class="op">/</span>u.s</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> c <span class="op">=</span> [<span class="dv">2</span>, <span class="dv">85</span>, <span class="dv">9</span>]<span class="op">*</span>u.g</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> u.savetxt(<span class="st">&quot;sphere.dat&quot;</span>, [a,b,c], header<span class="op">=</span><span class="st">&#39;My sphere stuff&#39;</span>,</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>...           delimiter<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
<p>and <code>unyt.loadtxt</code> to read them back in:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> vel, mass <span class="op">=</span> u.loadtxt(<span class="st">&quot;sphere.dat&quot;</span>, usecols<span class="op">=</span>(<span class="dv">1</span>,<span class="dv">2</span>), delimiter<span class="op">=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
<p><code>unyt_array</code>s can be written to HDF5 format using the <code>write_hdf5</code> method, where one can optionally specify the HDF5 dataset and group where it can be stored:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># Store the &quot;b&quot; array to the dataset &quot;array_data&quot; at the top of the file</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> b.write_hdf5(<span class="st">&quot;velocity1.hdf5&quot;</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># Store the &quot;b&quot; array to the dataset &quot;velocity&quot; at the top of the file</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> b.write_hdf5(<span class="st">&quot;velocity2.hdf5&quot;</span>, dataset_name<span class="op">=</span><span class="st">&quot;velocity&quot;</span>)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="co"># Store the &quot;b&quot; array to the dataset &quot;velocity&quot; in the group &quot;fields&quot;</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> b.write_hdf5(<span class="st">&quot;velocity3.hdf5&quot;</span>, dataset_name<span class="op">=</span><span class="st">&quot;velocity&quot;</span>, group_name<span class="op">=</span><span class="st">&quot;fields&quot;</span>)</span></code></pre></div>
<p>and read back in using <code>from_hdf5</code>:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">import</span> unyt <span class="im">as</span> u </span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> v1 <span class="op">=</span> u.read_hdf5(<span class="st">&quot;velocity1.hdf5&quot;</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> v2 <span class="op">=</span> u.read_hdf5(<span class="st">&quot;velocity2.hdf5&quot;</span>, dataset_name<span class="op">=</span><span class="st">&quot;velocity&quot;</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> v3 <span class="op">=</span> u.read_hdf5(<span class="st">&quot;velocity3.hdf5&quot;</span>, dataset_name<span class="op">=</span><span class="st">&quot;velocity&quot;</span>, group_name<span class="op">=</span><span class="st">&quot;fields&quot;</span>)</span></code></pre></div>
<h4 id="exporting-unyt_arrays-to-other-unit-packages">Exporting <code>unyt_array</code>s to other unit packages</h4>
<p>It is possible to export an <code>unyt_array</code> or an <code>unyt_quantity</code> to equivalent objects in two other symbolic unit packages: <a href="https://pint.readthedocs.io"><code>Pint</code></a> and <a href="https://www.astropy.org"><code>AstroPy</code></a>.
To convert to a <code>Pint</code> <code>Quantity</code> object, the <code>to_pint</code> method is used:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> cm, s</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> a <span class="op">=</span> <span class="dv">4</span><span class="op">*</span>cm<span class="op">**</span><span class="dv">2</span><span class="op">/</span>s</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">print</span>(a)</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> cm<span class="op">**</span><span class="dv">2</span><span class="op">/</span>s</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> a.to_pint()</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>Quantity(<span class="dv">4</span>, <span class="st">&#39;centimeter ** 2 / second&#39;</span>)<span class="op">&gt;</span></span></code></pre></div>
<p>Similarly, to convert to an <code>AstroPy</code> <code>Quantity</code> object, use <code>to_astropy</code>:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> unyt <span class="im">import</span> g, cm</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> data <span class="op">=</span> [<span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>]<span class="op">*</span>g<span class="op">/</span>cm<span class="op">**</span><span class="dv">3</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> data.to_astropy()</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>Quantity [<span class="fl">3.</span>, <span class="fl">4.</span>, <span class="fl">5.</span>] g <span class="op">/</span> cm3<span class="op">&gt;</span></span></code></pre></div>
<h2 id="user-friendliness">User-Friendliness</h2>
<p>Several different entrypoints exist for research codes.
For instance, an ontology of “end-users” in science has developed that emphasizes that typically “library” codes are only used implicitly by researchers.
In this framing, library codes primarily interface with application codes, rather than with researchers.
We have taken a slightly different approach with the design and development of yt, as it exists in a middle-ground as a library used <em>as an application</em> within the scripting language Python and also as a library for more complex analysis of data.
As an outgrowth of this, we have taken particular care with the “public-facing” API of yt.
We have attempted to abstract the API enough from the data structures yt uses internally that it is useful without detailed knowledge of yt’s internals.
By the same token, we have also attempted to provide low-level access to data, and to make as many of those methods accessible and usable as well.</p>
<p>In addition to API considerations, there are three areas that we note have additional care paid to the generation of figures, helpful error messages, and to integration with the omnipresent Jupyter environment.</p>
<h3 id="publication-ready-figures">Publication-Ready Figures</h3>
<p>Matplotlib <span class="citation" data-cites="1DeF7kDPM">[<a href="#ref-1DeF7kDPM" role="doc-biblioref">119</a>]</span> is a fully-featured mechanism for generating figures, with an incredible array of options to customize formatting, appearance, rendering of fonts and glyphs, selection of colorbars, calculation of tick locations, and appropriate bounds.
This degree of flexibility provides extremely fine-grained control over the appearance of figures for publication, and Matplotlib provides a large number of output formats for even the most discerning of journals.</p>
<p>yt utilizes Matplotlib as its primary rendering engine for visualizations; while 3D renderings (such as those described in <a href="#sec:software-volume-rendering">1.11.41</a>) typically generate raw pixel buffers that are saved as images directly, most other visualization functionality in yt relies on Matplotlib for generating images that are saved to disk.
Slices and projections are prepared as variable-resolution images that are pixelized (<a href="#sec:pixelization">1.11.39</a>) into fixed resolution buffers, then provided to Matplotlib in the <code>imshow</code> function.
Phase plots utilize the <code>pcolormesh</code> function.</p>
<p>All of these internal plots utilize Matplotlib’s “object-oriented” interface, wherein <code>Figure</code> and <code>Axes</code> objects are created directly.
yt then manages the internal state of these objects, and provides high-level access to operations that are aware of the nature of the visualizations; for instance, operations that zoom and pan, and that are unit-aware.
In the case of 2D spatial images, these are all generated as subclasses of a <code>PlotWindow</code> object, named such because it functions as a “window” onto the data, including automatically generating and managing multiple fields simultaneously.</p>
<p>Within the <code>PlotWindow</code> object, we set up two levels of visualization invalidation; when a characteristic of the image buffer data is changed (i.e., the viewing window, the resolution) then the <em>image data</em> is said to be invalidated.
When a characteristic of the visualization is changed (the label, the scaling of the colorbar, the colormap) then the <em>plot</em> is said to be invalidated.
This allows yt to minimize the overhead of conducting potentially expensive operations on the underlying (unprocessed) data, while maximizing its utilization of Matplotlib for generation of images.
The image buffers generated are also always available through the API, so in the common case that researchers wish to take the genrated images and perform detailed manipulations or plot layouts, they are able to do so, using the Matplotlib API they may already be familiar with.</p>
<p>In addition to providing data-aware operations for plot organization, a number of plot modifications are available that provide data-aware annotations.
As the plot window moves, the data-aware annotations are updated and move with the window.
For example, the process of generating vector overlays from variable-resolution data requires the same set of operations as generating a fixed resolution image buffer, so yt provides a method for overplotting vector fields.
This also requires correctly applying axis-ordering, so yt will automatically determine the appropriate x and y vectors in the image plane, and for an off-axis slice, it will generate in-plane vector fields by computing the appropriate set of dot products.
Additional data-aware operations include overplotting locations of gravitationally-bound clumps, contour plots of variable-resolution data, boundaries of data structures (such as grids in patch-based grid datasets, cells from patch and octree datasets, particle locations), streamlines, line integral convolution, and coordinate-aware annotations.</p>
<p>All of these figure construction and modification methods apply to cartesian (<span class="math inline">\(x\)</span>, <span class="math inline">\(y\)</span>, <span class="math inline">\(z\)</span>) datasets, but yt also supports them for curvilinear data such as that organized spherically (<span class="math inline">\(r\)</span>, <span class="math inline">\(\theta\)</span>, <span class="math inline">\(\phi\)</span>) and cylindrically (<span class="math inline">\(r\)</span>, <span class="math inline">\(z\)</span>, <span class="math inline">\(\theta\)</span>).</p>
<p>By focusing on the utilizing low-level yt operations to facilitate high-level interactions, we can embed best practices for accessibility and visualizations at as many levels as possible.
For instance, yt provides suggested colormaps for specific fields, will apply heuristics to determine the appropriate scaling (such as symmetric logarithm) and utilize the appropriate labels and units to the data.
This allows users of yt to access both the expansive functionality of Matplotlib as well as the more domain-specific tasks available in yt.</p>
<p>Finally, the <code>cmyt</code> package (‘colormaps from yt’) also provides several colormaps that were, until yt 4.0, included in the main yt distribution.
Many of these are developed by yt contributors, typically using the <a href="https://github.com/matplotlib/viscm"><code>viscm</code></a> package, and are designed to be perceptually uniform in colorspace.</p>
<div id="tbl:colormaps" class="tablenos">
<table id="tbl:colormaps">
<caption><span>Table 4:</span> Colormaps provided by <code>cmyt</code>. </caption>
<thead>
<tr class="header">
<th>Colormap Name</th>
<th>Colormap</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>algae</code></td>
<td><img src="images/colormaps/algae.png" title="fig:" /></td>
</tr>
<tr class="even">
<td><code>apricity</code></td>
<td><img src="images/colormaps/apricity.png" title="fig:" /></td>
</tr>
<tr class="odd">
<td><code>arbre</code></td>
<td><img src="images/colormaps/arbre.png" title="fig:" /></td>
</tr>
<tr class="even">
<td><code>dusk</code></td>
<td><img src="images/colormaps/dusk.png" title="fig:" /></td>
</tr>
<tr class="odd">
<td><code>kelp</code></td>
<td><img src="images/colormaps/kelp.png" title="fig:" /></td>
</tr>
<tr class="even">
<td><code>octarine</code></td>
<td><img src="images/colormaps/octarine.png" title="fig:" /></td>
</tr>
<tr class="odd">
<td><code>pastel</code></td>
<td><img src="images/colormaps/pastel.png" title="fig:" /></td>
</tr>
<tr class="even">
<td><code>pixel_blue</code></td>
<td><img src="images/colormaps/pixel_blue.png" title="fig:" /></td>
</tr>
<tr class="odd">
<td><code>pixel_green</code></td>
<td><img src="images/colormaps/pixel_green.png" title="fig:" /></td>
</tr>
<tr class="even">
<td><code>pixel_red</code></td>
<td><img src="images/colormaps/pixel_red.png" title="fig:" /></td>
</tr>
<tr class="odd">
<td><code>xray</code></td>
<td><img src="images/colormaps/xray.png" title="fig:" /></td>
</tr>
</tbody>
</table>
</div>
<p>We note in particular that in previous versions of yt, <code>algae</code> was known as <code>bds_highcontrast</code> (and was our default colormap) and that the <code>arbre</code> colormap is the current default colormap.
<code>pastel</code> was designed by Tune Kamae, and was previously referred to as <code>kamae</code>.</p>
<h3 id="enhanced-error-messages">Enhanced Error Messages</h3>
<p>Providing helpful error messages is almost an art form; projects like AstroPy <span class="citation" data-cites="XobbtsFq">[<a href="#ref-XobbtsFq" role="doc-biblioref">120</a>]</span> have developed complex, helpful and thoughtfully-designed methods of providing as much usable information as possible when something “goes wrong.”
In recent development history, we have spent a considerable amount of effort attempting to discern between errors that are “user-facing” and those that are strictly “internal.”
“Internal” errors, for the most part, are often highly-unexpected; they may result from malformed data, or data that yt (and the developers) do not know to expect, or from general software defects.
“User-facing” errors are those that can be anticipated, and they are the errors for which we endeavor to provide helpful and extensive responses.</p>
<p>Among others, some of the most useful error messages that yt provides are those related to accessing fields and loading datasets.
For instance, accessing the field <code>"x_velocity"</code> instead of <code>"velocity_x"</code> will produce an error, suggesting the <em>correct</em> field to access.
When loading a dataset, yt will attempt to determine what the format of the data is; if it is unable to unambiguously identify the data format, it will provide an error message that shows the different options, and indicates how to disambiguate.</p>
<p>While these may seem like simple, obvious changes to make, they can hide difficult technical challenges, and more importantly, have dramatically improved the user experience for people using yt.</p>
<h3 id="sec:jupyter_integration">Jupyter Integration</h3>
<p>Project Jupyter is an overarching term for a collection of related projects that provide an extensive, end-to-end suite for the user experience of developing code and narrative, as described in depth in (among other papers) <span class="citation" data-cites="lOgtbiQz">[<a href="#ref-lOgtbiQz" role="doc-biblioref">121</a>]</span> and <span class="citation" data-cites="16Nvi8gCt">[<a href="#ref-16Nvi8gCt" role="doc-biblioref">122</a>]</span>.
While many in the yt community utilize yt through python scripts executed on the command line or through submission queues on high-performance computing resources, a large fraction utilize Jupyter Notebooks for their data exploration.
In addition to enhancements in the user interface for unmodified Python libraries, Jupyter provides opportunities for libraries and applications to provide rich, enhanced interfaces with widgets, styling of text, complex layout and in-line visualizations.
An important aspect of Project Jupyter is the flexibility of the kernel system, which mediates communication between a frontend (often a web browser, but also a command-line application or native GUI) and a execution kernel, which can be running locally or on a remote resource.
This allows individuals to utilize a local web browser and execute their operations on remote resources; by means of its flexible transport layer, images and the like can be passed back to the web browser <em>inline</em> with the code that generated them, greatly speeding the process of examining and manipulating data.
For users of high-performance resources to which they do not have physical access, having visualizations inline with code can be transformative; rather than having toe <code>scp</code> or <code>rsync</code> plots back and forth to inspect, they are available with no additional steps.
Among the many different advantages that working in a Jupyter (Notebook) environment offers, this is perhaps the one that is the “stickiest” for researchers accustomed to working on systems without GUI toolkits installed by default, and the one that has led to the widespread adoption of Jupyter.
IPython, a Python-specific Jupyter subproject, also provides a number of quality-of-life improvements, such as tab-completion and “magic” commands that modify the interpretation and execution of code.</p>
<p>Under the broad umbrella of Project Jupyter is the <a href="https://github.com/jupyter-widgets/ipywidgets/"><code>ipywidgets</code></a> project.
The <code>ipywidgets</code> project provides GUI elements that are displayed in Jupyter Notebooks that are generated and represented by kernel-side Python code.
This allows a Python project to expose deeper information or interactive functionality without writing non-Python code, and also manages the data transport between the display and the kernel.</p>
<p>yt takes advantage of some of these facilities, particularly in those areas where users have in the past struggled with discoverability.
yt utilizes IPython tab-completion for field access on data objects, which greatly decreases the likelihood of typos for field <em>types</em> as well as field <em>names</em>.
This is generated on a per-dataset basis, to ensure that the fields are all specifically available for each individual dataset.
Providing tab-completion reduces the need for users to look up the full collection of fields (sorted by field type) before referencing them.
Because yt provides <em>so many</em> fields, in many cases with similar names, this substantially reduces the likelihood of typos and errors.</p>
<p>In addition to tab-completion, yt utilizes <code>ipywidgets</code> in two specific places.
The first of these is in display of three-element numerical fields; while this may seem like a rather niche application, these typically show up as properties of datasets that require unit conversion into meaningful values.
Many simulation codes in astrophysics, for instance, normalize the indexing units and then apply unit conversions to physical units.
All three-element numeric arrays associated with units are displayed as three (read-only) input boxes and a drop-down for unit conversion.</p>
<p>The other area that widgets are provided is in the field system.
Navigating the available fields (even mediated by tab-completion) can be overwhelming, and more importantly, ensuring that a field definition matches the expectations of a researcher is critical.
The fields widget, displayed whenever the <code>fields</code> attribute of a dataset is displayed, allows exploration of the field definitions, including the underlying source code, the expected units, display names, and so forth.
Because many of the fields in yt are defined programmatically, it also attempts to resolve variables defined in a closure.
For example, if the fields <code>momentum_x</code>, <code>momentum_y</code> and <code>momentum_z</code> are defined in a loop over the available axes (<code>x</code>, <code>y</code>, <code>z</code>), the source code explorer will attempt to show the loop variable’s current value for each definition.</p>
<p>These are the built-in widgets in yt; in addition to these, the <a href="https://github.com/yt-project/widgyts"><code>widgyts</code></a> package (described in <span class="citation" data-cites="v58DzoTC">[<a href="#ref-v58DzoTC" role="doc-biblioref">4</a>]</span>) provides a number of additional widgets.
Following the publication of its paper, it has expanded to include not only webassembly-based pixelization routines (to create in-browser ‘slippy maps’ of 2D data) but also PyThreeJS-based dataset exploration of simulations.
Packaging this separately from yt provides the opportunity for faster development and more experimental usage of external packages, but also greatly reduces discoverability and utilization.
We are exploring options for encouraging its uptake, particularly as we anticipate it will continue to grow and provide additional functionality.</p>
<h2 id="sec:halo_finding">Halo-Finding and Catalogs</h2>
<p>In cosmological simulations, dark matter is nearly always represented as a collection of collisionless particles. <strong>CITE SOMETHING ABOUT VLASOV GRID SOLUTIONS</strong>
This is well-suited to approximating it as a collisionless fluid; however, the identification of structures within that collection of particles can take on several forms, addressing different use cases to different degrees of accuracy. <strong>CITE HALO FINDING COMPARISON PAPER</strong>
Being able to identify halos, as well as their associated baryonic content, is necessary for rapid analysis of cosmological simulations.
Furthermore, convergence studies and cross-simulation comparisons requires a consistent method for identifying dark matter halos, as well as the ability to track their growth over time.</p>
<p>In past versions of <code>yt</code>, several specific halo finders were bundled and made available to work on any class of data <code>yt</code> was able to read.
These included the HOP halo finder, the classic Friends-of-Friends (FOF) halo finder <span class="citation" data-cites="ut5ogNdw">[<a href="#ref-ut5ogNdw" role="doc-biblioref">123</a>]</span>, a scalable and Parallel HOP <span class="citation" data-cites="1EbFITEAZ">[<a href="#ref-1EbFITEAZ" role="doc-biblioref">124</a>]</span>, and a wrapping of the ORIGAMI code <span class="citation" data-cites="Bz1lkIYo">[<a href="#ref-Bz1lkIYo" role="doc-biblioref">125</a>]</span> for filament identification.
To do so, <code>yt</code> would utilize direct in-memory connectors with these implementations; whereas typically data connectors are written for each individual dataset format for individual halo finding methods, this enabled a single connector to be written from <code>yt</code> to the halo finder.
In addition to these bundled halo finders, a direct in-memory interface with Rockstar <span class="citation" data-cites="TSiqnjZ6">[<a href="#ref-TSiqnjZ6" role="doc-biblioref">126</a>]</span> was developed that sidestepped Rockstar’s built in load-balancing to minimize data duplication and transfer.</p>
<p><code>yt</code> provides a unique set of functionality for accessing halo catalogs, as it provides the ability to query their values both <em>as</em> catalogs and as the original, underlying datasets.
This means that the same selection and analysis operations that can be conducted on a “primary” dataset can also be conducted on the halo catalog; furthermore, the halo catalog can be used as input to data selection operations.
This enables, for instance, querying original dark matter particle values in a halo (as defined by any characteristic radius of that halo) simultaneously with querying baryonic cells or particles included within.</p>
<p>As discussed in <span class="citation" data-cites="KKAuiCtl">[<a href="#ref-KKAuiCtl" role="doc-biblioref">127</a>]</span>, this can be used as input into other tools to provide sophisticated, graph-based queries of datasets and halo merger trees over cosmological time, such as with the <code>ytree</code> package.</p>
<p><code>yt</code> also includes an internal halo finding tool that is not widely exposed, built on its implementation of a union-find data structure for identifying topologically connected sets.
This implementation, a “particle contour tree,” uses this union-find data structure to connect (via percolation) particles into simple Friends-of-Friends collections.
These can then be used as input into more sophisticated phase-space finders, such as Rockstar, and as done in <span class="citation" data-cites="1DN6YqcpG">[<a href="#ref-1DN6YqcpG" role="doc-biblioref">128</a>]</span>.</p>
<h2 id="scaling-and-parallelism">Scaling and Parallelism</h2>
<p>To support cases where data volume results in long processing time or large memory requirements, yt operations have been parallelized using the Message Passing Interface (MPI; <span class="citation" data-cites="1B7zT2Jsm">[<a href="#ref-1B7zT2Jsm" role="doc-biblioref">129</a>]</span>).
When designing the parallel interface for yt, as discussed in <span class="citation" data-cites="1GOm01ggN">[<a href="#ref-1GOm01ggN" role="doc-biblioref">2</a>]</span>, the design goals included ensuring that scripts required little to no adjustments to be run in parallel.
In the intervening time, the parallel operation infrastructure has been rewritten in several key ways to enable individuals to apply multi-level parallelism to their analysis operations.</p>
<h3 id="structure-of-parallel-operations">Structure of Parallel Operations</h3>
<p>Almost all of the operations in yt that are conducted in parallel follow a straightforward method of decomposing work and consolidating results:</p>
<ol type="1">
<li>Identify which chunking method (see <a href="#sec:chunking">1.8.1</a>) is most appropriate for the operation.</li>
<li>Consolidate chunks according to IO minimization and assign to individual MPI tasks</li>
<li>Join (potentially applying reduction operations) final results to provide solution to <em>all tasks</em> in the group</li>
</ol>
<p>The final step, of joining across tasks, results in the final set of values being accessible to all tasks; this is not a universal “final step” in parallel operations, and in some cases results in substantial duplication of memory.
This compromise was accepted as a result of the design goals of ensuring that scripts can run unmodified.</p>
<p>The parallelism in yt heavily leans upon the “index” for a dataset either being available <em>already</em> at initiation time on all tasks, or that index being <em>accessible</em> through IO operations or fast generation.
This provides a degree of load-balancing that can be conducted, as estimates of memory and processing requirements are available on all tasks (and thus the load-balancing calculations are deterministic across all tasks).
In essence, this means that for grid-based datasets, the entire grid hierarchy is available on all processors; for octrees or particle datasets, it means that at least a rough estimate of the distribution of values must be available (and identical) on all processors.
This doesn’t prevent opaquely distributed datasets from being decomposed, but it does allow datasets whose distribution is well-described to be decomposed with greater precision.</p>
<h3 id="multi-level-parallelism">Multi-Level Parallelism</h3>
<p>In its original implementation of parallelism, yt utilized a single, global MPI communicator (<code>MPI_COMM_WORLD</code>).
This had the advantage of (counter-intuitively) not requiring a global state be tracked; however, it also greatly limited the degree to which tasks could be distributed.
Current versions of yt now implement a stack-based approach to MPI communicators, where subsets of MPI tasks are assigned to different collective operations.</p>
<p>For example, when conducting halo finding and analysis (see <a href="#sec:halo_finding">1.14</a>) yt can parallelize in groups of MPI tasks <em>across</em> halos, and then decompose the work <em>within</em> a given halo across MPI tasks inside that communicator.
This takes place by specifying a task size at the top level (or allowing yt’s internal heuristics to determine it) and then distributing work to sub-communicators, each of which is then used for decomposition inside that top-level task.</p>
<p>In addition to multi-level communicators, yt utilizes OpenMP constructs exposed in Cython in several places.
This includes in the software volume rendering (see <a href="#sec:software-volume-rendering">1.11.41</a>), in the pixelization operations for SPH data (see <a href="#sec:sph-analysis">1.8.4</a>), calculation of gravitational binding energy (see <a href="#sec:analysis-modules">1.16</a>) and for computing the bounding volume hierarchy for rendering finite element meshes (see <a href="#sec:unstructured_mesh">1.8.5</a>).
In some instances, the Cython interface to OpenMP has had unpredictable performance implications; owing to this, the usage of OpenMP within yt has been somewhat conservative.</p>
<h3 id="parallelism-interfaces">Parallelism Interfaces</h3>
<p>yt presents interfaces for implicit parallelism, wherein internal parallelism constructs are utilized by yt operations, as well as explicit operations that are conducted in parallel.
The former is woven throughout the fabric of all derived quantities, multi-dimensional profiling, and projection operations.
This parallelism is instrumented through the use of the yt “chunking” interface, and all derived quantities implement a fixed set of initialization, calculation, reduction and finalization operations.
The high-level interface to the <code>DerivedQuantity</code> subclasses computes the data chunks in the source data object and then assigns these to individual MPI tasks in the current top-level communicator.
Each initializes storage space for the intermediate values, iterates over its assigned chunks and constructs intermediate reductions, and then the finalization step involves broadcasting the values to all other tasks and completing the final set of operations.
For projections, the procedure is very similar; those datasets with an index duplicated across MPI tasks (such as patch-based grid datasets) are collapsed along a dimension and each MPI task fills in the values, which are then reduced through a broadcast operation.
Utilizing these operations requires <em>no</em> modifications to user-facing code other than a call to <code>yt.enable_parallelism()</code> at the start of the script.</p>
<p>The user-facing parallel constructs allow for somewhat greater flexibility in defining parallel task decomposition.
Many objects in yt, particularly those such as the <code>DatasetSeries</code> object, have constituent data objects on which analysis can be conducted.
These often provide a <code>piter</code> method, for “parallel iteration.”
This provides a shorthand method of applying the <code>parallel_objects</code> interface, described below, and allows for dynamic task allocation, per-item storage and specifying the number of processors in the communicator assigned to each object.</p>
<p>Finally, yt also provides a top-level <code>parallel_objects</code> command.
This accepts an iterable object (which typically must be able to provide length, so unknown-length generators are not suitable) and a desired sub-task size.
yt will group these objects into sub-communicators of the specified size, defaulting to a single MPI task per object.
Additionally, yt can provide a storage object to each individual sub-communicator, if that storage object is provided to the <code>parallel_objects</code> function call.
Each sub-communicator can then set a key and a value, and all tasks will receive the union of these keys and values following the completion of the loop.
For example:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>yt.enable_parallelism()</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>my_dictionary <span class="op">=</span> {}</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sto, dataset <span class="kw">in</span> dataset_series.piter(storage<span class="op">=</span>my_dictionary):</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>    ...  <span class="co"># process</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>    sto.result <span class="op">=</span> ...  <span class="co"># some information processed for this dataset</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>    sto.result_id <span class="op">=</span> ...  <span class="co"># some identifier for this dataset</span></span></code></pre></div>
<p>Following the completion of the loop, all tasks in the top-level communicator will have a full <code>my_dictionary</code> object, which can be used for plotting or subsequent analysis.
Common applications of this include iterating over datasets to identify quantities, conduct analysis operations, and so forth.
Using multiple levels of parallelism allows researchers to allocate a large processor count job on an HPC resource and dedicate individual portions of it to each dataset in a set of data outputs.
For many types of data analysis, particularly those operations conducted across a range of outputs, this allows much easier strong scaling.</p>
<h3 id="performance-of-operations">Performance of Operations</h3>
<h3 id="inline-analysis">Inline Analysis</h3>
<p>It is possible to instrument a simulation code to call Python routines inline during its execution.
Enzo has been instrumented in such a fashion; it accepts parameters that govern when and how frequently Python is called.
Prior to calling to Python, Enzo exposes views onto its data fields as numpy arrays.
<code>yt</code> has a special-purpose frontend that can access these views, as well as additional metadata passed through module-level objects, and then constructs appropriate <code>yt</code>-specific data objects around the data provided by the simulation.
In these cases, <code>yt</code> did not pass around datasets between MPI tasks, but rather decomposed under the assumption that data communication was not possible, and also that tasks would be broadly pre-load balanced by the simulation platform.
Within Enzo, all of the communication between Python and C++ was managed through Enzo’s usage of the C API.
This required some knowledge of how Python conducts garbage collection, and required ensuring that reference counting was managed correctly to avoid memory leaks.</p>
<p>This non-standardized approach to conducting <em>in situ</em> visualization led to the creation and development of the library <code>libyt</code> which serves as an intermediary layer between simulation codes and <code>yt</code> (and Python in general.)
This library encapsulates all Python API calls, manages references, and provides a systematic method for providing data pointers to Python.
<code>libyt</code> provides a stable C-based API, and is accessible from numerous different languages.
It also provides a custom-built <code>yt</code> frontend for accepting data.
A more complete description is outside the scope of this paper, and we refer the reader to (<strong>MJT: cite in prep manuscript</strong>).</p>
<h2 id="sec:analysis-modules">Analysis Modules</h2>
<p>For much of its development history, <code>yt</code> took the approach of bundling as many analysis modules as possible in the primary repository.
This provided the advantage of having all work be centralized, and ensuring that each download or installation of <code>yt</code> was a fully-featured system for analyzing a large swath of data, but it brought with it the development overhead of the entire <code>yt</code> package for what in many cases were isolated pieces of functionality with separable responsibilities.</p>
<p>As a result of the slowing in speed of development as a result of review requirements (and limited personnel to conduct those reviews), some of the analysis modules that were bundled with <code>yt</code> have been “spun out” into their own repository, <code>yt_astro_analysis</code> <span class="citation" data-cites="19u2YdIvE">[<a href="#ref-19u2YdIvE" role="doc-biblioref">130</a>]</span>.
This repository, which is developed, released and installed separately from <code>yt</code>, includes modules for cosmological observation (upon which Triden, which is discussed in <a href="#trident">1.17.1</a>, is based), dark matter halo finding and analysis, tools for interacting with position-position-velocity cubes, light cone generation <span class="citation" data-cites="Hd7r5YCP">[<a href="#ref-Hd7r5YCP" role="doc-biblioref">131</a>]</span> and a system for exporting from <code>yt</code> to RADMC-3D <span class="citation" data-cites="nfcCcNZX">[<a href="#ref-nfcCcNZX" role="doc-biblioref">132</a>]</span>.</p>
<h2 id="ecosystem-of-packages">Ecosystem of Packages</h2>
<p>Several packages have been developed that utilize <code>yt</code> as an infrastructural component.
In contrast to the analysis modules, which have typically been built as extensions of <code>yt</code>, these extensions utilize <code>yt</code> as a framework and typically only rely upon public APIs, often with no changes necessary to upstream <code>yt</code>.
In some cases, <code>yt</code> is also an <em>optional</em> dependency for the package.
In most cases, these are described in detail elsewhere, and citations to that external information are included here.</p>
<h3 id="trident">Trident</h3>
<p>Trident <span class="citation" data-cites="1DgKJy4lE">[<a href="#ref-1DgKJy4lE" role="doc-biblioref">133</a>]</span> is a Python-based open-source tool for post-processing hydrodynamical simulations to produce synthetic absorption spectra and related data.
In many ways, Trident is the first external package that utilizes <code>yt</code> to provide data access and numerical operations, but then builds on those to develop detailed, astrophysically-aware systems for processing and analyzing that data.</p>
<h3 id="powderday">Powderday</h3>
<p>Powderday <span class="citation" data-cites="g9AYex8L 1CgEbr6CQ">[<a href="#ref-g9AYex8L" role="doc-biblioref">134</a>,<a href="#ref-1CgEbr6CQ" role="doc-biblioref">135</a>]</span> is an open source package connecting <code>yt</code>, population synthesis modules and the monte carlo radiation transport code Hyperion <span class="citation" data-cites="6dD1NLpJ">[<a href="#ref-6dD1NLpJ" role="doc-biblioref">136</a>]</span> to develop synthetic observations of galaxies from astrophysics simulations.
Powderday uses <code>yt</code> to read and transform data, including high-fidelity gridding of SPH particles into an octree form, to prepare galaxy simulations to be processed.
Currently Powderday uses the Arepo, Enzo, Gadget, Gizmo, and Tipsy front ends to read in simulation results, and convert them to a common format in preparation for dust radiative transfer (using <code>unyt</code> to ensure common units).
Plans are in place to employ the wide range of front ends available in yt to broaden the applicability of powderday to a wider range of galaxy simulations.</p>
<h3 id="ytree">ytree</h3>
<p>Building on <code>yt</code> for access to halo catalogs, and implementing a similar system for derived fields as applied to graph datasets, ytree <span class="citation" data-cites="VB3RIyxd">[<a href="#ref-VB3RIyxd" role="doc-biblioref">137</a>]</span> is a system for analyzing merger trees from analysis of dark matter halos in cosmological simulations.</p>
<p>ytree provides flexibility in determining the path that a given analysis takes through the graph of merger trees; for instance, it enables the user to select if they wish to follow the “most massive” progenitor halo backwards in time, or even to set their own criteria for this.
Connecting this to the raw, unprocessed data from the simulation (such as the unsampled particle or cell content that comprises the halos) allows researchers to deepen and guide their analysis based on the physical characteristics of the merger history.</p>
<h3 id="pyxsim">pyxsim</h3>
<p>pyXSIM <span class="citation" data-cites="N2QcMTPv">[<a href="#ref-N2QcMTPv" role="doc-biblioref">138</a>]</span> is a python package for simulating X-ray observations from astrophysical sources.
It implements the PHOX algorithm <span class="citation" data-cites="gJwEappY nty3AM0C">[<a href="#ref-gJwEappY" role="doc-biblioref">139</a>,<a href="#ref-nty3AM0C" role="doc-biblioref">140</a>]</span> and is implemented to support grid-based codes, particle-based codes, and even data constructed from scratch in numpy arrays.
pyXSIM has been used to create synthetic observations from current and upcoming space telescopes.</p>
<h3 id="ytini">ytini</h3>
<p>While <code>yt</code> has functionality for volume rendering, it provides far from the scene-control, composition and camera positioning necessary for “cinematic visualization.”
The tool ytini <span class="citation" data-cites="3Gi4Hs1H">[<a href="#ref-3Gi4Hs1H" role="doc-biblioref">141</a>]</span> serves as an intermediary between <code>yt</code> and the visualization platform Houdini, often used for special effects and visual arts in industry and professional production environments.
ytini was developed to ease the process of converting data from simulation format into one suitable for production teams to utilize in Houdini and other visual effects software, and has been designed to enable deeper collaboration between researchers and visualization and outreach experts.</p>
<h2 id="future-directions">Future Directions</h2>
<ul>
<li>More integration with <em>in situ</em> analysis systems like <code>libyt</code></li>
<li>Much improved optimization</li>
<li>Integration with other domains besides astronomy</li>
<li>Refactoring for the long term</li>
<li>Static typing</li>
<li>Improving visual representation of <code>yt</code> objects</li>
<li>Testing infrastructure</li>
<li>Integration with external libraries such as pytorch-spatial, etc</li>
</ul>
<h2 id="sec:sustainability">Sustainability</h2>
<p>When discussing software projects, “Sustainability” is a complex, expansive topic that typically means different things to different people.
A number of different definitions have been proposed – ranging from pragmatic descriptions of responsivity to bug reports or changing hardware, to more idealistic descriptions of active development and ever-increasing functionality.
For the purposes of describing yt’s approach to sustainability, we will use these as path markers but not restrictions.</p>
<p>yt is supported through both peer-production and grant-funded development.
At times, this can pose challenges; the requirements imposed by grant-funded development naturally concentrates decision making, but within the yt community we have (so far) navigated this through deep engagement in community processes and interaction.
However, at the risk of belaboring a point that has been well-explored elsewhere, grant-funded development has traditionally focused on “new” or “innovative” development, rather than maintenance of a software project.</p>
<p>A tension exists, however, between support of an existing project and the support of new projects in an ecosystem.
By supporting an existing project, resources can tend to become concentrated; conversely, if a project supports a broader research agenda, that resource concentration can result in greater effort-multipliers for individuals who utilize the project.
We’re aware of this tension in yt; in fact, while yt has been grant-supported, most of the grant development has gone to a very small number of groups.
This grant funding has been provided through the National Science Foundation, the Gordon and Betty Moore Foundation, the Department of Energy, the Chan Zuckerberg Initiative and other sources. <span class="citation" data-cites="1AQkLoGRL">[<a href="#ref-1AQkLoGRL" role="doc-biblioref">142</a>]</span>, <span class="citation" data-cites="JmcaooMA">[<a href="#ref-JmcaooMA" role="doc-biblioref">143</a>]</span>, <span class="citation" data-cites="GSJFOpai">[<a href="#ref-GSJFOpai" role="doc-biblioref">144</a>]</span>.
Grants have supported the development of new features, including specific functionality for analysis routines and support for non-astronomical domains.</p>
<p>Into each of these grants has been explicit support for community building, constituted by the development of documentation, videos, and tutorials, as well as mentoring of new contributors and shepherding the growth of the project through code review and issue management.
While this does provide support for individuals who can provide dedicated, thoughtful attention to code review and bug reports, it is, quite frankly, insufficient without the engagement of non-funded community members who contribute their time and energy.</p>
<p>There are a number of avenues of development for yt, each of which draws different degrees of interest, urgency, and breadth of engagement.
A few of these are worth highlighting.</p>
<h3 id="platform-functionality">Platform Functionality</h3>
<p>Improving and developing the functionality of yt as a whole is a particular focus for investment of time and resources.
Typically, this is conducted via one of two avenues; the first is through explicit, funded development done for specific use cases.
The second mode of functionality development is when community members (either those who are already participants in the community or those who seek to) identify a feature that meets their particular use case, and work to develop it.</p>
<p>Community contributions usually fall into one of two categories – new functionality (such as supporting new datasets, applying new operations to datasets) or by scaling or improving the performance of yt for a use case or in general.
The former is more common than the latter, although both have occurred.
Expanding the featureset of yt tends to be more attractive than optimization for a large swath of the userbase; in some sense, optimization opens new avenues for scientific exploration (by taking formerly out-of-reach options and placing them within the realm of reasonable execution time), but adding new features certainly does.</p>
<p>A few examples of new functionality provided by community members include the volume rendering of octree datasets, support for new code frontends (such as AREPO), and many of the analysis modules.
New functionality provided by funded development has included long-term improvements such as the “demeshening” of particle datasets and support for non-astrophysical domains.
A few notable improvements to the efficiency and scalability of yt include multi-level parallelism operations and the initial implementation of dask as an array backend.</p>
<p>In general, we have found it difficult to move through large upheavals of code without large-scale effort on the part of the community.
This has resulted in a reluctance to investigate particularly invasive changes, and can provide a distortion of the cost/benefit analysis, leaning toward risk aversion.
The YTEP process mitigates this somewhat, but also provides additional opportunities for a ‘veto’ that can dampen enthusiasm and impede development.
At the same time, the safe guards provided by this can help to ensure stability for the community, particularly those not actively engaged in development.</p>
<h3 id="project-maintenance">Project Maintenance</h3>
<p>In addition to development of new features, the correction of problems or updates to old behavior are also critical to the sustained usage and applicability of a project.
In this category of “bug fixes” we certainly include faults or problems with yt itself, but it’s also important to note that in many cases, we can also use this category to describe improvements in behavior that bring the results greater accuracy or precision.</p>
<p>It is tough to estimate the cost of bug fixes, in terms of labor or in terms of community member attrition, but <span class="citation" data-cites="LRQOJg18">[<a href="#ref-LRQOJg18" role="doc-biblioref">145</a>]</span> suggests that large, established projects on github may spend up to 40% of their time on bugfixes rather than new features.
While we have not been able to verify this, examining the cadence of bug fix releases versus new feature releases suggests that this is indeed the case for yt as well.</p>
<p>Maintenance and bug fixes are critical to developing yt as a community-focused code.
The damage that can be done by ignoring a bug report, dismissing a suggestion, or outright rejecting proposed fixes is real, and can have measurable impact on community cohesion and growth.
As such, in the yt development process we strive to be as accommodating and welcoming as possible.
This requires balancing being welcoming and patient against the very real costs that come with responding to bug fixes and feature requests; often this means providing helpful insight that may already be covered by the documentation, or that may reflect known bugs.
This is a challenge, and one that benefits from having dedicated maintainers, or at least dedicated maintenance time from the developer community.
Furthermore, we have found that regular, collaborative co-working sessions can ease this burden, although they occasionally veer into too light of code review or constant reiteration of “standing” issues.</p>
<h3 id="ecosystem-maintenance">Ecosystem Maintenance</h3>
<p>yt exists inside a larger ecosystem of packages and infrastructure, composed of the Python language, the Scientific Python community, as well as the computational science communities (predominantly astrophysics).
Each of these communities, much like yt itself, has its own set of priorities, needs, and development schedules.
Changes in systems that yt depends on may require changes in yt; changes in what other systems expect from yt may likewise require changes in yt.</p>
<p>One of the more notable shifts in recent years has been in how Python libraries and applications are distributed.
At the time of the first yt method paper, yt was distributed typically in source form with its entire dependency stack to provide a uniform experience on supercomputer installations.
This was, at the time, “necessary” because Python was not as widely used in the computational science and astrophysics communities as it is now, and the versions of various libraries were often far out of date.
In the intervening time, however, a number of shifts have occurred, including the advent of conda (and conda-forge), wheels, wide-scale adoption of Python in the scientific community, and considerable standardization and improvement of package metadata and installation methods.
Ensuring yt maintains compatibility with the modern package ecosystem has taken, at times, a fairly-large degree of effort on the part of the developers <em>making</em> the changes and the developers who intend to continue developing.</p>
<p>In addition to the changes in yt <em>necessitated</em> by changes to packaging or updates in APIs or behavior, there are also those changes that are necessary to conform to community norms.
For example, utilizing code formatting tools, automated “linting” and other automated methods of applying “quality control” to incoming changesets.
Or, somewhat more invasively, inserting “type hints” that can provide automated analysis and inline suggestions within integrated development environments.
These developments may not provide as much value to long-time developers, but can be important for newcomers who are less familiar with the library or coding conventions.
The balance with these changes can be difficult to strike, to minimize disruption to existing developers and users while still expanding the accessibility of the library to others.</p>
<h3 id="enjoyment-maximization">Enjoyment Maximization</h3>
<p>Another aspect of sustainability is that for many people, developing software can be <em>fun</em>.
This is certainly true for many members of the yt community – sometimes in ways that align with the goals and requirements of academic research, and sometimes in ways that do not directly advance those goals.
For instance, the initial development of the software volume renderer was not aligned with immediate research goals, but it has gone on to support synthetic observations and to foster much broader research and education opportunities.</p>
<p>While development for the joy of developing can be an exciting and helpful process, it is also something that can require a delicate balance of attention and engagement from the rest of the community.
Accepting changes into the yt codebase requires careful review, and dedication of resources (physical and intellectual) that may already be allocated elsewhere.</p>
<p>Furthermore, not everyone derives pleasure from the same types of development.
Some individuals love to build new things and scaffold out plugins or visualization types; others have great satisfaction from optimization or documentation.
Accepting and supporting these different types of development is critical for a community built on respect, trust and gratitude.</p>
<h2 id="conclusions">Conclusions</h2>
<p>Since the publication of the last paper describing the methods and capabilities of <code>yt</code>, it has been dramatically transformed; while many of the underlying algorithms for processing grid-based data may remain similar or identical, it has been expanded considerably in scope to include data of many different forms.
Furthermore, each of these classes of data – most notably octree, smoothed particle hydrodynamics, and unstructured meshes – requires substantial care to ensure that the way that class of data is represented is a high-fidelity reflection of the underlying methods.
In this paper, we have presented our approach to making <code>yt</code> a functional, scientist-driven library to access, process, and visualize data.
This includes the selection of spatial (and non-spatial) regions, converting between spaces for representing data in <em>index</em> coordinates and <em>geometric</em> coordinates, the multi-dimensional reduction of data along spatial and non-spatial dimensions, production of publication-quality plots, volume rendering (both hardware and software) and the approaches we have taken toward developing a community of individuals using and developing <code>yt</code>.</p>
<p>We are grateful for the community of individuals who have participated in <code>yt</code>’s development; the authorship of this paper reflects a large, but not complete, fraction of those who have contributed changesets, although invitations have been extended to everyone we were able to reach.
The landscape of astrophysical computation is different now than it was when <code>yt</code> was created, and even since the publication of the first <code>yt</code> method paper.
Code development has become more open, conducted on platforms such as Github, and investment in community around software is largely recognized as <em>necessary</em>, rather than <em>supplemental</em>.
The usage of Python as a library, at least for high-level APIs, has become quite widespread, albeit not quite ubiquitous.</p>
<p>While the community of individuals who participate in <code>yt</code> usage and development is not as large as some in the astrophysical community, or in the broader pydata ecosystem, it is thriving.
The future development of <code>yt</code> will be focused on solidifying our existing community and growing to support the modern needs of scientists from with different approaches to and conceptions of data analysis and visualization.</p>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references csl-bib-body" role="doc-bibliography">
<div id="ref-cTN2TQIL" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">1. </div><div class="csl-right-inline"><strong>manubot/manubot</strong> <div class="csl-block">Manubot</div> (2025-01-12) <a href="https://github.com/manubot/manubot">https://github.com/manubot/manubot</a></div>
</div>
<div id="ref-1GOm01ggN" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">2. </div><div class="csl-right-inline"><strong>yt: A MULTI-CODE ANALYSIS TOOLKIT FOR ASTROPHYSICAL SIMULATION DATA</strong> <div class="csl-block">Matthew J Turk, Britton D Smith, Jeffrey S Oishi, Stephen Skory, Samuel W Skillman, Tom Abel, Michael L Norman</div> <em>The Astrophysical Journal Supplement Series</em> (2010-12-28) <a href="https://doi.org/ft6md2">https://doi.org/ft6md2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0067-0049/192/1/9">10.1088/0067-0049/192/1/9</a></div></div>
</div>
<div id="ref-XY0j7dm0" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">3. </div><div class="csl-right-inline"><strong>Ten simple rules for helping newcomers become contributors to open projects</strong> <div class="csl-block">Dan Sholler, Igor Steinmacher, Denae Ford, Mara Averick, Mike Hoye, Greg Wilson</div> <em>PLOS Computational Biology</em> (2019-09-12) <a href="https://doi.org/gf8gb9">https://doi.org/gf8gb9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1371/journal.pcbi.1007296">10.1371/journal.pcbi.1007296</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31513567">31513567</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6742214">PMC6742214</a></div></div>
</div>
<div id="ref-v58DzoTC" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">4. </div><div class="csl-right-inline"><strong>widgyts: Custom Jupyter Widgets for Interactive Data Exploration with yt</strong> <div class="csl-block">Madicken Munk, Matthew Turk</div> <em>Journal of Open Source Software</em> (2020-01-29) <a href="https://doi.org/gqbxbf">https://doi.org/gqbxbf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.21105/joss.01774">10.21105/joss.01774</a></div></div>
</div>
<div id="ref-9uqQt00R" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">5. </div><div class="csl-right-inline"><strong>Effects of Adopting Code Review Bots on Pull Requests to OSS Projects</strong> <div class="csl-block">Mairieli Wessel, Alexander Serebrenik, Igor Wiese, Igor Steinmacher, Marco A Gerosa</div> <em>2020 IEEE International Conference on Software Maintenance and Evolution (ICSME)</em> (2020-09) <a href="https://doi.org/grp6mq">https://doi.org/grp6mq</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/icsme46990.2020.00011">10.1109/icsme46990.2020.00011</a></div></div>
</div>
<div id="ref-fZ2Zii88" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">6. </div><div class="csl-right-inline"><strong>Smoothed particle hydrodynamics and magnetohydrodynamics</strong> <div class="csl-block">Daniel J Price</div> <em>Journal of Computational Physics</em> (2012-02) <a href="https://doi.org/d4zkq9">https://doi.org/d4zkq9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.jcp.2010.12.011">10.1016/j.jcp.2010.12.011</a></div></div>
</div>
<div id="ref-NSGlchZc" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">7. </div><div class="csl-right-inline"><strong>Smoothed Particle Hydrodynamics in Astrophysics</strong> <div class="csl-block">Volker Springel</div> <em>Annual Review of Astronomy and Astrophysics</em> (2010-08-01) <a href="https://doi.org/c59dm2">https://doi.org/c59dm2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1146/annurev-astro-081309-130914">10.1146/annurev-astro-081309-130914</a></div></div>
</div>
<div id="ref-1A6gLV5Q0" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">8. </div><div class="csl-right-inline"><strong>&lt;scp&gt;SPLASH&lt;/scp&gt;: An Interactive Visualisation Tool for Smoothed Particle Hydrodynamics Simulations</strong> <div class="csl-block">Daniel J Price</div> <em>Publications of the Astronomical Society of Australia</em> (2007) <a href="https://doi.org/ffsfkx">https://doi.org/ffsfkx</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1071/as07022">10.1071/as07022</a></div></div>
</div>
<div id="ref-Yng3Bhre" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">9. </div><div class="csl-right-inline"><strong>Improving convergence in smoothed particle hydrodynamics simulations without pairing instability</strong> <div class="csl-block">Walter Dehnen, Hossam Aly</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2012-08-07) <a href="https://doi.org/f38hjj">https://doi.org/f38hjj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1111/j.1365-2966.2012.21439.x">10.1111/j.1365-2966.2012.21439.x</a></div></div>
</div>
<div id="ref-GC8jwjG1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">10. </div><div class="csl-right-inline"><strong>Inconsistencies arising from the coupling of galaxy formation sub-grid models to pressure-smoothed particle hydrodynamics</strong> <div class="csl-block">Josh Borrow, Matthieu Schaller, Richard G Bower</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2021-05-19) <a href="https://doi.org/gsj6tn">https://doi.org/gsj6tn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stab1423">10.1093/mnras/stab1423</a></div></div>
</div>
<div id="ref-LwVDbUsQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">11. </div><div class="csl-right-inline"><strong>SciTools/cartopy: v0.21.1</strong> <div class="csl-block">Phil Elson, Elliott Sales De Andrade, Greg Lucas, Ryan May, Richard Hattersley, Ed Campbell, Andrew Dawson, Stephane Raynaud, Scmc72, Bill Little, … Daryl Herzmann</div> <em>Zenodo</em> (2022-12-12) <a href="https://doi.org/gr96pg">https://doi.org/gr96pg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5281/zenodo.7430317">10.5281/zenodo.7430317</a></div></div>
</div>
<div id="ref-190ab0NKz" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">12. </div><div class="csl-right-inline"><strong>Magnetic outflows from turbulent accretion disks</strong> <div class="csl-block">J Jacquemin-Ide, G Lesur, J Ferreira</div> <em>Astronomy &amp;amp; Astrophysics</em> (2021-03) <a href="https://doi.org/gsgh4s">https://doi.org/gsgh4s</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/202039322">10.1051/0004-6361/202039322</a></div></div>
</div>
<div id="ref-xbCL1tS1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">13. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Petascale Cosmology: Simulations of Structure Formation&lt;/span&gt;</strong> <div class="csl-block">Rupert Croft, Tiziana Di Matteo, Nishikanta Khandai</div> <em>Comput. Sci. Eng.</em> (2015-03) <a href="http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=7006381">http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=7006381</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/MCSE.2015.5">10.1109/mcse.2015.5</a></div></div>
</div>
<div id="ref-gw5rMEhN" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">14. </div><div class="csl-right-inline"><strong>The LSST Data Management System</strong> <div class="csl-block">Mario Jurić, Jeffrey Kantor, K-T Lim, Robert H Lupton, Gregory Dubois-Felsmann, Tim Jenness, Tim S Axelrod, Jovan Aleksić, Roberta A Allsman, Yusra AlSayyad, … LSST Project</div> <em>eprint arXiv:1512.07914</em> (2015) <a href="http://adsabs.harvard.edu/abs/2015arXiv151207914J">http://adsabs.harvard.edu/abs/2015arXiv151207914J</a></div>
</div>
<div id="ref-PSH5f9xg" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">15. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;A computer oriented geodetic data base and a new technique in file sequencing&lt;/span&gt;</strong> <div class="csl-block">GM Morton</div> <em>IBM Ltd.</em> (1996)</div>
</div>
<div id="ref-oIYFC4kD" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">16. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Gesammelte Abhandlungen: Band III: Analysis - Grundlagen der Mathematik Physik - Verschiedenes Lebensgeschichte&lt;/span&gt;</strong> <div class="csl-block">David Hilbert</div> <em>Springer Berlin Heidelberg</em> (1970) <a href="http://dx.doi.org/10.1007/978-3-662-25726-5{\_}1">http://dx.doi.org/10.1007/978-3-662-25726-5{\_}1</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/978-3-662-25726-5_1">10.1007/978-3-662-25726-5_1</a> · ISBN: 978-3-662-25726-5</div></div>
</div>
<div id="ref-I01oLUIj" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">17. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;The cosmological simulation code gadget-2&lt;/span&gt;</strong> <div class="csl-block">Volker Springel</div> <em>Mon. Not. R. Astron. Soc.</em> (2005-12) <a href="http://adsabs.harvard.edu/abs/2005MNRAS.364.1105S">http://adsabs.harvard.edu/abs/2005MNRAS.364.1105S</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1111/j.1365-2966.2005.09655.x">10.1111/j.1365-2966.2005.09655.x</a></div></div>
</div>
<div id="ref-1D5nekPGx" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">18. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Cosmological Hydrodynamics with Adaptive Mesh Refinement: a new high resolution code called RAMSES&lt;/span&gt;</strong> <div class="csl-block">Romain Teyssier</div> <em>Astron. Astrophys. v.385, p.337-364</em> (2001-11) <a href="http://arxiv.org/abs/astro-ph/0111367 http://dx.doi.org/10.1051/0004-6361:20011817">http://arxiv.org/abs/astro-ph/0111367 http://dx.doi.org/10.1051/0004-6361:20011817</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361:20011817">10.1051/0004-6361:20011817</a></div></div>
</div>
<div id="ref-LpXfagAT" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">19. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Speeding up construction of PMR quadtree-based spatial indexes&lt;/span&gt;</strong> <div class="csl-block">Gisli R Hjaltason, Hanan Samet</div> <em>VLDB J. Int. J. Very Large Data Bases</em> (2002-10) <a href="http://link.springer.com/10.1007/s00778-002-0067-8">http://link.springer.com/10.1007/s00778-002-0067-8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/s00778-002-0067-8">10.1007/s00778-002-0067-8</a></div></div>
</div>
<div id="ref-12vJ72aHG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">20. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Fast construction of k-nearest neighbor graphs for point clouds.&lt;/span&gt;</strong> <div class="csl-block">Michael Connor, Piyush Kumar</div> <em>IEEE Trans. Vis. Comput. Graph.</em> (2010-01) <a href="http://www.ncbi.nlm.nih.gov/pubmed/20467058">http://www.ncbi.nlm.nih.gov/pubmed/20467058</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/TVCG.2010.9">10.1109/tvcg.2010.9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/20467058">20467058</a></div></div>
</div>
<div id="ref-4mimKxQe" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">21. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;A class of data structures for associative searching&lt;/span&gt;</strong> <div class="csl-block">JA Orenstein, TH Merrett</div> <em>Proc. 3rd acm sigact-sigmod symp. Princ. Database syst. - pod. ’84</em> (1984-04) <a href="http://dl.acm.org/citation.cfm?id=588011.588037">http://dl.acm.org/citation.cfm?id=588011.588037</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/588011.588037">10.1145/588011.588037</a> · ISBN: 0897911288</div></div>
</div>
<div id="ref-f1rS5tZk" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">22. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Encoded bitmap indexing for data warehouses&lt;/span&gt;</strong> <div class="csl-block">Ming-Chuan Wu, AP Buchmann</div> <em>Proc. 14th int. Conf. Data eng.</em> (1998) <a href="http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=655780">http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=655780</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/ICDE.1998.655780">10.1109/icde.1998.655780</a> · ISBN: 0-8186-8289-2</div></div>
</div>
<div id="ref-YV4eT8wZ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">23. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Bitmap index design and evaluation&lt;/span&gt;</strong> <div class="csl-block">Chee-Yong Chan, Yannis E Ioannidis</div> <em>ACM SIGMOD Rec.</em> (1998-06) <a href="http://dl.acm.org/citation.cfm?id=276305.276336">http://dl.acm.org/citation.cfm?id=276305.276336</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/276305.276336">10.1145/276305.276336</a> · ISBN: 0-89791-995-5</div></div>
</div>
<div id="ref-DMkimxcD" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">24. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;An efficient bitmap encoding scheme for selection queries&lt;/span&gt;</strong> <div class="csl-block">Chee-Yong Chan, Yannis E Ioannidis</div> <em>ACM SIGMOD Rec.</em> (1999-06) <a href="http://dl.acm.org/citation.cfm?id=304181.304201">http://dl.acm.org/citation.cfm?id=304181.304201</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/304181.304201">10.1145/304181.304201</a> · ISBN: 1-58113-084-8</div></div>
</div>
<div id="ref-itaUSTKC" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">25. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Evaluating Geospatial Geometry and Proximity Queries Using Distributed Hash Tables&lt;/span&gt;</strong> <div class="csl-block">Matthew Malensek, Sangmi Pallickara, Shrideep Pallickara</div> <em>Comput. Sci. Eng.</em> (2014-07) <a href="http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=6785924">http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=6785924</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/MCSE.2014.48">10.1109/mcse.2014.48</a></div></div>
</div>
<div id="ref-QMaCupmb" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">26. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Bitmap indexes for large scientific data sets: a case study&lt;/span&gt;</strong> <div class="csl-block">RR Sinha, S Mitra, M Winslett</div> <em>Proc. 20th ieee int. Parallel distrib. Process. Symp.</em> (2006) <a href="http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=1639304">http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=1639304</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/IPDPS.2006.1639304">10.1109/ipdps.2006.1639304</a> · ISBN: 1-4244-0054-6</div></div>
</div>
<div id="ref-mMJ3UeWp" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">27. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Multi-resolution bitmap indexes for scientific data&lt;/span&gt;</strong> <div class="csl-block">Rishi Rakesh Sinha, Marianne Winslett</div> <em>ACM Trans. Database Syst.</em> (2007-08) <a href="http://portal.acm.org/citation.cfm?doid=1272743.1272746">http://portal.acm.org/citation.cfm?doid=1272743.1272746</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/1272743.1272746">10.1145/1272743.1272746</a></div></div>
</div>
<div id="ref-VzYAeBg3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">28. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Improving the Performance of High-Energy Physics Analysis through Bitmap Indices&lt;/span&gt;</strong> <div class="csl-block">Kurt Stockinger, Dirk Duellmann, Wolfgang Hoschek, Erich Schikuta</div> <em>Database expert syst. Appl. 11th int. Conf. DEXA 2000 london, uk, sept. 4–8, 2000 proc.</em> (2000-06) <a href="http://link.springer.com/10.1007/3-540-44469-6">http://link.springer.com/10.1007/3-540-44469-6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/3-540-44469-6">10.1007/3-540-44469-6</a> · ISBN: 978-3-540-67978-3</div></div>
</div>
<div id="ref-AFyCFou3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">29. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Using bitmap index for interactive exploration of large datasets&lt;/span&gt;</strong> <div class="csl-block">Kesheng Wu, W Koegler, J Chen, A Shoshani</div> <em>15th int. Conf. Sci. Stat. Database manag. 2003.</em> (2003) <a href="http://ieeexplore.ieee.org/articleDetails.jsp?arnumber=1214955">http://ieeexplore.ieee.org/articleDetails.jsp?arnumber=1214955</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/SSDM.2003.1214955">10.1109/ssdm.2003.1214955</a> · ISBN: 0-7695-1964-4</div></div>
</div>
<div id="ref-Sju7gfgZ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">30. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Range-based bitmap indexing for high cardinality attributes with skew&lt;/span&gt;</strong> <div class="csl-block">PS Yu</div> <em>Proceedings. Twenty-second annu. Int. Comput. Softw. Appl. Conf. (Compsac ’98) (cat. No.98CB 36241)</em> (1998) <a href="http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=716637">http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=716637</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/CMPSAC.1998.716637">10.1109/cmpsac.1998.716637</a> · ISBN: 0-8186-8585-9</div></div>
</div>
<div id="ref-uibMnWfE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">31. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Multidimensional indexing and query coordination for tertiary storage management&lt;/span&gt;</strong> <div class="csl-block">A Shoshani, LM Bernardo, H Nordberg, D Rotem, A Sim</div> <em>Proceedings. Elev. Int. Conf. Sci. Stat. Database manag.</em> (1999) <a href="http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=787637">http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=787637</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/SSDM.1999.787637">10.1109/ssdm.1999.787637</a> · ISBN: 0-7695-0046-3</div></div>
</div>
<div id="ref-pe5S32KG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">32. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Evaluation Strategies for Bitmap Indices with Binning&lt;/span&gt;</strong> <div class="csl-block">Kurt Stockinger, Kesheng Wu, Arie Shoshani</div> <em>Database expert syst. Appl. 15th int. Conf. DEXA 2004, zaragoza, spain, august 30-september 3, 2004. Proc.</em> (2004) <a href="http://dx.doi.org/10.1007/978-3-540-30075-5{\_}12">http://dx.doi.org/10.1007/978-3-540-30075-5{\_}12</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/978-3-540-30075-5_12">10.1007/978-3-540-30075-5_12</a> · ISBN: 978-3-540-30075-5</div></div>
</div>
<div id="ref-yuSdhiDQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">33. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Compressed bitmap indices for efficient query processing&lt;/span&gt;</strong> <div class="csl-block">Kesheng Wu, Ekow Otoo, Arie Shoshani</div> <em>Lawrence Berkeley Natl. Lab.</em> (2001-09) <a href="http://escholarship.org/uc/item/8k22w7q2">http://escholarship.org/uc/item/8k22w7q2</a></div>
</div>
<div id="ref-zXsFmNjY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">34. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Sorting improves word-aligned bitmap indexes&lt;/span&gt;</strong> <div class="csl-block">Daniel Lemire, Owen Kaser, Kamel Aouiche</div> <em>Data Knowl. Eng.</em> (2010-01) <a href="http://dl.acm.org/citation.cfm?id=1663645.1663682">http://dl.acm.org/citation.cfm?id=1663645.1663682</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.datak.2009.08.006">10.1016/j.datak.2009.08.006</a></div></div>
</div>
<div id="ref-2qfATif3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">35. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Compressed bitmap indexes: beyond unions and intersections&lt;/span&gt;</strong> <div class="csl-block">Owen Kaser, Daniel Lemire</div> <em>Softw. Pract. Exp.</em> (2016-02) <a href="http://doi.wiley.com/10.1002/spe.2289">http://doi.wiley.com/10.1002/spe.2289</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1002/spe.2289">10.1002/spe.2289</a></div></div>
</div>
<div id="ref-1C6sv2xA4" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">36. </div><div class="csl-right-inline"><strong>Smoothed Particle Hydrodynamics</strong> <div class="csl-block">JJ Monaghan</div> <em>Annu. Rev. Astron. Astrophys.</em> (1992-09) <a href="http://adsabs.harvard.edu/abs/1992ARA{\&amp;}A..30..543M">http://adsabs.harvard.edu/abs/1992ARA{\&amp;}A..30..543M</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1146/annurev.aa.30.090192.002551">10.1146/annurev.aa.30.090192.002551</a></div></div>
</div>
<div id="ref-DQa79Dfo" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">37. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;GADGET: a code for collisionless and gasdynamical cosmological simulations&lt;/span&gt;</strong> <div class="csl-block">Volker Springel, Naoki Yoshida, Simon DM White</div> <em>New Astron.</em> (2001-04) <a href="http://adsabs.harvard.edu/abs/2001NewA....6...79S">http://adsabs.harvard.edu/abs/2001NewA....6...79S</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/S1384-1076(01)00042-2">10.1016/s1384-1076(01)00042-2</a></div></div>
</div>
<div id="ref-18bL4U0kM" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">38. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;A new class of accurate, mesh-free hydrodynamic simulation methods&lt;/span&gt;</strong> <div class="csl-block">PF Hopkins</div> <em>Mon. Not. R. Astron. Soc.</em> (2015-04) <a href="http://adsabs.harvard.edu/abs/2014arXiv1409.7395H">http://adsabs.harvard.edu/abs/2014arXiv1409.7395H</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stv195">10.1093/mnras/stv195</a></div></div>
</div>
<div id="ref-RNeSxk82" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">39. </div><div class="csl-right-inline"><strong>&lt;span class="nocase"&gt;Cython: The Best of Both Worlds&lt;/span&gt;</strong> <div class="csl-block">Stefan Behnel, Robert Bradshaw, Craig Citro, Lisandro Dalcin, Dag Sverre Seljebotn, Kurt Smith</div> <em>Comput. Sci. Eng.</em> (2011-03) <a href="http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=5582062">http://ieeexplore.ieee.org/lpdocs/epic03/wrapper.htm?arnumber=5582062</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/MCSE.2010.118">10.1109/mcse.2010.118</a></div></div>
</div>
<div id="ref-zZcl1XfZ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">40. </div><div class="csl-right-inline"><strong>Adaptive Mesh Refinement for conservative systems: multi-dimensional efficiency evaluation</strong> <div class="csl-block">R Keppens, M Nool, G Tóth, JP Goedbloed</div> <em>Computer Physics Communications</em> (2003-07) <a href="https://doi.org/ck627b">https://doi.org/ck627b</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/s0010-4655(03)00139-5">10.1016/s0010-4655(03)00139-5</a></div></div>
</div>
<div id="ref-aP30JL4t" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">41. </div><div class="csl-right-inline"><strong>MPI-AMRVAC FOR SOLAR AND ASTROPHYSICS</strong> <div class="csl-block">O Porth, C Xia, T Hendrix, SP Moschou, R Keppens</div> <em>The Astrophysical Journal Supplement Series</em> (2014-08-20) <a href="https://doi.org/ggcgbb">https://doi.org/ggcgbb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0067-0049/214/1/4">10.1088/0067-0049/214/1/4</a></div></div>
</div>
<div id="ref-ooR7tjmM" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">42. </div><div class="csl-right-inline"><strong>MPI-AMRVAC 2.0 for Solar and Astrophysical Applications</strong> <div class="csl-block">C Xia, J Teunissen, IEl Mellah, E Chané, R Keppens</div> <em>The Astrophysical Journal Supplement Series</em> (2018-02-01) <a href="https://doi.org/gkh2v6">https://doi.org/gkh2v6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4365/aaa6c8">10.3847/1538-4365/aaa6c8</a></div></div>
</div>
<div id="ref-qhbp5GQg" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">43. </div><div class="csl-right-inline"><strong>Thermal instabilities: Fragmentation and field misalignment of filament fine structure</strong> <div class="csl-block">N Claes, R Keppens, C Xia</div> <em>Astronomy &amp;amp; Astrophysics</em> (2020-04) <a href="https://doi.org/gqbk65">https://doi.org/gqbk65</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/202037616">10.1051/0004-6361/202037616</a></div></div>
</div>
<div id="ref-13jLWXZs6" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">44. </div><div class="csl-right-inline"><strong>Prominence formation by levitation-condensation at extreme resolutions</strong> <div class="csl-block">JM Jenkins, R Keppens</div> <em>Astronomy &amp;amp; Astrophysics</em> (2021-02) <a href="https://doi.org/gqbk66">https://doi.org/gqbk66</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/202039630">10.1051/0004-6361/202039630</a></div></div>
</div>
<div id="ref-rbiXHtk4" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">45. </div><div class="csl-right-inline"><strong>Dynamical signatures of Rossby vortices in cavity-hosting disks</strong> <div class="csl-block">CMT Robert, H Méheut, F Ménard</div> <em>Astronomy &amp;amp; Astrophysics</em> (2020-09) <a href="https://doi.org/gqbk64">https://doi.org/gqbk64</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/201937414">10.1051/0004-6361/201937414</a></div></div>
</div>
<div id="ref-11kNsPLrg" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">46. </div><div class="csl-right-inline"><strong>High Angular Momentum Halo Gas: A Feedback and Code-independent Prediction of LCDM</strong> <div class="csl-block">Kyle R Stewart, Ariyeh H Maller, Jose Oñorbe, James S Bullock, MRyan Joung, Julien Devriendt, Daniel Ceverino, Dušan Kereš, Philip F Hopkins, Claude-André Faucher-Giguère</div> <em>The Astrophysical Journal</em> (2017-06-30) <a href="https://doi.org/gf48kq">https://doi.org/gf48kq</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/aa6dff">10.3847/1538-4357/aa6dff</a></div></div>
</div>
<div id="ref-K2xKlXYH" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">47. </div><div class="csl-right-inline"><strong>Fast galaxy bars continue to challenge standard cosmology</strong> <div class="csl-block">Mahmood Roshan, Neda Ghafourian, Tahere Kashfi, Indranil Banik, Moritz Haslbauer, Virginia Cuomo, Benoit Famaey, Pavel Kroupa</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2021-09-11) <a href="https://doi.org/gqsz56">https://doi.org/gqsz56</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stab2553">10.1093/mnras/stab2553</a></div></div>
</div>
<div id="ref-19CnTqnJi" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">48. </div><div class="csl-right-inline"><strong>O &lt;scp&gt;vi&lt;/scp&gt; traces photoionized streams with collisionally ionized boundaries in cosmological simulations of &lt;i&gt;z&lt;/i&gt; ∼ 1 massive galaxies</strong> <div class="csl-block">Clayton Strawn, Santi Roca-Fàbrega, Nir Mandelker, Joel Primack, Jonathan Stern, Daniel Ceverino, Avishai Dekel, Bryan Wang, Rishi Dange</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2020-12-24) <a href="https://doi.org/grk7k2">https://doi.org/grk7k2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/staa3972">10.1093/mnras/staa3972</a></div></div>
</div>
<div id="ref-8yudbbM3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">49. </div><div class="csl-right-inline"><strong>Testing feedback from star clusters in simulations of the Milky Way formation</strong> <div class="csl-block">Gillen Brown, Oleg Y Gnedin</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2022-04-28) <a href="https://doi.org/grk7k3">https://doi.org/grk7k3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stac1164">10.1093/mnras/stac1164</a></div></div>
</div>
<div id="ref-12IBqVsGq" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">50. </div><div class="csl-right-inline"><strong>Modeling UV Radiation Feedback from Massive Stars. I. Implementation of Adaptive Ray-tracing Method and Tests</strong> <div class="csl-block">Jeong-Gyu Kim, Woong-Tae Kim, Eve C Ostriker, MAaron Skinner</div> <em>The Astrophysical Journal</em> (2017-12-15) <a href="https://doi.org/gqsz6c">https://doi.org/gqsz6c</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/aa9b80">10.3847/1538-4357/aa9b80</a></div></div>
</div>
<div id="ref-159ZGwZYu" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">51. </div><div class="csl-right-inline"><strong>A Magnetized, Moon-forming Giant Impact</strong> <div class="csl-block">PD Mullen, CF Gammie</div> <em>The Astrophysical Journal Letters</em> (2020-10-29) <a href="https://doi.org/gqsz6j">https://doi.org/gqsz6j</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/2041-8213/abbffd">10.3847/2041-8213/abbffd</a></div></div>
</div>
<div id="ref-juPUBGqN" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">52. </div><div class="csl-right-inline"><strong>AMReX: a framework for block-structured adaptive mesh refinement</strong> <div class="csl-block">Weiqun Zhang, Ann Almgren, Vince Beckner, John Bell, Johannes Blaschke, Cy Chan, Marcus Day, Brian Friesen, Kevin Gott, Daniel Graves, … Michael Zingale</div> <em>Journal of Open Source Software</em> (2019-05-12) <a href="https://doi.org/ghr39d">https://doi.org/ghr39d</a> <div class="csl-block">DOI: <a href="https://doi.org/10.21105/joss.01370">10.21105/joss.01370</a></div></div>
</div>
<div id="ref-ETcy2ZP4" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">53. </div><div class="csl-right-inline"><strong>Local adaptive mesh refinement for shock hydrodynamics</strong> <div class="csl-block">MJ Berger, P Colella</div> <em>Journal of Computational Physics</em> (1989-05) <a href="https://doi.org/d6dh8m">https://doi.org/d6dh8m</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/0021-9991(89)90035-1">10.1016/0021-9991(89)90035-1</a></div></div>
</div>
<div id="ref-9y1W7z3a" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">54. </div><div class="csl-right-inline"><strong>Meeting the Challenges of Modeling Astrophysical Thermonuclear Explosions: Castro, Maestro, and the AMReX Astrophysics Suite</strong> <div class="csl-block">M Zingale, AS Almgren, MG Barrios Sazo, VE Beckner, JB Bell, B Friesen, AM Jacobs, MP Katz, CM Malone, AJ Nonaka, … W Zhang</div> <em>Journal of Physics: Conference Series</em> (2018-05) <a href="https://doi.org/gqsz53">https://doi.org/gqsz53</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/1742-6596/1031/1/012024">10.1088/1742-6596/1031/1/012024</a></div></div>
</div>
<div id="ref-Ul6Zyy8M" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">55. </div><div class="csl-right-inline"><strong>CASTRO: A Massively Parallel Compressible Astrophysics Simulation Code</strong> <div class="csl-block">Ann Almgren, Maria Sazo, John Bell, Alice Harpole, Max Katz, Jean Sexton, Donald Willcox, Weiqun Zhang, Michael Zingale</div> <em>Journal of Open Source Software</em> (2020-10-23) <a href="https://doi.org/grk7k8">https://doi.org/grk7k8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.21105/joss.02513">10.21105/joss.02513</a></div></div>
</div>
<div id="ref-15s0VliuP" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">56. </div><div class="csl-right-inline"><strong>MAESTROeX: A Massively Parallel Low Mach Number Astrophysical Solver</strong> <div class="csl-block">Duoming Fan, Andrew Nonaka, Ann Almgren, Donald Willcox, Alice Harpole, Michael Zingale</div> <em>Journal of Open Source Software</em> (2019-11-21) <a href="https://doi.org/grk7k7">https://doi.org/grk7k7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.21105/joss.01757">10.21105/joss.01757</a></div></div>
</div>
<div id="ref-1HoHM1Gnp" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">57. </div><div class="csl-right-inline"><strong>&lt;scp&gt;quokka&lt;/scp&gt;: a code for two-moment AMR radiation hydrodynamics on GPUs</strong> <div class="csl-block">Benjamin D Wibking, Mark R Krumholz</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2022-02-18) <a href="https://doi.org/gpsbw2">https://doi.org/gpsbw2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stac439">10.1093/mnras/stac439</a></div></div>
</div>
<div id="ref-Igp8qi2S" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">58. </div><div class="csl-right-inline"><strong>Numerical simulations of buoyancy-driven flows using adaptive mesh refinement: structure and dynamics of a large-scale helium plume</strong> <div class="csl-block">Nicholas T Wimer, Marcus S Day, Caelan Lapointe, Michael A Meehan, Amanda S Makowiecki, Jeffrey F Glusman, John W Daily, Gregory B Rieker, Peter E Hamlington</div> <em>Theoretical and Computational Fluid Dynamics</em> (2020-09-01) <a href="https://doi.org/gk8ssc">https://doi.org/gk8ssc</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/s00162-020-00548-6">10.1007/s00162-020-00548-6</a></div></div>
</div>
<div id="ref-1HiWhAc3l" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">59. </div><div class="csl-right-inline"><strong>Formation and mass growth of axion stars in axion miniclusters</strong> <div class="csl-block">Benedikt Eggemeier, Jens C Niemeyer</div> <em>Physical Review D</em> (2019-09-23) <a href="https://doi.org/grp7h3">https://doi.org/grp7h3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1103/physrevd.100.063528">10.1103/physrevd.100.063528</a></div></div>
</div>
<div id="ref-rfkfEHN0" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">60. </div><div class="csl-right-inline"><strong>Formation of inflaton halos after inflation</strong> <div class="csl-block">Benedikt Eggemeier, Jens C Niemeyer, Richard Easther</div> <em>Physical Review D</em> (2021-03-22) <a href="https://doi.org/f3g3">https://doi.org/f3g3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1103/physrevd.103.063525">10.1103/physrevd.103.063525</a></div></div>
</div>
<div id="ref-aETt4JEy" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">61. </div><div class="csl-right-inline"><strong>Gravitational collapse in the postinflationary Universe</strong> <div class="csl-block">Benedikt Eggemeier, Bodo Schwabe, Jens C Niemeyer, Richard Easther</div> <em>Physical Review D</em> (2022-01-12) <a href="https://doi.org/grp7hm">https://doi.org/grp7hm</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1103/physrevd.105.023516">10.1103/physrevd.105.023516</a></div></div>
</div>
<div id="ref-whh3mxKo" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">62. </div><div class="csl-right-inline"><strong>A hybrid nodal-staggered pseudo-spectral electromagnetic particle-in-cell method with finite-order centering</strong> <div class="csl-block">Edoardo Zoni, Remi Lehe, Olga Shapoval, Daniel Belkin, Neil Zaïm, Luca Fedeli, Henri Vincenti, Jean-Luc Vay</div> <em>Computer Physics Communications</em> (2022-10) <a href="https://doi.org/grp7hk">https://doi.org/grp7hk</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.cpc.2022.108457">10.1016/j.cpc.2022.108457</a></div></div>
</div>
<div id="ref-rUJIl3i2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">63. </div><div class="csl-right-inline"><strong>CHOLLA: A NEW MASSIVELY PARALLEL HYDRODYNAMICS CODE FOR ASTROPHYSICAL SIMULATION</strong> <div class="csl-block">Evan E Schneider, Brant E Robertson</div> <em>The Astrophysical Journal Supplement Series</em> (2015-04-10) <a href="https://doi.org/gr2xqz">https://doi.org/gr2xqz</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0067-0049/217/2/24">10.1088/0067-0049/217/2/24</a></div></div>
</div>
<div id="ref-MDEDwL4n" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">64. </div><div class="csl-right-inline"><strong>The Role of Outflows, Radiation Pressure, and Magnetic Fields in Massive Star Formation</strong> <div class="csl-block">Anna L Rosen, Mark R Krumholz</div> <em>The Astronomical Journal</em> (2020-07-23) <a href="https://doi.org/grk7k9">https://doi.org/grk7k9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-3881/ab9abf">10.3847/1538-3881/ab9abf</a></div></div>
</div>
<div id="ref-Mz86NzmT" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">65. </div><div class="csl-right-inline"><strong>Massive-star Formation via the Collapse of Subvirial and Virialized Turbulent Massive Cores</strong> <div class="csl-block">Anna L Rosen, Pak Shing Li, Qizhou Zhang, Blakesley Burkhart</div> <em>The Astrophysical Journal</em> (2019-12-16) <a href="https://doi.org/grk7mc">https://doi.org/grk7mc</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ab54c6">10.3847/1538-4357/ab54c6</a></div></div>
</div>
<div id="ref-pHJo1lO1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">66. </div><div class="csl-right-inline"><strong>An unstable truth: how massive stars get their mass</strong> <div class="csl-block">Anna L Rosen, Mark R Krumholz, Christopher F McKee, Richard I Klein</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2016-08-25) <a href="https://doi.org/f9rwrm">https://doi.org/f9rwrm</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stw2153">10.1093/mnras/stw2153</a></div></div>
</div>
<div id="ref-158VPhStY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">67. </div><div class="csl-right-inline"><strong>Revealing the physical properties of gas accreting to haloes in the EAGLE simulations</strong> <div class="csl-block">Ruby J Wright, Claudia del P Lagos, Chris Power, Camila A Correa</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2021-04-17) <a href="https://doi.org/gqsz55">https://doi.org/gqsz55</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stab1057">10.1093/mnras/stab1057</a></div></div>
</div>
<div id="ref-tZ9VJJVi" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">68. </div><div class="csl-right-inline"><strong>Barred spiral galaxies in modified gravity theories</strong> <div class="csl-block">Mahmood Roshan, Indranil Banik, Neda Ghafourian, Ingo Thies, Benoit Famaey, Elena Asencio, Pavel Kroupa</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2021-03-08) <a href="https://doi.org/gqsz57">https://doi.org/gqsz57</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stab651">10.1093/mnras/stab651</a></div></div>
</div>
<div id="ref-qsP6SRrG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">69. </div><div class="csl-right-inline"><strong>Stellar splashback: the edge of the intracluster light</strong> <div class="csl-block">Alis J Deason, Kyle A Oman, Azadeh Fattahi, Matthieu Schaller, Mathilde Jauzac, Yuanyuan Zhang, Mireia Montes, Yannick M Bahé, Claudio Dalla Vecchia, Scott T Kay, Tilly A Evans</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2020-11-19) <a href="https://doi.org/gqsz54">https://doi.org/gqsz54</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/staa3590">10.1093/mnras/staa3590</a></div></div>
</div>
<div id="ref-1AOjjAzwL" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">70. </div><div class="csl-right-inline"><strong>ENZO: AN ADAPTIVE MESH REFINEMENT CODE FOR ASTROPHYSICS</strong> <div class="csl-block">Greg L Bryan, Michael L Norman, Brian W O'Shea, Tom Abel, John H Wise, Matthew J Turk, Daniel R Reynolds, David C Collins, Peng Wang, Samuel W Skillman, … </div> <em>The Astrophysical Journal Supplement Series</em> (2014-03-20) <a href="https://doi.org/gfx7v2">https://doi.org/gfx7v2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0067-0049/211/2/19">10.1088/0067-0049/211/2/19</a></div></div>
</div>
<div id="ref-12pekKHkD" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">71. </div><div class="csl-right-inline"><strong>COLLAPSE AND FRAGMENTATION OF MAGNETIC MOLECULAR CLOUD CORES WITH THE ENZO AMR MHD CODE. I. UNIFORM DENSITY SPHERES</strong> <div class="csl-block">Alan P Boss, Sandra A Keiser</div> <em>The Astrophysical Journal</em> (2013-02-01) <a href="https://doi.org/gkmgb4">https://doi.org/gkmgb4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0004-637x/764/2/136">10.1088/0004-637x/764/2/136</a></div></div>
</div>
<div id="ref-XEFlUFJ9" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">72. </div><div class="csl-right-inline"><strong>Predicting Localized Primordial Star Formation with Deep Convolutional Neural Networks</strong> <div class="csl-block">Azton I Wells, Michael L Norman</div> <em>The Astrophysical Journal Supplement Series</em> (2021-06-01) <a href="https://doi.org/grk7mf">https://doi.org/grk7mf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4365/abfa17">10.3847/1538-4365/abfa17</a></div></div>
</div>
<div id="ref-EKVFk61S" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">73. </div><div class="csl-right-inline"><strong>Connecting Primordial Star-forming Regions and Second-generation Star Formation in the Phoenix Simulations</strong> <div class="csl-block">Azton I Wells, Michael L Norman</div> <em>The Astrophysical Journal</em> (2022-06-01) <a href="https://doi.org/grk7md">https://doi.org/grk7md</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ac6c87">10.3847/1538-4357/ac6c87</a></div></div>
</div>
<div id="ref-QmIwzRzQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">74. </div><div class="csl-right-inline"><strong>The birth of a galaxy – III. Propelling reionization with the faintest galaxies</strong> <div class="csl-block">John H Wise, Vasiliy G Demchenko, Martin T Halicek, Michael L Norman, Matthew J Turk, Tom Abel, Britton D Smith</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2014-06-26) <a href="https://doi.org/grk7k6">https://doi.org/grk7k6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stu979">10.1093/mnras/stu979</a></div></div>
</div>
<div id="ref-nKZ32a6" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">75. </div><div class="csl-right-inline"><strong>Evolution of Primordial Magnetic Fields during Large-scale Structure Formation</strong> <div class="csl-block">Salome Mtchedlidze, Paola Domínguez-Fernández, Xiaolong Du, Axel Brandenburg, Tina Kahniashvili, Shane O’Sullivan, Wolfram Schmidt, Marcus Brüggen</div> <em>The Astrophysical Journal</em> (2022-04-01) <a href="https://doi.org/gp32mc">https://doi.org/gp32mc</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ac5960">10.3847/1538-4357/ac5960</a></div></div>
</div>
<div id="ref-AQXulJGy" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">76. </div><div class="csl-right-inline"><strong>A Simple Model for Mixing and Cooling in Cloud–Wind Interactions</strong> <div class="csl-block">Matthew W Abruzzo, Greg L Bryan, Drummond B Fielding</div> <em>The Astrophysical Journal</em> (2022-02-01) <a href="https://doi.org/gqsz6g">https://doi.org/gqsz6g</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ac3c48">10.3847/1538-4357/ac3c48</a></div></div>
</div>
<div id="ref-1CqhBqTLw" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">77. </div><div class="csl-right-inline"><strong>Type Ia supernova ejecta–donor interaction: explosion model comparison</strong> <div class="csl-block">C McCutcheon, Y Zeng, Z-W Liu, RG Izzard, K-C Pan, H-L Chen, Z Han</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2022-05-10) <a href="https://doi.org/gqsz58">https://doi.org/gqsz58</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stac1275">10.1093/mnras/stac1275</a></div></div>
</div>
<div id="ref-1BmyNSium" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">78. </div><div class="csl-right-inline"><strong>Exploring Fundamentally Three-dimensional Phenomena in High-fidelity Simulations of Core-collapse Supernovae</strong> <div class="csl-block">Evan P O’Connor, Sean M Couch</div> <em>The Astrophysical Journal</em> (2018-09-25) <a href="https://doi.org/gjctnx">https://doi.org/gjctnx</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/aadcf7">10.3847/1538-4357/aadcf7</a></div></div>
</div>
<div id="ref-3RHUErzU" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">79. </div><div class="csl-right-inline"><strong>Implementing primordial binaries in simulations of star cluster formation with a hybrid MHD and direct &lt;i&gt;N&lt;/i&gt;-body method</strong> <div class="csl-block">Claude Cournoyer-Cloutier, Aaron Tran, Sean Lewis, Joshua E Wall, William E Harris, Mordecai-Mark Mac Low, Stephen LW McMillan, Simon Portegies Zwart, Alison Sills</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2020-12-18) <a href="https://doi.org/grk7kz">https://doi.org/grk7kz</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/staa3902">10.1093/mnras/staa3902</a></div></div>
</div>
<div id="ref-m3lUExr1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">80. </div><div class="csl-right-inline"><strong>Collisional N-body Dynamics Coupled to Self-gravitating Magnetohydrodynamics Reveals Dynamical Binary Formation</strong> <div class="csl-block">Joshua E Wall, Stephen LW McMillan, Mordecai-Mark Mac Low, Ralf S Klessen, Simon Portegies Zwart</div> <em>The Astrophysical Journal</em> (2019-12-10) <a href="https://doi.org/grk7mb">https://doi.org/grk7mb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ab4db1">10.3847/1538-4357/ab4db1</a></div></div>
</div>
<div id="ref-dqw1JuL" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">81. </div><div class="csl-right-inline"><strong>Modeling of the Effects of Stellar Feedback during Star Cluster Formation Using a Hybrid Gas and N-Body Method</strong> <div class="csl-block">Joshua E Wall, Mordecai-Mark Mac Low, Stephen LW McMillan, Ralf S Klessen, Simon Portegies Zwart, Andrew Pellegrino</div> <em>The Astrophysical Journal</em> (2020-12-01) <a href="https://doi.org/gphr2s">https://doi.org/gphr2s</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/abc011">10.3847/1538-4357/abc011</a></div></div>
</div>
<div id="ref-wzYhh0PP" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">82. </div><div class="csl-right-inline"><strong>Winds in Star Clusters Drive Kolmogorov Turbulence</strong> <div class="csl-block">Monica Gallegos-Garcia, Blakesley Burkhart, Anna L Rosen, Jill P Naiman, Enrico Ramirez-Ruiz</div> <em>The Astrophysical Journal Letters</em> (2020-08-01) <a href="https://doi.org/grk7mg">https://doi.org/grk7mg</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/2041-8213/ababae">10.3847/2041-8213/ababae</a></div></div>
</div>
<div id="ref-UiYJOB8x" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">83. </div><div class="csl-right-inline"><strong>The AGORA High-resolution Galaxy Simulations Comparison Project. III. Cosmological Zoom-in Simulation of a Milky Way–mass Halo</strong> <div class="csl-block">Santi Roca-Fàbrega, Ji-hoon Kim, Loic Hausammann, Kentaro Nagamine, Alessandro Lupi, Johnny W Powell, Ikkoh Shimizu, Daniel Ceverino, Joel R Primack, Thomas R Quinn, … </div> <em>The Astrophysical Journal</em> (2021-08-01) <a href="https://doi.org/gqsz6d">https://doi.org/gqsz6d</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ac088a">10.3847/1538-4357/ac088a</a></div></div>
</div>
<div id="ref-12qB7HhOD" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">84. </div><div class="csl-right-inline"><strong>gamer-2: a GPU-accelerated adaptive mesh refinement code – accuracy, performance, and scalability</strong> <div class="csl-block">Hsi-Yu Schive, John A ZuHone, Nathan J Goldbaum, Matthew J Turk, Massimo Gaspari, Chin-Yu Cheng</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2018-09-24) <a href="https://doi.org/gkvtzd">https://doi.org/gkvtzd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/sty2586">10.1093/mnras/sty2586</a></div></div>
</div>
<div id="ref-19Yvj8KDs" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">85. </div><div class="csl-right-inline"><strong>The galactic dust-up: modelling dust evolution in FIRE</strong> <div class="csl-block">Caleb R Choban, Dušan Kereš, Philip F Hopkins, Karin M Sandstrom, Christopher C Hayward, Claude-André Faucher-Giguère</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2022-06-07) <a href="https://doi.org/gqsz59">https://doi.org/gqsz59</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stac1542">10.1093/mnras/stac1542</a></div></div>
</div>
<div id="ref-P48PyrZ1" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">86. </div><div class="csl-right-inline"><strong>How the super-Eddington regime regulates black hole growth in high-redshift galaxies</strong> <div class="csl-block">Warren Massonneau, Marta Volonteri, Yohan Dubois, Ricarda S Beckmann</div> <em>Astronomy &amp;amp; Astrophysics</em> (2023-02) <a href="https://doi.org/g8wcs3">https://doi.org/g8wcs3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/202243170">10.1051/0004-6361/202243170</a></div></div>
</div>
<div id="ref-19Fc0UYAA" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">87. </div><div class="csl-right-inline"><strong>The OBELISK simulation: Galaxies contribute more than AGN to H I reionization of protoclusters</strong> <div class="csl-block">Maxime Trebitsch, Yohan Dubois, Marta Volonteri, Hugo Pfister, Corentin Cadiou, Harley Katz, Joakim Rosdahl, Taysun Kimm, Christophe Pichon, Ricarda S Beckmann, … Adrianne Slyz</div> <em>Astronomy &amp;amp; Astrophysics</em> (2021-09) <a href="https://doi.org/gnh2vh">https://doi.org/gnh2vh</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/202037698">10.1051/0004-6361/202037698</a></div></div>
</div>
<div id="ref-WeeFmczr" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">88. </div><div class="csl-right-inline"><strong>Dense gas formation and destruction in a simulated Perseus-like galaxy cluster with spin-driven black hole feedback</strong> <div class="csl-block">RS Beckmann, Y Dubois, P Guillard, P Salome, V Olivares, F Polles, C Cadiou, F Combes, S Hamer, MD Lehnert, G Pineau des Forets</div> <em>Astronomy &amp;amp; Astrophysics</em> (2019-10-21) <a href="https://doi.org/grk7kw">https://doi.org/grk7kw</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/201936188">10.1051/0004-6361/201936188</a></div></div>
</div>
<div id="ref-zrKJQN1c" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">89. </div><div class="csl-right-inline"><strong>The bridge: a transient phenomenon of forming stellar multiples</strong> <div class="csl-block">M Kuffmeier, H Calcutt, LE Kristensen</div> <em>Astronomy &amp;amp; Astrophysics</em> (2019-08) <a href="https://doi.org/gh824h">https://doi.org/gh824h</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/201935504">10.1051/0004-6361/201935504</a></div></div>
</div>
<div id="ref-KeH9yqn6" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">90. </div><div class="csl-right-inline"><strong>Chaos and variance in galaxy formation</strong> <div class="csl-block">BW Keller, JW Wadsley, L Wang, JMD Kruijssen</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2018-10-23) <a href="https://doi.org/g8wcs5">https://doi.org/g8wcs5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/sty2859">10.1093/mnras/sty2859</a></div></div>
</div>
<div id="ref-14vlGPzZS" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">91. </div><div class="csl-right-inline"><strong>Episodic accretion: the interplay of infall and disc instabilities</strong> <div class="csl-block">Michael Kuffmeier, Søren Frimann, Sigurd S Jensen, Troels Haugbølle</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2018-01-09) <a href="https://doi.org/gc8hr3">https://doi.org/gc8hr3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/sty024">10.1093/mnras/sty024</a></div></div>
</div>
<div id="ref-Gj37qtQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">92. </div><div class="csl-right-inline"><strong>Following the Cosmic Evolution of Pristine Gas. II. The Search for Pop III–bright Galaxies</strong> <div class="csl-block">Richard Sarmento, Evan Scannapieco, Seth Cohen</div> <em>The Astrophysical Journal</em> (2018-02-10) <a href="https://doi.org/gkh3zb">https://doi.org/gkh3zb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/aa989a">10.3847/1538-4357/aa989a</a></div></div>
</div>
<div id="ref-OgdghMdw" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">93. </div><div class="csl-right-inline"><strong>Simulating neutron star mergers as r-process sources in ultrafaint dwarf galaxies</strong> <div class="csl-block">Mohammadtaher Safarzadeh, Evan Scannapieco</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2017-07-08) <a href="https://doi.org/gbwzg9">https://doi.org/gbwzg9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stx1706">10.1093/mnras/stx1706</a></div></div>
</div>
<div id="ref-1f52Kyt" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">94. </div><div class="csl-right-inline"><strong>THE AGORA HIGH-RESOLUTION GALAXY SIMULATIONS COMPARISON PROJECT. II. ISOLATED DISK TEST</strong> <div class="csl-block">Ji-hoon Kim, Oscar Agertz, Romain Teyssier, Michael J Butler, Daniel Ceverino, Jun-Hwan Choi, Robert Feldmann, Ben W Keller, Alessandro Lupi, Thomas Quinn, … </div> <em>The Astrophysical Journal</em> (2016-12-19) <a href="https://doi.org/f9r5cx">https://doi.org/f9r5cx</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/833/2/202">10.3847/1538-4357/833/2/202</a></div></div>
</div>
<div id="ref-lBI1SKpf" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">95. </div><div class="csl-right-inline"><strong>TRACKING THE DISTRIBUTION OF &lt;sup&gt;26&lt;/sup&gt;Al AND &lt;sup&gt;60&lt;/sup&gt;Fe DURING THE EARLY PHASES OF STAR AND DISK EVOLUTION</strong> <div class="csl-block">Michael Kuffmeier, Troels Frostholm Mogensen, Troels Haugbølle, Martin Bizzarro, Åke Nordlund</div> <em>The Astrophysical Journal</em> (2016-07-18) <a href="https://doi.org/gthjr7">https://doi.org/gthjr7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/0004-637x/826/1/22">10.3847/0004-637x/826/1/22</a></div></div>
</div>
<div id="ref-jwp8Rkvm" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">96. </div><div class="csl-right-inline"><strong>THE FORMATION OF FILAMENTARY BUNDLES IN TURBULENT MOLECULAR CLOUDS</strong> <div class="csl-block">Nickolas Moeckel, Andreas Burkert</div> <em>The Astrophysical Journal</em> (2015-07-01) <a href="https://doi.org/g8wcs4">https://doi.org/g8wcs4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0004-637x/807/1/67">10.1088/0004-637x/807/1/67</a></div></div>
</div>
<div id="ref-39ZLqJPZ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">97. </div><div class="csl-right-inline"><strong>THE AGORA HIGH-RESOLUTION GALAXY SIMULATIONS COMPARISON PROJECT</strong> <div class="csl-block">Ji-hoon Kim, Tom Abel, Oscar Agertz, Greg L Bryan, Daniel Ceverino, Charlotte Christensen, Charlie Conroy, Avishai Dekel, Nickolay Y Gnedin, Nathan J Goldbaum, … </div> <em>The Astrophysical Journal Supplement Series</em> (2013-12-24) <a href="https://doi.org/gjc275">https://doi.org/gjc275</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0067-0049/210/1/14">10.1088/0067-0049/210/1/14</a></div></div>
</div>
<div id="ref-aLNHrylH" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">98. </div><div class="csl-right-inline"><strong>A model for cosmological simulations of galaxy formation physics</strong> <div class="csl-block">Mark Vogelsberger, Shy Genel, Debora Sijacki, Paul Torrey, Volker Springel, Lars Hernquist</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2013-10-23) <a href="https://doi.org/f5kh8c">https://doi.org/f5kh8c</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stt1789">10.1093/mnras/stt1789</a></div></div>
</div>
<div id="ref-11r9ZRs87" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">99. </div><div class="csl-right-inline"><strong>Nyx: A MASSIVELY PARALLEL AMR CODE FOR COMPUTATIONAL COSMOLOGY</strong> <div class="csl-block">Ann S Almgren, John B Bell, Mike J Lijewski, Zarija Lukić, Ethan Van Andel</div> <em>The Astrophysical Journal</em> (2013-02-13) <a href="https://doi.org/gkmgcq">https://doi.org/gkmgcq</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0004-637x/765/1/39">10.1088/0004-637x/765/1/39</a></div></div>
</div>
<div id="ref-RMe7IYaY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">100. </div><div class="csl-right-inline"><strong>Stellar angular momentum can be controlled from cosmological initial conditions</strong> <div class="csl-block">Corentin Cadiou, Andrew Pontzen, Hiranya V Peiris</div> <em>arXiv</em> (2022) <a href="https://doi.org/grk7mh">https://doi.org/grk7mh</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.2206.11913">10.48550/arxiv.2206.11913</a></div></div>
</div>
<div id="ref-8ElS8qaE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">101. </div><div class="csl-right-inline"><strong>Accurate tracer particles of baryon dynamics in the adaptive mesh refinement code Ramses</strong> <div class="csl-block">Corentin Cadiou, Yohan Dubois, Christophe Pichon</div> <em>Astronomy &amp;amp; Astrophysics</em> (2019-01) <a href="https://doi.org/gf42rj">https://doi.org/gf42rj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/201834496">10.1051/0004-6361/201834496</a></div></div>
</div>
<div id="ref-NLdlIida" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">102. </div><div class="csl-right-inline"><strong>Gravitational torques dominate the dynamics of accreted gas at &lt;i&gt;z&lt;/i&gt; &amp;amp;gt; 2</strong> <div class="csl-block">Corentin Cadiou, Yohan Dubois, Christophe Pichon</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2022-06-17) <a href="https://doi.org/grk7k4">https://doi.org/grk7k4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stac1663">10.1093/mnras/stac1663</a></div></div>
</div>
<div id="ref-JV9vVQtk" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">103. </div><div class="csl-right-inline"><strong>Gas accretion and ram pressure stripping of haloes in void walls</strong> <div class="csl-block">BB Thompson, R Smith, K Kraljic</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2022-10-19) <a href="https://doi.org/grk7k5">https://doi.org/grk7k5</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stac2963">10.1093/mnras/stac2963</a></div></div>
</div>
<div id="ref-Z9ORyk4e" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">104. </div><div class="csl-right-inline"><strong>The Gaia-ESO Survey: matching chemodynamical simulations to observations of the Milky Way</strong> <div class="csl-block">BB Thompson, CG Few, M Bergemann, BK Gibson, BA MacFarlane, A Serenelli, G Gilmore, S Randich, A Vallenari, EJ Alfaro, … S Zaggia</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2017-09-11) <a href="https://doi.org/gc8ws7">https://doi.org/gc8ws7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stx2316">10.1093/mnras/stx2316</a></div></div>
</div>
<div id="ref-1Ce48o4Xl" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">105. </div><div class="csl-right-inline"><strong>Turnaround radius of galaxy clusters in &lt;i&gt;N&lt;/i&gt;-body simulations</strong> <div class="csl-block">Giorgos Korkidis, Vasiliki Pavlidou, Konstantinos Tassis, Evangelia Ntormousi, Theodore N Tomaras, Konstantinos Kovlakas</div> <em>Astronomy &amp;amp; Astrophysics</em> (2020-07) <a href="https://doi.org/gqsz52">https://doi.org/gqsz52</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/201937337">10.1051/0004-6361/201937337</a></div></div>
</div>
<div id="ref-VVJcMOza" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">106. </div><div class="csl-right-inline"><strong>3D volume rendering of geophysical data using the yt platform</strong> <div class="csl-block">Christopher Havlin, Benjamin Holtzman, Kacper Kowalik, Madicken Munk, Samantha Walkow, Matthew Turk</div> <em>Wiley</em> (2021-02-10) <a href="https://doi.org/grk7kv">https://doi.org/grk7kv</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1002/essoar.10506118.2">10.1002/essoar.10506118.2</a></div></div>
</div>
<div id="ref-16ZVBcAT2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">107. </div><div class="csl-right-inline"><strong>Understanding Heating in Active Region Cores through Machine Learning. I. Numerical Modeling and Predicted Observables</strong> <div class="csl-block">WT Barnes, SJ Bradshaw, NM Viall</div> <em>The Astrophysical Journal</em> (2019-07-20) <a href="https://doi.org/grk7m6">https://doi.org/grk7m6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ab290c">10.3847/1538-4357/ab290c</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31920207">31920207</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6951256">PMC6951256</a></div></div>
</div>
<div id="ref-10pVv5dtC" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">108. </div><div class="csl-right-inline"><strong>swiftsimio: A Python library for reading SWIFT data</strong> <div class="csl-block">Josh Borrow, Alexei Borrisov</div> <em>Journal of Open Source Software</em> (2020-08-01) <a href="https://doi.org/gqsz6b">https://doi.org/gqsz6b</a> <div class="csl-block">DOI: <a href="https://doi.org/10.21105/joss.02430">10.21105/joss.02430</a></div></div>
</div>
<div id="ref-17ydwlTdn" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">109. </div><div class="csl-right-inline"><strong>sígame v3: Gas Fragmentation in Postprocessing of Cosmological Simulations for More Accurate Infrared Line Emission Modeling</strong> <div class="csl-block">Karen Pardos Olsen, Blakesley Burkhart, Mordecai-Mark Mac Low, Robin G Treß, Thomas R Greve, David Vizgan, Jay Motka, Josh Borrow, Gergö Popping, Romeel Davé, … Desika Narayanan</div> <em>The Astrophysical Journal</em> (2021-11-01) <a href="https://doi.org/gqsz6f">https://doi.org/gqsz6f</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ac20d4">10.3847/1538-4357/ac20d4</a></div></div>
</div>
<div id="ref-100I0ZLek" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">110. </div><div class="csl-right-inline"><strong>Sensitivity of non-radiative cloud-wind interactions to the hydrodynamics solver</strong> <div class="csl-block">Joey Braspenning, Joop Schaye, Josh Borrow, Matthieu Schaller</div> <em>arXiv</em> (2022) <a href="https://doi.org/grp7hn">https://doi.org/grp7hn</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.2203.13915">10.48550/arxiv.2203.13915</a></div></div>
</div>
<div id="ref-YCT49Lcv" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">111. </div><div class="csl-right-inline"><strong>AstroBlend: An astrophysical visualization package for Blender</strong> <div class="csl-block">JP Naiman</div> <em>Astronomy and Computing</em> (2016-04) <a href="https://doi.org/f8phr9">https://doi.org/f8phr9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.ascom.2016.02.002">10.1016/j.ascom.2016.02.002</a></div></div>
</div>
<div id="ref-100MO1Ryn" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">112. </div><div class="csl-right-inline"><strong>The anatomy of a star-forming galaxy: pressure-driven regulation of star formation in simulated galaxies</strong> <div class="csl-block">SM Benincasa, J Wadsley, HMP Couchman, BW Keller</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2016-07-20) <a href="https://doi.org/f86kmr">https://doi.org/f86kmr</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/stw1741">10.1093/mnras/stw1741</a></div></div>
</div>
<div id="ref-kWd16rLo" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">113. </div><div class="csl-right-inline"><strong>The growth of black holes from Population III remnants in the Renaissance simulations</strong> <div class="csl-block">Britton D Smith, John A Regan, Turlough P Downes, Michael L Norman, Brian W O’Shea, John H Wise</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2018-08-02) <a href="https://doi.org/gfn3dp">https://doi.org/gfn3dp</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/sty2103">10.1093/mnras/sty2103</a></div></div>
</div>
<div id="ref-kEEsXbtq" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">114. </div><div class="csl-right-inline"><strong>Parthenon—a performance portable block-structured adaptive mesh refinement framework</strong> <div class="csl-block">Philipp Grete, Joshua C Dolence, Jonah M Miller, Joshua Brown, Ben Ryan, Andrew Gaspar, Forrest Glines, Sriram Swaminarayan, Jonas Lippuner, Clell J Solomon, … Luke F Roberts</div> <em>The International Journal of High Performance Computing Applications</em> (2022-12-13) <a href="https://doi.org/grzcrd">https://doi.org/grzcrd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1177/10943420221143775">10.1177/10943420221143775</a></div></div>
</div>
<div id="ref-8BOQ3JKi" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">115. </div><div class="csl-right-inline"><strong>A Fast Voxel Traversal Algorithm for Ray Tracing</strong> <div class="csl-block">John Amanatides, Andrew Woo</div> <em>EG 1987-Technical Papers</em> (1987) <a href="https://doi.org/gq2qd9">https://doi.org/gq2qd9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.2312/egtp.19871000">10.2312/egtp.19871000</a></div></div>
</div>
<div id="ref-FQhLVjwO" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">116. </div><div class="csl-right-inline"><strong>An Efficient Parametric Algorithm for Octree Traversal</strong> <div class="csl-block">Jorge Revelles, Carlos Ureña, Miguel Lastra</div> <em>Journal of WSCG</em> <a href="http://wscg.zcu.cz/wscg2000/Papers_2000/X31.pdf">http://wscg.zcu.cz/wscg2000/Papers_2000/X31.pdf</a></div>
</div>
<div id="ref-zRE8zhVC" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">117. </div><div class="csl-right-inline"><strong>Gravitational torques dominate the dynamics of angular momentum of the accreted gas at $z&gt;2$</strong> <div class="csl-block">Corentin Cadiou, Yohan Dubois, Christophe Pichon</div> <em>in prep.</em></div>
</div>
<div id="ref-QcHRZTY3" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">118. </div><div class="csl-right-inline"><strong>unyt: Handle, manipulate, and convert data with units in Python</strong> <div class="csl-block">Nathan J. Goldbaum, John A. ZuHone, Matthew J. Turk, Kacper Kowalik, Anna L. Rosen</div> <em>Journal of Open Source Software</em> (2018-08-14) <a href="https://doi.org/gsnf9k">https://doi.org/gsnf9k</a> <div class="csl-block">DOI: <a href="https://doi.org/10.21105/joss.00809">10.21105/joss.00809</a></div></div>
</div>
<div id="ref-1DeF7kDPM" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">119. </div><div class="csl-right-inline"><strong>matplotlib/matplotlib: REL: v3.7.0rc1</strong> <div class="csl-block">Thomas A Caswell, Antony Lee, Elliott Sales De Andrade, Michael Droettboom, Tim Hoffmann, Jody Klymak, John Hunter, Eric Firing, David Stansby, Nelle Varoquaux, … Nikita Kniazev</div> <em>Zenodo</em> (2023-01-25) <a href="https://doi.org/grqcmh">https://doi.org/grqcmh</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5281/zenodo.7570264">10.5281/zenodo.7570264</a></div></div>
</div>
<div id="ref-XobbtsFq" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">120. </div><div class="csl-right-inline"><strong>The Astropy Project: Sustaining and Growing a Community-oriented Open-source Project and the Latest Major Release (v5.0) of the Core Package</strong> <div class="csl-block">Astropy Collaboration, Adrian M Price-Whelan, Pey Lian Lim, Nicholas Earl, Nathaniel Starkman, Larry Bradley, David L Shupe, Aarya A Patil, Lia Corrales, CE Brasseur, … Astropy Project Contributors</div> (2022-08) <a href="https://arxiv.org/abs/2206.14220">https://arxiv.org/abs/2206.14220</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/ac7c74">10.3847/1538-4357/ac7c74</a></div></div>
</div>
<div id="ref-lOgtbiQz" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">121. </div><div class="csl-right-inline"><strong>Jupyter: Thinking and Storytelling With Code and Data</strong> <div class="csl-block">Brian E Granger, Fernando Perez</div> <em>Computing in Science &amp;amp; Engineering</em> (2021-03-01) <a href="https://doi.org/gjkwx2">https://doi.org/gjkwx2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/mcse.2021.3059263">10.1109/mcse.2021.3059263</a></div></div>
</div>
<div id="ref-16Nvi8gCt" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">122. </div><div class="csl-right-inline"><strong>Jupyter notebooks ? A publishing format for reproducible computational workflows</strong> <div class="csl-block">Thomas Kluyver, Benjamin Ragan-Kelley, Fernando Pérez, Brian Granger, Matthias Bussonnier, Jonathan Frederic, Kyle Kelley, Jessica Hamrick, Jason Grout, Sylvain Corlay, … Jupyter development team</div> <em>Positioning and power in academic publishing: Players, agents and agendas</em> (2016) <a href="https://eprints.soton.ac.uk/403913/">https://eprints.soton.ac.uk/403913/</a></div>
</div>
<div id="ref-ut5ogNdw" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">123. </div><div class="csl-right-inline"><strong>Numerical techniques for large cosmological N-body simulations</strong> <div class="csl-block">G Efstathiou, M Davis, SDM White, CS Frenk</div> <em>The Astrophysical Journal Supplement Series</em> (1985-02) <a href="https://doi.org/d62mzw">https://doi.org/d62mzw</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1086/191003">10.1086/191003</a></div></div>
</div>
<div id="ref-1EbFITEAZ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">124. </div><div class="csl-right-inline"><strong>HOP: A New Group‐finding Algorithm for&lt;i&gt;N&lt;/i&gt;‐Body Simulations</strong> <div class="csl-block">Daniel J Eisenstein, Piet Hut</div> <em>The Astrophysical Journal</em> (1998-05) <a href="https://doi.org/d4frsz">https://doi.org/d4frsz</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1086/305535">10.1086/305535</a></div></div>
</div>
<div id="ref-Bz1lkIYo" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">125. </div><div class="csl-right-inline"><strong>ORIGAMI: DELINEATING COSMIC STRUCTURES WITH PHASE-SPACE FOLDS</strong> <div class="csl-block">MARK C NEYRINCK, BRIDGET L FALCK, ALEX S SZALAY</div> <em>The Thirteenth Marcel Grossmann Meeting</em> (2015-01-29) <a href="https://doi.org/gtrcq6">https://doi.org/gtrcq6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1142/9789814623995_0378">10.1142/9789814623995_0378</a></div></div>
</div>
<div id="ref-TSiqnjZ6" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">126. </div><div class="csl-right-inline"><strong>THE ROCKSTAR PHASE-SPACE TEMPORAL HALO FINDER AND THE VELOCITY OFFSETS OF CLUSTER CORES</strong> <div class="csl-block">Peter S Behroozi, Risa H Wechsler, Hao-Yi Wu</div> <em>The Astrophysical Journal</em> (2012-12-20) <a href="https://doi.org/ggz96g">https://doi.org/ggz96g</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0004-637x/762/2/109">10.1088/0004-637x/762/2/109</a></div></div>
</div>
<div id="ref-KKAuiCtl" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">127. </div><div class="csl-right-inline"><strong>ytree-project/ytree: ytree 3.2.1 Release</strong> <div class="csl-block">Britton Smith, Meagan Lang, John Wise, Juanjo Bazán</div> <em>Zenodo</em> (2023-09-15) <a href="https://doi.org/gtrcq8">https://doi.org/gtrcq8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5281/zenodo.8349044">10.5281/zenodo.8349044</a></div></div>
</div>
<div id="ref-1DN6YqcpG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">128. </div><div class="csl-right-inline"><strong>Dark Sky Simulations: Early Data Release</strong> <div class="csl-block">Samuel W Skillman, Michael S Warren, Matthew J Turk, Risa H Wechsler, Daniel E Holz, PM Sutter</div> <em>arXiv</em> (2014) <a href="https://doi.org/gtrcq7">https://doi.org/gtrcq7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.48550/arxiv.1407.2600">10.48550/arxiv.1407.2600</a></div></div>
</div>
<div id="ref-1B7zT2Jsm" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">129. </div><div class="csl-right-inline"><strong>MPI: A message-passing interface standard version 4.0</strong> <div class="csl-block">Message Passing Interface Forum</div> (2021-06) <a href="https://www.mpi-forum.org/docs/mpi-4.0/mpi40-report.pdf">https://www.mpi-forum.org/docs/mpi-4.0/mpi40-report.pdf</a></div>
</div>
<div id="ref-19u2YdIvE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">130. </div><div class="csl-right-inline"><strong>yt-project/yt_astro_analysis: yt_astro_analysis-1.1.3</strong> <div class="csl-block">Britton Smith, Matthew Turk, John ZuHone, Clément Robert, Stephen Skory, Cameron Hummels, Andrew Myers, Kacper Kowalik, Eganhila, Sam Skillman, … Stonnes</div> <em>Zenodo</em> (2023-10-11) <a href="https://doi.org/gtrcq9">https://doi.org/gtrcq9</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5281/zenodo.8431185">10.5281/zenodo.8431185</a></div></div>
</div>
<div id="ref-Hd7r5YCP" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">131. </div><div class="csl-right-inline"><strong>THE SANTA FE LIGHT CONE SIMULATION PROJECT. II. THE PROSPECTS FOR DIRECT DETECTION OF THE WHIM WITH SZE SURVEYS</strong> <div class="csl-block">Eric J Hallman, Brian W O'Shea, Britton D Smith, Jack O Burns, Michael L Norman</div> <em>The Astrophysical Journal</em> (2009-06-04) <a href="https://doi.org/bh8rmw">https://doi.org/bh8rmw</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/0004-637x/698/2/1795">10.1088/0004-637x/698/2/1795</a></div></div>
</div>
<div id="ref-nfcCcNZX" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">132. </div><div class="csl-right-inline"><strong>ASCL.net - RADMC-3D: A multi-purpose radiative transfer tool</strong> <a href="http://ascl.net/1202.015">http://ascl.net/1202.015</a></div>
</div>
<div id="ref-1DgKJy4lE" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">133. </div><div class="csl-right-inline"><strong>Trident: A Universal Tool for Generating Synthetic Absorption Spectra from Astrophysical Simulations</strong> <div class="csl-block">Cameron B Hummels, Britton D Smith, Devin W Silvia</div> <em>The Astrophysical Journal</em> (2017-09-20) <a href="https://doi.org/gcx4th">https://doi.org/gcx4th</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4357/aa7e2d">10.3847/1538-4357/aa7e2d</a></div></div>
</div>
<div id="ref-g9AYex8L" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">134. </div><div class="csl-right-inline"><strong>powderday: Dust Radiative Transfer for Galaxy Simulations</strong> <div class="csl-block">Desika Narayanan, Matthew J Turk, Thomas Robitaille, Ashley J Kelly, BConnor McClellan, Ray S Sharma, Prerak Garg, Matthew Abruzzo, Ena Choi, Charlie Conroy, … John H Wise</div> <em>The Astrophysical Journal Supplement Series</em> (2021-01-01) <a href="https://doi.org/gpxcx8">https://doi.org/gpxcx8</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3847/1538-4365/abc487">10.3847/1538-4365/abc487</a></div></div>
</div>
<div id="ref-1CgEbr6CQ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">135. </div><div class="csl-right-inline"><strong>The formation of submillimetre-bright galaxies from gas infall over a billion years</strong> <div class="csl-block">Desika Narayanan, Matthew Turk, Robert Feldmann, Thomas Robitaille, Philip Hopkins, Robert Thompson, Christopher Hayward, David Ball, Claude-André Faucher-Giguère, Dušan Kereš</div> <em>Nature</em> (2015-09-23) <a href="https://doi.org/f7r445">https://doi.org/f7r445</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/nature15383">10.1038/nature15383</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26399829">26399829</a></div></div>
</div>
<div id="ref-6dD1NLpJ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">136. </div><div class="csl-right-inline"><strong>HYPERION: an open-source parallelized three-dimensional dust continuum radiative transfer code</strong> <div class="csl-block">TP Robitaille</div> <em>Astronomy &amp;amp; Astrophysics</em> (2011-12) <a href="https://doi.org/ckp63w">https://doi.org/ckp63w</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1051/0004-6361/201117150">10.1051/0004-6361/201117150</a></div></div>
</div>
<div id="ref-VB3RIyxd" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">137. </div><div class="csl-right-inline"><strong>ytree: merger-tree toolkit</strong> <div class="csl-block">Britton Smith, Meagan Lang</div> <em>Zenodo</em> (2018-02-16) <a href="https://doi.org/gcx4x2">https://doi.org/gcx4x2</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5281/zenodo.1174374">10.5281/zenodo.1174374</a></div></div>
</div>
<div id="ref-N2QcMTPv" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">138. </div><div class="csl-right-inline"><strong>Simulating X-ray Observations with Python</strong> <div class="csl-block">John ZuHone, Veronica Biffi, Eric Hallman, Scott Randall, Adam Foster, Christian Schmid</div> <em>Proceedings of the Python in Science Conference</em> (2014) <a href="https://doi.org/gpxcx7">https://doi.org/gpxcx7</a> <div class="csl-block">DOI: <a href="https://doi.org/10.25080/majora-14bd3278-010">10.25080/majora-14bd3278-010</a></div></div>
</div>
<div id="ref-gJwEappY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">139. </div><div class="csl-right-inline"><strong>Observing simulated galaxy clusters with phox: a novel X-ray photon simulator</strong> <div class="csl-block">V Biffi, K Dolag, H Böhringer, G Lemson</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2012-01) <a href="https://doi.org/fz38qm">https://doi.org/fz38qm</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1111/j.1365-2966.2011.20278.x">10.1111/j.1365-2966.2011.20278.x</a></div></div>
</div>
<div id="ref-nty3AM0C" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">140. </div><div class="csl-right-inline"><strong>Investigating the velocity structure and X-ray observable properties of simulated galaxy clusters with PHOX</strong> <div class="csl-block">V Biffi, K Dolag, H Böhringer</div> <em>Monthly Notices of the Royal Astronomical Society</em> (2012-10-29) <a href="https://doi.org/f42dgv">https://doi.org/f42dgv</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1093/mnras/sts120">10.1093/mnras/sts120</a></div></div>
</div>
<div id="ref-3Gi4Hs1H" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">141. </div><div class="csl-right-inline"><strong>Houdini for Astrophysical Visualization</strong> <div class="csl-block">JP Naiman, Kalina Borkiewicz, AJ Christensen</div> <em>Publications of the Astronomical Society of the Pacific</em> (2017-04-18) <a href="https://doi.org/f96hcf">https://doi.org/f96hcf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1088/1538-3873/aa51b3">10.1088/1538-3873/aa51b3</a></div></div>
</div>
<div id="ref-1AQkLoGRL" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">142. </div><div class="csl-right-inline"><strong>SI2-SSI: Inquiry-Focused Volumetric Data Analysis Across Scientific Domains: Sustaining and Expanding the yt Community</strong> <div class="csl-block">Matthew Turk, Nathan Goldbaum, 0000-0002-6226-7689, Leigh Orf</div> <em>figshare</em> (2017) <a href="https://doi.org/grmmjc">https://doi.org/grmmjc</a> <div class="csl-block">DOI: <a href="https://doi.org/10.6084/m9.figshare.2061465.v1">10.6084/m9.figshare.2061465.v1</a></div></div>
</div>
<div id="ref-JmcaooMA" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">143. </div><div class="csl-right-inline"><strong>SI2-SSE: yt: Reusable Components for Simulating, Analyzing and Visualizing Astrophysical Systems</strong> <div class="csl-block">Matthew Turk</div> <em>figshare</em> (2014) <a href="https://doi.org/grmmjd">https://doi.org/grmmjd</a> <div class="csl-block">DOI: <a href="https://doi.org/10.6084/m9.figshare.909413.v1">10.6084/m9.figshare.909413.v1</a></div></div>
</div>
<div id="ref-GSJFOpai" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">144. </div><div class="csl-right-inline"><strong>Collaborative research: Elements: Shared data-delivery infrastructure to enable discovery with the next generation of Dark Matter and Computational Astrophysics experiments</strong> <div class="csl-block">Amy Roberts, Matthew Turk</div> <em>Zenodo</em> (2020-10-28) <a href="https://doi.org/grmmjb">https://doi.org/grmmjb</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5281/zenodo.4158589">10.5281/zenodo.4158589</a></div></div>
</div>
<div id="ref-LRQOJg18" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">145. </div><div class="csl-right-inline"><strong>Corrective commit probability: a measure of the effort invested in bug fixing</strong> <div class="csl-block">Idan Amit, Dror G Feitelson</div> <em>Software Quality Journal</em> (2021-08-05) <a href="https://doi.org/grp6mj">https://doi.org/grp6mj</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1007/s11219-021-09564-z">10.1007/s11219-021-09564-z</a></div></div>
</div>
</div>
<h2 id="acknowledgments">Acknowledgments</h2>
<p>The authors of this paper would like to extend their deepest gratitude to the many, many individual and institutions that have contributed, directly or indirectly, to the growth of both <code>yt</code> and the <code>yt</code> community.</p>
<p>We particularly thank KIPAC and SLAC at Stanford, the University of California at San Diego and Santa Cruz, the High-Performance Astro Computing Center, Columbia University, the University of Illinois, University of Colorado at Boulder, University of Edinburgh, the scientific Python community, NumFOCUS,</p>
<p>MT dedicates this paper to the memory of Michael S. Warren, who was both friend and mentor to several of us on the development team.
Mike passed away during the final stages of its completion, but his influence was felt deeply along the way.
Mike was a truly inspirational person, and provided deep mentorship and guidance to multiple of the authors of this paper – and stepped in with encouragement, advice and friendship at two specific, critical junctures in the development history of yt.
Thank you, Mike, and we miss you.
Tomorrow, we’re gonna do Stonehenge.</p>
<!-- default theme -->

<style>
  /* import google fonts */
  @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
  @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

  /* -------------------------------------------------- */
  /* global */
  /* -------------------------------------------------- */

  /* all elements */
  * {
    /* force sans-serif font unless specified otherwise */
    font-family: "Open Sans", "Helvetica", sans-serif;

    /* prevent text inflation on some mobile browsers */
    -webkit-text-size-adjust: none !important;
    -moz-text-size-adjust: none !important;
    -o-text-size-adjust: none !important;
    text-size-adjust: none !important;
  }

  @media only screen {
    /* "page" element */
    body {
      position: relative;
      box-sizing: border-box;
      font-size: 12pt;
      line-height: 1.5;
      max-width: 8.5in;
      margin: 20px auto;
      padding: 40px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* "page" element */
    body {
      padding: 20px;
      margin: 0;
      border-radius: 0;
      border: none;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
      background: none;
    }
  }

  /* -------------------------------------------------- */
  /* headings */
  /* -------------------------------------------------- */

  /* all headings */
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 20px 0;
    padding: 0;
    font-weight: bold;
  }

  /* biggest heading */
  h1 {
    margin: 40px 0;
    text-align: center;
  }

  /* second biggest heading */
  h2 {
    margin-top: 30px;
    padding-bottom: 5px;
    border-bottom: solid 1px #bdbdbd;
  }

  /* heading font sizes */
  h1 {
    font-size: 2em;
  }
  h2 {
    font-size: 1.5em;
  }
  h3 {
    font-size: 1.35em;
  }
  h4 {
    font-size: 1.25em;
  }
  h5 {
    font-size: 1.15em;
  }
  h6 {
    font-size: 1em;
  }

  /* -------------------------------------------------- */
  /* manuscript header */
  /* -------------------------------------------------- */

  /* manuscript title */
  header > h1 {
    margin: 0;
  }

  /* manuscript title caption text (ie "automatically generated on") */
  header + p {
    text-align: center;
    margin-top: 10px;
  }

  /* -------------------------------------------------- */
  /* text elements */
  /* -------------------------------------------------- */

  /* links */
  a {
    color: #2196f3;
    overflow-wrap: break-word;
  }

  /* superscripts and subscripts */
  sub,
  sup {
    /* prevent from affecting line height */
    line-height: 0;
  }

  /* unordered and ordered lists*/
  ul,
  ol {
    padding-left: 20px;
  }

  /* class for styling text semibold */
  .semibold {
    font-weight: 600;
  }

  /* class for styling elements horizontally left aligned */
  .left {
    display: block;
    text-align: left;
    margin-left: auto;
    margin-right: 0;
    justify-content: left;
  }

  /* class for styling elements horizontally centered */
  .center {
    display: block;
    text-align: center;
    margin-left: auto;
    margin-right: auto;
    justify-content: center;
  }

  /* class for styling elements horizontally right aligned */
  .right {
    display: block;
    text-align: right;
    margin-left: 0;
    margin-right: auto;
    justify-content: right;
  }

  /* -------------------------------------------------- */
  /* section elements */
  /* -------------------------------------------------- */

  /* horizontal divider line */
  hr {
    border: none;
    height: 1px;
    background: #bdbdbd;
  }

  /* paragraphs, horizontal dividers, figures, tables, code */
  p,
  hr,
  figure,
  table,
  pre {
    /* treat all as "paragraphs", with consistent vertical margins */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* figures */
  /* -------------------------------------------------- */

  /* figure */
  figure {
    max-width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure caption */
  figcaption {
    padding: 0;
    padding-top: 10px;
  }

  /* figure image element */
  figure > img,
  figure > svg {
    max-width: 100%;
    display: block;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure auto-number */
  img + figcaption > span:first-of-type,
  svg + figcaption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* tables */
  /* -------------------------------------------------- */

  /* table */
  table {
    border-collapse: collapse;
    border-spacing: 0;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
    font-size: 10pt;
  }

  /* table cells */
  th,
  td {
    border: solid 1px #bdbdbd;
    padding: 10px;
    /* squash table if too wide for page by forcing line breaks */
    overflow-wrap: break-word;
    /* I am disabling this because the YTEP table gets a bit confused. */
    /*word-break: break-word;*/
  }

  /* header row and even rows */
  th,
  tr:nth-child(2n) {
    background-color: #fafafa;
  }

  /* odd rows */
  tr:nth-child(2n + 1) {
    background-color: #ffffff;
  }

  /* table caption */
  caption {
    text-align: left;
    padding: 0;
    padding-bottom: 10px;
  }

  /* table auto-number */
  table > caption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* code */
  /* -------------------------------------------------- */

  /* multi-line code block */
  pre {
    padding: 10px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
    break-inside: avoid;
    text-align: left;
  }

  /* inline code, ie code within normal text */
  :not(pre) > code {
    padding: 0 4px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
  }

  /* code text */
  /* apply all children, to reach syntax highlighting sub-elements */
  code,
  code * {
    /* force monospace font */
    font-family: "Source Code Pro", "Courier New", monospace;
  }

  /* -------------------------------------------------- */
  /* quotes */
  /* -------------------------------------------------- */

  /* quoted text */
  blockquote {
    margin: 0;
    padding: 0;
    border-left: 4px solid #bdbdbd;
    padding-left: 16px;
    break-inside: avoid;
  }

  /* -------------------------------------------------- */
  /* banners */
  /* -------------------------------------------------- */

  /* info banners */
  .banner {
    box-sizing: border-box;
    display: block;
    position: relative;
    width: 100%;
    margin-top: 20px;
    margin-bottom: 20px;
    padding: 20px;
    text-align: center;
  }

  /* paragraph in banner */
  .banner > p {
    margin: 0;
  }

  /* -------------------------------------------------- */
  /* highlight colors */
  /* -------------------------------------------------- */

  .white {
    background: #ffffff;
  }
  .lightgrey {
    background: #eeeeee;
  }
  .grey {
    background: #757575;
  }
  .darkgrey {
    background: #424242;
  }
  .black {
    background: #000000;
  }
  .lightred {
    background: #ffcdd2;
  }
  .lightyellow {
    background: #ffecb3;
  }
  .lightgreen {
    background: #dcedc8;
  }
  .lightblue {
    background: #e3f2fd;
  }
  .lightpurple {
    background: #f3e5f5;
  }
  .red {
    background: #f44336;
  }
  .orange {
    background: #ff9800;
  }
  .yellow {
    background: #ffeb3b;
  }
  .green {
    background: #4caf50;
  }
  .blue {
    background: #2196f3;
  }
  .purple {
    background: #9c27b0;
  }
  .white,
  .lightgrey,
  .lightred,
  .lightyellow,
  .lightgreen,
  .lightblue,
  .lightpurple,
  .orange,
  .yellow,
  .white a,
  .lightgrey a,
  .lightred a,
  .lightyellow a,
  .lightgreen a,
  .lightblue a,
  .lightpurple a,
  .orange a,
  .yellow a {
    color: #000000;
  }
  .grey,
  .darkgrey,
  .black,
  .red,
  .green,
  .blue,
  .purple,
  .grey a,
  .darkgrey a,
  .black a,
  .red a,
  .green a,
  .blue a,
  .purple a {
    color: #ffffff;
  }

  /* -------------------------------------------------- */
  /* buttons */
  /* -------------------------------------------------- */

  /* class for styling links like buttons */
  .button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin: 5px;
    padding: 10px 20px;
    font-size: 0.75em;
    font-weight: 600;
    text-transform: uppercase;
    text-decoration: none;
    letter-spacing: 1px;
    background: none;
    color: #2196f3;
    border: solid 1px #bdbdbd;
    border-radius: 5px;
  }

  /* buttons when hovered */
  .button:hover:not([disabled]),
  .icon_button:hover:not([disabled]) {
    cursor: pointer;
    background: #f5f5f5;
  }

  /* buttons when disabled */
  .button[disabled],
  .icon_button[disabled] {
    opacity: 0.35;
    pointer-events: none;
  }

  /* class for styling buttons containg only single icon */
  .icon_button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    margin: 0;
    padding: 0;
    background: none;
    border-radius: 5px;
    border: none;
    width: 20px;
    height: 20px;
    min-width: 20px;
    min-height: 20px;
  }

  /* icon button inner svg image */
  .icon_button > svg {
    height: 16px;
  }

  /* -------------------------------------------------- */
  /* icons */
  /* -------------------------------------------------- */

  /* class for styling icons inline with text */
  .inline_icon {
    height: 1em;
    position: relative;
    top: 0.125em;
  }

  /* -------------------------------------------------- */
  /* references */
  /* -------------------------------------------------- */

  .csl-entry {
    margin-top: 15px;
    margin-bottom: 15px;
  }

  /* -------------------------------------------------- */
  /* print control */
  /* -------------------------------------------------- */

  @media print {
    @page {
      /* suggested printing margin */
      margin: 0.5in;
    }

    /* document and "page" elements */
    html,
    body {
      margin: 0;
      padding: 0;
      width: 100%;
      height: 100%;
    }

    /* "page" element */
    body {
      font-size: 11pt !important;
      line-height: 1.35;
    }

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: 15px 0;
    }

    /* figures and tables */
    figure,
    table {
      font-size: 0.85em;
    }

    /* table cells */
    th,
    td {
      padding: 5px;
    }

    /* shrink font awesome icons */
    i.fas,
    i.fab,
    i.far,
    i.fal {
      transform: scale(0.85);
    }

    /* decrease banner margins */
    .banner {
      margin-top: 15px;
      margin-bottom: 15px;
      padding: 15px;
    }

    /* class for centering an element vertically on its own page */
    .page_center {
      margin: auto;
      width: 100%;
      height: 100%;
      display: flex;
      align-items: center;
      vertical-align: middle;
      break-before: page;
      break-after: page;
    }

    /* always insert a page break before the element */
    .page_break_before {
      break-before: page;
    }

    /* always insert a page break after the element */
    .page_break_after {
      break-after: page;
    }

    /* avoid page break before the element */
    .page_break_before_avoid {
      break-before: avoid;
    }

    /* avoid page break after the element */
    .page_break_after_avoid {
      break-after: avoid;
    }

    /* avoid page break inside the element */
    .page_break_inside_avoid {
      break-inside: avoid;
    }
  }

  /* -------------------------------------------------- */
  /* override pandoc css quirks */
  /* -------------------------------------------------- */

  .sourceCode {
    /* prevent unsightly overflow in wide code blocks */
    overflow: auto !important;
  }

  div.sourceCode {
    /* prevent background fill on top-most code block  container */
    background: none !important;
  }

  .sourceCode * {
    /* force consistent line spacing */
    line-height: 1.5 !important;
  }

  div.sourceCode {
    /* style code block margins same as <pre> element */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* tablenos */
  /* -------------------------------------------------- */

  /* tablenos wrapper */
  .tablenos {
    width: 100%;
    margin: 20px 0;
  }

  .tablenos > table {
    /* move margins from table to table_wrapper to allow margin collapsing */
    margin: 0;
  }

  @media only screen {
    /* tablenos wrapper */
    .tablenos {
      /* show scrollbar on tables if necessary to prevent overflow */
      overflow-x: auto !important;
    }

    .tablenos th,
    .tablenos td {
      overflow-wrap: unset !important;
      word-break: unset !important;
    }

    /* table in wrapper */
    .tablenos table,
    .tablenos table * {
      /* don't break table words */
      overflow-wrap: normal !important;
    }
  }
</style>
<!-- 
    Plugin Core

    Functions needed for and shared across all first-party plugins.
-->

<script>
  // get element that is target of hash (from link element or url)
  function getHashTarget(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector(`[id="${id}"]`);
    if (!target) return;

    // if figure or table, modify target to get expected element
    if (id.indexOf("fig:") === 0) target = target.querySelector("figure");
    if (id.indexOf("tbl:") === 0) target = target.querySelector("table");

    return target;
  }

  // get position/dimensions of element or viewport
  function getRectInView(element) {
    let rect = {};
    rect.left = 0;
    rect.top = 0;
    rect.right = document.documentElement.clientWidth;
    rect.bottom = document.documentElement.clientHeight;
    let style = {};

    if (element instanceof HTMLElement) {
      rect = element.getBoundingClientRect();
      style = window.getComputedStyle(element);
    }

    const margin = {};
    margin.left = parseFloat(style.marginLeftWidth) || 0;
    margin.top = parseFloat(style.marginTopWidth) || 0;
    margin.right = parseFloat(style.marginRightWidth) || 0;
    margin.bottom = parseFloat(style.marginBottomWidth) || 0;

    const border = {};
    border.left = parseFloat(style.borderLeftWidth) || 0;
    border.top = parseFloat(style.borderTopWidth) || 0;
    border.right = parseFloat(style.borderRightWidth) || 0;
    border.bottom = parseFloat(style.borderBottomWidth) || 0;

    const newRect = {};
    newRect.left = rect.left + margin.left + border.left;
    newRect.top = rect.top + margin.top + border.top;
    newRect.right = rect.right + margin.right + border.right;
    newRect.bottom = rect.bottom + margin.bottom + border.bottom;
    newRect.width = newRect.right - newRect.left;
    newRect.height = newRect.bottom - newRect.top;

    return newRect;
  }

  // get position of element relative to page
  function getRectInPage(element) {
    const rect = getRectInView(element);
    const body = getRectInView(document.body);

    const newRect = {};
    newRect.left = rect.left - body.left;
    newRect.top = rect.top - body.top;
    newRect.right = rect.right - body.left;
    newRect.bottom = rect.bottom - body.top;
    newRect.width = rect.width;
    newRect.height = rect.height;

    return newRect;
  }

  // get closest element before specified element that matches query
  function firstBefore(element, query) {
    while (element && element !== document.body && !element.matches(query))
      element = element.previousElementSibling || element.parentNode;

    return element;
  }

  // check if element is part of collapsed heading
  function isCollapsed(element) {
    while (element && element !== document.body) {
      if (element.dataset.collapsed === "true") return true;
      element = element.parentNode;
    }
    return false;
  }

  // expand any collapsed parent containers of element if necessary
  function expandElement(element) {
    if (isCollapsed(element)) {
      // accordion plugin
      const heading = firstBefore(element, "h2");
      if (heading) heading.click();
      // details/summary HTML element
      const summary = firstBefore(element, "summary");
      if (summary) summary.click();
    }
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);

    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // get list of elements after a start element up to element matching query
  function nextUntil(element, query, exclude) {
    const elements = [];
    while (((element = element.nextElementSibling), element)) {
      if (element.matches(query)) break;
      if (!element.matches(exclude)) elements.push(element);
    }
    return elements;
  }
</script>
<!--
  Accordion Plugin

  Allows sections of content under h2 headings to be collapsible.
-->

<script type="module">
  // whether to always start expanded ('false'), always start collapsed
  // ('true'), or start collapsed when screen small ('auto')
  const startCollapsed = "auto";

  // start script
  function start() {
    // run through each <h2> heading
    const headings = document.querySelectorAll("h2");
    for (const heading of headings) {
      addArrow(heading);

      // start expanded/collapsed based on option
      if (
        startCollapsed === "true" ||
        (startCollapsed === "auto" && isSmallScreen()) ||
        heading.dataset.collapsed === "true"
      )
        collapseHeading(heading);
      else expandElement(heading);
    }

    // attach hash change listener to window
    window.addEventListener("hashchange", onHashChange);
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) goToElement(target);
  }

  // add arrow to heading
  function addArrow(heading) {
    // add arrow button
    const arrow = document.createElement("button");
    arrow.innerHTML = document.querySelector(".icon_angle_down").innerHTML;
    arrow.classList.add("icon_button", "accordion_arrow");
    heading.insertBefore(arrow, heading.firstChild);

    // attach click listener to heading and button
    heading.addEventListener("click", onHeadingClick);
    arrow.addEventListener("click", onArrowClick);
  }

  // determine if on mobile-like device with small screen
  function isSmallScreen() {
    return Math.min(window.innerWidth, window.innerHeight) < 480;
  }

  // when <h2> heading is clicked
  function onHeadingClick(event) {
    // only collapse if <h2> itself is target of click (eg, user did
    // not click on anchor within <h2>)
    if (event.target === this) toggleCollapse(this);
  }

  // when arrow button is clicked
  function onArrowClick() {
    toggleCollapse(this.parentNode);
  }

  // collapse section if expanded, expand if collapsed
  function toggleCollapse(heading) {
    if (heading.dataset.collapsed === "false") collapseHeading(heading);
    else expandElement(heading);
  }

  // elements to exclude from collapse, such as table of contents panel,
  // hypothesis panel, etc
  const exclude = "#toc_panel, div.annotator-frame, #lightbox_overlay";

  // collapse section
  function collapseHeading(heading) {
    heading.setAttribute("data-collapsed", "true");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "true");
  }

  // expand section
  function expandElement(heading) {
    heading.setAttribute("data-collapsed", "false");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "false");
  }

  // get list of elements between this <h2> and next <h2> or <h1>
  // ("children" of the <h2> section)
  function getChildren(heading) {
    return nextUntil(heading, "h2, h1", exclude);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
  <!-- modified from: https://fontawesome.com/icons/angle-down -->
  <svg width="16" height="16" viewBox="0 0 448 512">
    <path
      fill="currentColor"
      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* accordion arrow button */
    .accordion_arrow {
      margin-right: 10px;
    }

    /* arrow icon when <h2> data-collapsed attribute true */
    h2[data-collapsed="true"] > .accordion_arrow > svg {
      transform: rotate(-90deg);
    }

    /* all elements (except <h2>'s) when data-collapsed attribute true */
    *:not(h2)[data-collapsed="true"] {
      display: none;
    }

    /* accordion arrow button when hovered and <h2>'s when hovered */
    .accordion_arrow:hover,
    h2[data-collapsed="true"]:hover,
    h2[data-collapsed="false"]:hover {
      cursor: pointer;
    }
  }

  /* always hide accordion arrow button on print */
  @media only print {
    .accordion_arrow {
      display: none;
    }
  }
</style>
<!--
  Anchors Plugin

  Adds an anchor next to each of a certain type of element that provides a
  human-readable url to that specific item/position in the document (e.g.
  "manuscript.html#abstract"). It also makes it such that scrolling out of view
  of a target removes its identifier from the url.
-->

<script type="module">
  // which types of elements to add anchors next to, in "document.querySelector"
  // format
  const typesQuery =
    'h1, h2, h3, div[id^="fig:"], div[id^="tbl:"], span[id^="eq:"]';

  // start script
  function start() {
    // add anchor to each element of specified types
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) addAnchor(element);

    // attach scroll listener to window
    window.addEventListener("scroll", onScroll);
  }

  // when window is scrolled
  function onScroll() {
    // if url has hash and user has scrolled out of view of hash
    // target, remove hash from url
    const tolerance = 100;
    const target = getHashTarget();
    if (target) {
      if (
        target.getBoundingClientRect().top > window.innerHeight + tolerance ||
        target.getBoundingClientRect().bottom < 0 - tolerance
      )
        history.pushState(null, null, " ");
    }
  }

  // add anchor to element
  function addAnchor(element) {
    let addTo; // element to add anchor button to

    // if figure or table, modify withId and addTo to get expected
    // elements
    if (element.id.indexOf("fig:") === 0) {
      addTo = element.querySelector("figcaption");
    } else if (element.id.indexOf("tbl:") === 0) {
      addTo = element.querySelector("caption");
    } else if (element.id.indexOf("eq:") === 0) {
      addTo = element.querySelector(".eqnos-number");
    }

    addTo = addTo || element;
    const id = element.id || null;

    // do not add anchor if element doesn't have assigned id.
    // id is generated by pandoc and is assumed to be unique and
    // human-readable
    if (!id) return;

    // create anchor button
    const anchor = document.createElement("a");
    anchor.innerHTML = document.querySelector(".icon_link").innerHTML;
    anchor.title = "Link to this part of the document";
    anchor.classList.add("icon_button", "anchor");
    anchor.dataset.ignore = "true";
    anchor.href = "#" + id;
    addTo.appendChild(anchor);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- link icon -->

<template class="icon_link">
  <!-- modified from: https://fontawesome.com/icons/link -->
  <svg width="16" height="16" viewBox="0 0 512 512">
    <path
      fill="currentColor"
      d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* anchor button */
    .anchor {
      opacity: 0;
      margin-left: 5px;
    }

    /* anchor buttons within <h2>'s */
    h2 .anchor {
      margin-left: 10px;
    }

    /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
    *:hover > .anchor,
    .anchor:hover,
    .anchor:focus {
      opacity: 1;
    }

    /* anchor button when hovered */
    .anchor:hover {
      cursor: pointer;
    }
  }

  /* always show anchor button on devices with no mouse/hover ability */
  @media (hover: none) {
    .anchor {
      opacity: 1;
    }
  }

  /* always hide anchor button on print */
  @media only print {
    .anchor {
      display: none;
    }
  }
</style>
<!-- 
    Attributes Plugin

    Allows arbitrary HTML attributes to be attached to (almost) any element.
    Place an HTML comment inside or next to the desired element with the content:
    $attribute="value"
-->

<script type="module">
  // start script
  function start() {
    // get list of comments in document
    const comments = findComments();

    for (const comment of comments)
      if (comment.parentElement)
        addAttributes(comment.parentElement, comment.nodeValue.trim());
  }

  // add html attributes to specified element based on string of
  // html attributes and values
  function addAttributes(element, text) {
    // regex's for finding attribute/value pairs in the format of
    // attribute="value" or attribute='value
    const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
    const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

    // loop through attribute/value pairs
    let match;
    while ((match = text.match(regex2) || text.match(regex1))) {
      // get attribute and value from regex capture groups
      let attribute = match[1];
      let value = match[2];

      // remove from string
      text = text.substring(match.index + match[0].length);

      if (!attribute || !value) break;

      // set attribute of parent element
      try {
        element.setAttribute(attribute, value);
      } catch (error) {
        console.log(error);
      }

      // special case for colspan
      if (attribute === "colspan") removeTableCells(element, value);
    }
  }

  // get list of comment elements in document
  function findComments() {
    const comments = [];

    // iterate over comment nodes in document
    function acceptNode(node) {
      return NodeFilter.FILTER_ACCEPT;
    }
    const iterator = document.createNodeIterator(
      document.body,
      NodeFilter.SHOW_COMMENT,
      acceptNode
    );
    let node;
    while ((node = iterator.nextNode())) comments.push(node);

    return comments;
  }

  // remove certain number of cells after specified cell
  function removeTableCells(cell, number) {
    number = parseInt(number);
    if (!number) return;

    // remove elements
    for (; number > 1; number--) {
      if (cell.nextElementSibling) cell.nextElementSibling.remove();
    }
  }

  // start script on DOMContentLoaded instead of load to ensure this plugins
  // runs before other plugins
  window.addEventListener("DOMContentLoaded", start);
</script>
<!--
  Inline Svg Plugin

  Fetches the text source code of SVGs in <img> tags, and inserts it into the
  document in-place. This provides many advantages, such as being able to style
  SVGs globally from the CSS themes, and being able to make SVGs interactive
  with JavaScript and D3. If you have a script that expects SVGs to be inlined,
  wait for an "SVGLoaded" event before running it.

  Note: This requires the page to be served from a server to work
  See: https://stackoverflow.com/a/11063963/2180570
  Tip: Use the manubot commands to open the page in a local webserver
  See: https://github.com/manubot/rootstock#local-execution
-->

<script type="module">
  // start script
  async function start() {
    // get all <img> tags that have "src" attributes that end with ".svg"
    const imgs = document.querySelectorAll('img[src$=".svg"]');
    const inlineSvgs = Array.from(imgs).map(inlineSvg);
    await Promise.all(inlineSvgs);

    // dispatch "inline finished" event
    document.dispatchEvent(new Event("SVGLoaded"));
  }

  // take an svg <img> tag, fetch its source code, and replace it
  async function inlineSvg(img) {
    try {
      // fetch svg source code
      const response = await fetch(img.src);
      if (!response.ok) throw new Error("Couldn't get SVG source");
      const source = await response.text();
      if (!source.trim()) throw new Error("No SVG source");
      // parse specifically as svg, create new (hidden) dom node
      const svg = new DOMParser()
        .parseFromString(source, "image/svg+xml")
        .querySelector("svg");
      if (!svg) throw new Error("Couldn't parse SVG");
      // transfer original img attributes into new svg
      for (const { name, value } of img.attributes)
        svg.setAttribute(name, value);
      // replace original image with new svg
      img.outerHTML = svg.outerHTML;
    } catch (error) {
      console.log(error);
    }
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>
<!--
  Jump to First Plugin

  Adds a button next to each reference entry, figure, and table that jumps the
  page to the first occurrence of a link to that item in the manuscript.
-->

<script type="module">
  // whether to add buttons next to reference entries
  const references = "true";
  // whether to add buttons next to figures
  const figures = "true";
  // whether to add buttons next to tables
  const tables = "true";

  // start script
  function start() {
    if (references !== "false")
      makeButtons(`div[id^="ref-"]`, ".csl-left-margin", "reference");
    if (figures !== "false")
      makeButtons(`div[id^="fig:"]`, "figcaption", "figure");
    if (tables !== "false") makeButtons(`div[id^="tbl:"]`, "caption", "table");
  }

  // when jump button clicked
  function onButtonClick() {
    const first = getFirstOccurrence(this.dataset.id);
    if (!first) return;

    // update url hash so navigating "back" in history will return user to button
    window.location.hash = this.dataset.id;
    // scroll to link
    const timeout = function () {
      goToElement(first, window.innerHeight * 0.5);
    };
    window.setTimeout(timeout, 0);
  }

  // get first occurrence of link to item in document
  function getFirstOccurrence(id) {
    let query = "a";
    query += '[href="#' + id + '"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelector(query);
  }

  // add button next to each reference entry, figure, or table
  function makeButtons(query, containerQuery, subject) {
    const elements = document.querySelectorAll(query);
    for (const element of elements) {
      const id = element.id;
      const buttonContainer = element.querySelector(containerQuery);
      const first = getFirstOccurrence(id);

      // if can't find link to reference or place to put button, ignore
      if (!first || !buttonContainer) continue;

      // make jump button
      let button = document.createElement("button");
      button.classList.add("icon_button", "jump_arrow");
      button.title = `Jump to the first occurrence of this ${subject} in the document`;
      const icon = document.querySelector(".icon_angle_double_up");
      button.innerHTML = icon.innerHTML;
      button.dataset.id = id;
      button.dataset.ignore = "true";
      button.addEventListener("click", onButtonClick);
      buttonContainer.prepend(button);
    }
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
  <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
  <svg width="16" height="16" viewBox="0 0 320 512">
    <path
      fill="currentColor"
      d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* jump button */
    .jump_arrow {
      position: relative;
      top: 0.125em;
      margin-right: 5px;
    }
  }

  /* always hide jump button on print */
  @media only print {
    .jump_arrow {
      display: none;
    }
  }
</style>
<!-- 
    Lightbox Plugin

    Makes it such that when a user clicks on an image, the image fills the
    screen and the user can pan/drag/zoom the image and navigate between other
    images in the document.
-->

<script type="module">
  // list of possible zoom/scale factors
  const zooms =
    "0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8";
  // whether to fit image to view ('fit'), display at 100% and shrink if
  // necessary ('shrink'), or always display at 100% ('100')
  const defaultZoom = "fit";
  // whether to zoom in/out toward center of view ('true') or mouse ('false')
  const centerZoom = "false";

  // start script
  function start() {
    // run through each <img> element
    const imgs = document.querySelectorAll("figure > img");
    let count = 1;
    for (const img of imgs) {
      img.classList.add("lightbox_document_img");
      img.dataset.number = count;
      img.dataset.total = imgs.length;
      img.addEventListener("click", openLightbox);
      count++;
    }

    // attach mouse and key listeners to window
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("keyup", onKeyUp);
  }

  // when mouse is moved anywhere in window
  function onWindowMouseMove(event) {
    window.mouseX = event.clientX;
    window.mouseY = event.clientY;
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("lightbox_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("lightbox_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeLightbox();
        break;
    }
  }

  // open lightbox
  function openLightbox() {
    const lightbox = makeLightbox(this);
    if (!lightbox) return;

    blurBody(lightbox);
    document.body.appendChild(lightbox);
  }

  // make lightbox
  function makeLightbox(img) {
    // delete lightbox if it exists, start fresh
    closeLightbox();

    // create screen overlay containing lightbox
    const overlay = document.createElement("div");
    overlay.id = "lightbox_overlay";

    // create image info boxes
    const numberInfo = document.createElement("div");
    const zoomInfo = document.createElement("div");
    numberInfo.id = "lightbox_number_info";
    zoomInfo.id = "lightbox_zoom_info";

    // create container for image
    const imageContainer = document.createElement("div");
    imageContainer.id = "lightbox_image_container";
    const lightboxImg = makeLightboxImg(
      img,
      imageContainer,
      numberInfo,
      zoomInfo
    );
    imageContainer.appendChild(lightboxImg);

    // create bottom container for caption and navigation buttons
    const bottomContainer = document.createElement("div");
    bottomContainer.id = "lightbox_bottom_container";
    const caption = makeCaption(img);
    const prevButton = makePrevButton(img);
    const nextButton = makeNextButton(img);
    bottomContainer.appendChild(prevButton);
    bottomContainer.appendChild(caption);
    bottomContainer.appendChild(nextButton);

    // attach top middle and bottom to overlay
    overlay.appendChild(numberInfo);
    overlay.appendChild(zoomInfo);
    overlay.appendChild(imageContainer);
    overlay.appendChild(bottomContainer);

    return overlay;
  }

  // make <img> object that is intuitively draggable and zoomable
  function makeLightboxImg(sourceImg, container, numberInfoBox, zoomInfoBox) {
    // create copy of source <img>
    const img = sourceImg.cloneNode(true);
    img.classList.remove("lightbox_document_img");
    img.removeAttribute("id");
    img.removeAttribute("width");
    img.removeAttribute("height");
    img.style.position = "unset";
    img.style.margin = "0";
    img.style.padding = "0";
    img.style.width = "";
    img.style.height = "";
    img.style.minWidth = "";
    img.style.minHeight = "";
    img.style.maxWidth = "";
    img.style.maxHeight = "";
    img.id = "lightbox_img";

    // build sorted list of zoomSteps
    const zoomSteps = zooms.split(/[^0-9.]/).map((step) => parseFloat(step));
    zoomSteps.sort((a, b) => a - b);

    // <img> object property variables
    let zoom = 1;
    let translateX = 0;
    let translateY = 0;
    let clickMouseX = undefined;
    let clickMouseY = undefined;
    let clickTranslateX = undefined;
    let clickTranslateY = undefined;

    updateNumberInfo();

    // update image numbers displayed in info box
    function updateNumberInfo() {
      numberInfoBox.innerHTML =
        sourceImg.dataset.number + " of " + sourceImg.dataset.total;
    }

    // update zoom displayed in info box
    function updateZoomInfo() {
      let zoomInfo = zoom * 100;
      if (!Number.isInteger(zoomInfo)) zoomInfo = zoomInfo.toFixed(2);
      zoomInfoBox.innerHTML = zoomInfo + "%";
    }

    // move to closest zoom step above current zoom
    const zoomIn = function () {
      for (const zoomStep of zoomSteps) {
        if (zoomStep > zoom) {
          zoom = zoomStep;
          break;
        }
      }
      updateTransform();
    };

    // move to closest zoom step above current zoom
    const zoomOut = function () {
      zoomSteps.reverse();
      for (const zoomStep of zoomSteps) {
        if (zoomStep < zoom) {
          zoom = zoomStep;
          break;
        }
      }
      zoomSteps.reverse();

      updateTransform();
    };

    // update display of <img> based on scale/translate properties
    const updateTransform = function () {
      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      // get new width/height after scale
      const rect = img.getBoundingClientRect();
      // limit translate
      translateX = Math.max(translateX, -rect.width / 2);
      translateX = Math.min(translateX, rect.width / 2);
      translateY = Math.max(translateY, -rect.height / 2);
      translateY = Math.min(translateY, rect.height / 2);

      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      updateZoomInfo();
    };

    // fit <img> to container
    const fit = function () {
      // no x/y offset, 100% zoom by default
      translateX = 0;
      translateY = 0;
      zoom = 1;

      // widths of <img> and container
      const imgWidth = img.naturalWidth;
      const imgHeight = img.naturalHeight;
      const containerWidth = parseFloat(
        window.getComputedStyle(container).width
      );
      const containerHeight = parseFloat(
        window.getComputedStyle(container).height
      );

      // how much zooming is needed to fit <img> to container
      const xRatio = imgWidth / containerWidth;
      const yRatio = imgHeight / containerHeight;
      const maxRatio = Math.max(xRatio, yRatio);
      const newZoom = 1 / maxRatio;

      // fit <img> to container according to option
      if (defaultZoom === "shrink") {
        if (maxRatio > 1) zoom = newZoom;
      } else if (defaultZoom === "fit") zoom = newZoom;

      updateTransform();
    };

    // when mouse wheel is rolled anywhere in container
    const onContainerWheel = function (event) {
      if (!event) return;

      // let ctrl + mouse wheel to zoom behave as normal
      if (event.ctrlKey) return;

      // prevent normal scroll behavior
      event.preventDefault();
      event.stopPropagation();

      // point around which to scale img
      const viewRect = container.getBoundingClientRect();
      const viewX = (viewRect.left + viewRect.right) / 2;
      const viewY = (viewRect.top + viewRect.bottom) / 2;
      const originX = centerZoom === "true" ? viewX : mouseX;
      const originY = centerZoom === "true" ? viewY : mouseY;

      // get point on image under origin
      const oldRect = img.getBoundingClientRect();
      const oldPercentX = (originX - oldRect.left) / oldRect.width;
      const oldPercentY = (originY - oldRect.top) / oldRect.height;

      // increment/decrement zoom
      if (event.deltaY < 0) zoomIn();
      if (event.deltaY > 0) zoomOut();

      // get offset between previous image point and origin
      const newRect = img.getBoundingClientRect();
      const offsetX = originX - (newRect.left + newRect.width * oldPercentX);
      const offsetY = originY - (newRect.top + newRect.height * oldPercentY);

      // translate image to keep image point under origin
      translateX += offsetX;
      translateY += offsetY;

      // perform translate
      updateTransform();
    };

    // when container is clicked
    function onContainerClick(event) {
      // if container itself is target of click, and not other
      // element above it
      if (event.target === this) closeLightbox();
    }

    // when mouse button is pressed on image
    const onImageMouseDown = function (event) {
      // store original mouse position relative to image
      clickMouseX = window.mouseX;
      clickMouseY = window.mouseY;
      clickTranslateX = translateX;
      clickTranslateY = translateY;
      event.stopPropagation();
      event.preventDefault();
    };

    // when mouse button is released anywhere in window
    const onWindowMouseUp = function (event) {
      // reset original mouse position
      clickMouseX = undefined;
      clickMouseY = undefined;
      clickTranslateX = undefined;
      clickTranslateY = undefined;

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mouseup", onWindowMouseUp);
    };

    // when mouse is moved anywhere in window
    const onWindowMouseMove = function (event) {
      if (
        clickMouseX === undefined ||
        clickMouseY === undefined ||
        clickTranslateX === undefined ||
        clickTranslateY === undefined
      )
        return;

      // offset image based on original and current mouse position
      translateX = clickTranslateX + window.mouseX - clickMouseX;
      translateY = clickTranslateY + window.mouseY - clickMouseY;
      updateTransform();
      event.preventDefault();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mousemove", onWindowMouseMove);
    };

    // when window is resized
    const onWindowResize = function (event) {
      fit();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("resize", onWindowResize);
    };

    // attach the necessary event listeners
    img.addEventListener("dblclick", fit);
    img.addEventListener("mousedown", onImageMouseDown);
    container.addEventListener("wheel", onContainerWheel);
    container.addEventListener("mousedown", onContainerClick);
    container.addEventListener("touchstart", onContainerClick);
    window.addEventListener("mouseup", onWindowMouseUp);
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("resize", onWindowResize);

    // run fit() after lightbox atttached to document and <img> Loaded
    // so needed container and img dimensions available
    img.addEventListener("load", fit);

    return img;
  }

  // make caption
  function makeCaption(img) {
    const caption = document.createElement("div");
    caption.id = "lightbox_caption";
    const captionSource = img.nextElementSibling;
    if (captionSource.tagName.toLowerCase() === "figcaption") {
      const captionCopy = makeCopy(captionSource);
      caption.innerHTML = captionCopy.innerHTML;
    }

    caption.addEventListener("touchstart", function (event) {
      event.stopPropagation();
    });

    return caption;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // make button to jump to previous image in document
  function makePrevButton(img) {
    const prevButton = document.createElement("button");
    prevButton.id = "lightbox_prev_button";
    prevButton.title = "Jump to the previous image in the document [←]";
    prevButton.classList.add("icon_button", "lightbox_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;

    // attach click listeners to button
    prevButton.addEventListener("click", function () {
      getPrevImg(img).click();
    });

    return prevButton;
  }

  // make button to jump to next image in document
  function makeNextButton(img) {
    const nextButton = document.createElement("button");
    nextButton.id = "lightbox_next_button";
    nextButton.title = "Jump to the next image in the document [→]";
    nextButton.classList.add("icon_button", "lightbox_button");
    nextButton.innerHTML = document.querySelector(
      ".icon_caret_right"
    ).innerHTML;

    // attach click listeners to button
    nextButton.addEventListener("click", function () {
      getNextImg(img).click();
    });

    return nextButton;
  }

  // get previous image in document
  function getPrevImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if < 1
    if (index - 1 >= 0) index--;
    else index = imgs.length - 1;
    return imgs[index];
  }

  // get next image in document
  function getNextImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if > total
    if (index + 1 <= imgs.length - 1) index++;
    else index = 0;
    return imgs[index];
  }

  // close lightbox
  function closeLightbox() {
    focusBody();

    const lightbox = document.getElementById("lightbox_overlay");
    if (lightbox) lightbox.remove();
  }

  // make all elements behind lightbox non-focusable
  function blurBody(overlay) {
    const all = document.querySelectorAll("*");
    for (const element of all) element.tabIndex = -1;
    document.body.classList.add("body_no_scroll");
  }

  // make all elements focusable again
  function focusBody() {
    const all = document.querySelectorAll("*");
    for (const element of all) element.removeAttribute("tabIndex");
    document.body.classList.remove("body_no_scroll");
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* regular <img> in document when hovered */
    img.lightbox_document_img:hover {
      cursor: pointer;
    }

    .body_no_scroll {
      overflow: hidden !important;
    }

    /* screen overlay */
    #lightbox_overlay {
      display: flex;
      flex-direction: column;
      position: fixed;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      background: rgba(0, 0, 0, 0.75);
      z-index: 3;
    }

    /* middle area containing lightbox image */
    #lightbox_image_container {
      flex-grow: 1;
      display: flex;
      justify-content: center;
      align-items: center;
      overflow: hidden;
      position: relative;
      padding: 20px;
    }

    /* bottom area containing caption */
    #lightbox_bottom_container {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 100px;
      min-height: 100px;
      max-height: 100px;
      background: rgba(0, 0, 0, 0.5);
    }

    /* image number info text box */
    #lightbox_number_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      left: 2px;
      top: 0;
      z-index: 4;
    }

    /* zoom info text box */
    #lightbox_zoom_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      right: 2px;
      top: 0;
      z-index: 4;
    }

    /* copy of image caption */
    #lightbox_caption {
      box-sizing: border-box;
      display: inline-block;
      width: 100%;
      max-height: 100%;
      padding: 10px 0;
      text-align: center;
      overflow-y: auto;
      color: #ffffff;
    }

    /* navigation previous/next button */
    .lightbox_button {
      width: 100px;
      height: 100%;
      min-width: 100px;
      min-height: 100%;
      color: #ffffff;
    }

    /* navigation previous/next button when hovered */
    .lightbox_button:hover {
      background: none !important;
    }

    /* navigation button icon */
    .lightbox_button > svg {
      height: 25px;
    }

    /* figure auto-number */
    #lightbox_caption > span:first-of-type {
      font-weight: bold;
      margin-right: 5px;
    }

    /* lightbox image when hovered */
    #lightbox_img:hover {
      cursor: grab;
    }

    /* lightbox image when grabbed */
    #lightbox_img:active {
      cursor: grabbing;
    }
  }

  /* when on screen < 480px wide */
  @media only screen and (max-width: 480px) {
    /* make navigation buttons skinnier on small screens to make more room for caption text */
    .lightbox_button {
      width: 50px;
      min-width: 50px;
    }
  }

  /* always hide lightbox on print */
  @media only print {
    #lightbox_overlay {
      display: none;
    }
  }
</style>
<!-- 
  Link Highlight Plugin

  Makes it such that when a user hovers or focuses a link, other links that have
  the same target will be highlighted. It also makes it such that when clicking
  a link, the target of the link (eg reference, figure, table) is briefly
  highlighted.
-->

<script type="module">
  // whether to also highlight links that go to external urls
  const externalLinks = "false";
  // whether user must click off to unhighlight instead of just
  // un-hovering
  const clickUnhighlight = "false";
  // whether to also highlight links that are unique
  const highlightUnique = "true";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach mouse and focus listeners to link
      link.addEventListener("mouseenter", onLinkFocus);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("mouseleave", onLinkUnhover);
    }

    // attach click and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("hashchange", onHashChange);

    // run hash change on window load in case user has navigated
    // directly to hash
    onHashChange();
  }

  // when link is focused (tabbed to) or hovered
  function onLinkFocus() {
    highlight(this);
  }

  // when link is unhovered
  function onLinkUnhover() {
    if (clickUnhighlight !== "true") unhighlightAll();
  }

  // when the mouse is clicked anywhere in window
  function onClick(event) {
    unhighlightAll();
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) glowElement(target);
  }

  // start glow sequence on an element
  function glowElement(element) {
    const startGlow = function () {
      onGlowEnd();
      element.dataset.glow = "true";
      element.addEventListener("animationend", onGlowEnd);
    };
    const onGlowEnd = function () {
      element.removeAttribute("data-glow");
      element.removeEventListener("animationend", onGlowEnd);
    };
    startGlow();
  }

  // highlight link and all others with same target
  function highlight(link) {
    // force unhighlight all to start fresh
    unhighlightAll();

    // get links with same target
    if (!link) return;
    const sameLinks = getSameLinks(link);

    // if link unique and option is off, exit and don't highlight
    if (sameLinks.length <= 1 && highlightUnique !== "true") return;

    // highlight all same links, and "select" (special highlight) this
    // one
    for (const sameLink of sameLinks) {
      if (sameLink === link) sameLink.setAttribute("data-selected", "true");
      else sameLink.setAttribute("data-highlighted", "true");
    }
  }

  // unhighlight all links
  function unhighlightAll() {
    const links = getLinks();
    for (const link of links) {
      link.setAttribute("data-selected", "false");
      link.setAttribute("data-highlighted", "false");
    }
  }

  // get links with same target
  function getSameLinks(link) {
    const results = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        results.push(otherLink);
    }
    return results;
  }

  // get all links of types we wish to handle
  function getLinks() {
    let query = "a";
    if (externalLinks !== "true") query += '[href^="#"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelectorAll(query);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<style>
  @media only screen {
    /* anything with data-highlighted attribute true */
    [data-highlighted="true"] {
      background: #ffeb3b;
    }

    /* anything with data-selected attribute true */
    [data-selected="true"] {
      background: #ff8a65 !important;
    }

    /* animation definition for glow */
    @keyframes highlight_glow {
      0% {
        background: none;
      }
      10% {
        background: #bbdefb;
      }
      100% {
        background: none;
      }
    }

    /* anything with data-glow attribute true */
    [data-glow="true"] {
      animation: highlight_glow 2s;
    }
  }
</style>
<!--
  Table of Contents Plugin

  Provides a "table of contents" (toc) panel on the side of the document that
  allows the user to conveniently navigate between sections of the document.
-->

<script type="module">
  // which types of elements to add links for, in "document.querySelector" format
  const typesQuery = "h1, h2, h3";
  // whether toc starts open. use 'true' or 'false', or 'auto' to
  // use 'true' behavior when screen wide enough and 'false' when not
  const startOpen = "false";
  // whether toc closes when clicking on toc link. use 'true' or
  // 'false', or 'auto' to use 'false' behavior when screen wide
  // enough and 'true' when not
  const clickClose = "auto";
  // if list item is more than this many characters, text will be
  // truncated
  const charLimit = "50";
  // whether or not to show bullets next to each toc item
  const bullets = "false";

  // start script
  function start() {
    // make toc panel and populate with entries (links to document
    // sections)
    const panel = makePanel();
    if (!panel) return;
    makeEntries(panel);
    // attach panel to document after making entries, so 'toc' heading
    // in panel isn't included in toc
    document.body.insertBefore(panel, document.body.firstChild);

    // initial panel state
    if (startOpen === "true" || (startOpen === "auto" && !isSmallScreen()))
      openPanel();
    else closePanel();

    // attach click, scroll, and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("scroll", onScroll);
    window.addEventListener("hashchange", onScroll);
    window.addEventListener("keyup", onKeyUp);
    onScroll();

    // add class to push document body down out of way of toc button
    document.body.classList.add("toc_body_nudge");
  }

  // determine if screen wide enough to fit toc panel
  function isSmallScreen() {
    // in default theme:
    // 816px = 8.5in = width of "page" (<body>) element
    // 260px = min width of toc panel (*2 for both sides of <body>)
    return window.innerWidth < 816 + 260 * 2;
  }

  // when mouse is clicked anywhere in window
  function onClick() {
    if (isSmallScreen()) closePanel();
  }

  // when window is scrolled or hash changed
  function onScroll() {
    highlightViewed();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    // close on esc
    if (event.key === "Escape") closePanel();
  }

  // find entry of currently viewed document section in toc and highlight
  function highlightViewed() {
    const firstId = getFirstInView(typesQuery);

    // get toc entries (links), unhighlight all, then highlight viewed
    const list = document.getElementById("toc_list");
    if (!firstId || !list) return;
    const links = list.querySelectorAll("a");
    for (const link of links) link.dataset.viewing = "false";
    const link = list.querySelector('a[href="#' + firstId + '"]');
    if (!link) return;
    link.dataset.viewing = "true";
  }

  // get first or previous toc listed element in top half of view
  function getFirstInView(query) {
    // get all elements matching query and with id
    const elements = document.querySelectorAll(query);
    const elementsWithIds = [];
    for (const element of elements) {
      if (element.id) elementsWithIds.push(element);
    }

    // get first or previous element in top half of view
    for (let i = 0; i < elementsWithIds.length; i++) {
      const element = elementsWithIds[i];
      const prevElement = elementsWithIds[Math.max(0, i - 1)];
      if (element.getBoundingClientRect().top >= 0) {
        if (element.getBoundingClientRect().top < window.innerHeight / 2)
          return element.id;
        else return prevElement.id;
      }
    }
  }

  // make panel
  function makePanel() {
    // create panel
    const panel = document.createElement("div");
    panel.id = "toc_panel";
    if (bullets === "true") panel.dataset.bullets = "true";

    // create header
    const header = document.createElement("div");
    header.id = "toc_header";

    // create toc button
    const button = document.createElement("button");
    button.id = "toc_button";
    button.innerHTML = document.querySelector(".icon_th_list").innerHTML;
    button.title = "Table of Contents";
    button.classList.add("icon_button");

    // create header text
    const text = document.createElement("h4");
    text.innerHTML = "Table of Contents";

    // create container for toc list
    const list = document.createElement("div");
    list.id = "toc_list";

    // attach click listeners
    panel.addEventListener("click", onPanelClick);
    header.addEventListener("click", onHeaderClick);
    button.addEventListener("click", onButtonClick);

    // attach elements
    header.appendChild(button);
    header.appendChild(text);
    panel.appendChild(header);
    panel.appendChild(list);

    return panel;
  }

  // create toc entries (links) to each element of the specified types
  function makeEntries(panel) {
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) {
      // do not add link if element doesn't have assigned id
      if (!element.id) continue;

      // create link/list item
      const link = document.createElement("a");
      link.classList.add("toc_link");
      switch (element.tagName.toLowerCase()) {
        case "h1":
          link.dataset.level = "1";
          break;
        case "h2":
          link.dataset.level = "2";
          break;
        case "h3":
          link.dataset.level = "3";
          break;
        case "h4":
          link.dataset.level = "4";
          break;
      }
      link.title = element.innerText;
      let text = element.innerText;
      if (text.length > charLimit) text = text.slice(0, charLimit) + "...";
      link.innerHTML = text;
      link.href = "#" + element.id;
      link.addEventListener("click", onLinkClick);

      // attach link
      panel.querySelector("#toc_list").appendChild(link);
    }
  }

  // when panel is clicked
  function onPanelClick(event) {
    // stop click from propagating to window/document and closing panel
    event.stopPropagation();
  }

  // when header itself is clicked
  function onHeaderClick(event) {
    togglePanel();
  }

  // when button is clicked
  function onButtonClick(event) {
    togglePanel();
    // stop header underneath button from also being clicked
    event.stopPropagation();
  }

  // when link is clicked
  function onLinkClick(event) {
    if (clickClose === "true" || (clickClose === "auto" && isSmallScreen()))
      closePanel();
    else openPanel();
  }

  // open panel if closed, close if opened
  function togglePanel() {
    const panel = document.getElementById("toc_panel");
    if (!panel) return;

    if (panel.dataset.open === "true") closePanel();
    else openPanel();
  }

  // open panel
  function openPanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "true";
  }

  // close panel
  function closePanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "false";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- th list icon -->

<template class="icon_th_list">
  <!-- modified from: https://fontawesome.com/icons/th-list -->
  <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
    <path
      fill="currentColor"
      d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* toc panel */
    #toc_panel {
      box-sizing: border-box;
      position: fixed;
      top: 0;
      left: 0;
      background: #ffffff;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      z-index: 2;
    }

    /* toc panel when closed */
    #toc_panel[data-open="false"] {
      min-width: 60px;
      width: 60px;
      height: 60px;
      border-right: solid 1px #bdbdbd;
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc panel when open */
    #toc_panel[data-open="true"] {
      min-width: 260px;
      max-width: 480px;
      /* keep panel edge consistent distance away from "page" edge */
      width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
      bottom: 0;
      border-right: solid 1px #bdbdbd;
    }

    /* toc panel header */
    #toc_header {
      box-sizing: border-box;
      display: flex;
      flex-direction: row;
      align-items: center;
      height: 60px;
      margin: 0;
      padding: 20px;
    }

    /* toc panel header when hovered */
    #toc_header:hover {
      cursor: pointer;
    }

    /* toc panel header when panel open */
    #toc_panel[data-open="true"] > #toc_header {
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc open/close header button */
    #toc_button {
      margin-right: 20px;
    }

    /* hide toc list and header text when closed */
    #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
    #toc_panel[data-open="false"] > #toc_list {
      display: none;
    }

    /* toc list of entries */
    #toc_list {
      box-sizing: border-box;
      width: 100%;
      padding: 20px;
      position: absolute;
      top: calc(60px + 1px);
      bottom: 0;
      overflow: auto;
    }

    /* toc entry, link to section in document */
    .toc_link {
      display: block;
      padding: 5px;
      position: relative;
      font-weight: 600;
      text-decoration: none;
    }

    /* toc entry when hovered or when "viewed" */
    .toc_link:hover,
    .toc_link[data-viewing="true"] {
      background: #f5f5f5;
    }

    /* toc entry, level 1 indentation */
    .toc_link[data-level="1"] {
      margin-left: 0;
    }

    /* toc entry, level 2 indentation */
    .toc_link[data-level="2"] {
      margin-left: 20px;
    }

    /* toc entry, level 3 indentation */
    .toc_link[data-level="3"] {
      margin-left: 40px;
    }

    /* toc entry, level 4 indentation */
    .toc_link[data-level="4"] {
      margin-left: 60px;
    }

    /* toc entry bullets */
    #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
      position: absolute;
      left: -15px;
      top: -1px;
      font-size: 1.5em;
    }

    /* toc entry, level 2 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
      content: "\2022";
    }

    /* toc entry, level 3 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
      content: "\25AB";
    }

    /* toc entry, level 4 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
      content: "-";
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* push <body> ("page") element down to make room for toc icon */
    .toc_body_nudge {
      padding-top: 60px;
    }

    /* toc icon when panel closed and not hovered */
    #toc_panel[data-open="false"]:not(:hover) {
      background: rgba(255, 255, 255, 0.75);
    }
  }

  /* always hide toc panel on print */
  @media only print {
    #toc_panel {
      display: none;
    }
  }
</style>
<!-- 
  Tooltips Plugin

  Makes it such that when the user hovers or focuses a link to a citation or
  figure, a tooltip appears with a preview of the reference content, along with
  arrows to navigate between instances of the same reference in the document.
-->

<script type="module">
  // whether user must click off to close tooltip instead of just un-hovering
  const clickClose = "false";
  // delay (in ms) between opening and closing tooltip
  const delay = "100";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach hover and focus listeners to link
      link.addEventListener("mouseover", onLinkHover);
      link.addEventListener("mouseleave", onLinkUnhover);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("touchend", onLinkTouch);
    }

    // attach mouse, key, and resize listeners to window
    window.addEventListener("mousedown", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("keyup", onKeyUp);
    window.addEventListener("resize", onResize);
  }

  // when link is hovered
  function onLinkHover() {
    // function to open tooltip
    const delayOpenTooltip = function () {
      openTooltip(this);
    }.bind(this);

    // run open function after delay
    this.openTooltipTimer = window.setTimeout(delayOpenTooltip, delay);
  }

  // when mouse leaves link
  function onLinkUnhover() {
    // cancel opening tooltip
    window.clearTimeout(this.openTooltipTimer);

    // don't close on unhover if option specifies
    if (clickClose === "true") return;

    // function to close tooltip
    const delayCloseTooltip = function () {
      // if tooltip open and if mouse isn't over tooltip, close
      const tooltip = document.getElementById("tooltip");
      if (tooltip && !tooltip.matches(":hover")) closeTooltip();
    };

    // run close function after delay
    this.closeTooltipTimer = window.setTimeout(delayCloseTooltip, delay);
  }

  // when link is focused (tabbed to)
  function onLinkFocus(event) {
    openTooltip(this);
  }

  // when link is touched on touch screen
  function onLinkTouch(event) {
    // attempt to force hover state on first tap always, and trigger
    // regular link click (and navigation) on second tap
    if (event.target === document.activeElement) event.target.click();
    else {
      document.activeElement.blur();
      event.target.focus();
    }
    if (event.cancelable) event.preventDefault();
    event.stopPropagation();
    return false;
  }

  // when mouse is clicked anywhere in window
  function onClick(event) {
    closeTooltip();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("tooltip_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("tooltip_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeTooltip();
        break;
    }
  }

  // when window is resized or zoomed
  function onResize() {
    closeTooltip();
  }

  // get all links of types we wish to handle
  function getLinks() {
    const queries = [];
    // exclude buttons, anchor links, toc links, etc
    const exclude =
      ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    queries.push('a[href^="#ref-"]' + exclude); // citation links
    queries.push('a[href^="#fig:"]' + exclude); // figure links
    const query = queries.join(", ");
    return document.querySelectorAll(query);
  }

  // get links with same target, get index of link in set, get total
  // same links
  function getSameLinks(link) {
    const sameLinks = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        sameLinks.push(otherLink);
    }

    return {
      elements: sameLinks,
      index: sameLinks.indexOf(link),
      total: sameLinks.length,
    };
  }

  // open tooltip
  function openTooltip(link) {
    // delete tooltip if it exists, start fresh
    closeTooltip();

    // make tooltip element
    const tooltip = makeTooltip(link);

    // if source couldn't be found and tooltip not made, exit
    if (!tooltip) return;

    // make navbar elements
    const navBar = makeNavBar(link);
    if (navBar) tooltip.firstElementChild.appendChild(navBar);

    // attach tooltip to page
    document.body.appendChild(tooltip);

    // position tooltip
    const position = function () {
      positionTooltip(link);
    };
    position();

    // if tooltip contains images, position again after they've loaded
    const imgs = tooltip.querySelectorAll("img");
    for (const img of imgs) img.addEventListener("load", position);
  }

  // close (delete) tooltip
  function closeTooltip() {
    const tooltip = document.getElementById("tooltip");
    if (tooltip) tooltip.remove();
  }

  // make tooltip
  function makeTooltip(link) {
    // get target element that link points to
    const source = getSource(link);

    // if source can't be found, exit
    if (!source) return;

    // create new tooltip
    const tooltip = document.createElement("div");
    tooltip.id = "tooltip";
    const tooltipContent = document.createElement("div");
    tooltipContent.id = "tooltip_content";
    tooltip.appendChild(tooltipContent);

    // make copy of source node and put in tooltip
    const sourceCopy = makeCopy(source);
    tooltipContent.appendChild(sourceCopy);

    // attach mouse event listeners
    tooltip.addEventListener("click", onTooltipClick);
    tooltip.addEventListener("mousedown", onTooltipClick);
    tooltip.addEventListener("touchstart", onTooltipClick);
    tooltip.addEventListener("mouseleave", onTooltipUnhover);

    // (for interaction with lightbox plugin)
    // transfer click on tooltip copied img to original img
    const sourceImg = source.querySelector("img");
    const sourceCopyImg = sourceCopy.querySelector("img");
    if (sourceImg && sourceCopyImg) {
      const clickImg = function () {
        sourceImg.click();
        closeTooltip();
      };
      sourceCopyImg.addEventListener("click", clickImg);
    }

    return tooltip;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
      "class",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // when tooltip is clicked
  function onTooltipClick(event) {
    // when user clicks on tooltip, stop click from transferring
    // outside of tooltip (eg, click off to close tooltip, or eg click
    // off to unhighlight same refs)
    event.stopPropagation();
  }

  // when tooltip is unhovered
  function onTooltipUnhover(event) {
    if (clickClose === "true") return;

    // make sure new mouse/touch/focus no longer over tooltip or any
    // element within it
    const tooltip = document.getElementById("tooltip");
    if (!tooltip) return;
    if (this.contains(event.relatedTarget)) return;

    closeTooltip();
  }

  // make nav bar to go betwen prev/next instances of same reference
  function makeNavBar(link) {
    // find other links to the same source
    const sameLinks = getSameLinks(link);

    // don't show nav bar when singular reference
    if (sameLinks.total <= 1) return;

    // find prev/next links with same target
    const prevLink = getPrevLink(link, sameLinks);
    const nextLink = getNextLink(link, sameLinks);

    // create nav bar
    const navBar = document.createElement("div");
    navBar.id = "tooltip_nav_bar";
    const text = sameLinks.index + 1 + " of " + sameLinks.total;

    // create nav bar prev/next buttons
    const prevButton = document.createElement("button");
    const nextButton = document.createElement("button");
    prevButton.id = "tooltip_prev_button";
    nextButton.id = "tooltip_next_button";
    prevButton.title =
      "Jump to the previous occurence of this item in the document [←]";
    nextButton.title =
      "Jump to the next occurence of this item in the document [→]";
    prevButton.classList.add("icon_button");
    nextButton.classList.add("icon_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;
    nextButton.innerHTML =
      document.querySelector(".icon_caret_right").innerHTML;
    navBar.appendChild(prevButton);
    navBar.appendChild(document.createTextNode(text));
    navBar.appendChild(nextButton);

    // attach click listeners to buttons
    prevButton.addEventListener("click", function () {
      onPrevNextClick(link, prevLink);
    });
    nextButton.addEventListener("click", function () {
      onPrevNextClick(link, nextLink);
    });

    return navBar;
  }

  // get previous link with same target
  function getPrevLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if < 1
    let index;
    if (sameLinks.index - 1 >= 0) index = sameLinks.index - 1;
    else index = sameLinks.total - 1;
    return sameLinks.elements[index];
  }

  // get next link with same target
  function getNextLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if > total
    let index;
    if (sameLinks.index + 1 <= sameLinks.total - 1) index = sameLinks.index + 1;
    else index = 0;
    return sameLinks.elements[index];
  }

  // get element that is target of link or url hash
  function getSource(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector('[id="' + id + '"]');
    if (!target) return;

    // if ref or figure, modify target to get expected element
    if (id.indexOf("ref-") === 0) target = target.querySelector(":nth-child(2)");
    else if (id.indexOf("fig:") === 0) target = target.querySelector("figure");

    return target;
  }

  // when prev/next arrow button is clicked
  function onPrevNextClick(link, prevNextLink) {
    if (link && prevNextLink)
      goToElement(prevNextLink, window.innerHeight * 0.5);
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);
    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // determine position to place tooltip based on link position in
  // viewport and tooltip size
  function positionTooltip(link, left, top) {
    const tooltipElement = document.getElementById("tooltip");
    if (!tooltipElement) return;

    // get convenient vars for position/dimensions of
    // link/tooltip/page/view
    link = getRectInPage(link);
    const tooltip = getRectInPage(tooltipElement);
    const view = getRectInPage();

    // horizontal positioning
    if (left)
      // use explicit value
      left = left;
    else if (link.left + tooltip.width < view.right)
      // fit tooltip to right of link
      left = link.left;
    else if (link.right - tooltip.width > view.left)
      // fit tooltip to left of link
      left = link.right - tooltip.width;
    // center tooltip in view
    else left = (view.right - view.left) / 2 - tooltip.width / 2;

    // vertical positioning
    if (top)
      // use explicit value
      top = top;
    else if (link.top - tooltip.height > view.top)
      // fit tooltip above link
      top = link.top - tooltip.height;
    else if (link.bottom + tooltip.height < view.bottom)
      // fit tooltip below link
      top = link.bottom;
    else {
      // center tooltip in view
      top = view.top + view.height / 2 - tooltip.height / 2;
      // nudge off of link to left/right if possible
      if (link.right + tooltip.width < view.right) left = link.right;
      else if (link.left - tooltip.width > view.left)
        left = link.left - tooltip.width;
    }

    tooltipElement.style.left = left + "px";
    tooltipElement.style.top = top + "px";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* tooltip container */
    #tooltip {
      position: absolute;
      width: 50%;
      min-width: 240px;
      max-width: 75%;
      z-index: 1;
    }

    /* tooltip content */
    #tooltip_content {
      margin-bottom: 5px;
      padding: 20px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
      overflow-wrap: break-word;
    }

    /* tooltip copy of paragraphs and figures */
    #tooltip_content > p,
    #tooltip_content > figure {
      margin: 0;
      max-height: 320px;
      overflow-y: auto;
    }

    /* tooltip copy of <img> */
    #tooltip_content > figure > img,
    #tooltip_content > figure > svg {
      max-height: 260px;
    }

    /* navigation bar */
    #tooltip_nav_bar {
      margin-top: 10px;
      text-align: center;
    }

    /* navigation bar previous/next buton */
    #tooltip_nav_bar > .icon_button {
      position: relative;
      top: 3px;
    }

    /* navigation bar previous button */
    #tooltip_nav_bar > .icon_button:first-of-type {
      margin-right: 5px;
    }

    /* navigation bar next button */
    #tooltip_nav_bar > .icon_button:last-of-type {
      margin-left: 5px;
    }
  }

  /* always hide tooltip on print */
  @media only print {
    #tooltip {
      display: none;
    }
  }
</style>
<!--
  Analytics Plugin (third-party) 
  
  Copy and paste code from Google Analytics or similar service here.
-->
<!-- 
  Annotations Plugin

  Allows public annotation of the  manuscript. See https://web.hypothes.is/.
-->

<script type="module">
  // configuration
  window.hypothesisConfig = function () {
    return {
      branding: {
        accentColor: "#2196f3",
        appBackgroundColor: "#f8f8f8",
        ctaBackgroundColor: "#f8f8f8",
        ctaTextColor: "#000000",
        selectionFontFamily: "Open Sans, Helvetica, sans serif",
        annotationFontFamily: "Open Sans, Helvetica, sans serif",
      },
    };
  };

  // hypothesis client script
  const embed = "https://hypothes.is/embed.js";
  // hypothesis annotation count query url
  const query = "https://api.hypothes.is/api/search?limit=0&url=";

  // start script
  function start() {
    const button = makeButton();
    document.body.insertBefore(button, document.body.firstChild);
    insertCount(button);
  }

  // make button
  function makeButton() {
    // create button
    const button = document.createElement("button");
    button.id = "hypothesis_button";
    button.innerHTML = document.querySelector(".icon_hypothesis").innerHTML;
    button.title = "Hypothesis annotations";
    button.classList.add("icon_button");

    function onClick(event) {
      onButtonClick(event, button);
    }

    // attach click listeners
    button.addEventListener("click", onClick);

    return button;
  }

  // insert annotations count
  async function insertCount(button) {
    // get annotation count from Hypothesis based on url
    let count = "-";
    try {
      const canonical = document.querySelector('link[rel="canonical"]');
      const location = window.location;
      const url = encodeURIComponent((canonical || location).href);
      const response = await fetch(query + url);
      const json = await response.json();
      count = json.total || "-";
    } catch (error) {
      console.log(error);
    }

    // put count into button
    const counter = document.createElement("span");
    counter.id = "hypothesis_count";
    counter.innerHTML = count;
    button.title = "View " + count + " Hypothesis annotations";
    button.append(counter);
  }

  // when button is clicked
  function onButtonClick(event, button) {
    const script = document.createElement("script");
    script.src = embed;
    document.body.append(script);
    button.remove();
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
  <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
  <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
    <path
      fill="currentColor"
      d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  /* hypothesis activation button */
  #hypothesis_button {
    box-sizing: border-box;
    position: fixed;
    top: 0;
    right: 0;
    width: 60px;
    height: 60px;
    background: #ffffff;
    border-radius: 0;
    border-left: solid 1px #bdbdbd;
    border-bottom: solid 1px #bdbdbd;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
    z-index: 2;
  }

  /* hypothesis button svg */
  #hypothesis_button > svg {
    position: relative;
    top: -4px;
  }

  /* hypothesis annotation count */
  #hypothesis_count {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 5px;
  }

  /* side panel */
  .annotator-frame {
    width: 280px !important;
  }

  /* match highlight color to rest of theme */
  .annotator-highlights-always-on .annotator-hl {
    background-color: #ffeb3b !important;
  }

  /* match focused color to rest of theme */
  .annotator-hl.annotator-hl-focused {
    background-color: #ff8a65 !important;
  }

  /* match bucket bar color to rest of theme */
  .annotator-bucket-bar {
    background: #f5f5f5 !important;
  }

  /* always hide button, toolbar, and tooltip on print */
  @media only print {
    #hypothesis_button {
      display: none;
    }

    .annotator-frame {
      display: none !important;
    }

    hypothesis-adder {
      display: none !important;
    }
  }
</style>
<!-- 
  Mathjax Plugin (third-party) 

  Allows the proper rendering of math/equations written in LaTeX.
  See https://www.mathjax.org/.
-->

<script type="text/x-mathjax-config">
  // configuration
  MathJax.Hub.Config({
    "CommonHTML": { linebreaks: { automatic: true } },
    "HTML-CSS": { linebreaks: { automatic: true } },
    "SVG": { linebreaks: { automatic: true } },
    "fast-preview": { disabled: true }
  });
</script>

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
  integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
  crossorigin="anonymous"
></script>

<style>
  /* mathjax containers */
  .math.display > span:not(.MathJax_Preview) {
    /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
    display: flex !important;
    overflow-x: auto !important;
    overflow-y: hidden !important;
    justify-content: center;
    align-items: center;
    margin: 0 !important;
  }

  /* right click menu */
  .MathJax_Menu {
    border-radius: 5px !important;
    border: solid 1px #bdbdbd !important;
    box-shadow: none !important;
  }

  /* equation auto-number */
  span[id^="eq:"] > span.math.display + span {
    font-weight: 600;
  }

  /* equation */
  span[id^="eq:"] > span.math.display > span {
    /* nudge to make room for equation auto-number and anchor */
    margin-right: 60px !important;
  }
</style>
</body>
</html>
